{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas, copy, numpy,sys\n",
    "from gumpy.genome import Genome\n",
    "pandas.set_option('display.max_columns', 999)\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "numpy.set_printoptions(threshold=sys.maxsize)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.25.1'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pandas.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "TEST_CASE_DIR = \"tests/test-cases/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference=Genome(genbank_file=\"config/NC_004148.2.gbk\",name=\"HMPV\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "sample_01=copy.deepcopy(reference)\n",
    "sample_01.apply_vcf_file(vcf_file=TEST_CASE_DIR+\"01.vcf\",ignore_status=True,ignore_filter=True,metadata_fields=['GT_CONF','GT_CONF_PERCENTILE'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['4687t>c', '4725t>c', '13333c>z', '13335t>a', '4730_indel', '4735_indel']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_01.list_variants_wrt(reference)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df=sample_01.table_variants_wrt(reference)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VARIANT</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENOME_INDEX</th>\n",
       "      <th>GENE</th>\n",
       "      <th>ELEMENT_TYPE</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>NUCLEOTIDE_NUMBER</th>\n",
       "      <th>AMINO_ACID_NUMBER</th>\n",
       "      <th>ASSOCIATED_WITH_GENE</th>\n",
       "      <th>IN_PROMOTER</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IS_INDEL</th>\n",
       "      <th>IS_HET</th>\n",
       "      <th>IS_NULL</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_1</th>\n",
       "      <th>INDEL_2</th>\n",
       "      <th>COVERAGE</th>\n",
       "      <th>HET_VARIANT_0</th>\n",
       "      <th>HET_VARIANT_1</th>\n",
       "      <th>HET_COVERAGE_0</th>\n",
       "      <th>HET_COVERAGE_1</th>\n",
       "      <th>HET_INDEL_LENGTH_0</th>\n",
       "      <th>HET_INDEL_LENGTH_1</th>\n",
       "      <th>HET_REF</th>\n",
       "      <th>HET_ALT_0</th>\n",
       "      <th>HET_ALT_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>t4687c</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>4687</td>\n",
       "      <td>M2</td>\n",
       "      <td>GENE</td>\n",
       "      <td>SNP</td>\n",
       "      <td>-37</td>\n",
       "      <td>-37</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>68</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>t4725c</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>4725</td>\n",
       "      <td>M2</td>\n",
       "      <td>GENE</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>68</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>c13333z</td>\n",
       "      <td>c</td>\n",
       "      <td>z</td>\n",
       "      <td>13333</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>119</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>68</td>\n",
       "      <td>51</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>t</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>t13335a</td>\n",
       "      <td>t</td>\n",
       "      <td>a</td>\n",
       "      <td>13335</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>68</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>4730_indel</td>\n",
       "      <td>cg</td>\n",
       "      <td>g</td>\n",
       "      <td>4730</td>\n",
       "      <td>M2</td>\n",
       "      <td>GENE</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>4730_del</td>\n",
       "      <td>4730_del_1</td>\n",
       "      <td>68</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>4735_indel</td>\n",
       "      <td>g</td>\n",
       "      <td>gcc</td>\n",
       "      <td>4735</td>\n",
       "      <td>M2</td>\n",
       "      <td>GENE</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>4735_ins</td>\n",
       "      <td>4735_ins_2</td>\n",
       "      <td>68</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      VARIANT REF  ALT  GENOME_INDEX GENE ELEMENT_TYPE MUTATION_TYPE  \\\n",
       "0      t4687c   t    c          4687   M2         GENE           SNP   \n",
       "1      t4725c   t    c          4725   M2         GENE           SNP   \n",
       "2     c13333z   c    z         13333                             SNP   \n",
       "3     t13335a   t    a         13335                             SNP   \n",
       "4  4730_indel  cg    g          4730   M2         GENE         INDEL   \n",
       "5  4735_indel   g  gcc          4735   M2         GENE         INDEL   \n",
       "\n",
       "   POSITION  NUCLEOTIDE_NUMBER  AMINO_ACID_NUMBER  ASSOCIATED_WITH_GENE  \\\n",
       "0       -37                -37                NaN                  True   \n",
       "1         1                  2                  1                  True   \n",
       "2       NaN                NaN                NaN                  True   \n",
       "3       NaN                NaN                NaN                  True   \n",
       "4         7                  7                  3                  True   \n",
       "5        12                 12                  4                  True   \n",
       "\n",
       "   IN_PROMOTER  IN_CDS  IS_SNP  IS_INDEL  IS_HET  IS_NULL  INDEL_LENGTH  \\\n",
       "0         True   False    True     False   False    False           NaN   \n",
       "1        False    True    True     False   False    False           NaN   \n",
       "2        False   False    True     False    True    False           NaN   \n",
       "3        False   False    True     False   False    False           NaN   \n",
       "4        False    True   False      True   False    False            -1   \n",
       "5        False    True   False      True   False    False             2   \n",
       "\n",
       "    INDEL_1     INDEL_2  COVERAGE HET_VARIANT_0 HET_VARIANT_1  HET_COVERAGE_0  \\\n",
       "0      None        None        68                                         NaN   \n",
       "1      None        None        68                                         NaN   \n",
       "2      None        None       119             a             g              68   \n",
       "3      None        None        68                                         NaN   \n",
       "4  4730_del  4730_del_1        68                                         NaN   \n",
       "5  4735_ins  4735_ins_2        68                                         NaN   \n",
       "\n",
       "   HET_COVERAGE_1  HET_INDEL_LENGTH_0  HET_INDEL_LENGTH_1 HET_REF HET_ALT_0  \\\n",
       "0             NaN                 NaN                 NaN                     \n",
       "1             NaN                 NaN                 NaN                     \n",
       "2              51                 NaN                 NaN       t         a   \n",
       "3             NaN                 NaN                 NaN                     \n",
       "4             NaN                 NaN                 NaN                     \n",
       "5             NaN                 NaN                 NaN                     \n",
       "\n",
       "  HET_ALT_1  \n",
       "0            \n",
       "1            \n",
       "2         g  \n",
       "3            \n",
       "4            \n",
       "5            "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "m2_ref=reference.genes['M2']\n",
    "m2_sam=sample_01.genes['M2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('int64')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m2_ref.amino_acid_numbering.dtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>AMINO_ACID_NUMBER</th>\n",
       "      <th>GENOME_INDEX</th>\n",
       "      <th>NUCLEOTIDE_NUMBER</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IS_INDEL</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>IN_PROMOTER</th>\n",
       "      <th>ELEMENT_TYPE</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_1</th>\n",
       "      <th>INDEL_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>M2</td>\n",
       "      <td>M1T</td>\n",
       "      <td>atg</td>\n",
       "      <td>acg</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>M2</td>\n",
       "      <td>t-37c</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>-37</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4687</td>\n",
       "      <td>-37</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>M2</td>\n",
       "      <td>7_indel</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>4730</td>\n",
       "      <td>7</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>GENE</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>-1</td>\n",
       "      <td>7_del</td>\n",
       "      <td>7_del_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>M2</td>\n",
       "      <td>12_indel</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>4735</td>\n",
       "      <td>12</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>GENE</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>2</td>\n",
       "      <td>12_ins</td>\n",
       "      <td>12_ins_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  GENE  MUTATION   REF   ALT  POSITION  AMINO_ACID_NUMBER  GENOME_INDEX  \\\n",
       "0   M2       M1T   atg   acg         1                  1           NaN   \n",
       "1   M2     t-37c     t     c       -37                NaN          4687   \n",
       "2   M2   7_indel  None  None         7                  3          4730   \n",
       "3   M2  12_indel  None  None        12                  4          4735   \n",
       "\n",
       "   NUCLEOTIDE_NUMBER  IS_SNP  IS_INDEL  IN_CDS  IN_PROMOTER ELEMENT_TYPE  \\\n",
       "0                NaN    True     False    True        False         GENE   \n",
       "1                -37    True     False   False         True         GENE   \n",
       "2                  7   False      True    True        False         GENE   \n",
       "3                 12   False      True    True        False         GENE   \n",
       "\n",
       "  MUTATION_TYPE  INDEL_LENGTH INDEL_1   INDEL_2  \n",
       "0           AAM           NaN    None      None  \n",
       "1           SNP           NaN    None      None  \n",
       "2         INDEL            -1   7_del   7_del_1  \n",
       "3         INDEL             2  12_ins  12_ins_2  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m2_sam.table_mutations_wrt(m2_ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['M1T', 't-37c', '7_indel', '12_indel']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m2_sam.list_mutations_wrt(m2_ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "m2_sam.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "a=pandas.read_csv(\"/Users/fowler/Dropbox/mutations.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True     1418\n",
       "False     120\n",
       "Name: IS_SNP, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a=a.astype({'IS_SNP':'bool','IS_INDEL':'bool','IN_CDS':'bool','IN_PROMOTER':'bool'})\n",
    "a.IS_SNP.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>AMINO_ACID_NUMBER</th>\n",
       "      <th>GENOME_INDEX</th>\n",
       "      <th>NUCLEOTIDE_NUMBER</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IS_INDEL</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>IN_PROMOTER</th>\n",
       "      <th>IS_SYNONYMOUS</th>\n",
       "      <th>IS_NONSYNONYMOUS</th>\n",
       "      <th>IS_HET</th>\n",
       "      <th>IS_NULL</th>\n",
       "      <th>ELEMENT_TYPE</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_1</th>\n",
       "      <th>INDEL_2</th>\n",
       "      <th>SITEID</th>\n",
       "      <th>NUMBER_NUCLEOTIDE_CHANGES</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>site.00.subj.1066561.lab.H113540015.iso.1</td>\n",
       "      <td>PE1</td>\n",
       "      <td>G286A</td>\n",
       "      <td>286</td>\n",
       "      <td>286.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ggg</td>\n",
       "      <td>gcg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>site.00.subj.1066561.lab.H113540015.iso.1</td>\n",
       "      <td>PE1</td>\n",
       "      <td>G369R</td>\n",
       "      <td>369</td>\n",
       "      <td>369.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ggc</td>\n",
       "      <td>cgc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>site.00.subj.1066561.lab.H113540015.iso.1</td>\n",
       "      <td>PE1</td>\n",
       "      <td>L485L</td>\n",
       "      <td>485</td>\n",
       "      <td>485.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>site.00.subj.1066561.lab.H113540015.iso.1</td>\n",
       "      <td>PE17</td>\n",
       "      <td>T36T</td>\n",
       "      <td>36</td>\n",
       "      <td>36.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>acg</td>\n",
       "      <td>aca</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>site.00.subj.1066561.lab.H113540015.iso.1</td>\n",
       "      <td>PE17</td>\n",
       "      <td>T285I</td>\n",
       "      <td>285</td>\n",
       "      <td>285.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aca</td>\n",
       "      <td>ata</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1533</td>\n",
       "      <td>site.00.subj.1066561.lab.H113540015.iso.1</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>-74_indel</td>\n",
       "      <td>-74</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4338595.0</td>\n",
       "      <td>-74.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GENE</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-74_del</td>\n",
       "      <td>-74_del_1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1534</td>\n",
       "      <td>site.00.subj.1066561.lab.H113540015.iso.1</td>\n",
       "      <td>yrbE1B</td>\n",
       "      <td>G248C</td>\n",
       "      <td>248</td>\n",
       "      <td>248.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ggc</td>\n",
       "      <td>tgc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1535</td>\n",
       "      <td>site.00.subj.1066561.lab.H113540015.iso.1</td>\n",
       "      <td>yrbE2A</td>\n",
       "      <td>A111A</td>\n",
       "      <td>111</td>\n",
       "      <td>111.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gcc</td>\n",
       "      <td>gcg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1536</td>\n",
       "      <td>site.00.subj.1066561.lab.H113540015.iso.1</td>\n",
       "      <td>yrbE2A</td>\n",
       "      <td>L160L</td>\n",
       "      <td>160</td>\n",
       "      <td>160.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctt</td>\n",
       "      <td>ctc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1537</td>\n",
       "      <td>site.00.subj.1066561.lab.H113540015.iso.1</td>\n",
       "      <td>zwf1</td>\n",
       "      <td>Q456!</td>\n",
       "      <td>456</td>\n",
       "      <td>456.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cag</td>\n",
       "      <td>tag</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1538 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       UNIQUEID    GENE   MUTATION  POSITION  \\\n",
       "0     site.00.subj.1066561.lab.H113540015.iso.1     PE1      G286A       286   \n",
       "1     site.00.subj.1066561.lab.H113540015.iso.1     PE1      G369R       369   \n",
       "2     site.00.subj.1066561.lab.H113540015.iso.1     PE1      L485L       485   \n",
       "3     site.00.subj.1066561.lab.H113540015.iso.1    PE17       T36T        36   \n",
       "4     site.00.subj.1066561.lab.H113540015.iso.1    PE17      T285I       285   \n",
       "...                                         ...     ...        ...       ...   \n",
       "1533  site.00.subj.1066561.lab.H113540015.iso.1   whiB6  -74_indel       -74   \n",
       "1534  site.00.subj.1066561.lab.H113540015.iso.1  yrbE1B      G248C       248   \n",
       "1535  site.00.subj.1066561.lab.H113540015.iso.1  yrbE2A      A111A       111   \n",
       "1536  site.00.subj.1066561.lab.H113540015.iso.1  yrbE2A      L160L       160   \n",
       "1537  site.00.subj.1066561.lab.H113540015.iso.1    zwf1      Q456!       456   \n",
       "\n",
       "      AMINO_ACID_NUMBER  GENOME_INDEX  NUCLEOTIDE_NUMBER  REF  ALT  IS_SNP  \\\n",
       "0                 286.0           NaN                NaN  ggg  gcg    True   \n",
       "1                 369.0           NaN                NaN  ggc  cgc    True   \n",
       "2                 485.0           NaN                NaN  ctg  ttg    True   \n",
       "3                  36.0           NaN                NaN  acg  aca    True   \n",
       "4                 285.0           NaN                NaN  aca  ata    True   \n",
       "...                 ...           ...                ...  ...  ...     ...   \n",
       "1533                NaN     4338595.0              -74.0  NaN  NaN   False   \n",
       "1534              248.0           NaN                NaN  ggc  tgc    True   \n",
       "1535              111.0           NaN                NaN  gcc  gcg    True   \n",
       "1536              160.0           NaN                NaN  ctt  ctc    True   \n",
       "1537              456.0           NaN                NaN  cag  tag    True   \n",
       "\n",
       "      IS_INDEL  IN_CDS  IN_PROMOTER  IS_SYNONYMOUS  IS_NONSYNONYMOUS  IS_HET  \\\n",
       "0        False    True        False          False              True   False   \n",
       "1        False    True        False          False              True   False   \n",
       "2        False    True        False           True             False   False   \n",
       "3        False    True        False           True             False   False   \n",
       "4        False    True        False          False              True   False   \n",
       "...        ...     ...          ...            ...               ...     ...   \n",
       "1533      True   False         True          False             False   False   \n",
       "1534     False    True        False          False              True   False   \n",
       "1535     False    True        False           True             False   False   \n",
       "1536     False    True        False           True             False   False   \n",
       "1537     False    True        False          False              True   False   \n",
       "\n",
       "      IS_NULL ELEMENT_TYPE MUTATION_TYPE  INDEL_LENGTH  INDEL_1    INDEL_2  \\\n",
       "0       False         GENE           AAM           NaN      NaN        NaN   \n",
       "1       False         GENE           AAM           NaN      NaN        NaN   \n",
       "2       False         GENE           AAM           NaN      NaN        NaN   \n",
       "3       False         GENE           AAM           NaN      NaN        NaN   \n",
       "4       False         GENE           AAM           NaN      NaN        NaN   \n",
       "...       ...          ...           ...           ...      ...        ...   \n",
       "1533    False         GENE         INDEL          -1.0  -74_del  -74_del_1   \n",
       "1534    False         GENE           AAM           NaN      NaN        NaN   \n",
       "1535    False         GENE           AAM           NaN      NaN        NaN   \n",
       "1536    False         GENE           AAM           NaN      NaN        NaN   \n",
       "1537    False         GENE           AAM           NaN      NaN        NaN   \n",
       "\n",
       "      SITEID  NUMBER_NUCLEOTIDE_CHANGES  \n",
       "0          0                          1  \n",
       "1          0                          1  \n",
       "2          0                          1  \n",
       "3          0                          1  \n",
       "4          0                          1  \n",
       "...      ...                        ...  \n",
       "1533       0                          0  \n",
       "1534       0                          1  \n",
       "1535       0                          1  \n",
       "1536       0                          1  \n",
       "1537       0                          1  \n",
       "\n",
       "[1538 rows x 24 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0    1418\n",
       "0.0     120\n",
       "Name: IS_SNP, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "b=pandas.read_csv(\"/Users/fowler/Dropbox/mutations.csv\")\n",
    "b.IS_SNP.value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "36"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g=\"0e9d82e8-c49f-4e54-b2f5-40ec8ab69b60\"\n",
    "len(g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "not empty\n"
     ]
    }
   ],
   "source": [
    "l={}\n",
    "if l is not None:\n",
    "    print(\"not empty\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
