
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../difference/">
      
      
        <link rel="next" href="../genome/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.30">
    
    
      
        <title>gene - Gumpy Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.3cba04c6.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#gene" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
      </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Gumpy Documentation" class="md-header__button md-logo" aria-label="Gumpy Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Gumpy Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              gene
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/oxfordmmm/gumpy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    gumpy
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../readme/" class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../" class="md-tabs__link">
          
  
  Code Reference

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Gumpy Documentation" class="md-nav__button md-logo" aria-label="Gumpy Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Gumpy Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/oxfordmmm/gumpy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    gumpy
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../readme/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Code Reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gumpy
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../difference/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    difference
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    gene
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    gene
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#gene" class="md-nav__link">
    <span class="md-ellipsis">
      gene
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gene.Gene" class="md-nav__link">
    <span class="md-ellipsis">
      Gene
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Gene">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gene.Gene.__adjust_dels" class="md-nav__link">
    <span class="md-ellipsis">
      __adjust_dels
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.__duplicate" class="md-nav__link">
    <span class="md-ellipsis">
      __duplicate
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.__duplicate_index" class="md-nav__link">
    <span class="md-ellipsis">
      __duplicate_index
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.__eq__" class="md-nav__link">
    <span class="md-ellipsis">
      __eq__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.__len__" class="md-nav__link">
    <span class="md-ellipsis">
      __len__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.__repr__" class="md-nav__link">
    <span class="md-ellipsis">
      __repr__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.__revcomp_indel" class="md-nav__link">
    <span class="md-ellipsis">
      __revcomp_indel
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.__sub__" class="md-nav__link">
    <span class="md-ellipsis">
      __sub__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.minority_populations_GARC" class="md-nav__link">
    <span class="md-ellipsis">
      minority_populations_GARC
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.valid_variant" class="md-nav__link">
    <span class="md-ellipsis">
      valid_variant
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genome/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    genome
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../variantfile/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    variantfile
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#gene" class="md-nav__link">
    <span class="md-ellipsis">
      gene
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gene.Gene" class="md-nav__link">
    <span class="md-ellipsis">
      Gene
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Gene">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gene.Gene.__adjust_dels" class="md-nav__link">
    <span class="md-ellipsis">
      __adjust_dels
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.__duplicate" class="md-nav__link">
    <span class="md-ellipsis">
      __duplicate
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.__duplicate_index" class="md-nav__link">
    <span class="md-ellipsis">
      __duplicate_index
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.__eq__" class="md-nav__link">
    <span class="md-ellipsis">
      __eq__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.__len__" class="md-nav__link">
    <span class="md-ellipsis">
      __len__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.__repr__" class="md-nav__link">
    <span class="md-ellipsis">
      __repr__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.__revcomp_indel" class="md-nav__link">
    <span class="md-ellipsis">
      __revcomp_indel
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.__sub__" class="md-nav__link">
    <span class="md-ellipsis">
      __sub__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.minority_populations_GARC" class="md-nav__link">
    <span class="md-ellipsis">
      minority_populations_GARC
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene.Gene.valid_variant" class="md-nav__link">
    <span class="md-ellipsis">
      valid_variant
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>gene</h1>

<div class="doc doc-object doc-module">



<a id="gene"></a>
    <div class="doc doc-contents first">

      <p>Gene object</p>



  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="gene.Gene" class="doc doc-heading">
            <code>Gene</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code>object</code></p>


      <p>Gene object that uses underlying numpy arrays</p>

              <details class="quote">
                <summary>Source code in <code>gumpy/gene.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span>
<span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span>
<span class="normal">847</span>
<span class="normal">848</span>
<span class="normal">849</span>
<span class="normal">850</span>
<span class="normal">851</span>
<span class="normal">852</span>
<span class="normal">853</span>
<span class="normal">854</span>
<span class="normal">855</span>
<span class="normal">856</span>
<span class="normal">857</span>
<span class="normal">858</span>
<span class="normal">859</span>
<span class="normal">860</span>
<span class="normal">861</span>
<span class="normal">862</span>
<span class="normal">863</span>
<span class="normal">864</span>
<span class="normal">865</span>
<span class="normal">866</span>
<span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span>
<span class="normal">873</span>
<span class="normal">874</span>
<span class="normal">875</span>
<span class="normal">876</span>
<span class="normal">877</span>
<span class="normal">878</span>
<span class="normal">879</span>
<span class="normal">880</span>
<span class="normal">881</span>
<span class="normal">882</span>
<span class="normal">883</span>
<span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span>
<span class="normal">899</span>
<span class="normal">900</span>
<span class="normal">901</span>
<span class="normal">902</span>
<span class="normal">903</span>
<span class="normal">904</span>
<span class="normal">905</span>
<span class="normal">906</span>
<span class="normal">907</span>
<span class="normal">908</span>
<span class="normal">909</span>
<span class="normal">910</span>
<span class="normal">911</span>
<span class="normal">912</span>
<span class="normal">913</span>
<span class="normal">914</span>
<span class="normal">915</span>
<span class="normal">916</span>
<span class="normal">917</span>
<span class="normal">918</span>
<span class="normal">919</span>
<span class="normal">920</span>
<span class="normal">921</span>
<span class="normal">922</span>
<span class="normal">923</span>
<span class="normal">924</span>
<span class="normal">925</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Gene</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Gene object that uses underlying numpy arrays&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">nucleotide_sequence</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">nucleotide_index</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">nucleotide_number</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">is_cds</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">is_promoter</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">is_indel</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">indel_length</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">indel_nucleotides</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">reverse_complement</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
        <span class="n">codes_protein</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
        <span class="n">feature_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">ribosomal_shifts</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span>
        <span class="n">minority_populations</span><span class="p">:</span> <span class="n">List</span><span class="p">,</span>
        <span class="n">is_deleted</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">vcf_evidence</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Constructor for the Gene object.</span>

<span class="sd">        Args:</span>
<span class="sd">            name (str, optional): Name of the gene. Defaults to None</span>
<span class="sd">            nucleotide_sequence (numpy.array, optional): Numpy array of the nucleotide</span>
<span class="sd">                sequence. Defaults to None</span>
<span class="sd">            nucleotide_index (numpy.array, optional): Numpy array of the gene indices.</span>
<span class="sd">                Defaults to None</span>
<span class="sd">            nucleotide_number (numpy.array, optional): Numpy array of the gene</span>
<span class="sd">                numbering. Defaults to None</span>
<span class="sd">            is_cds (numpy.array, optional): Numpy array to act as a mask for whether</span>
<span class="sd">                given elements are codons. Defaults to None</span>
<span class="sd">            is_promoter (numpy.array, optional): Numpy array to act as a mask for</span>
<span class="sd">                whether given elements are promoters. Defaults to None</span>
<span class="sd">            is_indel (numpy.array, optional): Numpy array to act as a mask for whether</span>
<span class="sd">                given elements are indels. Defaults to None</span>
<span class="sd">            indel_length (numpy.array, optional): Numpy array denoting the lengths of</span>
<span class="sd">                the indels whenever is_indel==True. Defaults to None</span>
<span class="sd">            indel_nucleotides (numpy.array, optional): Numpy array describing the</span>
<span class="sd">                nucleotides inserted or deleted whenever is_indel==True.</span>
<span class="sd">                    Defaults to None</span>
<span class="sd">            reverse_complement (boolean, optional): Boolean showing whether this gene</span>
<span class="sd">                is reverse complement. Defaults to False</span>
<span class="sd">            codes_protein (boolean, optional): Boolean showing whether this gene codes</span>
<span class="sd">                a protein. Defaults to True</span>
<span class="sd">            feature_type (str, optional): The name of the type of feature that this</span>
<span class="sd">                gene represents. Defaults to None</span>
<span class="sd">            ribosomal_shifts (list(int), optional): Indices of repeated bases due to</span>
<span class="sd">                ribosomal frame shifting. Defaults to []</span>
<span class="sd">            minority_populations ([int, str, str|(str,str), int, float], optional):</span>
<span class="sd">                List of minor populations. Each minor population is defined as</span>
<span class="sd">                [position, type, bases - either str or tuple of (ref, alt),</span>
<span class="sd">                depth supporting this, fractional read support]</span>
<span class="sd">            is_deleted (numpy.array, optional): Numpy array of booleans showing if a</span>
<span class="sd">                given nucleotide index is deleted. Defaults to None</span>
<span class="sd">            vcf_evidence (dict, optional): Dictionary tracking genome indices which</span>
<span class="sd">                have VCF evidence to support these calls. Defaults to None</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Using [] as a default value is dangerous, so convert from None</span>
        <span class="k">if</span> <span class="n">ribosomal_shifts</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">ribosomal_shifts</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">is_deleted</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">is_deleted</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">nucleotide_index</span><span class="p">]</span>

        <span class="k">assert</span> <span class="n">name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;must provide a gene name!&quot;</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>

        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">feature_type</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">feature_type</span> <span class="o">=</span> <span class="n">feature_type</span>
        <span class="k">assert</span> <span class="n">codes_protein</span> <span class="ow">in</span> <span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">],</span> <span class="p">(</span>
            <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;: codes_protein must be True or False!&quot;</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span> <span class="o">=</span> <span class="n">codes_protein</span>

        <span class="k">assert</span> <span class="n">reverse_complement</span> <span class="ow">in</span> <span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">],</span> <span class="p">(</span>
            <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;: reverse_complement must be True or False!&quot;</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reverse_complement</span> <span class="o">=</span> <span class="n">reverse_complement</span>

        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">nucleotide_sequence</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">),</span> <span class="p">(</span>
            <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;: sequence of bases must be a Numpy array!&quot;</span>
        <span class="p">)</span>

        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">issubdtype</span><span class="p">(</span>
            <span class="n">nucleotide_index</span><span class="o">.</span><span class="n">dtype</span><span class="o">.</span><span class="n">type</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">integer</span>
        <span class="p">),</span> <span class="p">(</span><span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;: genome indices must be a Numpy array of integers!&quot;</span><span class="p">)</span>

        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">nucleotide_number</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">),</span> <span class="p">(</span>
            <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;: gene numbering must be a Numpy array of integers!&quot;</span>
        <span class="p">)</span>

        <span class="c1"># check it is a list, and if it is that it is made up of integers and</span>
        <span class="c1">#   only integers</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ribosomal_shifts</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ribosomal_shifts</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">assert</span> <span class="nb">all</span><span class="p">([</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ribosomal_shifts</span><span class="p">])</span>

        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
            <span class="n">nucleotide_sequence</span>
        <span class="p">),</span> <span class="s2">&quot;all inputs arrays must be the same length!&quot;</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
            <span class="n">nucleotide_number</span>
        <span class="p">),</span> <span class="s2">&quot;all inputs arrays must be the same length!&quot;</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
            <span class="n">is_cds</span>
        <span class="p">),</span> <span class="s2">&quot;all inputs arrays must be the same length!&quot;</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
            <span class="n">is_promoter</span>
        <span class="p">),</span> <span class="s2">&quot;all inputs arrays must be the same length!&quot;</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
            <span class="n">is_indel</span>
        <span class="p">),</span> <span class="s2">&quot;all inputs arrays must be the same length!&quot;</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
            <span class="n">indel_length</span>
        <span class="p">),</span> <span class="s2">&quot;all inputs arrays must be the same length!&quot;</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
            <span class="n">indel_nucleotides</span>
        <span class="p">),</span> <span class="s2">&quot;all inputs arrays must be the same length!&quot;</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
            <span class="n">is_deleted</span>
        <span class="p">),</span> <span class="s2">&quot;all inputs arrays must be the same length!&quot;</span>

        <span class="n">nucleotide_sequence</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">char</span><span class="o">.</span><span class="n">lower</span><span class="p">(</span><span class="n">nucleotide_sequence</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">numpy</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span>
            <span class="n">numpy</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">nucleotide_sequence</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;t&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">,</span> <span class="s2">&quot;g&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">,</span> <span class="s2">&quot;o&quot;</span><span class="p">])</span>
        <span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_sequence</span><span class="p">),</span> <span class="p">(</span>
            <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;: sequence can only contain a,t,c,g,z,x&quot;</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">=</span> <span class="n">nucleotide_sequence</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">=</span> <span class="n">nucleotide_index</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">=</span> <span class="n">nucleotide_number</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span> <span class="o">=</span> <span class="n">is_cds</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span> <span class="o">=</span> <span class="n">is_promoter</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">=</span> <span class="n">is_indel</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="n">indel_nucleotides</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">=</span> <span class="n">indel_length</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">minority_populations</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">[]</span> <span class="k">if</span> <span class="n">minority_populations</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">minority_populations</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_deleted</span> <span class="o">=</span> <span class="n">is_deleted</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{}</span> <span class="k">if</span> <span class="n">vcf_evidence</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">vcf_evidence</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">minor_nc_changes</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="c1"># Track the original nucleotide position of each indel if this gene is revcomp</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">revcomp_indel_nc_index</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="c1"># As revcomp changes some of the positions for indels, track separately</span>
        <span class="c1"># so we can track the genome position they came from</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_index</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">__adjust_dels</span><span class="p">()</span>

        <span class="c1"># Make appropriate changes to the arrays to encorporate the frame shift</span>
        <span class="k">for</span> <span class="n">shift</span> <span class="ow">in</span> <span class="n">ribosomal_shifts</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">__duplicate</span><span class="p">(</span><span class="n">shift</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_complement</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">__revcomp_indel</span><span class="p">()</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">codon_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">floor_divide</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">]</span> <span class="o">+</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span>
                <span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span>
                    <span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">],</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">codon_number</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">codon_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">floor_divide</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">]</span> <span class="o">+</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span>
                <span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span>
                    <span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">],</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">codon_number</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">total_number_nucleotides</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_sequence</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_setup_conversion_dicts</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_translate_sequence</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_convert_minority_populations</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">_convert_minority_populations</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Convert the minority population details to gene indices + revcomp as req&quot;&quot;&quot;</span>
        <span class="n">fixed</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">minority_populations</span><span class="p">:</span>
            <span class="c1"># Update genome index to gene index</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="n">pop</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">==</span> <span class="n">pos</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">type_</span> <span class="o">=</span> <span class="n">pop</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">bases</span> <span class="o">=</span> <span class="n">pop</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">:</span>
                <span class="c1"># Is revcomp so update bases to reflect this</span>
                <span class="k">if</span> <span class="n">type_</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;ref&quot;</span><span class="p">,</span> <span class="s2">&quot;snp&quot;</span><span class="p">]:</span>
                    <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_complement</span><span class="p">(</span><span class="n">bases</span><span class="p">)</span>
                    <span class="n">bases</span> <span class="o">=</span> <span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Is an indel so a little more complex</span>
                    <span class="n">bases</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_complement</span><span class="p">(</span><span class="n">bases</span><span class="p">))[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">type_</span> <span class="o">==</span> <span class="s2">&quot;del&quot;</span><span class="p">:</span>
                        <span class="n">pos</span> <span class="o">=</span> <span class="n">pos</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">bases</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">fixed</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">pos</span><span class="p">,</span> <span class="n">type_</span><span class="p">,</span> <span class="n">bases</span><span class="p">,</span> <span class="n">pop</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span> <span class="n">pop</span><span class="p">[</span><span class="mi">4</span><span class="p">]])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">minority_populations</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">fixed</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

    <span class="k">def</span> <span class="nf">__adjust_dels</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adjust some of the deletions, such as deletions starting/ending in another</span>
<span class="sd">            gene.</span>
<span class="sd">        Also adjust the vcf evidences so they aren&#39;t present unless it is the start</span>
<span class="sd">            of the del</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Starting in another gene</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># Find first N contiguous deletions</span>
            <span class="n">n</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">i</span><span class="p">:</span>
                    <span class="k">break</span>
                <span class="n">n</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="n">n</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[:</span><span class="n">n</span><span class="p">])</span>

        <span class="c1"># Ending in another (truncating indel_length)</span>
        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">indel</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">indel</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="n">indel</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">idx</span><span class="p">:]):</span>
                <span class="c1"># Too long so truncate</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">idx</span><span class="p">:])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">idx</span><span class="p">:])</span>

        <span class="c1"># Clean up vcf_evidences so only the start of a del has the evidence (to avoid</span>
        <span class="c1">#   erroneous evidences)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">deleted</span><span class="p">,</span> <span class="n">length</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">deleted</span> <span class="ow">and</span> <span class="n">length</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># Deleted but this isn&#39;t the start</span>
                <span class="n">nucleotide_index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="c1"># Delete if exists</span>
                    <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">[</span><span class="n">nucleotide_index</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">minority_populations_GARC</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">interpretation</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;reads&quot;</span><span class="p">,</span> <span class="n">reference</span><span class="o">=</span><span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Fetch the mutations caused by minority populations in GARC</span>

<span class="sd">        Args:</span>
<span class="sd">            interpretation (str, optional): Which way to interpret the coverage calls.</span>
<span class="sd">                &#39;reads&#39; gives absolute read coverage. &#39;percentage&#39; gives fractional</span>
<span class="sd">                read support. Defaults to &#39;reads&#39;.</span>
<span class="sd">            reference (gumpy.Gene, optional): The reference to denote mutations from.</span>
<span class="sd">                Defaults to self.</span>

<span class="sd">        Returns:</span>
<span class="sd">            [str]: List of mutations in GARC</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Depending on interpretation requested, we either want [3] or [4]</span>
        <span class="k">if</span> <span class="n">interpretation</span> <span class="o">==</span> <span class="s2">&quot;percentage&quot;</span><span class="p">:</span>
            <span class="n">coverage</span> <span class="o">=</span> <span class="mi">4</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">coverage</span> <span class="o">=</span> <span class="mi">3</span>

        <span class="c1"># Make sure we have a reference to compare against</span>
        <span class="k">if</span> <span class="n">reference</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">reference</span> <span class="o">=</span> <span class="bp">self</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">reference</span><span class="o">.</span><span class="n">minority_populations</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">,</span> <span class="p">(</span>
                <span class="s2">&quot;Minority populations can only be compared when 1 Gene &quot;</span>
                <span class="s2">&quot;does not have them!&quot;</span>
            <span class="p">)</span>

        <span class="c1"># Map gene_position--&gt;[minor populations at this position]</span>
        <span class="c1"># This also groups codons together</span>
        <span class="n">gene_pos_map</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">population</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">minority_populations</span><span class="p">:</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">==</span> <span class="n">population</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">gene_pos_map</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">population</span><span class="p">)</span>

        <span class="n">mutations</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">gene_pos</span> <span class="ow">in</span> <span class="n">gene_pos_map</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">populations</span> <span class="o">=</span> <span class="n">gene_pos_map</span><span class="p">[</span><span class="n">gene_pos</span><span class="p">]</span>
            <span class="c1"># As these don&#39;t have alt codons for Z calls, track the number of</span>
            <span class="c1">#   nucleotide changes separately</span>
            <span class="n">nc_changes</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">populations</span> <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;snp&quot;</span><span class="p">])</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">populations</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="c1"># We have exactly one so return it</span>
                <span class="n">pos</span><span class="p">,</span> <span class="n">type_</span><span class="p">,</span> <span class="n">bases</span><span class="p">,</span> <span class="n">cov</span><span class="p">,</span> <span class="n">frs</span> <span class="o">=</span> <span class="n">populations</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">type_</span> <span class="o">==</span> <span class="s2">&quot;snp&quot;</span><span class="p">:</span>
                    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span> <span class="ow">and</span> <span class="n">gene_pos</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">minor_nc_changes</span><span class="p">[</span><span class="n">gene_pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">nc_changes</span>
                        <span class="c1"># Get the ref codon to build the ref/alt AAs</span>
                        <span class="n">ref_codon</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
                            <span class="n">reference</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span> <span class="o">==</span> <span class="n">gene_pos</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                        <span class="p">)</span>
                        <span class="n">codon_idx</span> <span class="o">=</span> <span class="p">(</span><span class="n">pos</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span>
                        <span class="n">alt_codon</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">ref_codon</span><span class="p">)</span>
                        <span class="n">alt_codon</span><span class="p">[</span><span class="n">codon_idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">bases</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                        <span class="n">ref</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">codon_to_amino_acid</span><span class="p">[</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ref_codon</span><span class="p">)]</span>
                        <span class="n">alt</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">codon_to_amino_acid</span><span class="p">[</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alt_codon</span><span class="p">)]</span>
                        <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                            <span class="n">ref</span>
                            <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene_pos</span><span class="p">)</span>
                            <span class="o">+</span> <span class="n">alt</span>
                            <span class="o">+</span> <span class="s2">&quot;:&quot;</span>
                            <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">populations</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">coverage</span><span class="p">])</span>
                        <span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">minor_nc_changes</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">nc_changes</span>
                        <span class="n">ref</span> <span class="o">=</span> <span class="n">reference</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">==</span> <span class="n">pos</span>
                        <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                        <span class="n">alt</span> <span class="o">=</span> <span class="n">bases</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                        <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                            <span class="n">ref</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span> <span class="o">+</span> <span class="n">alt</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">populations</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">coverage</span><span class="p">])</span>
                        <span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">minor_nc_changes</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">nc_changes</span>
                    <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="nb">str</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
                        <span class="o">+</span> <span class="s2">&quot;_&quot;</span>
                        <span class="o">+</span> <span class="n">type_</span>
                        <span class="o">+</span> <span class="s2">&quot;_&quot;</span>
                        <span class="o">+</span> <span class="n">bases</span>
                        <span class="o">+</span> <span class="s2">&quot;:&quot;</span>
                        <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">populations</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">coverage</span><span class="p">])</span>
                    <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">minor_nc_changes</span><span class="p">[</span><span class="n">gene_pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">nc_changes</span>
                <span class="c1"># We have a mixture here so report as Zs for SNPs, indel for indels,</span>
                <span class="c1">#   and mixed for mixed indels and SNPs</span>
                <span class="n">c</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span>
                    <span class="p">[</span>
                        <span class="p">(</span>
                            <span class="p">(</span><span class="n">nc_idx</span><span class="p">,</span> <span class="s2">&quot;indel&quot;</span><span class="p">)</span>
                            <span class="k">if</span> <span class="n">type_</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;ins&quot;</span><span class="p">,</span> <span class="s2">&quot;del&quot;</span><span class="p">]</span>
                            <span class="k">else</span> <span class="p">(</span><span class="n">nc_idx</span><span class="p">,</span> <span class="n">type_</span><span class="p">)</span>
                        <span class="p">)</span>
                        <span class="k">for</span> <span class="n">nc_idx</span><span class="p">,</span> <span class="n">type_</span><span class="p">,</span> <span class="n">bases</span><span class="p">,</span> <span class="n">cov</span><span class="p">,</span> <span class="n">frs</span> <span class="ow">in</span> <span class="n">gene_pos_map</span><span class="p">[</span><span class="n">gene_pos</span><span class="p">]</span>
                    <span class="p">]</span>
                <span class="p">)</span>

                <span class="c1"># Get the number of indels within this codon/base</span>
                <span class="n">indels</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">count</span> <span class="k">for</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">t</span><span class="p">),</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">c</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">t</span> <span class="o">==</span> <span class="s2">&quot;indel&quot;</span><span class="p">])</span>
                <span class="n">snps</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">count</span> <span class="k">for</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">t</span><span class="p">),</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">c</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">t</span> <span class="o">==</span> <span class="s2">&quot;snp&quot;</span><span class="p">])</span>

                <span class="k">if</span> <span class="n">indels</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">snps</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="c1"># Mixed within the codon/base so return &lt;pos&gt;_mixed</span>
                    <span class="n">cov</span> <span class="o">=</span> <span class="nb">max</span><span class="p">([</span><span class="n">pop</span><span class="p">[</span><span class="n">coverage</span><span class="p">]</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">populations</span><span class="p">])</span>
                    <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">gene_pos</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_mixed:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cov</span><span class="p">))</span>
                <span class="k">elif</span> <span class="n">indels</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="c1"># Just indels, so return of form &lt;pos&gt;_indel</span>
                    <span class="c1"># Little bit more difficult here as we have to check each</span>
                    <span class="c1">#   nucleotide_number this way, if there is only 1 indel per</span>
                    <span class="c1">#   nucleotide, we can be specific</span>
                    <span class="k">for</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">t</span><span class="p">),</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">c</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                        <span class="k">if</span> <span class="n">count</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                            <span class="c1"># Exactly 1 indel at this position so be specific</span>
                            <span class="c1"># Use the nc_idx instead of gene_pos as they may not be</span>
                            <span class="c1">#   equal</span>
                            <span class="n">minor</span> <span class="o">=</span> <span class="p">[</span>
                                <span class="p">(</span><span class="n">nc_idx</span><span class="p">,</span> <span class="n">type_</span><span class="p">,</span> <span class="n">bases</span><span class="p">,</span> <span class="n">cov</span><span class="p">,</span> <span class="n">frs</span><span class="p">)</span>
                                <span class="k">for</span> <span class="n">nc_idx</span><span class="p">,</span> <span class="n">type_</span><span class="p">,</span> <span class="n">bases</span><span class="p">,</span> <span class="n">cov</span><span class="p">,</span> <span class="n">frs</span> <span class="ow">in</span> <span class="n">gene_pos_map</span><span class="p">[</span>
                                    <span class="n">gene_pos</span>
                                <span class="p">]</span>
                                <span class="k">if</span> <span class="n">nc_idx</span> <span class="o">==</span> <span class="n">i</span>
                            <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">minor_nc_changes</span><span class="p">[</span><span class="n">minor</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="mi">0</span>
                            <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                                <span class="nb">str</span><span class="p">(</span><span class="n">minor</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                                <span class="o">+</span> <span class="s2">&quot;_&quot;</span>
                                <span class="o">+</span> <span class="n">minor</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                                <span class="o">+</span> <span class="s2">&quot;_&quot;</span>
                                <span class="o">+</span> <span class="n">minor</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
                                <span class="o">+</span> <span class="s2">&quot;:&quot;</span>
                                <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">minor</span><span class="p">[</span><span class="n">coverage</span><span class="p">])</span>
                            <span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="c1"># &gt;1 indel here, so `&lt;pos&gt;_indel` as as accurate as we</span>
                            <span class="c1">#   can be</span>
                            <span class="n">minors</span> <span class="o">=</span> <span class="p">[</span>
                                <span class="p">(</span><span class="n">nc_idx</span><span class="p">,</span> <span class="n">type_</span><span class="p">,</span> <span class="n">bases</span><span class="p">,</span> <span class="n">cov</span><span class="p">,</span> <span class="n">frs</span><span class="p">)</span>
                                <span class="k">for</span> <span class="n">nc_idx</span><span class="p">,</span> <span class="n">type_</span><span class="p">,</span> <span class="n">bases</span><span class="p">,</span> <span class="n">cov</span><span class="p">,</span> <span class="n">frs</span> <span class="ow">in</span> <span class="n">gene_pos_map</span><span class="p">[</span>
                                    <span class="n">gene_pos</span>
                                <span class="p">]</span>
                                <span class="k">if</span> <span class="n">nc_idx</span> <span class="o">==</span> <span class="n">i</span>
                            <span class="p">]</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">minor_nc_changes</span><span class="p">[</span><span class="n">minors</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="mi">0</span>
                            <span class="n">cov</span> <span class="o">=</span> <span class="nb">max</span><span class="p">([</span><span class="n">pop</span><span class="p">[</span><span class="n">coverage</span><span class="p">]</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">minors</span><span class="p">])</span>
                            <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">minors</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;_indel:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cov</span><span class="p">))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Just SNPs, so return of form &lt;ref&gt;&lt;pos&gt;(z|Z) as appropriate</span>
                    <span class="n">cov</span> <span class="o">=</span> <span class="nb">max</span><span class="p">([</span><span class="n">pop</span><span class="p">[</span><span class="n">coverage</span><span class="p">]</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">populations</span><span class="p">])</span>
                    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span> <span class="ow">and</span> <span class="n">gene_pos</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="c1"># These need a little extra effort to pull out the ref AA</span>
                        <span class="n">ref</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">codon_to_amino_acid</span><span class="p">[</span>
                            <span class="n">reference</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span> <span class="o">==</span> <span class="n">gene_pos</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                        <span class="p">]</span>
                        <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ref</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene_pos</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;Z:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cov</span><span class="p">))</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">ref</span> <span class="o">=</span> <span class="n">reference</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span> <span class="o">==</span> <span class="n">gene_pos</span>
                        <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                        <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ref</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene_pos</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;z:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cov</span><span class="p">))</span>
        <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">mutations</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__revcomp_indel</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Make some adjustments for deletions within revcomp genes.</span>
<span class="sd">        The largest of which is that the gene position of the deletion needs adjusting</span>
<span class="sd">            for revcomp.</span>
<span class="sd">        In a normal gene, the position x says delete to the left y bases.</span>
<span class="sd">        In a revcomp gene we want to say the position x says delete to the right</span>
<span class="sd">            y bases. So the pos needs changing, and indel_nucleotides need revcomping.</span>

<span class="sd">        Because its going the opposite direction, the same indel will refer to a</span>
<span class="sd">            different nucleotide position.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Reverse the indel arrays</span>
        <span class="n">is_indel</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">indel_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">fixed_indel_index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_index</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_deleted</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

        <span class="c1"># Check for positions where there actually is an indel</span>
        <span class="n">positions</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">indel_length</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># No indels to adjust, so return</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">=</span> <span class="n">is_indel</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">=</span> <span class="n">indel_length</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="n">indel_nucleotides</span>
            <span class="k">return</span>

        <span class="c1"># Fixed arrays to avoid collision cases</span>
        <span class="n">fixed_indel_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="kc">None</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">])</span>
        <span class="n">fixed_is_indel</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="kc">False</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">is_indel</span><span class="p">])</span>
        <span class="n">fixed_indel_length</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">indel_length</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">positions</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># An insertion at this pos so only revcomp the inserted bases</span>
                <span class="c1"># Because this is backwards now, pos needs adjusting to point to the</span>
                <span class="c1">#   right base for inserting after</span>
                <span class="n">old_pos</span> <span class="o">=</span> <span class="n">pos</span>
                <span class="n">pos</span> <span class="o">=</span> <span class="n">pos</span> <span class="o">-</span> <span class="mi">1</span>
                <span class="n">fixed_indel_nucleotides</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_complement</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">old_pos</span><span class="p">][::</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="p">)</span>
                <span class="n">fixed_is_indel</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="n">fixed_indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">indel_length</span><span class="p">[</span><span class="n">old_pos</span><span class="p">]</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">pos</span><span class="p">]]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">[</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">old_pos</span><span class="p">]</span>
                <span class="p">]</span>
                <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">old_pos</span><span class="p">]]</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">revcomp_indel_nc_index</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">pos</span><span class="p">])]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">old_pos</span><span class="p">]</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># A deletion at this pos so adjust position</span>
                <span class="n">new_pos</span> <span class="o">=</span> <span class="n">pos</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">pos</span><span class="p">])</span> <span class="o">+</span> <span class="mi">1</span>
                <span class="n">fixed_indel_nucleotides</span><span class="p">[</span><span class="n">new_pos</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_complement</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">pos</span><span class="p">][::</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="p">)</span>
                <span class="c1"># Only set is_indel if it was set before</span>
                <span class="c1"># In cases of large deletions, we don&#39;t want this set</span>
                <span class="k">if</span> <span class="n">is_indel</span><span class="p">[</span><span class="n">pos</span><span class="p">]:</span>
                    <span class="n">fixed_is_indel</span><span class="p">[</span><span class="n">new_pos</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="n">fixed_indel_length</span><span class="p">[</span><span class="n">new_pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                <span class="n">fixed_indel_index</span><span class="p">[</span><span class="n">new_pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">fixed_indel_index</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">revcomp_indel_nc_index</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">new_pos</span><span class="p">])]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                <span class="p">)</span>

        <span class="c1"># Update instance variables</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">=</span> <span class="n">fixed_is_indel</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">=</span> <span class="n">fixed_indel_length</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="n">fixed_indel_nucleotides</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_index</span> <span class="o">=</span> <span class="n">fixed_indel_index</span>

    <span class="k">def</span> <span class="nf">__duplicate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">int</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Duplicate all indices of important arrays to add the ribosomal shift</span>

<span class="sd">        Args:</span>
<span class="sd">            index (int): Gene index to duplicate in all arrays</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Convert to gene array index</span>
        <span class="n">index</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">==</span> <span class="n">index</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="mi">1</span>

        <span class="c1"># Update the nucelotide_numbers so they include the duplicate</span>
        <span class="c1"># Check for promoters before the codons</span>
        <span class="n">first_half</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">index</span><span class="p">)]</span>
        <span class="n">second_half</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span>
                <span class="k">else</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">))</span>
        <span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="n">first_half</span> <span class="o">+</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">index</span><span class="p">]]</span> <span class="o">+</span> <span class="n">second_half</span>
        <span class="p">)</span>
        <span class="c1"># Update all</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__duplicate_index</span><span class="p">(</span>
            <span class="n">index</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__duplicate_index</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__duplicate_index</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__duplicate_index</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__duplicate_index</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__duplicate_index</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__duplicate_index</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_index</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__duplicate_index</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__duplicate_index</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">array</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Duplicates an element at a given index and returns the new array</span>

<span class="sd">        Args:</span>
<span class="sd">            index (int): Index of the array to duplicate</span>
<span class="sd">            array (numpy.array): Array of items</span>
<span class="sd">        Returns:</span>
<span class="sd">            numpy.array: Array with duplicated item</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">first_half</span> <span class="o">=</span> <span class="p">[</span><span class="n">array</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">index</span><span class="p">)]</span>
        <span class="n">second_half</span> <span class="o">=</span> <span class="p">[</span><span class="n">array</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">array</span><span class="p">))]</span>
        <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">first_half</span> <span class="o">+</span> <span class="p">[</span><span class="n">array</span><span class="p">[</span><span class="n">index</span><span class="p">]]</span> <span class="o">+</span> <span class="n">second_half</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Overloading the equality operator to provide a method for determining if</span>
<span class="sd">            two genes are the same</span>

<span class="sd">        Args:</span>
<span class="sd">            other (gumpy.Gene) : The other gene object to compare against</span>

<span class="sd">        Returns:</span>
<span class="sd">            bool : Boolean showing equality</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Default to true</span>
        <span class="n">check</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bool_</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bool_</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="c1"># Check all fields</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">name</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">nucleotide_sequence</span>
        <span class="p">)</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">)</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">)</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">is_cds</span><span class="p">)</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">)</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">is_indel</span><span class="p">)</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">indel_length</span><span class="p">)</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reverse_complement</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">)</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">)</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">feature_type</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">feature_type</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">:</span>
            <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_sequence</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">amino_acid_sequence</span>
            <span class="p">)</span>
            <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">codons</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">codons</span><span class="p">)</span>

        <span class="k">return</span> <span class="nb">bool</span><span class="p">(</span><span class="n">check</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_complement</span><span class="p">(</span><span class="n">nucleotides_array</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Simple private method for returning the complement of an array of bases.</span>

<span class="sd">        * Note that takes account of HET and NULL calls via z and x, respectively</span>

<span class="sd">        Args:</span>
<span class="sd">            nucleotides_array (Iterable(str)): Some iterable of nucleotide bases.</span>
<span class="sd">                Usually a numpy.array or list</span>

<span class="sd">        Returns:</span>
<span class="sd">            numpy.array: Array of complemented bases</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">complementary_bases</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;a&quot;</span><span class="p">:</span> <span class="s2">&quot;t&quot;</span><span class="p">,</span>
            <span class="s2">&quot;c&quot;</span><span class="p">:</span> <span class="s2">&quot;g&quot;</span><span class="p">,</span>
            <span class="s2">&quot;g&quot;</span><span class="p">:</span> <span class="s2">&quot;c&quot;</span><span class="p">,</span>
            <span class="s2">&quot;t&quot;</span><span class="p">:</span> <span class="s2">&quot;a&quot;</span><span class="p">,</span>
            <span class="s2">&quot;x&quot;</span><span class="p">:</span> <span class="s2">&quot;x&quot;</span><span class="p">,</span>
            <span class="s2">&quot;z&quot;</span><span class="p">:</span> <span class="s2">&quot;z&quot;</span><span class="p">,</span>
            <span class="s2">&quot;o&quot;</span><span class="p">:</span> <span class="s2">&quot;o&quot;</span><span class="p">,</span>
            <span class="s2">&quot;n&quot;</span><span class="p">:</span> <span class="s2">&quot;n&quot;</span><span class="p">,</span>
            <span class="s2">&quot;r&quot;</span><span class="p">:</span> <span class="s2">&quot;y&quot;</span><span class="p">,</span>
            <span class="s2">&quot;y&quot;</span><span class="p">:</span> <span class="s2">&quot;r&quot;</span><span class="p">,</span>
            <span class="s2">&quot;s&quot;</span><span class="p">:</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span>
            <span class="s2">&quot;w&quot;</span><span class="p">:</span> <span class="s2">&quot;s&quot;</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="n">complement</span> <span class="o">=</span> <span class="p">[</span><span class="n">complementary_bases</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">nucleotides_array</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">complement</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_translate_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Translate the coding sequence into amino acids&quot;&quot;&quot;</span>

        <span class="c1"># this will ensure that only amino acids with all three bases present</span>
        <span class="n">unique</span><span class="p">,</span> <span class="n">counts</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">codon_number</span><span class="p">,</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span> <span class="o">=</span> <span class="n">unique</span><span class="p">[</span><span class="n">counts</span> <span class="o">==</span> <span class="mi">3</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

        <span class="n">cds_sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">]</span>

        <span class="n">number_codons</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">)</span> <span class="o">/</span> <span class="mi">3</span><span class="p">)</span>

        <span class="n">stacked_codons</span> <span class="o">=</span> <span class="n">cds_sequence</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="n">number_codons</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>

        <span class="n">codons</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">char</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">stacked_codons</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">stacked_codons</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">codons</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">char</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">codons</span><span class="p">,</span> <span class="n">stacked_codons</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">])</span>

        <span class="c1"># now translate the triplets into amino acids using this new dictionary</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_sequence</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">codon_to_amino_acid</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">codons</span><span class="p">]</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_setup_conversion_dicts</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Create the conversion dictionary for converting codon -&gt; amino acid&quot;&quot;&quot;</span>
        <span class="n">bases</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;t&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;g&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">,</span> <span class="s2">&quot;o&quot;</span><span class="p">]</span>
        <span class="n">aminoacids</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;FFLLXZOSSSSXZOYY!!XZOCC!WXZOXXXXXXXZZZZXZOOOOOXOOLLLLXZOPPPPXZOHHQQXZORRRRX&quot;</span>
            <span class="s2">&quot;ZOXXXXXXXZZZZXZOOOOOXOOIIIMXZOTTTTXZONNKKXZOSSRRXZOXXXXXXXZZZZXZOOOOOXOOVVV&quot;</span>
            <span class="s2">&quot;VXZOAAAAXZODDEEXZOGGGGXZOXXXXXXXZZZZXZOOOOOXOOXXXXXXXXXXXXXXXXXXXXXXXXXXXXX&quot;</span>
            <span class="s2">&quot;XXXXXXXXXXXXXXXXXXXXZZZZXZOZZZZXZOZZZZXZOZZZZXZOXXXXXXXZZZZXZOOOOOXOOOOOOXO&quot;</span>
            <span class="s2">&quot;OOOOOXOOOOOOXOOOOOOXOOXXXXXXXOOOOXOOOOOOXOO&quot;</span>
        <span class="p">)</span>
        <span class="n">all_codons</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span><span class="n">a</span> <span class="o">+</span> <span class="n">b</span> <span class="o">+</span> <span class="n">c</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">bases</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">bases</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">bases</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">codon_to_amino_acid</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">all_codons</span><span class="p">,</span> <span class="n">aminoacids</span><span class="p">))</span>

    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Overload the print function to write a summary of the Gene.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: String describing the Gene</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">string_length</span> <span class="o">=</span> <span class="mi">5</span>

        <span class="n">output</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">+</span> <span class="s2">&quot; gene</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="si">%i</span><span class="s2"> nucleotides&quot;</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">total_number_nucleotides</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">:</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;, codes for protein</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">]</span><span class="o">.</span><span class="n">size</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># updated to use numpy.printoptions to control the formatting</span>
            <span class="c1"># does return as `[&#39;a&#39; &#39;c&#39; &#39;t&#39; &#39;g&#39;]`` rather than `actg` but fixes</span>
            <span class="c1">#   formatting issues from old code such as repeating elements if</span>
            <span class="c1">#   len(promoter) &lt;= string_length</span>
            <span class="k">with</span> <span class="n">numpy</span><span class="o">.</span><span class="n">printoptions</span><span class="p">(</span><span class="n">threshold</span><span class="o">=</span><span class="n">string_length</span> <span class="o">*</span> <span class="mi">2</span><span class="p">):</span>
                <span class="n">output</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="n">output</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;promoter likely in adjacent gene(s)</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">numpy</span><span class="o">.</span><span class="n">printoptions</span><span class="p">(</span><span class="n">threshold</span><span class="o">=</span><span class="n">string_length</span> <span class="o">*</span> <span class="mi">2</span><span class="p">):</span>
                <span class="n">output</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_sequence</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="n">output</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">numpy</span><span class="o">.</span><span class="n">printoptions</span><span class="p">(</span><span class="n">threshold</span><span class="o">=</span><span class="n">string_length</span> <span class="o">*</span> <span class="mi">2</span><span class="p">):</span>
                <span class="n">output</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="n">output</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">])</span>

        <span class="k">return</span> <span class="n">output</span>

    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the number of nucleotides in the coding region (i.e. ignoring any</span>
<span class="sd">            assumed promoter)</span>

<span class="sd">        Returns:</span>
<span class="sd">            int</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">])</span>

    <span class="k">def</span> <span class="fm">__sub__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GeneDifference</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Generate a GeneDifference object for an in-depth examination of the</span>
<span class="sd">            difference between two genes.</span>

<span class="sd">        Args:</span>
<span class="sd">            other (gumpy.Gene): Other Gene object</span>

<span class="sd">        Returns:</span>
<span class="sd">            gumpy.GeneDifference: The GeneDifference object to detail changes at all</span>
<span class="sd">                levels such as indel and amino acid.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">Gene</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">GeneDifference</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">valid_variant</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variant</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Determines if a given variant is valid for this gene</span>

<span class="sd">        Args:</span>
<span class="sd">            variant (str): String of a mutation in GARC</span>

<span class="sd">        Returns:</span>
<span class="sd">            bool: True when variant is valid, False otherwise</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="n">variant</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;No Variant given!&quot;</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">variant</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;Variant must be at least 2 characters e.g. A=&quot;</span>

        <span class="c1"># Match mutation formats using regex</span>

        <span class="c1"># Check for minority populations (and remove before continuing)</span>
        <span class="k">if</span> <span class="s2">&quot;:&quot;</span> <span class="ow">in</span> <span class="n">variant</span><span class="p">:</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">variant_</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)</span>

            <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">variant_</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span>

            <span class="c1"># Use duck typing to ensure it gives a valid number</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="nb">float</span><span class="p">(</span><span class="n">variant_</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                <span class="k">return</span> <span class="kc">False</span>

            <span class="c1"># 0 is not valid either</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="nb">float</span><span class="p">(</span><span class="n">variant_</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">0</span>

            <span class="k">if</span> <span class="n">valid</span><span class="p">:</span>
                <span class="c1"># Valid minority population, so continue without it</span>
                <span class="n">variant</span> <span class="o">=</span> <span class="n">variant_</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="kc">False</span>

        <span class="c1"># Match promoter/non-coding SNP format</span>
        <span class="n">promoter</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
<span class="w">            </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                ([a-zA-Z_0-9.()]+@)? #Possibly a leading gene name</span>
<span class="sd">                ([acgtzx]) #Reference base</span>
<span class="sd">                (-?[0-9]+) #Position</span>
<span class="sd">                ([acgtzx]) #Alt base</span>
<span class="sd">                &quot;&quot;&quot;</span><span class="p">,</span>
            <span class="n">re</span><span class="o">.</span><span class="n">VERBOSE</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">promoter_match</span> <span class="o">=</span> <span class="n">promoter</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">promoter_match</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># The variant is either promoter or non-coding</span>
            <span class="c1"># Check that the mutation is valid</span>
            <span class="n">name</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">alt</span> <span class="o">=</span> <span class="n">promoter_match</span><span class="o">.</span><span class="n">groups</span><span class="p">()</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="c1"># Strip &#39;@&#39; from gene name and compare</span>
            <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">name</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">name</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span>
            <span class="c1"># Check that the pos is in the correct range</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span>
            <span class="c1"># Check that the ref matches the seq at the given pos</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">valid</span>
                <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">==</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)]</span> <span class="o">==</span> <span class="n">ref</span>
            <span class="p">)</span>
            <span class="c1"># Mutation should not have same ref and alt</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">ref</span> <span class="o">!=</span> <span class="n">alt</span>
            <span class="k">return</span> <span class="n">valid</span>
        <span class="c1"># Match amino acid SNP</span>
        <span class="n">snp</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
<span class="w">            </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                    ([a-zA-Z_0-9.()]+@)? #Possibly leading gene name</span>
<span class="sd">                    ([A-Z!]) #Reference amino acid</span>
<span class="sd">                    ([0-9]+) #Position</span>
<span class="sd">                    ([A-Z!]) #Alt amino acid</span>
<span class="sd">                    &quot;&quot;&quot;</span><span class="p">,</span>
            <span class="n">re</span><span class="o">.</span><span class="n">VERBOSE</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">snp_match</span> <span class="o">=</span> <span class="n">snp</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span> <span class="ow">and</span> <span class="n">snp_match</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># The variant is an amino acid SNP</span>
            <span class="c1"># Check it is valid</span>
            <span class="n">name</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">alt</span> <span class="o">=</span> <span class="n">snp_match</span><span class="o">.</span><span class="n">groups</span><span class="p">()</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="c1"># Strip &#39;@&#39; from gene name and compare</span>
            <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">name</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">name</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span>
            <span class="c1"># Check pos is in correct range</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span>
            <span class="c1"># Check ref matches the aa seq at the given pos</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">valid</span>
                <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_sequence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span> <span class="o">==</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)]</span> <span class="o">==</span> <span class="n">ref</span>
            <span class="p">)</span>
            <span class="k">return</span> <span class="n">valid</span>

        <span class="c1"># Match amino acid synon-mutation</span>
        <span class="n">synon</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
<span class="w">            </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                        ([a-zA-Z_0-9.()]+@)? #Possibly leading gene name</span>
<span class="sd">                        ([0-9]+) #Pos</span>
<span class="sd">                        = #Synonymous amino acid</span>
<span class="sd">                        &quot;&quot;&quot;</span><span class="p">,</span>
            <span class="n">re</span><span class="o">.</span><span class="n">VERBOSE</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">synon_match</span> <span class="o">=</span> <span class="n">synon</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span> <span class="ow">and</span> <span class="n">synon_match</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># Variant is a synonymous mutation</span>
            <span class="c1"># Check it is valid</span>
            <span class="n">name</span><span class="p">,</span> <span class="n">pos</span> <span class="o">=</span> <span class="n">synon_match</span><span class="o">.</span><span class="n">groups</span><span class="p">()</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">name</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">name</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span>
            <span class="k">return</span> <span class="n">valid</span>

        <span class="c1"># Match indel</span>
        <span class="n">indel</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
<span class="w">            </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                    ([a-zA-Z_0-9.()]+@)? #Possibly leading gene name</span>
<span class="sd">                    (-?[0-9]+) #Position</span>
<span class="sd">                    _(ins|del|indel|mixed)_? #Type</span>
<span class="sd">                    ([0-9]+|[acgtzx]+)? #Bases deleted/inserted</span>
<span class="sd">                    &quot;&quot;&quot;</span><span class="p">,</span>
            <span class="n">re</span><span class="o">.</span><span class="n">VERBOSE</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">indel_match</span> <span class="o">=</span> <span class="n">indel</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">indel_match</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># Variant is an indel</span>
            <span class="c1"># Check it is valid</span>
            <span class="n">name</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">type_</span><span class="p">,</span> <span class="n">bases</span> <span class="o">=</span> <span class="n">indel_match</span><span class="o">.</span><span class="n">groups</span><span class="p">()</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
            <span class="n">end</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">)</span>
            <span class="c1"># Check the names match</span>
            <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">name</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">name</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span>
            <span class="c1"># Check pos in correct range</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span>
            <span class="k">if</span> <span class="n">type_</span> <span class="o">==</span> <span class="s2">&quot;indel&quot;</span> <span class="ow">or</span> <span class="n">type_</span> <span class="o">==</span> <span class="s2">&quot;mixed&quot;</span><span class="p">:</span>
                <span class="c1"># If a mutation is given as `indel` or `mixed`, no length/bases should</span>
                <span class="c1">#   be given</span>
                <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="p">(</span><span class="n">bases</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">bases</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">type_</span> <span class="o">==</span> <span class="s2">&quot;del&quot;</span> <span class="ow">and</span> <span class="n">bases</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">bases</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="c1"># Mutation was del, so check if the bases given match the ref</span>
                <span class="k">if</span> <span class="n">bases</span><span class="o">.</span><span class="n">isnumeric</span><span class="p">():</span>
                    <span class="c1"># A length was given rather than bases so just check that all bases</span>
                    <span class="c1">#   are in the correct range</span>
                    <span class="k">if</span> <span class="n">pos</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="c1"># Is a promoter so be careful about pos+bases not being 0</span>
                        <span class="n">pos</span> <span class="o">-=</span> <span class="mi">1</span>
                    <span class="n">offset</span> <span class="o">=</span> <span class="mi">0</span>
                    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">bases</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
                        <span class="k">if</span> <span class="n">pos</span> <span class="o">+</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                            <span class="n">offset</span> <span class="o">=</span> <span class="mi">1</span>
                        <span class="c1"># Ignore checking if the del passes the 5&#39;end of the gene</span>
                        <span class="k">if</span> <span class="n">pos</span> <span class="o">+</span> <span class="n">i</span> <span class="o">+</span> <span class="n">offset</span> <span class="o">&gt;</span> <span class="n">end</span><span class="p">:</span>
                            <span class="k">continue</span>
                        <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">pos</span> <span class="o">+</span> <span class="n">i</span> <span class="o">+</span> <span class="n">offset</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Bases were given rather than a length, so check for equality</span>
                    <span class="c1">#   against base seq</span>
                    <span class="n">offset</span> <span class="o">=</span> <span class="mi">0</span>
                    <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">base</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">bases</span><span class="p">):</span>
                        <span class="k">if</span> <span class="n">pos</span> <span class="o">+</span> <span class="n">index</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                            <span class="n">offset</span> <span class="o">=</span> <span class="mi">1</span>
                        <span class="k">if</span> <span class="n">pos</span> <span class="o">+</span> <span class="n">index</span> <span class="o">+</span> <span class="n">offset</span> <span class="o">&gt;</span> <span class="n">end</span><span class="p">:</span>
                            <span class="c1"># Ignore checking if the del passes the 5&#39; end of the gene</span>
                            <span class="k">continue</span>
                        <span class="n">valid</span> <span class="o">=</span> <span class="p">(</span>
                            <span class="n">valid</span>
                            <span class="ow">and</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span> <span class="o">+</span> <span class="n">index</span> <span class="o">+</span> <span class="n">offset</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span>
                        <span class="p">)</span>
                        <span class="n">valid</span> <span class="o">=</span> <span class="p">(</span>
                            <span class="n">valid</span>
                            <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">==</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span> <span class="o">+</span> <span class="n">index</span> <span class="o">+</span> <span class="n">offset</span>
                            <span class="p">]</span>
                            <span class="o">==</span> <span class="n">base</span>
                        <span class="p">)</span>
            <span class="k">if</span> <span class="n">type_</span> <span class="o">==</span> <span class="s2">&quot;ins&quot;</span><span class="p">:</span>
                <span class="c1"># Just check that the position specified lies within the gene</span>
                <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">pos</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span>
            <span class="k">return</span> <span class="n">valid</span>
        <span class="c1"># Checking for percentage deletion</span>
        <span class="n">deletion</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
<span class="w">            </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                            ([a-zA-Z_0-9.()]+@)? #Possibly leading gene name</span>
<span class="sd">                            del_ #Deletion</span>
<span class="sd">                            ([01]\.[0-9]+)</span>
<span class="sd">                            &quot;&quot;&quot;</span><span class="p">,</span>
            <span class="n">re</span><span class="o">.</span><span class="n">VERBOSE</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">deletion_match</span> <span class="o">=</span> <span class="n">deletion</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">deletion_match</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">name</span><span class="p">,</span> <span class="n">percent</span> <span class="o">=</span> <span class="n">deletion_match</span><span class="o">.</span><span class="n">groups</span><span class="p">()</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">name</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span>
            <span class="n">percent</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">percent</span><span class="p">)</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">percent</span> <span class="o">&lt;=</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">percent</span> <span class="o">&gt;=</span> <span class="mi">0</span>
            <span class="k">return</span> <span class="n">valid</span>
        <span class="c1"># If none of the mutations have matched, return False</span>
        <span class="k">return</span> <span class="kc">False</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="gene.Gene.__adjust_dels" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__adjust_dels</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Adjust some of the deletions, such as deletions starting/ending in another
    gene.
Also adjust the vcf evidences so they aren't present unless it is the start
    of the del</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/gene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__adjust_dels</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adjust some of the deletions, such as deletions starting/ending in another</span>
<span class="sd">        gene.</span>
<span class="sd">    Also adjust the vcf evidences so they aren&#39;t present unless it is the start</span>
<span class="sd">        of the del</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Starting in another gene</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Find first N contiguous deletions</span>
        <span class="n">n</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">i</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="n">n</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="n">n</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[:</span><span class="n">n</span><span class="p">])</span>

    <span class="c1"># Ending in another (truncating indel_length)</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">indel</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">indel</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="n">indel</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">idx</span><span class="p">:]):</span>
            <span class="c1"># Too long so truncate</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">idx</span><span class="p">:])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">idx</span><span class="p">:])</span>

    <span class="c1"># Clean up vcf_evidences so only the start of a del has the evidence (to avoid</span>
    <span class="c1">#   erroneous evidences)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">deleted</span><span class="p">,</span> <span class="n">length</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">)):</span>
        <span class="k">if</span> <span class="n">deleted</span> <span class="ow">and</span> <span class="n">length</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># Deleted but this isn&#39;t the start</span>
            <span class="n">nucleotide_index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="c1"># Delete if exists</span>
                <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">[</span><span class="n">nucleotide_index</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gene.Gene.__duplicate" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__duplicate</span><span class="p">(</span><span class="n">index</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Duplicate all indices of important arrays to add the ribosomal shift</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>index</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Gene index to duplicate in all arrays</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/gene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__duplicate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">int</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Duplicate all indices of important arrays to add the ribosomal shift</span>

<span class="sd">    Args:</span>
<span class="sd">        index (int): Gene index to duplicate in all arrays</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Convert to gene array index</span>
    <span class="n">index</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">==</span> <span class="n">index</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="mi">1</span>

    <span class="c1"># Update the nucelotide_numbers so they include the duplicate</span>
    <span class="c1"># Check for promoters before the codons</span>
    <span class="n">first_half</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">index</span><span class="p">)]</span>
    <span class="n">second_half</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span>
            <span class="k">else</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">))</span>
    <span class="p">]</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
        <span class="n">first_half</span> <span class="o">+</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">index</span><span class="p">]]</span> <span class="o">+</span> <span class="n">second_half</span>
    <span class="p">)</span>
    <span class="c1"># Update all</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__duplicate_index</span><span class="p">(</span>
        <span class="n">index</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span>
    <span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__duplicate_index</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__duplicate_index</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__duplicate_index</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__duplicate_index</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__duplicate_index</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">indel_index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__duplicate_index</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_index</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__duplicate_index</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gene.Gene.__duplicate_index" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__duplicate_index</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="n">array</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Duplicates an element at a given index and returns the new array</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>index</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Index of the array to duplicate</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>array</code></td>
            <td>
                  <code><span title="numpy.array">array</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array of items</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
      <p>Returns:
    numpy.array: Array with duplicated item</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/gene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__duplicate_index</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">array</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Duplicates an element at a given index and returns the new array</span>

<span class="sd">    Args:</span>
<span class="sd">        index (int): Index of the array to duplicate</span>
<span class="sd">        array (numpy.array): Array of items</span>
<span class="sd">    Returns:</span>
<span class="sd">        numpy.array: Array with duplicated item</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">first_half</span> <span class="o">=</span> <span class="p">[</span><span class="n">array</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">index</span><span class="p">)]</span>
    <span class="n">second_half</span> <span class="o">=</span> <span class="p">[</span><span class="n">array</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">array</span><span class="p">))]</span>
    <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">first_half</span> <span class="o">+</span> <span class="p">[</span><span class="n">array</span><span class="p">[</span><span class="n">index</span><span class="p">]]</span> <span class="o">+</span> <span class="n">second_half</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gene.Gene.__eq__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__eq__</span><span class="p">(</span><span class="n">other</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Overloading the equality operator to provide a method for determining if
    two genes are the same</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>other</code></td>
            <td>
                  <code>gumpy.Gene) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The other gene object to compare against</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>bool</code></td>            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Boolean showing equality</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/gene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Overloading the equality operator to provide a method for determining if</span>
<span class="sd">        two genes are the same</span>

<span class="sd">    Args:</span>
<span class="sd">        other (gumpy.Gene) : The other gene object to compare against</span>

<span class="sd">    Returns:</span>
<span class="sd">        bool : Boolean showing equality</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Default to true</span>
    <span class="n">check</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bool_</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bool_</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># Check all fields</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">name</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">nucleotide_sequence</span>
    <span class="p">)</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">)</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">)</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">is_cds</span><span class="p">)</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">)</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">is_indel</span><span class="p">)</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">indel_length</span><span class="p">)</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reverse_complement</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">)</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">)</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">feature_type</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">feature_type</span><span class="p">)</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">:</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_sequence</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">amino_acid_sequence</span>
        <span class="p">)</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">codons</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">codons</span><span class="p">)</span>

    <span class="k">return</span> <span class="nb">bool</span><span class="p">(</span><span class="n">check</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gene.Gene.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">nucleotide_sequence</span><span class="p">,</span> <span class="n">nucleotide_index</span><span class="p">,</span> <span class="n">nucleotide_number</span><span class="p">,</span> <span class="n">is_cds</span><span class="p">,</span> <span class="n">is_promoter</span><span class="p">,</span> <span class="n">is_indel</span><span class="p">,</span> <span class="n">indel_length</span><span class="p">,</span> <span class="n">indel_nucleotides</span><span class="p">,</span> <span class="n">reverse_complement</span><span class="p">,</span> <span class="n">codes_protein</span><span class="p">,</span> <span class="n">feature_type</span><span class="p">,</span> <span class="n">ribosomal_shifts</span><span class="p">,</span> <span class="n">minority_populations</span><span class="p">,</span> <span class="n">is_deleted</span><span class="p">,</span> <span class="n">vcf_evidence</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Constructor for the Gene object.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>name</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the gene. Defaults to None</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>nucleotide_sequence</code></td>
            <td>
                  <code><span title="numpy.array">array</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Numpy array of the nucleotide
sequence. Defaults to None</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>nucleotide_index</code></td>
            <td>
                  <code><span title="numpy.array">array</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Numpy array of the gene indices.
Defaults to None</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>nucleotide_number</code></td>
            <td>
                  <code><span title="numpy.array">array</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Numpy array of the gene
numbering. Defaults to None</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>is_cds</code></td>
            <td>
                  <code><span title="numpy.array">array</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Numpy array to act as a mask for whether
given elements are codons. Defaults to None</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>is_promoter</code></td>
            <td>
                  <code><span title="numpy.array">array</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Numpy array to act as a mask for
whether given elements are promoters. Defaults to None</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>is_indel</code></td>
            <td>
                  <code><span title="numpy.array">array</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Numpy array to act as a mask for whether
given elements are indels. Defaults to None</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>indel_length</code></td>
            <td>
                  <code><span title="numpy.array">array</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Numpy array denoting the lengths of
the indels whenever is_indel==True. Defaults to None</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>indel_nucleotides</code></td>
            <td>
                  <code><span title="numpy.array">array</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Numpy array describing the
nucleotides inserted or deleted whenever is_indel==True.
    Defaults to None</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>reverse_complement</code></td>
            <td>
                  <code>boolean</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Boolean showing whether this gene
is reverse complement. Defaults to False</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codes_protein</code></td>
            <td>
                  <code>boolean</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Boolean showing whether this gene codes
a protein. Defaults to True</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>feature_type</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the type of feature that this
gene represents. Defaults to None</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>ribosomal_shifts</code></td>
            <td>
                  <code>list(int)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Indices of repeated bases due to
ribosomal frame shifting. Defaults to []</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>minority_populations</code></td>
            <td>
                  <code>[int, str, str | (str, str), int, float]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of minor populations. Each minor population is defined as
[position, type, bases - either str or tuple of (ref, alt),
depth supporting this, fractional read support]</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>is_deleted</code></td>
            <td>
                  <code><span title="numpy.array">array</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Numpy array of booleans showing if a
given nucleotide index is deleted. Defaults to None</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>vcf_evidence</code></td>
            <td>
                  <code>dict</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary tracking genome indices which
have VCF evidence to support these calls. Defaults to None</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/gene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">nucleotide_sequence</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">nucleotide_index</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">nucleotide_number</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">is_cds</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">is_promoter</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">is_indel</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">indel_length</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">indel_nucleotides</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">reverse_complement</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">codes_protein</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">feature_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">ribosomal_shifts</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span>
    <span class="n">minority_populations</span><span class="p">:</span> <span class="n">List</span><span class="p">,</span>
    <span class="n">is_deleted</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">vcf_evidence</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Constructor for the Gene object.</span>

<span class="sd">    Args:</span>
<span class="sd">        name (str, optional): Name of the gene. Defaults to None</span>
<span class="sd">        nucleotide_sequence (numpy.array, optional): Numpy array of the nucleotide</span>
<span class="sd">            sequence. Defaults to None</span>
<span class="sd">        nucleotide_index (numpy.array, optional): Numpy array of the gene indices.</span>
<span class="sd">            Defaults to None</span>
<span class="sd">        nucleotide_number (numpy.array, optional): Numpy array of the gene</span>
<span class="sd">            numbering. Defaults to None</span>
<span class="sd">        is_cds (numpy.array, optional): Numpy array to act as a mask for whether</span>
<span class="sd">            given elements are codons. Defaults to None</span>
<span class="sd">        is_promoter (numpy.array, optional): Numpy array to act as a mask for</span>
<span class="sd">            whether given elements are promoters. Defaults to None</span>
<span class="sd">        is_indel (numpy.array, optional): Numpy array to act as a mask for whether</span>
<span class="sd">            given elements are indels. Defaults to None</span>
<span class="sd">        indel_length (numpy.array, optional): Numpy array denoting the lengths of</span>
<span class="sd">            the indels whenever is_indel==True. Defaults to None</span>
<span class="sd">        indel_nucleotides (numpy.array, optional): Numpy array describing the</span>
<span class="sd">            nucleotides inserted or deleted whenever is_indel==True.</span>
<span class="sd">                Defaults to None</span>
<span class="sd">        reverse_complement (boolean, optional): Boolean showing whether this gene</span>
<span class="sd">            is reverse complement. Defaults to False</span>
<span class="sd">        codes_protein (boolean, optional): Boolean showing whether this gene codes</span>
<span class="sd">            a protein. Defaults to True</span>
<span class="sd">        feature_type (str, optional): The name of the type of feature that this</span>
<span class="sd">            gene represents. Defaults to None</span>
<span class="sd">        ribosomal_shifts (list(int), optional): Indices of repeated bases due to</span>
<span class="sd">            ribosomal frame shifting. Defaults to []</span>
<span class="sd">        minority_populations ([int, str, str|(str,str), int, float], optional):</span>
<span class="sd">            List of minor populations. Each minor population is defined as</span>
<span class="sd">            [position, type, bases - either str or tuple of (ref, alt),</span>
<span class="sd">            depth supporting this, fractional read support]</span>
<span class="sd">        is_deleted (numpy.array, optional): Numpy array of booleans showing if a</span>
<span class="sd">            given nucleotide index is deleted. Defaults to None</span>
<span class="sd">        vcf_evidence (dict, optional): Dictionary tracking genome indices which</span>
<span class="sd">            have VCF evidence to support these calls. Defaults to None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Using [] as a default value is dangerous, so convert from None</span>
    <span class="k">if</span> <span class="n">ribosomal_shifts</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">ribosomal_shifts</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">is_deleted</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">is_deleted</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">nucleotide_index</span><span class="p">]</span>

    <span class="k">assert</span> <span class="n">name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;must provide a gene name!&quot;</span>
    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>

    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">feature_type</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">feature_type</span> <span class="o">=</span> <span class="n">feature_type</span>
    <span class="k">assert</span> <span class="n">codes_protein</span> <span class="ow">in</span> <span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">],</span> <span class="p">(</span>
        <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;: codes_protein must be True or False!&quot;</span>
    <span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span> <span class="o">=</span> <span class="n">codes_protein</span>

    <span class="k">assert</span> <span class="n">reverse_complement</span> <span class="ow">in</span> <span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">],</span> <span class="p">(</span>
        <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;: reverse_complement must be True or False!&quot;</span>
    <span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">reverse_complement</span> <span class="o">=</span> <span class="n">reverse_complement</span>

    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">nucleotide_sequence</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">),</span> <span class="p">(</span>
        <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;: sequence of bases must be a Numpy array!&quot;</span>
    <span class="p">)</span>

    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">issubdtype</span><span class="p">(</span>
        <span class="n">nucleotide_index</span><span class="o">.</span><span class="n">dtype</span><span class="o">.</span><span class="n">type</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">integer</span>
    <span class="p">),</span> <span class="p">(</span><span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;: genome indices must be a Numpy array of integers!&quot;</span><span class="p">)</span>

    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">nucleotide_number</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">),</span> <span class="p">(</span>
        <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;: gene numbering must be a Numpy array of integers!&quot;</span>
    <span class="p">)</span>

    <span class="c1"># check it is a list, and if it is that it is made up of integers and</span>
    <span class="c1">#   only integers</span>
    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ribosomal_shifts</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ribosomal_shifts</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">assert</span> <span class="nb">all</span><span class="p">([</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ribosomal_shifts</span><span class="p">])</span>

    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
        <span class="n">nucleotide_sequence</span>
    <span class="p">),</span> <span class="s2">&quot;all inputs arrays must be the same length!&quot;</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
        <span class="n">nucleotide_number</span>
    <span class="p">),</span> <span class="s2">&quot;all inputs arrays must be the same length!&quot;</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
        <span class="n">is_cds</span>
    <span class="p">),</span> <span class="s2">&quot;all inputs arrays must be the same length!&quot;</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
        <span class="n">is_promoter</span>
    <span class="p">),</span> <span class="s2">&quot;all inputs arrays must be the same length!&quot;</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
        <span class="n">is_indel</span>
    <span class="p">),</span> <span class="s2">&quot;all inputs arrays must be the same length!&quot;</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
        <span class="n">indel_length</span>
    <span class="p">),</span> <span class="s2">&quot;all inputs arrays must be the same length!&quot;</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
        <span class="n">indel_nucleotides</span>
    <span class="p">),</span> <span class="s2">&quot;all inputs arrays must be the same length!&quot;</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
        <span class="n">is_deleted</span>
    <span class="p">),</span> <span class="s2">&quot;all inputs arrays must be the same length!&quot;</span>

    <span class="n">nucleotide_sequence</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">char</span><span class="o">.</span><span class="n">lower</span><span class="p">(</span><span class="n">nucleotide_sequence</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">numpy</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span>
        <span class="n">numpy</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">nucleotide_sequence</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;t&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">,</span> <span class="s2">&quot;g&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">,</span> <span class="s2">&quot;o&quot;</span><span class="p">])</span>
    <span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_sequence</span><span class="p">),</span> <span class="p">(</span>
        <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;: sequence can only contain a,t,c,g,z,x&quot;</span>
    <span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">=</span> <span class="n">nucleotide_sequence</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">=</span> <span class="n">nucleotide_index</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">=</span> <span class="n">nucleotide_number</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span> <span class="o">=</span> <span class="n">is_cds</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span> <span class="o">=</span> <span class="n">is_promoter</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">=</span> <span class="n">is_indel</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="n">indel_nucleotides</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">=</span> <span class="n">indel_length</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">minority_populations</span> <span class="o">=</span> <span class="p">(</span>
        <span class="p">[]</span> <span class="k">if</span> <span class="n">minority_populations</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">minority_populations</span>
    <span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_deleted</span> <span class="o">=</span> <span class="n">is_deleted</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{}</span> <span class="k">if</span> <span class="n">vcf_evidence</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">vcf_evidence</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">minor_nc_changes</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># Track the original nucleotide position of each indel if this gene is revcomp</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">revcomp_indel_nc_index</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># As revcomp changes some of the positions for indels, track separately</span>
    <span class="c1"># so we can track the genome position they came from</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">indel_index</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">__adjust_dels</span><span class="p">()</span>

    <span class="c1"># Make appropriate changes to the arrays to encorporate the frame shift</span>
    <span class="k">for</span> <span class="n">shift</span> <span class="ow">in</span> <span class="n">ribosomal_shifts</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">__duplicate</span><span class="p">(</span><span class="n">shift</span><span class="p">)</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_complement</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">__revcomp_indel</span><span class="p">()</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">codon_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">floor_divide</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">]</span> <span class="o">+</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span>
                <span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">],</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">codon_number</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">codon_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">floor_divide</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">]</span> <span class="o">+</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span>
                <span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">],</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">codon_number</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">total_number_nucleotides</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_sequence</span><span class="p">)</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_setup_conversion_dicts</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_translate_sequence</span><span class="p">()</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_convert_minority_populations</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gene.Gene.__len__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__len__</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Return the number of nucleotides in the coding region (i.e. ignoring any
    assumed promoter)</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>int</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/gene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return the number of nucleotides in the coding region (i.e. ignoring any</span>
<span class="sd">        assumed promoter)</span>

<span class="sd">    Returns:</span>
<span class="sd">        int</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">])</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gene.Gene.__repr__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__repr__</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Overload the print function to write a summary of the Gene.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>String describing the Gene</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/gene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Overload the print function to write a summary of the Gene.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: String describing the Gene</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">string_length</span> <span class="o">=</span> <span class="mi">5</span>

    <span class="n">output</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">+</span> <span class="s2">&quot; gene</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="si">%i</span><span class="s2"> nucleotides&quot;</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">total_number_nucleotides</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">:</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;, codes for protein</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">]</span><span class="o">.</span><span class="n">size</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># updated to use numpy.printoptions to control the formatting</span>
        <span class="c1"># does return as `[&#39;a&#39; &#39;c&#39; &#39;t&#39; &#39;g&#39;]`` rather than `actg` but fixes</span>
        <span class="c1">#   formatting issues from old code such as repeating elements if</span>
        <span class="c1">#   len(promoter) &lt;= string_length</span>
        <span class="k">with</span> <span class="n">numpy</span><span class="o">.</span><span class="n">printoptions</span><span class="p">(</span><span class="n">threshold</span><span class="o">=</span><span class="n">string_length</span> <span class="o">*</span> <span class="mi">2</span><span class="p">):</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;promoter likely in adjacent gene(s)</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">numpy</span><span class="o">.</span><span class="n">printoptions</span><span class="p">(</span><span class="n">threshold</span><span class="o">=</span><span class="n">string_length</span> <span class="o">*</span> <span class="mi">2</span><span class="p">):</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_sequence</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">numpy</span><span class="o">.</span><span class="n">printoptions</span><span class="p">(</span><span class="n">threshold</span><span class="o">=</span><span class="n">string_length</span> <span class="o">*</span> <span class="mi">2</span><span class="p">):</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">output</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gene.Gene.__revcomp_indel" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__revcomp_indel</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Make some adjustments for deletions within revcomp genes.
The largest of which is that the gene position of the deletion needs adjusting
    for revcomp.
In a normal gene, the position x says delete to the left y bases.
In a revcomp gene we want to say the position x says delete to the right
    y bases. So the pos needs changing, and indel_nucleotides need revcomping.</p>
<p>Because its going the opposite direction, the same indel will refer to a
    different nucleotide position.</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/gene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__revcomp_indel</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Make some adjustments for deletions within revcomp genes.</span>
<span class="sd">    The largest of which is that the gene position of the deletion needs adjusting</span>
<span class="sd">        for revcomp.</span>
<span class="sd">    In a normal gene, the position x says delete to the left y bases.</span>
<span class="sd">    In a revcomp gene we want to say the position x says delete to the right</span>
<span class="sd">        y bases. So the pos needs changing, and indel_nucleotides need revcomping.</span>

<span class="sd">    Because its going the opposite direction, the same indel will refer to a</span>
<span class="sd">        different nucleotide position.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Reverse the indel arrays</span>
    <span class="n">is_indel</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">indel_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">fixed_indel_index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_index</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_deleted</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="c1"># Check for positions where there actually is an indel</span>
    <span class="n">positions</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">indel_length</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># No indels to adjust, so return</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">=</span> <span class="n">is_indel</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">=</span> <span class="n">indel_length</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="n">indel_nucleotides</span>
        <span class="k">return</span>

    <span class="c1"># Fixed arrays to avoid collision cases</span>
    <span class="n">fixed_indel_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="kc">None</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">])</span>
    <span class="n">fixed_is_indel</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="kc">False</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">is_indel</span><span class="p">])</span>
    <span class="n">fixed_indel_length</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">indel_length</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">positions</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
        <span class="k">if</span> <span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># An insertion at this pos so only revcomp the inserted bases</span>
            <span class="c1"># Because this is backwards now, pos needs adjusting to point to the</span>
            <span class="c1">#   right base for inserting after</span>
            <span class="n">old_pos</span> <span class="o">=</span> <span class="n">pos</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="n">pos</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="n">fixed_indel_nucleotides</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_complement</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">old_pos</span><span class="p">][::</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
            <span class="p">)</span>
            <span class="n">fixed_is_indel</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">fixed_indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">indel_length</span><span class="p">[</span><span class="n">old_pos</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">pos</span><span class="p">]]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">[</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">old_pos</span><span class="p">]</span>
            <span class="p">]</span>
            <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">old_pos</span><span class="p">]]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">revcomp_indel_nc_index</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">pos</span><span class="p">])]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">old_pos</span><span class="p">]</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># A deletion at this pos so adjust position</span>
            <span class="n">new_pos</span> <span class="o">=</span> <span class="n">pos</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">pos</span><span class="p">])</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">fixed_indel_nucleotides</span><span class="p">[</span><span class="n">new_pos</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_complement</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">pos</span><span class="p">][::</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
            <span class="p">)</span>
            <span class="c1"># Only set is_indel if it was set before</span>
            <span class="c1"># In cases of large deletions, we don&#39;t want this set</span>
            <span class="k">if</span> <span class="n">is_indel</span><span class="p">[</span><span class="n">pos</span><span class="p">]:</span>
                <span class="n">fixed_is_indel</span><span class="p">[</span><span class="n">new_pos</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">fixed_indel_length</span><span class="p">[</span><span class="n">new_pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
            <span class="n">fixed_indel_index</span><span class="p">[</span><span class="n">new_pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">fixed_indel_index</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">revcomp_indel_nc_index</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">new_pos</span><span class="p">])]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
            <span class="p">)</span>

    <span class="c1"># Update instance variables</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">=</span> <span class="n">fixed_is_indel</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">=</span> <span class="n">fixed_indel_length</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="n">fixed_indel_nucleotides</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">indel_index</span> <span class="o">=</span> <span class="n">fixed_indel_index</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gene.Gene.__sub__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__sub__</span><span class="p">(</span><span class="n">other</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Generate a GeneDifference object for an in-depth examination of the
    difference between two genes.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>other</code></td>
            <td>
                  <code><span title="gumpy.Gene">Gene</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Other Gene object</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="gumpy.GeneDifference">GeneDifference</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>gumpy.GeneDifference: The GeneDifference object to detail changes at all
levels such as indel and amino acid.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/gene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__sub__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GeneDifference</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate a GeneDifference object for an in-depth examination of the</span>
<span class="sd">        difference between two genes.</span>

<span class="sd">    Args:</span>
<span class="sd">        other (gumpy.Gene): Other Gene object</span>

<span class="sd">    Returns:</span>
<span class="sd">        gumpy.GeneDifference: The GeneDifference object to detail changes at all</span>
<span class="sd">            levels such as indel and amino acid.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">Gene</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">GeneDifference</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gene.Gene.minority_populations_GARC" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">minority_populations_GARC</span><span class="p">(</span><span class="n">interpretation</span><span class="o">=</span><span class="s1">&#39;reads&#39;</span><span class="p">,</span> <span class="n">reference</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Fetch the mutations caused by minority populations in GARC</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>interpretation</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Which way to interpret the coverage calls.
'reads' gives absolute read coverage. 'percentage' gives fractional
read support. Defaults to 'reads'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;reads&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>reference</code></td>
            <td>
                  <code><span title="gumpy.Gene">Gene</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The reference to denote mutations from.
Defaults to self.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.List">List</span>[str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>[str]: List of mutations in GARC</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/gene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">minority_populations_GARC</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span> <span class="n">interpretation</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;reads&quot;</span><span class="p">,</span> <span class="n">reference</span><span class="o">=</span><span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Fetch the mutations caused by minority populations in GARC</span>

<span class="sd">    Args:</span>
<span class="sd">        interpretation (str, optional): Which way to interpret the coverage calls.</span>
<span class="sd">            &#39;reads&#39; gives absolute read coverage. &#39;percentage&#39; gives fractional</span>
<span class="sd">            read support. Defaults to &#39;reads&#39;.</span>
<span class="sd">        reference (gumpy.Gene, optional): The reference to denote mutations from.</span>
<span class="sd">            Defaults to self.</span>

<span class="sd">    Returns:</span>
<span class="sd">        [str]: List of mutations in GARC</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Depending on interpretation requested, we either want [3] or [4]</span>
    <span class="k">if</span> <span class="n">interpretation</span> <span class="o">==</span> <span class="s2">&quot;percentage&quot;</span><span class="p">:</span>
        <span class="n">coverage</span> <span class="o">=</span> <span class="mi">4</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">coverage</span> <span class="o">=</span> <span class="mi">3</span>

    <span class="c1"># Make sure we have a reference to compare against</span>
    <span class="k">if</span> <span class="n">reference</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">reference</span> <span class="o">=</span> <span class="bp">self</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">reference</span><span class="o">.</span><span class="n">minority_populations</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">,</span> <span class="p">(</span>
            <span class="s2">&quot;Minority populations can only be compared when 1 Gene &quot;</span>
            <span class="s2">&quot;does not have them!&quot;</span>
        <span class="p">)</span>

    <span class="c1"># Map gene_position--&gt;[minor populations at this position]</span>
    <span class="c1"># This also groups codons together</span>
    <span class="n">gene_pos_map</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">population</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">minority_populations</span><span class="p">:</span>
        <span class="n">pos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">==</span> <span class="n">population</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">gene_pos_map</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">population</span><span class="p">)</span>

    <span class="n">mutations</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">gene_pos</span> <span class="ow">in</span> <span class="n">gene_pos_map</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">populations</span> <span class="o">=</span> <span class="n">gene_pos_map</span><span class="p">[</span><span class="n">gene_pos</span><span class="p">]</span>
        <span class="c1"># As these don&#39;t have alt codons for Z calls, track the number of</span>
        <span class="c1">#   nucleotide changes separately</span>
        <span class="n">nc_changes</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">populations</span> <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;snp&quot;</span><span class="p">])</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">populations</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="c1"># We have exactly one so return it</span>
            <span class="n">pos</span><span class="p">,</span> <span class="n">type_</span><span class="p">,</span> <span class="n">bases</span><span class="p">,</span> <span class="n">cov</span><span class="p">,</span> <span class="n">frs</span> <span class="o">=</span> <span class="n">populations</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">type_</span> <span class="o">==</span> <span class="s2">&quot;snp&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span> <span class="ow">and</span> <span class="n">gene_pos</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">minor_nc_changes</span><span class="p">[</span><span class="n">gene_pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">nc_changes</span>
                    <span class="c1"># Get the ref codon to build the ref/alt AAs</span>
                    <span class="n">ref_codon</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
                        <span class="n">reference</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span> <span class="o">==</span> <span class="n">gene_pos</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                    <span class="p">)</span>
                    <span class="n">codon_idx</span> <span class="o">=</span> <span class="p">(</span><span class="n">pos</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span>
                    <span class="n">alt_codon</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">ref_codon</span><span class="p">)</span>
                    <span class="n">alt_codon</span><span class="p">[</span><span class="n">codon_idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">bases</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                    <span class="n">ref</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">codon_to_amino_acid</span><span class="p">[</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ref_codon</span><span class="p">)]</span>
                    <span class="n">alt</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">codon_to_amino_acid</span><span class="p">[</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alt_codon</span><span class="p">)]</span>
                    <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="n">ref</span>
                        <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene_pos</span><span class="p">)</span>
                        <span class="o">+</span> <span class="n">alt</span>
                        <span class="o">+</span> <span class="s2">&quot;:&quot;</span>
                        <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">populations</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">coverage</span><span class="p">])</span>
                    <span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">minor_nc_changes</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">nc_changes</span>
                    <span class="n">ref</span> <span class="o">=</span> <span class="n">reference</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">==</span> <span class="n">pos</span>
                    <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">alt</span> <span class="o">=</span> <span class="n">bases</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                    <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="n">ref</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span> <span class="o">+</span> <span class="n">alt</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">populations</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">coverage</span><span class="p">])</span>
                    <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">minor_nc_changes</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">nc_changes</span>
                <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="nb">str</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
                    <span class="o">+</span> <span class="s2">&quot;_&quot;</span>
                    <span class="o">+</span> <span class="n">type_</span>
                    <span class="o">+</span> <span class="s2">&quot;_&quot;</span>
                    <span class="o">+</span> <span class="n">bases</span>
                    <span class="o">+</span> <span class="s2">&quot;:&quot;</span>
                    <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">populations</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">coverage</span><span class="p">])</span>
                <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">minor_nc_changes</span><span class="p">[</span><span class="n">gene_pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">nc_changes</span>
            <span class="c1"># We have a mixture here so report as Zs for SNPs, indel for indels,</span>
            <span class="c1">#   and mixed for mixed indels and SNPs</span>
            <span class="n">c</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span>
                <span class="p">[</span>
                    <span class="p">(</span>
                        <span class="p">(</span><span class="n">nc_idx</span><span class="p">,</span> <span class="s2">&quot;indel&quot;</span><span class="p">)</span>
                        <span class="k">if</span> <span class="n">type_</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;ins&quot;</span><span class="p">,</span> <span class="s2">&quot;del&quot;</span><span class="p">]</span>
                        <span class="k">else</span> <span class="p">(</span><span class="n">nc_idx</span><span class="p">,</span> <span class="n">type_</span><span class="p">)</span>
                    <span class="p">)</span>
                    <span class="k">for</span> <span class="n">nc_idx</span><span class="p">,</span> <span class="n">type_</span><span class="p">,</span> <span class="n">bases</span><span class="p">,</span> <span class="n">cov</span><span class="p">,</span> <span class="n">frs</span> <span class="ow">in</span> <span class="n">gene_pos_map</span><span class="p">[</span><span class="n">gene_pos</span><span class="p">]</span>
                <span class="p">]</span>
            <span class="p">)</span>

            <span class="c1"># Get the number of indels within this codon/base</span>
            <span class="n">indels</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">count</span> <span class="k">for</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">t</span><span class="p">),</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">c</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">t</span> <span class="o">==</span> <span class="s2">&quot;indel&quot;</span><span class="p">])</span>
            <span class="n">snps</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">count</span> <span class="k">for</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">t</span><span class="p">),</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">c</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">t</span> <span class="o">==</span> <span class="s2">&quot;snp&quot;</span><span class="p">])</span>

            <span class="k">if</span> <span class="n">indels</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">snps</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># Mixed within the codon/base so return &lt;pos&gt;_mixed</span>
                <span class="n">cov</span> <span class="o">=</span> <span class="nb">max</span><span class="p">([</span><span class="n">pop</span><span class="p">[</span><span class="n">coverage</span><span class="p">]</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">populations</span><span class="p">])</span>
                <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">gene_pos</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_mixed:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cov</span><span class="p">))</span>
            <span class="k">elif</span> <span class="n">indels</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># Just indels, so return of form &lt;pos&gt;_indel</span>
                <span class="c1"># Little bit more difficult here as we have to check each</span>
                <span class="c1">#   nucleotide_number this way, if there is only 1 indel per</span>
                <span class="c1">#   nucleotide, we can be specific</span>
                <span class="k">for</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">t</span><span class="p">),</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">c</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="k">if</span> <span class="n">count</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="c1"># Exactly 1 indel at this position so be specific</span>
                        <span class="c1"># Use the nc_idx instead of gene_pos as they may not be</span>
                        <span class="c1">#   equal</span>
                        <span class="n">minor</span> <span class="o">=</span> <span class="p">[</span>
                            <span class="p">(</span><span class="n">nc_idx</span><span class="p">,</span> <span class="n">type_</span><span class="p">,</span> <span class="n">bases</span><span class="p">,</span> <span class="n">cov</span><span class="p">,</span> <span class="n">frs</span><span class="p">)</span>
                            <span class="k">for</span> <span class="n">nc_idx</span><span class="p">,</span> <span class="n">type_</span><span class="p">,</span> <span class="n">bases</span><span class="p">,</span> <span class="n">cov</span><span class="p">,</span> <span class="n">frs</span> <span class="ow">in</span> <span class="n">gene_pos_map</span><span class="p">[</span>
                                <span class="n">gene_pos</span>
                            <span class="p">]</span>
                            <span class="k">if</span> <span class="n">nc_idx</span> <span class="o">==</span> <span class="n">i</span>
                        <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">minor_nc_changes</span><span class="p">[</span><span class="n">minor</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="mi">0</span>
                        <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                            <span class="nb">str</span><span class="p">(</span><span class="n">minor</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                            <span class="o">+</span> <span class="s2">&quot;_&quot;</span>
                            <span class="o">+</span> <span class="n">minor</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                            <span class="o">+</span> <span class="s2">&quot;_&quot;</span>
                            <span class="o">+</span> <span class="n">minor</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
                            <span class="o">+</span> <span class="s2">&quot;:&quot;</span>
                            <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">minor</span><span class="p">[</span><span class="n">coverage</span><span class="p">])</span>
                        <span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># &gt;1 indel here, so `&lt;pos&gt;_indel` as as accurate as we</span>
                        <span class="c1">#   can be</span>
                        <span class="n">minors</span> <span class="o">=</span> <span class="p">[</span>
                            <span class="p">(</span><span class="n">nc_idx</span><span class="p">,</span> <span class="n">type_</span><span class="p">,</span> <span class="n">bases</span><span class="p">,</span> <span class="n">cov</span><span class="p">,</span> <span class="n">frs</span><span class="p">)</span>
                            <span class="k">for</span> <span class="n">nc_idx</span><span class="p">,</span> <span class="n">type_</span><span class="p">,</span> <span class="n">bases</span><span class="p">,</span> <span class="n">cov</span><span class="p">,</span> <span class="n">frs</span> <span class="ow">in</span> <span class="n">gene_pos_map</span><span class="p">[</span>
                                <span class="n">gene_pos</span>
                            <span class="p">]</span>
                            <span class="k">if</span> <span class="n">nc_idx</span> <span class="o">==</span> <span class="n">i</span>
                        <span class="p">]</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">minor_nc_changes</span><span class="p">[</span><span class="n">minors</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="mi">0</span>
                        <span class="n">cov</span> <span class="o">=</span> <span class="nb">max</span><span class="p">([</span><span class="n">pop</span><span class="p">[</span><span class="n">coverage</span><span class="p">]</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">minors</span><span class="p">])</span>
                        <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">minors</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;_indel:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cov</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Just SNPs, so return of form &lt;ref&gt;&lt;pos&gt;(z|Z) as appropriate</span>
                <span class="n">cov</span> <span class="o">=</span> <span class="nb">max</span><span class="p">([</span><span class="n">pop</span><span class="p">[</span><span class="n">coverage</span><span class="p">]</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">populations</span><span class="p">])</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span> <span class="ow">and</span> <span class="n">gene_pos</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="c1"># These need a little extra effort to pull out the ref AA</span>
                    <span class="n">ref</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">codon_to_amino_acid</span><span class="p">[</span>
                        <span class="n">reference</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span> <span class="o">==</span> <span class="n">gene_pos</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                    <span class="p">]</span>
                    <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ref</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene_pos</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;Z:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cov</span><span class="p">))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">ref</span> <span class="o">=</span> <span class="n">reference</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span> <span class="o">==</span> <span class="n">gene_pos</span>
                    <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ref</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene_pos</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;z:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cov</span><span class="p">))</span>
    <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">mutations</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gene.Gene.valid_variant" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">valid_variant</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Determines if a given variant is valid for this gene</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>variant</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>String of a mutation in GARC</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>bool</code></td>            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True when variant is valid, False otherwise</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/gene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span>
<span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span>
<span class="normal">847</span>
<span class="normal">848</span>
<span class="normal">849</span>
<span class="normal">850</span>
<span class="normal">851</span>
<span class="normal">852</span>
<span class="normal">853</span>
<span class="normal">854</span>
<span class="normal">855</span>
<span class="normal">856</span>
<span class="normal">857</span>
<span class="normal">858</span>
<span class="normal">859</span>
<span class="normal">860</span>
<span class="normal">861</span>
<span class="normal">862</span>
<span class="normal">863</span>
<span class="normal">864</span>
<span class="normal">865</span>
<span class="normal">866</span>
<span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span>
<span class="normal">873</span>
<span class="normal">874</span>
<span class="normal">875</span>
<span class="normal">876</span>
<span class="normal">877</span>
<span class="normal">878</span>
<span class="normal">879</span>
<span class="normal">880</span>
<span class="normal">881</span>
<span class="normal">882</span>
<span class="normal">883</span>
<span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span>
<span class="normal">899</span>
<span class="normal">900</span>
<span class="normal">901</span>
<span class="normal">902</span>
<span class="normal">903</span>
<span class="normal">904</span>
<span class="normal">905</span>
<span class="normal">906</span>
<span class="normal">907</span>
<span class="normal">908</span>
<span class="normal">909</span>
<span class="normal">910</span>
<span class="normal">911</span>
<span class="normal">912</span>
<span class="normal">913</span>
<span class="normal">914</span>
<span class="normal">915</span>
<span class="normal">916</span>
<span class="normal">917</span>
<span class="normal">918</span>
<span class="normal">919</span>
<span class="normal">920</span>
<span class="normal">921</span>
<span class="normal">922</span>
<span class="normal">923</span>
<span class="normal">924</span>
<span class="normal">925</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">valid_variant</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variant</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Determines if a given variant is valid for this gene</span>

<span class="sd">    Args:</span>
<span class="sd">        variant (str): String of a mutation in GARC</span>

<span class="sd">    Returns:</span>
<span class="sd">        bool: True when variant is valid, False otherwise</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">variant</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;No Variant given!&quot;</span>
    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">variant</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;Variant must be at least 2 characters e.g. A=&quot;</span>

    <span class="c1"># Match mutation formats using regex</span>

    <span class="c1"># Check for minority populations (and remove before continuing)</span>
    <span class="k">if</span> <span class="s2">&quot;:&quot;</span> <span class="ow">in</span> <span class="n">variant</span><span class="p">:</span>
        <span class="n">valid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">variant_</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)</span>

        <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">variant_</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span>

        <span class="c1"># Use duck typing to ensure it gives a valid number</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="nb">float</span><span class="p">(</span><span class="n">variant_</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">False</span>

        <span class="c1"># 0 is not valid either</span>
        <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="nb">float</span><span class="p">(</span><span class="n">variant_</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">0</span>

        <span class="k">if</span> <span class="n">valid</span><span class="p">:</span>
            <span class="c1"># Valid minority population, so continue without it</span>
            <span class="n">variant</span> <span class="o">=</span> <span class="n">variant_</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">False</span>

    <span class="c1"># Match promoter/non-coding SNP format</span>
    <span class="n">promoter</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            ([a-zA-Z_0-9.()]+@)? #Possibly a leading gene name</span>
<span class="sd">            ([acgtzx]) #Reference base</span>
<span class="sd">            (-?[0-9]+) #Position</span>
<span class="sd">            ([acgtzx]) #Alt base</span>
<span class="sd">            &quot;&quot;&quot;</span><span class="p">,</span>
        <span class="n">re</span><span class="o">.</span><span class="n">VERBOSE</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">promoter_match</span> <span class="o">=</span> <span class="n">promoter</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">promoter_match</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># The variant is either promoter or non-coding</span>
        <span class="c1"># Check that the mutation is valid</span>
        <span class="n">name</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">alt</span> <span class="o">=</span> <span class="n">promoter_match</span><span class="o">.</span><span class="n">groups</span><span class="p">()</span>
        <span class="n">valid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="c1"># Strip &#39;@&#39; from gene name and compare</span>
        <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">name</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">name</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span>
        <span class="c1"># Check that the pos is in the correct range</span>
        <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span>
        <span class="c1"># Check that the ref matches the seq at the given pos</span>
        <span class="n">valid</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">valid</span>
            <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">==</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)]</span> <span class="o">==</span> <span class="n">ref</span>
        <span class="p">)</span>
        <span class="c1"># Mutation should not have same ref and alt</span>
        <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">ref</span> <span class="o">!=</span> <span class="n">alt</span>
        <span class="k">return</span> <span class="n">valid</span>
    <span class="c1"># Match amino acid SNP</span>
    <span class="n">snp</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                ([a-zA-Z_0-9.()]+@)? #Possibly leading gene name</span>
<span class="sd">                ([A-Z!]) #Reference amino acid</span>
<span class="sd">                ([0-9]+) #Position</span>
<span class="sd">                ([A-Z!]) #Alt amino acid</span>
<span class="sd">                &quot;&quot;&quot;</span><span class="p">,</span>
        <span class="n">re</span><span class="o">.</span><span class="n">VERBOSE</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">snp_match</span> <span class="o">=</span> <span class="n">snp</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span> <span class="ow">and</span> <span class="n">snp_match</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># The variant is an amino acid SNP</span>
        <span class="c1"># Check it is valid</span>
        <span class="n">name</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">alt</span> <span class="o">=</span> <span class="n">snp_match</span><span class="o">.</span><span class="n">groups</span><span class="p">()</span>
        <span class="n">valid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="c1"># Strip &#39;@&#39; from gene name and compare</span>
        <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">name</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">name</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span>
        <span class="c1"># Check pos is in correct range</span>
        <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span>
        <span class="c1"># Check ref matches the aa seq at the given pos</span>
        <span class="n">valid</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">valid</span>
            <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_sequence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span> <span class="o">==</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)]</span> <span class="o">==</span> <span class="n">ref</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">valid</span>

    <span class="c1"># Match amino acid synon-mutation</span>
    <span class="n">synon</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                    ([a-zA-Z_0-9.()]+@)? #Possibly leading gene name</span>
<span class="sd">                    ([0-9]+) #Pos</span>
<span class="sd">                    = #Synonymous amino acid</span>
<span class="sd">                    &quot;&quot;&quot;</span><span class="p">,</span>
        <span class="n">re</span><span class="o">.</span><span class="n">VERBOSE</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">synon_match</span> <span class="o">=</span> <span class="n">synon</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span> <span class="ow">and</span> <span class="n">synon_match</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Variant is a synonymous mutation</span>
        <span class="c1"># Check it is valid</span>
        <span class="n">name</span><span class="p">,</span> <span class="n">pos</span> <span class="o">=</span> <span class="n">synon_match</span><span class="o">.</span><span class="n">groups</span><span class="p">()</span>
        <span class="n">valid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">name</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">name</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span>
        <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span>
        <span class="k">return</span> <span class="n">valid</span>

    <span class="c1"># Match indel</span>
    <span class="n">indel</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                ([a-zA-Z_0-9.()]+@)? #Possibly leading gene name</span>
<span class="sd">                (-?[0-9]+) #Position</span>
<span class="sd">                _(ins|del|indel|mixed)_? #Type</span>
<span class="sd">                ([0-9]+|[acgtzx]+)? #Bases deleted/inserted</span>
<span class="sd">                &quot;&quot;&quot;</span><span class="p">,</span>
        <span class="n">re</span><span class="o">.</span><span class="n">VERBOSE</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">indel_match</span> <span class="o">=</span> <span class="n">indel</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">indel_match</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Variant is an indel</span>
        <span class="c1"># Check it is valid</span>
        <span class="n">name</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">type_</span><span class="p">,</span> <span class="n">bases</span> <span class="o">=</span> <span class="n">indel_match</span><span class="o">.</span><span class="n">groups</span><span class="p">()</span>
        <span class="n">valid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
        <span class="n">end</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">)</span>
        <span class="c1"># Check the names match</span>
        <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">name</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">name</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span>
        <span class="c1"># Check pos in correct range</span>
        <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span>
        <span class="k">if</span> <span class="n">type_</span> <span class="o">==</span> <span class="s2">&quot;indel&quot;</span> <span class="ow">or</span> <span class="n">type_</span> <span class="o">==</span> <span class="s2">&quot;mixed&quot;</span><span class="p">:</span>
            <span class="c1"># If a mutation is given as `indel` or `mixed`, no length/bases should</span>
            <span class="c1">#   be given</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="p">(</span><span class="n">bases</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">bases</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">type_</span> <span class="o">==</span> <span class="s2">&quot;del&quot;</span> <span class="ow">and</span> <span class="n">bases</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">bases</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
            <span class="c1"># Mutation was del, so check if the bases given match the ref</span>
            <span class="k">if</span> <span class="n">bases</span><span class="o">.</span><span class="n">isnumeric</span><span class="p">():</span>
                <span class="c1"># A length was given rather than bases so just check that all bases</span>
                <span class="c1">#   are in the correct range</span>
                <span class="k">if</span> <span class="n">pos</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="c1"># Is a promoter so be careful about pos+bases not being 0</span>
                    <span class="n">pos</span> <span class="o">-=</span> <span class="mi">1</span>
                <span class="n">offset</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">bases</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">pos</span> <span class="o">+</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="n">offset</span> <span class="o">=</span> <span class="mi">1</span>
                    <span class="c1"># Ignore checking if the del passes the 5&#39;end of the gene</span>
                    <span class="k">if</span> <span class="n">pos</span> <span class="o">+</span> <span class="n">i</span> <span class="o">+</span> <span class="n">offset</span> <span class="o">&gt;</span> <span class="n">end</span><span class="p">:</span>
                        <span class="k">continue</span>
                    <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">pos</span> <span class="o">+</span> <span class="n">i</span> <span class="o">+</span> <span class="n">offset</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Bases were given rather than a length, so check for equality</span>
                <span class="c1">#   against base seq</span>
                <span class="n">offset</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">base</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">bases</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">pos</span> <span class="o">+</span> <span class="n">index</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="n">offset</span> <span class="o">=</span> <span class="mi">1</span>
                    <span class="k">if</span> <span class="n">pos</span> <span class="o">+</span> <span class="n">index</span> <span class="o">+</span> <span class="n">offset</span> <span class="o">&gt;</span> <span class="n">end</span><span class="p">:</span>
                        <span class="c1"># Ignore checking if the del passes the 5&#39; end of the gene</span>
                        <span class="k">continue</span>
                    <span class="n">valid</span> <span class="o">=</span> <span class="p">(</span>
                        <span class="n">valid</span>
                        <span class="ow">and</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span> <span class="o">+</span> <span class="n">index</span> <span class="o">+</span> <span class="n">offset</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span>
                    <span class="p">)</span>
                    <span class="n">valid</span> <span class="o">=</span> <span class="p">(</span>
                        <span class="n">valid</span>
                        <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">==</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span> <span class="o">+</span> <span class="n">index</span> <span class="o">+</span> <span class="n">offset</span>
                        <span class="p">]</span>
                        <span class="o">==</span> <span class="n">base</span>
                    <span class="p">)</span>
        <span class="k">if</span> <span class="n">type_</span> <span class="o">==</span> <span class="s2">&quot;ins&quot;</span><span class="p">:</span>
            <span class="c1"># Just check that the position specified lies within the gene</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">pos</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span>
        <span class="k">return</span> <span class="n">valid</span>
    <span class="c1"># Checking for percentage deletion</span>
    <span class="n">deletion</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                        ([a-zA-Z_0-9.()]+@)? #Possibly leading gene name</span>
<span class="sd">                        del_ #Deletion</span>
<span class="sd">                        ([01]\.[0-9]+)</span>
<span class="sd">                        &quot;&quot;&quot;</span><span class="p">,</span>
        <span class="n">re</span><span class="o">.</span><span class="n">VERBOSE</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">deletion_match</span> <span class="o">=</span> <span class="n">deletion</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">deletion_match</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">name</span><span class="p">,</span> <span class="n">percent</span> <span class="o">=</span> <span class="n">deletion_match</span><span class="o">.</span><span class="n">groups</span><span class="p">()</span>
        <span class="n">valid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">name</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span>
        <span class="n">percent</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">percent</span><span class="p">)</span>
        <span class="n">valid</span> <span class="o">=</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">percent</span> <span class="o">&lt;=</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">percent</span> <span class="o">&gt;=</span> <span class="mi">0</span>
        <span class="k">return</span> <span class="n">valid</span>
    <span class="c1"># If none of the mutations have matched, return False</span>
    <span class="k">return</span> <span class="kc">False</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../difference/" class="md-footer__link md-footer__link--prev" aria-label="Previous: difference">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                difference
              </div>
            </div>
          </a>
        
        
          
          <a href="../genome/" class="md-footer__link md-footer__link--next" aria-label="Next: genome">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                genome
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.footer", "navigation.instant", "navigation.sections", "navigation.tabs", "navigation.tracking"], "search": "../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "jeremy"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.fe8b6f2b.min.js"></script>
      
    
  </body>
</html>