
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../gene/">
      
      
        <link rel="next" href="../variantfile/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.30">
    
    
      
        <title>genome - Gumpy Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.3cba04c6.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#genome" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
      </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Gumpy Documentation" class="md-header__button md-logo" aria-label="Gumpy Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Gumpy Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              genome
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/oxfordmmm/gumpy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    gumpy
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../readme/" class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../" class="md-tabs__link">
          
  
  Code Reference

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Gumpy Documentation" class="md-nav__button md-logo" aria-label="Gumpy Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Gumpy Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/oxfordmmm/gumpy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    gumpy
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../readme/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Code Reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gumpy
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../difference/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    difference
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gene/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gene
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    genome
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    genome
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#genome" class="md-nav__link">
    <span class="md-ellipsis">
      genome
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#genome.Genome" class="md-nav__link">
    <span class="md-ellipsis">
      Genome
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Genome">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__add__" class="md-nav__link">
    <span class="md-ellipsis">
      __add__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__add_empty_row" class="md-nav__link">
    <span class="md-ellipsis">
      __add_empty_row
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__assign_deleted" class="md-nav__link">
    <span class="md-ellipsis">
      __assign_deleted
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__assign_promoter_regions" class="md-nav__link">
    <span class="md-ellipsis">
      __assign_promoter_regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__build_genome_variable_length_string" class="md-nav__link">
    <span class="md-ellipsis">
      __build_genome_variable_length_string
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__convert_references" class="md-nav__link">
    <span class="md-ellipsis">
      __convert_references
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__eq__" class="md-nav__link">
    <span class="md-ellipsis">
      __eq__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__find_overlaps" class="md-nav__link">
    <span class="md-ellipsis">
      __find_overlaps
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__fit_gene" class="md-nav__link">
    <span class="md-ellipsis">
      __fit_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__handle_rev_comp" class="md-nav__link">
    <span class="md-ellipsis">
      __handle_rev_comp
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__insert_newlines" class="md-nav__link">
    <span class="md-ellipsis">
      __insert_newlines
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__len__" class="md-nav__link">
    <span class="md-ellipsis">
      __len__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__parse_genbank_file" class="md-nav__link">
    <span class="md-ellipsis">
      __parse_genbank_file
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__repr__" class="md-nav__link">
    <span class="md-ellipsis">
      __repr__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__setup_arrays" class="md-nav__link">
    <span class="md-ellipsis">
      __setup_arrays
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__sub__" class="md-nav__link">
    <span class="md-ellipsis">
      __sub__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.at_index" class="md-nav__link">
    <span class="md-ellipsis">
      at_index
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.build_gene" class="md-nav__link">
    <span class="md-ellipsis">
      build_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.build_genome_string" class="md-nav__link">
    <span class="md-ellipsis">
      build_genome_string
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.contains_gene" class="md-nav__link">
    <span class="md-ellipsis">
      contains_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.minority_populations_GARC" class="md-nav__link">
    <span class="md-ellipsis">
      minority_populations_GARC
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.save_fasta" class="md-nav__link">
    <span class="md-ellipsis">
      save_fasta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.save_sequence" class="md-nav__link">
    <span class="md-ellipsis">
      save_sequence
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../variantfile/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    variantfile
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#genome" class="md-nav__link">
    <span class="md-ellipsis">
      genome
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#genome.Genome" class="md-nav__link">
    <span class="md-ellipsis">
      Genome
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Genome">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__add__" class="md-nav__link">
    <span class="md-ellipsis">
      __add__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__add_empty_row" class="md-nav__link">
    <span class="md-ellipsis">
      __add_empty_row
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__assign_deleted" class="md-nav__link">
    <span class="md-ellipsis">
      __assign_deleted
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__assign_promoter_regions" class="md-nav__link">
    <span class="md-ellipsis">
      __assign_promoter_regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__build_genome_variable_length_string" class="md-nav__link">
    <span class="md-ellipsis">
      __build_genome_variable_length_string
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__convert_references" class="md-nav__link">
    <span class="md-ellipsis">
      __convert_references
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__eq__" class="md-nav__link">
    <span class="md-ellipsis">
      __eq__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__find_overlaps" class="md-nav__link">
    <span class="md-ellipsis">
      __find_overlaps
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__fit_gene" class="md-nav__link">
    <span class="md-ellipsis">
      __fit_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__handle_rev_comp" class="md-nav__link">
    <span class="md-ellipsis">
      __handle_rev_comp
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__insert_newlines" class="md-nav__link">
    <span class="md-ellipsis">
      __insert_newlines
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__len__" class="md-nav__link">
    <span class="md-ellipsis">
      __len__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__parse_genbank_file" class="md-nav__link">
    <span class="md-ellipsis">
      __parse_genbank_file
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__repr__" class="md-nav__link">
    <span class="md-ellipsis">
      __repr__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__setup_arrays" class="md-nav__link">
    <span class="md-ellipsis">
      __setup_arrays
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.__sub__" class="md-nav__link">
    <span class="md-ellipsis">
      __sub__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.at_index" class="md-nav__link">
    <span class="md-ellipsis">
      at_index
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.build_gene" class="md-nav__link">
    <span class="md-ellipsis">
      build_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.build_genome_string" class="md-nav__link">
    <span class="md-ellipsis">
      build_genome_string
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.contains_gene" class="md-nav__link">
    <span class="md-ellipsis">
      contains_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.minority_populations_GARC" class="md-nav__link">
    <span class="md-ellipsis">
      minority_populations_GARC
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.save_fasta" class="md-nav__link">
    <span class="md-ellipsis">
      save_fasta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome.Genome.save_sequence" class="md-nav__link">
    <span class="md-ellipsis">
      save_sequence
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>genome</h1>

<div class="doc doc-object doc-module">



<a id="genome"></a>
    <div class="doc doc-contents first">

      <p>Genome object</p>



  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="genome.Genome" class="doc doc-heading">
            <code>Genome</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code>object</code></p>


      <p>Genome object</p>

              <details class="quote">
                <summary>Source code in <code>gumpy/genome.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  19</span>
<span class="normal">  20</span>
<span class="normal">  21</span>
<span class="normal">  22</span>
<span class="normal">  23</span>
<span class="normal">  24</span>
<span class="normal">  25</span>
<span class="normal">  26</span>
<span class="normal">  27</span>
<span class="normal">  28</span>
<span class="normal">  29</span>
<span class="normal">  30</span>
<span class="normal">  31</span>
<span class="normal">  32</span>
<span class="normal">  33</span>
<span class="normal">  34</span>
<span class="normal">  35</span>
<span class="normal">  36</span>
<span class="normal">  37</span>
<span class="normal">  38</span>
<span class="normal">  39</span>
<span class="normal">  40</span>
<span class="normal">  41</span>
<span class="normal">  42</span>
<span class="normal">  43</span>
<span class="normal">  44</span>
<span class="normal">  45</span>
<span class="normal">  46</span>
<span class="normal">  47</span>
<span class="normal">  48</span>
<span class="normal">  49</span>
<span class="normal">  50</span>
<span class="normal">  51</span>
<span class="normal">  52</span>
<span class="normal">  53</span>
<span class="normal">  54</span>
<span class="normal">  55</span>
<span class="normal">  56</span>
<span class="normal">  57</span>
<span class="normal">  58</span>
<span class="normal">  59</span>
<span class="normal">  60</span>
<span class="normal">  61</span>
<span class="normal">  62</span>
<span class="normal">  63</span>
<span class="normal">  64</span>
<span class="normal">  65</span>
<span class="normal">  66</span>
<span class="normal">  67</span>
<span class="normal">  68</span>
<span class="normal">  69</span>
<span class="normal">  70</span>
<span class="normal">  71</span>
<span class="normal">  72</span>
<span class="normal">  73</span>
<span class="normal">  74</span>
<span class="normal">  75</span>
<span class="normal">  76</span>
<span class="normal">  77</span>
<span class="normal">  78</span>
<span class="normal">  79</span>
<span class="normal">  80</span>
<span class="normal">  81</span>
<span class="normal">  82</span>
<span class="normal">  83</span>
<span class="normal">  84</span>
<span class="normal">  85</span>
<span class="normal">  86</span>
<span class="normal">  87</span>
<span class="normal">  88</span>
<span class="normal">  89</span>
<span class="normal">  90</span>
<span class="normal">  91</span>
<span class="normal">  92</span>
<span class="normal">  93</span>
<span class="normal">  94</span>
<span class="normal">  95</span>
<span class="normal">  96</span>
<span class="normal">  97</span>
<span class="normal">  98</span>
<span class="normal">  99</span>
<span class="normal"> 100</span>
<span class="normal"> 101</span>
<span class="normal"> 102</span>
<span class="normal"> 103</span>
<span class="normal"> 104</span>
<span class="normal"> 105</span>
<span class="normal"> 106</span>
<span class="normal"> 107</span>
<span class="normal"> 108</span>
<span class="normal"> 109</span>
<span class="normal"> 110</span>
<span class="normal"> 111</span>
<span class="normal"> 112</span>
<span class="normal"> 113</span>
<span class="normal"> 114</span>
<span class="normal"> 115</span>
<span class="normal"> 116</span>
<span class="normal"> 117</span>
<span class="normal"> 118</span>
<span class="normal"> 119</span>
<span class="normal"> 120</span>
<span class="normal"> 121</span>
<span class="normal"> 122</span>
<span class="normal"> 123</span>
<span class="normal"> 124</span>
<span class="normal"> 125</span>
<span class="normal"> 126</span>
<span class="normal"> 127</span>
<span class="normal"> 128</span>
<span class="normal"> 129</span>
<span class="normal"> 130</span>
<span class="normal"> 131</span>
<span class="normal"> 132</span>
<span class="normal"> 133</span>
<span class="normal"> 134</span>
<span class="normal"> 135</span>
<span class="normal"> 136</span>
<span class="normal"> 137</span>
<span class="normal"> 138</span>
<span class="normal"> 139</span>
<span class="normal"> 140</span>
<span class="normal"> 141</span>
<span class="normal"> 142</span>
<span class="normal"> 143</span>
<span class="normal"> 144</span>
<span class="normal"> 145</span>
<span class="normal"> 146</span>
<span class="normal"> 147</span>
<span class="normal"> 148</span>
<span class="normal"> 149</span>
<span class="normal"> 150</span>
<span class="normal"> 151</span>
<span class="normal"> 152</span>
<span class="normal"> 153</span>
<span class="normal"> 154</span>
<span class="normal"> 155</span>
<span class="normal"> 156</span>
<span class="normal"> 157</span>
<span class="normal"> 158</span>
<span class="normal"> 159</span>
<span class="normal"> 160</span>
<span class="normal"> 161</span>
<span class="normal"> 162</span>
<span class="normal"> 163</span>
<span class="normal"> 164</span>
<span class="normal"> 165</span>
<span class="normal"> 166</span>
<span class="normal"> 167</span>
<span class="normal"> 168</span>
<span class="normal"> 169</span>
<span class="normal"> 170</span>
<span class="normal"> 171</span>
<span class="normal"> 172</span>
<span class="normal"> 173</span>
<span class="normal"> 174</span>
<span class="normal"> 175</span>
<span class="normal"> 176</span>
<span class="normal"> 177</span>
<span class="normal"> 178</span>
<span class="normal"> 179</span>
<span class="normal"> 180</span>
<span class="normal"> 181</span>
<span class="normal"> 182</span>
<span class="normal"> 183</span>
<span class="normal"> 184</span>
<span class="normal"> 185</span>
<span class="normal"> 186</span>
<span class="normal"> 187</span>
<span class="normal"> 188</span>
<span class="normal"> 189</span>
<span class="normal"> 190</span>
<span class="normal"> 191</span>
<span class="normal"> 192</span>
<span class="normal"> 193</span>
<span class="normal"> 194</span>
<span class="normal"> 195</span>
<span class="normal"> 196</span>
<span class="normal"> 197</span>
<span class="normal"> 198</span>
<span class="normal"> 199</span>
<span class="normal"> 200</span>
<span class="normal"> 201</span>
<span class="normal"> 202</span>
<span class="normal"> 203</span>
<span class="normal"> 204</span>
<span class="normal"> 205</span>
<span class="normal"> 206</span>
<span class="normal"> 207</span>
<span class="normal"> 208</span>
<span class="normal"> 209</span>
<span class="normal"> 210</span>
<span class="normal"> 211</span>
<span class="normal"> 212</span>
<span class="normal"> 213</span>
<span class="normal"> 214</span>
<span class="normal"> 215</span>
<span class="normal"> 216</span>
<span class="normal"> 217</span>
<span class="normal"> 218</span>
<span class="normal"> 219</span>
<span class="normal"> 220</span>
<span class="normal"> 221</span>
<span class="normal"> 222</span>
<span class="normal"> 223</span>
<span class="normal"> 224</span>
<span class="normal"> 225</span>
<span class="normal"> 226</span>
<span class="normal"> 227</span>
<span class="normal"> 228</span>
<span class="normal"> 229</span>
<span class="normal"> 230</span>
<span class="normal"> 231</span>
<span class="normal"> 232</span>
<span class="normal"> 233</span>
<span class="normal"> 234</span>
<span class="normal"> 235</span>
<span class="normal"> 236</span>
<span class="normal"> 237</span>
<span class="normal"> 238</span>
<span class="normal"> 239</span>
<span class="normal"> 240</span>
<span class="normal"> 241</span>
<span class="normal"> 242</span>
<span class="normal"> 243</span>
<span class="normal"> 244</span>
<span class="normal"> 245</span>
<span class="normal"> 246</span>
<span class="normal"> 247</span>
<span class="normal"> 248</span>
<span class="normal"> 249</span>
<span class="normal"> 250</span>
<span class="normal"> 251</span>
<span class="normal"> 252</span>
<span class="normal"> 253</span>
<span class="normal"> 254</span>
<span class="normal"> 255</span>
<span class="normal"> 256</span>
<span class="normal"> 257</span>
<span class="normal"> 258</span>
<span class="normal"> 259</span>
<span class="normal"> 260</span>
<span class="normal"> 261</span>
<span class="normal"> 262</span>
<span class="normal"> 263</span>
<span class="normal"> 264</span>
<span class="normal"> 265</span>
<span class="normal"> 266</span>
<span class="normal"> 267</span>
<span class="normal"> 268</span>
<span class="normal"> 269</span>
<span class="normal"> 270</span>
<span class="normal"> 271</span>
<span class="normal"> 272</span>
<span class="normal"> 273</span>
<span class="normal"> 274</span>
<span class="normal"> 275</span>
<span class="normal"> 276</span>
<span class="normal"> 277</span>
<span class="normal"> 278</span>
<span class="normal"> 279</span>
<span class="normal"> 280</span>
<span class="normal"> 281</span>
<span class="normal"> 282</span>
<span class="normal"> 283</span>
<span class="normal"> 284</span>
<span class="normal"> 285</span>
<span class="normal"> 286</span>
<span class="normal"> 287</span>
<span class="normal"> 288</span>
<span class="normal"> 289</span>
<span class="normal"> 290</span>
<span class="normal"> 291</span>
<span class="normal"> 292</span>
<span class="normal"> 293</span>
<span class="normal"> 294</span>
<span class="normal"> 295</span>
<span class="normal"> 296</span>
<span class="normal"> 297</span>
<span class="normal"> 298</span>
<span class="normal"> 299</span>
<span class="normal"> 300</span>
<span class="normal"> 301</span>
<span class="normal"> 302</span>
<span class="normal"> 303</span>
<span class="normal"> 304</span>
<span class="normal"> 305</span>
<span class="normal"> 306</span>
<span class="normal"> 307</span>
<span class="normal"> 308</span>
<span class="normal"> 309</span>
<span class="normal"> 310</span>
<span class="normal"> 311</span>
<span class="normal"> 312</span>
<span class="normal"> 313</span>
<span class="normal"> 314</span>
<span class="normal"> 315</span>
<span class="normal"> 316</span>
<span class="normal"> 317</span>
<span class="normal"> 318</span>
<span class="normal"> 319</span>
<span class="normal"> 320</span>
<span class="normal"> 321</span>
<span class="normal"> 322</span>
<span class="normal"> 323</span>
<span class="normal"> 324</span>
<span class="normal"> 325</span>
<span class="normal"> 326</span>
<span class="normal"> 327</span>
<span class="normal"> 328</span>
<span class="normal"> 329</span>
<span class="normal"> 330</span>
<span class="normal"> 331</span>
<span class="normal"> 332</span>
<span class="normal"> 333</span>
<span class="normal"> 334</span>
<span class="normal"> 335</span>
<span class="normal"> 336</span>
<span class="normal"> 337</span>
<span class="normal"> 338</span>
<span class="normal"> 339</span>
<span class="normal"> 340</span>
<span class="normal"> 341</span>
<span class="normal"> 342</span>
<span class="normal"> 343</span>
<span class="normal"> 344</span>
<span class="normal"> 345</span>
<span class="normal"> 346</span>
<span class="normal"> 347</span>
<span class="normal"> 348</span>
<span class="normal"> 349</span>
<span class="normal"> 350</span>
<span class="normal"> 351</span>
<span class="normal"> 352</span>
<span class="normal"> 353</span>
<span class="normal"> 354</span>
<span class="normal"> 355</span>
<span class="normal"> 356</span>
<span class="normal"> 357</span>
<span class="normal"> 358</span>
<span class="normal"> 359</span>
<span class="normal"> 360</span>
<span class="normal"> 361</span>
<span class="normal"> 362</span>
<span class="normal"> 363</span>
<span class="normal"> 364</span>
<span class="normal"> 365</span>
<span class="normal"> 366</span>
<span class="normal"> 367</span>
<span class="normal"> 368</span>
<span class="normal"> 369</span>
<span class="normal"> 370</span>
<span class="normal"> 371</span>
<span class="normal"> 372</span>
<span class="normal"> 373</span>
<span class="normal"> 374</span>
<span class="normal"> 375</span>
<span class="normal"> 376</span>
<span class="normal"> 377</span>
<span class="normal"> 378</span>
<span class="normal"> 379</span>
<span class="normal"> 380</span>
<span class="normal"> 381</span>
<span class="normal"> 382</span>
<span class="normal"> 383</span>
<span class="normal"> 384</span>
<span class="normal"> 385</span>
<span class="normal"> 386</span>
<span class="normal"> 387</span>
<span class="normal"> 388</span>
<span class="normal"> 389</span>
<span class="normal"> 390</span>
<span class="normal"> 391</span>
<span class="normal"> 392</span>
<span class="normal"> 393</span>
<span class="normal"> 394</span>
<span class="normal"> 395</span>
<span class="normal"> 396</span>
<span class="normal"> 397</span>
<span class="normal"> 398</span>
<span class="normal"> 399</span>
<span class="normal"> 400</span>
<span class="normal"> 401</span>
<span class="normal"> 402</span>
<span class="normal"> 403</span>
<span class="normal"> 404</span>
<span class="normal"> 405</span>
<span class="normal"> 406</span>
<span class="normal"> 407</span>
<span class="normal"> 408</span>
<span class="normal"> 409</span>
<span class="normal"> 410</span>
<span class="normal"> 411</span>
<span class="normal"> 412</span>
<span class="normal"> 413</span>
<span class="normal"> 414</span>
<span class="normal"> 415</span>
<span class="normal"> 416</span>
<span class="normal"> 417</span>
<span class="normal"> 418</span>
<span class="normal"> 419</span>
<span class="normal"> 420</span>
<span class="normal"> 421</span>
<span class="normal"> 422</span>
<span class="normal"> 423</span>
<span class="normal"> 424</span>
<span class="normal"> 425</span>
<span class="normal"> 426</span>
<span class="normal"> 427</span>
<span class="normal"> 428</span>
<span class="normal"> 429</span>
<span class="normal"> 430</span>
<span class="normal"> 431</span>
<span class="normal"> 432</span>
<span class="normal"> 433</span>
<span class="normal"> 434</span>
<span class="normal"> 435</span>
<span class="normal"> 436</span>
<span class="normal"> 437</span>
<span class="normal"> 438</span>
<span class="normal"> 439</span>
<span class="normal"> 440</span>
<span class="normal"> 441</span>
<span class="normal"> 442</span>
<span class="normal"> 443</span>
<span class="normal"> 444</span>
<span class="normal"> 445</span>
<span class="normal"> 446</span>
<span class="normal"> 447</span>
<span class="normal"> 448</span>
<span class="normal"> 449</span>
<span class="normal"> 450</span>
<span class="normal"> 451</span>
<span class="normal"> 452</span>
<span class="normal"> 453</span>
<span class="normal"> 454</span>
<span class="normal"> 455</span>
<span class="normal"> 456</span>
<span class="normal"> 457</span>
<span class="normal"> 458</span>
<span class="normal"> 459</span>
<span class="normal"> 460</span>
<span class="normal"> 461</span>
<span class="normal"> 462</span>
<span class="normal"> 463</span>
<span class="normal"> 464</span>
<span class="normal"> 465</span>
<span class="normal"> 466</span>
<span class="normal"> 467</span>
<span class="normal"> 468</span>
<span class="normal"> 469</span>
<span class="normal"> 470</span>
<span class="normal"> 471</span>
<span class="normal"> 472</span>
<span class="normal"> 473</span>
<span class="normal"> 474</span>
<span class="normal"> 475</span>
<span class="normal"> 476</span>
<span class="normal"> 477</span>
<span class="normal"> 478</span>
<span class="normal"> 479</span>
<span class="normal"> 480</span>
<span class="normal"> 481</span>
<span class="normal"> 482</span>
<span class="normal"> 483</span>
<span class="normal"> 484</span>
<span class="normal"> 485</span>
<span class="normal"> 486</span>
<span class="normal"> 487</span>
<span class="normal"> 488</span>
<span class="normal"> 489</span>
<span class="normal"> 490</span>
<span class="normal"> 491</span>
<span class="normal"> 492</span>
<span class="normal"> 493</span>
<span class="normal"> 494</span>
<span class="normal"> 495</span>
<span class="normal"> 496</span>
<span class="normal"> 497</span>
<span class="normal"> 498</span>
<span class="normal"> 499</span>
<span class="normal"> 500</span>
<span class="normal"> 501</span>
<span class="normal"> 502</span>
<span class="normal"> 503</span>
<span class="normal"> 504</span>
<span class="normal"> 505</span>
<span class="normal"> 506</span>
<span class="normal"> 507</span>
<span class="normal"> 508</span>
<span class="normal"> 509</span>
<span class="normal"> 510</span>
<span class="normal"> 511</span>
<span class="normal"> 512</span>
<span class="normal"> 513</span>
<span class="normal"> 514</span>
<span class="normal"> 515</span>
<span class="normal"> 516</span>
<span class="normal"> 517</span>
<span class="normal"> 518</span>
<span class="normal"> 519</span>
<span class="normal"> 520</span>
<span class="normal"> 521</span>
<span class="normal"> 522</span>
<span class="normal"> 523</span>
<span class="normal"> 524</span>
<span class="normal"> 525</span>
<span class="normal"> 526</span>
<span class="normal"> 527</span>
<span class="normal"> 528</span>
<span class="normal"> 529</span>
<span class="normal"> 530</span>
<span class="normal"> 531</span>
<span class="normal"> 532</span>
<span class="normal"> 533</span>
<span class="normal"> 534</span>
<span class="normal"> 535</span>
<span class="normal"> 536</span>
<span class="normal"> 537</span>
<span class="normal"> 538</span>
<span class="normal"> 539</span>
<span class="normal"> 540</span>
<span class="normal"> 541</span>
<span class="normal"> 542</span>
<span class="normal"> 543</span>
<span class="normal"> 544</span>
<span class="normal"> 545</span>
<span class="normal"> 546</span>
<span class="normal"> 547</span>
<span class="normal"> 548</span>
<span class="normal"> 549</span>
<span class="normal"> 550</span>
<span class="normal"> 551</span>
<span class="normal"> 552</span>
<span class="normal"> 553</span>
<span class="normal"> 554</span>
<span class="normal"> 555</span>
<span class="normal"> 556</span>
<span class="normal"> 557</span>
<span class="normal"> 558</span>
<span class="normal"> 559</span>
<span class="normal"> 560</span>
<span class="normal"> 561</span>
<span class="normal"> 562</span>
<span class="normal"> 563</span>
<span class="normal"> 564</span>
<span class="normal"> 565</span>
<span class="normal"> 566</span>
<span class="normal"> 567</span>
<span class="normal"> 568</span>
<span class="normal"> 569</span>
<span class="normal"> 570</span>
<span class="normal"> 571</span>
<span class="normal"> 572</span>
<span class="normal"> 573</span>
<span class="normal"> 574</span>
<span class="normal"> 575</span>
<span class="normal"> 576</span>
<span class="normal"> 577</span>
<span class="normal"> 578</span>
<span class="normal"> 579</span>
<span class="normal"> 580</span>
<span class="normal"> 581</span>
<span class="normal"> 582</span>
<span class="normal"> 583</span>
<span class="normal"> 584</span>
<span class="normal"> 585</span>
<span class="normal"> 586</span>
<span class="normal"> 587</span>
<span class="normal"> 588</span>
<span class="normal"> 589</span>
<span class="normal"> 590</span>
<span class="normal"> 591</span>
<span class="normal"> 592</span>
<span class="normal"> 593</span>
<span class="normal"> 594</span>
<span class="normal"> 595</span>
<span class="normal"> 596</span>
<span class="normal"> 597</span>
<span class="normal"> 598</span>
<span class="normal"> 599</span>
<span class="normal"> 600</span>
<span class="normal"> 601</span>
<span class="normal"> 602</span>
<span class="normal"> 603</span>
<span class="normal"> 604</span>
<span class="normal"> 605</span>
<span class="normal"> 606</span>
<span class="normal"> 607</span>
<span class="normal"> 608</span>
<span class="normal"> 609</span>
<span class="normal"> 610</span>
<span class="normal"> 611</span>
<span class="normal"> 612</span>
<span class="normal"> 613</span>
<span class="normal"> 614</span>
<span class="normal"> 615</span>
<span class="normal"> 616</span>
<span class="normal"> 617</span>
<span class="normal"> 618</span>
<span class="normal"> 619</span>
<span class="normal"> 620</span>
<span class="normal"> 621</span>
<span class="normal"> 622</span>
<span class="normal"> 623</span>
<span class="normal"> 624</span>
<span class="normal"> 625</span>
<span class="normal"> 626</span>
<span class="normal"> 627</span>
<span class="normal"> 628</span>
<span class="normal"> 629</span>
<span class="normal"> 630</span>
<span class="normal"> 631</span>
<span class="normal"> 632</span>
<span class="normal"> 633</span>
<span class="normal"> 634</span>
<span class="normal"> 635</span>
<span class="normal"> 636</span>
<span class="normal"> 637</span>
<span class="normal"> 638</span>
<span class="normal"> 639</span>
<span class="normal"> 640</span>
<span class="normal"> 641</span>
<span class="normal"> 642</span>
<span class="normal"> 643</span>
<span class="normal"> 644</span>
<span class="normal"> 645</span>
<span class="normal"> 646</span>
<span class="normal"> 647</span>
<span class="normal"> 648</span>
<span class="normal"> 649</span>
<span class="normal"> 650</span>
<span class="normal"> 651</span>
<span class="normal"> 652</span>
<span class="normal"> 653</span>
<span class="normal"> 654</span>
<span class="normal"> 655</span>
<span class="normal"> 656</span>
<span class="normal"> 657</span>
<span class="normal"> 658</span>
<span class="normal"> 659</span>
<span class="normal"> 660</span>
<span class="normal"> 661</span>
<span class="normal"> 662</span>
<span class="normal"> 663</span>
<span class="normal"> 664</span>
<span class="normal"> 665</span>
<span class="normal"> 666</span>
<span class="normal"> 667</span>
<span class="normal"> 668</span>
<span class="normal"> 669</span>
<span class="normal"> 670</span>
<span class="normal"> 671</span>
<span class="normal"> 672</span>
<span class="normal"> 673</span>
<span class="normal"> 674</span>
<span class="normal"> 675</span>
<span class="normal"> 676</span>
<span class="normal"> 677</span>
<span class="normal"> 678</span>
<span class="normal"> 679</span>
<span class="normal"> 680</span>
<span class="normal"> 681</span>
<span class="normal"> 682</span>
<span class="normal"> 683</span>
<span class="normal"> 684</span>
<span class="normal"> 685</span>
<span class="normal"> 686</span>
<span class="normal"> 687</span>
<span class="normal"> 688</span>
<span class="normal"> 689</span>
<span class="normal"> 690</span>
<span class="normal"> 691</span>
<span class="normal"> 692</span>
<span class="normal"> 693</span>
<span class="normal"> 694</span>
<span class="normal"> 695</span>
<span class="normal"> 696</span>
<span class="normal"> 697</span>
<span class="normal"> 698</span>
<span class="normal"> 699</span>
<span class="normal"> 700</span>
<span class="normal"> 701</span>
<span class="normal"> 702</span>
<span class="normal"> 703</span>
<span class="normal"> 704</span>
<span class="normal"> 705</span>
<span class="normal"> 706</span>
<span class="normal"> 707</span>
<span class="normal"> 708</span>
<span class="normal"> 709</span>
<span class="normal"> 710</span>
<span class="normal"> 711</span>
<span class="normal"> 712</span>
<span class="normal"> 713</span>
<span class="normal"> 714</span>
<span class="normal"> 715</span>
<span class="normal"> 716</span>
<span class="normal"> 717</span>
<span class="normal"> 718</span>
<span class="normal"> 719</span>
<span class="normal"> 720</span>
<span class="normal"> 721</span>
<span class="normal"> 722</span>
<span class="normal"> 723</span>
<span class="normal"> 724</span>
<span class="normal"> 725</span>
<span class="normal"> 726</span>
<span class="normal"> 727</span>
<span class="normal"> 728</span>
<span class="normal"> 729</span>
<span class="normal"> 730</span>
<span class="normal"> 731</span>
<span class="normal"> 732</span>
<span class="normal"> 733</span>
<span class="normal"> 734</span>
<span class="normal"> 735</span>
<span class="normal"> 736</span>
<span class="normal"> 737</span>
<span class="normal"> 738</span>
<span class="normal"> 739</span>
<span class="normal"> 740</span>
<span class="normal"> 741</span>
<span class="normal"> 742</span>
<span class="normal"> 743</span>
<span class="normal"> 744</span>
<span class="normal"> 745</span>
<span class="normal"> 746</span>
<span class="normal"> 747</span>
<span class="normal"> 748</span>
<span class="normal"> 749</span>
<span class="normal"> 750</span>
<span class="normal"> 751</span>
<span class="normal"> 752</span>
<span class="normal"> 753</span>
<span class="normal"> 754</span>
<span class="normal"> 755</span>
<span class="normal"> 756</span>
<span class="normal"> 757</span>
<span class="normal"> 758</span>
<span class="normal"> 759</span>
<span class="normal"> 760</span>
<span class="normal"> 761</span>
<span class="normal"> 762</span>
<span class="normal"> 763</span>
<span class="normal"> 764</span>
<span class="normal"> 765</span>
<span class="normal"> 766</span>
<span class="normal"> 767</span>
<span class="normal"> 768</span>
<span class="normal"> 769</span>
<span class="normal"> 770</span>
<span class="normal"> 771</span>
<span class="normal"> 772</span>
<span class="normal"> 773</span>
<span class="normal"> 774</span>
<span class="normal"> 775</span>
<span class="normal"> 776</span>
<span class="normal"> 777</span>
<span class="normal"> 778</span>
<span class="normal"> 779</span>
<span class="normal"> 780</span>
<span class="normal"> 781</span>
<span class="normal"> 782</span>
<span class="normal"> 783</span>
<span class="normal"> 784</span>
<span class="normal"> 785</span>
<span class="normal"> 786</span>
<span class="normal"> 787</span>
<span class="normal"> 788</span>
<span class="normal"> 789</span>
<span class="normal"> 790</span>
<span class="normal"> 791</span>
<span class="normal"> 792</span>
<span class="normal"> 793</span>
<span class="normal"> 794</span>
<span class="normal"> 795</span>
<span class="normal"> 796</span>
<span class="normal"> 797</span>
<span class="normal"> 798</span>
<span class="normal"> 799</span>
<span class="normal"> 800</span>
<span class="normal"> 801</span>
<span class="normal"> 802</span>
<span class="normal"> 803</span>
<span class="normal"> 804</span>
<span class="normal"> 805</span>
<span class="normal"> 806</span>
<span class="normal"> 807</span>
<span class="normal"> 808</span>
<span class="normal"> 809</span>
<span class="normal"> 810</span>
<span class="normal"> 811</span>
<span class="normal"> 812</span>
<span class="normal"> 813</span>
<span class="normal"> 814</span>
<span class="normal"> 815</span>
<span class="normal"> 816</span>
<span class="normal"> 817</span>
<span class="normal"> 818</span>
<span class="normal"> 819</span>
<span class="normal"> 820</span>
<span class="normal"> 821</span>
<span class="normal"> 822</span>
<span class="normal"> 823</span>
<span class="normal"> 824</span>
<span class="normal"> 825</span>
<span class="normal"> 826</span>
<span class="normal"> 827</span>
<span class="normal"> 828</span>
<span class="normal"> 829</span>
<span class="normal"> 830</span>
<span class="normal"> 831</span>
<span class="normal"> 832</span>
<span class="normal"> 833</span>
<span class="normal"> 834</span>
<span class="normal"> 835</span>
<span class="normal"> 836</span>
<span class="normal"> 837</span>
<span class="normal"> 838</span>
<span class="normal"> 839</span>
<span class="normal"> 840</span>
<span class="normal"> 841</span>
<span class="normal"> 842</span>
<span class="normal"> 843</span>
<span class="normal"> 844</span>
<span class="normal"> 845</span>
<span class="normal"> 846</span>
<span class="normal"> 847</span>
<span class="normal"> 848</span>
<span class="normal"> 849</span>
<span class="normal"> 850</span>
<span class="normal"> 851</span>
<span class="normal"> 852</span>
<span class="normal"> 853</span>
<span class="normal"> 854</span>
<span class="normal"> 855</span>
<span class="normal"> 856</span>
<span class="normal"> 857</span>
<span class="normal"> 858</span>
<span class="normal"> 859</span>
<span class="normal"> 860</span>
<span class="normal"> 861</span>
<span class="normal"> 862</span>
<span class="normal"> 863</span>
<span class="normal"> 864</span>
<span class="normal"> 865</span>
<span class="normal"> 866</span>
<span class="normal"> 867</span>
<span class="normal"> 868</span>
<span class="normal"> 869</span>
<span class="normal"> 870</span>
<span class="normal"> 871</span>
<span class="normal"> 872</span>
<span class="normal"> 873</span>
<span class="normal"> 874</span>
<span class="normal"> 875</span>
<span class="normal"> 876</span>
<span class="normal"> 877</span>
<span class="normal"> 878</span>
<span class="normal"> 879</span>
<span class="normal"> 880</span>
<span class="normal"> 881</span>
<span class="normal"> 882</span>
<span class="normal"> 883</span>
<span class="normal"> 884</span>
<span class="normal"> 885</span>
<span class="normal"> 886</span>
<span class="normal"> 887</span>
<span class="normal"> 888</span>
<span class="normal"> 889</span>
<span class="normal"> 890</span>
<span class="normal"> 891</span>
<span class="normal"> 892</span>
<span class="normal"> 893</span>
<span class="normal"> 894</span>
<span class="normal"> 895</span>
<span class="normal"> 896</span>
<span class="normal"> 897</span>
<span class="normal"> 898</span>
<span class="normal"> 899</span>
<span class="normal"> 900</span>
<span class="normal"> 901</span>
<span class="normal"> 902</span>
<span class="normal"> 903</span>
<span class="normal"> 904</span>
<span class="normal"> 905</span>
<span class="normal"> 906</span>
<span class="normal"> 907</span>
<span class="normal"> 908</span>
<span class="normal"> 909</span>
<span class="normal"> 910</span>
<span class="normal"> 911</span>
<span class="normal"> 912</span>
<span class="normal"> 913</span>
<span class="normal"> 914</span>
<span class="normal"> 915</span>
<span class="normal"> 916</span>
<span class="normal"> 917</span>
<span class="normal"> 918</span>
<span class="normal"> 919</span>
<span class="normal"> 920</span>
<span class="normal"> 921</span>
<span class="normal"> 922</span>
<span class="normal"> 923</span>
<span class="normal"> 924</span>
<span class="normal"> 925</span>
<span class="normal"> 926</span>
<span class="normal"> 927</span>
<span class="normal"> 928</span>
<span class="normal"> 929</span>
<span class="normal"> 930</span>
<span class="normal"> 931</span>
<span class="normal"> 932</span>
<span class="normal"> 933</span>
<span class="normal"> 934</span>
<span class="normal"> 935</span>
<span class="normal"> 936</span>
<span class="normal"> 937</span>
<span class="normal"> 938</span>
<span class="normal"> 939</span>
<span class="normal"> 940</span>
<span class="normal"> 941</span>
<span class="normal"> 942</span>
<span class="normal"> 943</span>
<span class="normal"> 944</span>
<span class="normal"> 945</span>
<span class="normal"> 946</span>
<span class="normal"> 947</span>
<span class="normal"> 948</span>
<span class="normal"> 949</span>
<span class="normal"> 950</span>
<span class="normal"> 951</span>
<span class="normal"> 952</span>
<span class="normal"> 953</span>
<span class="normal"> 954</span>
<span class="normal"> 955</span>
<span class="normal"> 956</span>
<span class="normal"> 957</span>
<span class="normal"> 958</span>
<span class="normal"> 959</span>
<span class="normal"> 960</span>
<span class="normal"> 961</span>
<span class="normal"> 962</span>
<span class="normal"> 963</span>
<span class="normal"> 964</span>
<span class="normal"> 965</span>
<span class="normal"> 966</span>
<span class="normal"> 967</span>
<span class="normal"> 968</span>
<span class="normal"> 969</span>
<span class="normal"> 970</span>
<span class="normal"> 971</span>
<span class="normal"> 972</span>
<span class="normal"> 973</span>
<span class="normal"> 974</span>
<span class="normal"> 975</span>
<span class="normal"> 976</span>
<span class="normal"> 977</span>
<span class="normal"> 978</span>
<span class="normal"> 979</span>
<span class="normal"> 980</span>
<span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span>
<span class="normal">1022</span>
<span class="normal">1023</span>
<span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span>
<span class="normal">1103</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Genome</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Genome object&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">genbank_file_</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">show_progress_bar</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">gene_subset</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">max_promoter_length</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
        <span class="n">max_gene_name_length</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>
        <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">is_reference</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Constructor for the Genome object.</span>

<span class="sd">        Args:</span>
<span class="sd">            genbank_file_ (str) : The path to the genbank file.</span>
<span class="sd">            show_progress_bar (bool, optional) : Boolean as whether to show a progress</span>
<span class="sd">                bar when building Gene objects. Defaults to False.</span>
<span class="sd">            gene_subset (list, optional) : List of gene names used to extract just a</span>
<span class="sd">                subset of genes. Defaults to None</span>
<span class="sd">            max_promoter_length (int, optional) : Size of the default maximum number of</span>
<span class="sd">                upstream bases to consider the promoter of a gene. Defaults to 100</span>
<span class="sd">            max_gene_name_length (int, optional) : Length of the longest gene name.</span>
<span class="sd">                Defaults to 20</span>
<span class="sd">            verbose (bool, optional) : Give verbose statements? Defaults to False</span>
<span class="sd">            is_reference (bool, optional) : Is this a reference genome? i.e. mutations</span>
<span class="sd">                can be derived with respect to it? Defaults to False</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">show_progress_bar</span> <span class="o">=</span> <span class="n">show_progress_bar</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span> <span class="o">=</span> <span class="n">gene_subset</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_promoter_length</span> <span class="o">=</span> <span class="n">max_promoter_length</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_gene_name_length</span> <span class="o">=</span> <span class="n">max_gene_name_length</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span> <span class="o">=</span> <span class="n">verbose</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_reference</span> <span class="o">=</span> <span class="n">is_reference</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vcf_file</span><span class="p">:</span> <span class="n">VCFFile</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gumpy_version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="n">genbank_file</span> <span class="o">=</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">genbank_file_</span><span class="p">)</span>

        <span class="k">assert</span> <span class="n">genbank_file</span><span class="o">.</span><span class="n">is_file</span><span class="p">(),</span> <span class="s2">&quot;GenBank file does not exist!&quot;</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">max_promoter_length</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_promoter_length</span> <span class="o">&gt;=</span> <span class="mi">0</span>
        <span class="p">),</span> <span class="s2">&quot;the promoter length must be zero or a positive integer!&quot;</span>

        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">,</span> <span class="nb">bool</span><span class="p">)</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_reference</span><span class="p">,</span> <span class="nb">bool</span><span class="p">)</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">show_progress_bar</span><span class="p">,</span> <span class="nb">bool</span><span class="p">)</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">max_gene_name_length</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_gene_name_length</span> <span class="o">&gt;</span> <span class="mi">0</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># first check it is a list</span>
            <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span>
            <span class="c1"># then check all elements in the list are strings</span>
            <span class="k">assert</span> <span class="nb">all</span><span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="n">timings</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
            <span class="n">start_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">__parse_genbank_file</span><span class="p">(</span><span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">genbank_file</span><span class="p">))</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="n">timings</span><span class="p">[</span><span class="s2">&quot;parse genbank&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">start_time</span><span class="p">)</span>
            <span class="n">start_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">__setup_arrays</span><span class="p">()</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="n">timings</span><span class="p">[</span><span class="s2">&quot;define arrays&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">start_time</span><span class="p">)</span>
            <span class="n">start_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_promoter_length</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">__assign_promoter_regions</span><span class="p">()</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="n">timings</span><span class="p">[</span><span class="s2">&quot;promoter&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">start_time</span><span class="p">)</span>
            <span class="n">start_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">__convert_references</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">__convert_references</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Convert BIOPython Reference objects to normal dictionaries. They do not</span>
<span class="sd">        appear to have any greater application than storing structured data, so</span>
<span class="sd">        removing the object wrappers appears to be a clean way to combat the object&#39;s</span>
<span class="sd">        issues with serialization.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">reference</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;references&quot;</span><span class="p">]):</span>
            <span class="n">new_ref</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">vars</span><span class="p">(</span><span class="n">reference</span><span class="p">):</span>
                <span class="c1"># This key contains unhelpfully structured data</span>
                <span class="k">if</span> <span class="n">key</span> <span class="o">==</span> <span class="s2">&quot;location&quot;</span><span class="p">:</span>
                    <span class="n">new_loc</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">vars</span><span class="p">(</span><span class="n">reference</span><span class="p">)[</span><span class="n">key</span><span class="p">]:</span>
                        <span class="n">loc</span> <span class="o">=</span> <span class="p">{}</span>
                        <span class="k">for</span> <span class="n">item_key</span> <span class="ow">in</span> <span class="nb">vars</span><span class="p">(</span><span class="n">item</span><span class="p">):</span>
                            <span class="k">if</span> <span class="n">item_key</span> <span class="o">==</span> <span class="s2">&quot;_start&quot;</span> <span class="ow">or</span> <span class="n">item_key</span> <span class="o">==</span> <span class="s2">&quot;_end&quot;</span><span class="p">:</span>
                                <span class="c1"># These are the only ones we care about</span>
                                <span class="n">loc</span><span class="p">[</span><span class="n">item_key</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="n">item</span><span class="p">,</span> <span class="n">item_key</span><span class="p">))</span>
                        <span class="n">new_loc</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span>
                    <span class="n">new_ref</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_loc</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">new_ref</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="nb">vars</span><span class="p">(</span><span class="n">reference</span><span class="p">)[</span><span class="n">key</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;references&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_ref</span>

    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Overload the print function to write a summary of the genome.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str : String including attributes for the genome</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">):</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;id&quot;</span><span class="p">):</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;description&quot;</span><span class="p">):</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">description</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot; bases</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">6</span><span class="p">])</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;...&quot;</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="o">-</span><span class="mi">6</span><span class="p">:])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;metadata for all genes/loci have been included</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">10</span><span class="p">:</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="p">(</span>
                <span class="s2">&quot;the following &quot;</span>
                <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span><span class="p">))</span>
                <span class="o">+</span> <span class="s2">&quot; genes have been included: &quot;</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span><span class="p">:</span>
                <span class="n">output</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span><span class="p">))</span> <span class="o">+</span> <span class="s2">&quot; gene/loci have been included.&quot;</span>
        <span class="k">return</span> <span class="n">output</span>

    <span class="k">def</span> <span class="fm">__sub__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenomeDifference</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Generate a GenomeDifference object for a in-depth difference of the</span>
<span class="sd">            two Genomes</span>

<span class="sd">        Args:</span>
<span class="sd">            other (gumpy.Genome) : The other genome used in the subtraction</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomeDifference: object containing numpy arrays of the</span>
<span class="sd">                differences (variants)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">Genome</span><span class="p">),</span> <span class="s2">&quot;RHS must be a gumpy.Genome object&quot;</span>

        <span class="k">return</span> <span class="n">GenomeDifference</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Overloading the equality operator so two Genome objects can be compared</span>
<span class="sd">        directly. Checks for the equality based on fields, but does not check</span>
<span class="sd">        for filename equality</span>

<span class="sd">        Args:</span>
<span class="sd">            other (gumpy.Genome) : The other Genome object to compare to</span>

<span class="sd">        Returns:</span>
<span class="sd">            bool : Boolean showing equality of the objects</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">Genome</span><span class="p">)</span>

        <span class="n">check</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bool_</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bool_</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">name</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">id</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">description</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">description</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">nucleotide_sequence</span>
        <span class="p">)</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">)</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bool_</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">length</span><span class="p">)</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">stacked_gene_name</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">stacked_gene_name</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="nb">bool</span><span class="p">(</span><span class="n">check</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adding len functionality - len(genome) returns the length of the genome</span>

<span class="sd">        Returns:</span>
<span class="sd">            int : Length of the genome</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span>

    <span class="k">def</span> <span class="nf">contains_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Simply checks to see if the specified gene exists in the Genome object.</span>

<span class="sd">        Args:</span>
<span class="sd">            gene_name (str) : Name of the gene e.g. katG</span>

<span class="sd">        Returns:</span>
<span class="sd">            bool : Boolean showing if the genome contains a gene with that name</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span>
            <span class="n">gene_name</span><span class="p">,</span> <span class="nb">str</span>
        <span class="p">),</span> <span class="s2">&quot;Gene name must be string. Gene name provided was of type: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span>
            <span class="nb">type</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="c1"># Use of dict.get(obj) returns an object or None if obj does not exist in dict</span>
        <span class="c1"># bool(None) = False, bool(obj) = True</span>
        <span class="k">return</span> <span class="nb">bool</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene_name</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">at_index</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the name of any genome features (genes, loci) at a specified genome</span>
<span class="sd">            index (1-based).</span>

<span class="sd">        Args:</span>
<span class="sd">            index (int): Genome index to check for genes at.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[str] | None: list of gene_names or locus_tags at that index</span>
<span class="sd">                in the genome</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span> <span class="s2">&quot;index must be an integer!&quot;</span>
        <span class="k">assert</span> <span class="n">index</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;index must be a positive integer!&quot;</span>
        <span class="k">assert</span> <span class="n">index</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">,</span> <span class="s2">&quot;index must be less than the length of the genome!&quot;</span>

        <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span> <span class="o">==</span> <span class="n">index</span>

        <span class="n">foo</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stacked_gene_name</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>

        <span class="n">putative_genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">foo</span><span class="p">[</span><span class="n">foo</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">])</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">putative_genes</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">putative_genes</span>

    <span class="k">def</span> <span class="nf">save_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Save the genome as a compressed NPZ file (compressed internally using gzip).</span>

<span class="sd">        This is purely done because loading an NPZ file back into memory is FAST</span>
<span class="sd">            (~200s) so this could allow future analyses</span>

<span class="sd">        Args:</span>
<span class="sd">            filename (str): path of the output file without the file extension</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">numpy</span><span class="o">.</span><span class="n">savez_compressed</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__build_genome_variable_length_string</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">indices</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Build a string of the genome sequence, including indels - resulting in a</span>
<span class="sd">            variable length genome</span>

<span class="sd">        Args:</span>
<span class="sd">            indices (List[int]): List of the indices of indels</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: Genome sequence as a string</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">genome_string</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="c1"># work backwards as easier to deal with insertions/deletions when you&#39;ve</span>
        <span class="c1">#   already gone past them</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">indices</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
            <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">==</span> <span class="n">i</span>
            <span class="n">base</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">genome_string</span> <span class="o">=</span> <span class="n">base</span> <span class="o">+</span> <span class="n">genome_string</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[</span><span class="n">mask</span><span class="p">]:</span>
                <span class="n">indel_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">mask</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">indel_length</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">genome_string</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">mask</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">genome_string</span>
                <span class="k">elif</span> <span class="n">indel_length</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">genome_string</span> <span class="o">=</span> <span class="n">genome_string</span><span class="p">[</span><span class="nb">abs</span><span class="p">(</span><span class="n">indel_length</span><span class="p">)</span> <span class="p">:]</span>
        <span class="k">return</span> <span class="n">genome_string</span>

    <span class="k">def</span> <span class="nf">build_genome_string</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">fixed_length</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">nucleotide_index_range</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate a string of the nucleotides in the genome (positive strand if DNA).</span>

<span class="sd">        Args:</span>
<span class="sd">            fixed_length (bool): if True, then do not add insertions and deletions.</span>
<span class="sd">                Default False.</span>
<span class="sd">            nucleotide_index_range (tuple, ints): the 1-based positions of the sequence</span>
<span class="sd">                to return with start&lt;=index&lt;end.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (str): the genome as a string.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># create a string of the genome</span>
        <span class="k">if</span> <span class="n">fixed_length</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">nucleotide_index_range</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">nucleotide_index_range</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">)</span>
                <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">nucleotide_index_range</span>
                <span class="n">genome_string</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">genome_string</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">nucleotide_index_range</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">nucleotide_index_range</span>
                <span class="n">genome_string</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__build_genome_variable_length_string</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">genome_string</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__build_genome_variable_length_string</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span>
                <span class="p">)</span>

        <span class="k">return</span> <span class="n">genome_string</span>

    <span class="k">def</span> <span class="nf">save_fasta</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">filename</span><span class="p">,</span>
        <span class="n">fixed_length</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">nucleotide_index_range</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">compression</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">compresslevel</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
        <span class="n">chars_per_line</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">70</span><span class="p">,</span>
        <span class="n">nucleotides_uppercase</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">description</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">overwrite_existing</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Save the genome as a FASTA file.</span>

<span class="sd">        Args:</span>
<span class="sd">            filename (str): path of the output file</span>
<span class="sd">            fixed_length (bool): If True, ignore indels and only output a genome the</span>
<span class="sd">                same length as the reference but with SNPs. This is useful for</span>
<span class="sd">                phylogeny analyses and relatedness. If false, a genome including indels</span>
<span class="sd">                is produced. Default is false.</span>
<span class="sd">            nucleotide_index_range (tuple, optional): A tuple of (start,end)</span>
<span class="sd">                genome indices</span>
<span class="sd">            compression (bool): If True, save compressed using gzip. (bzip2 is too slow)</span>
<span class="sd">            compresslevel (0-9): the higher the number, the harder the algorithm tries</span>
<span class="sd">                to compress but it takes longer. Default is 2.</span>
<span class="sd">            chars_per_line (int): the number of characters per line. Default=70. Must</span>
<span class="sd">                be either a positive integer or None (i.e. no CRs)</span>
<span class="sd">            nucleotide_uppercase (bool): If True, provide the nucleotides in</span>
<span class="sd">                UPPER CASE. Default is True.</span>
<span class="sd">            description (str, optional): what to write on the header line of the FASTA</span>
<span class="sd">                file. If not provided, then a description will be automatically</span>
<span class="sd">                generated from the GenBank file metadata.</span>
<span class="sd">            overwrite_existing (bool): If False, then the code will refuse to overwrite</span>
<span class="sd">                a FASTA file already on disc. Default is True.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># check the arguments are well formed</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">overwrite_existing</span><span class="p">:</span>
            <span class="k">assert</span> <span class="ow">not</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span><span class="o">.</span><span class="n">is_file</span><span class="p">(),</span> <span class="p">(</span>
                <span class="s2">&quot;filename already exists! &quot;</span> <span class="o">+</span> <span class="n">filename</span>
            <span class="p">)</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">compression</span><span class="p">,</span> <span class="nb">bool</span><span class="p">)</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">fixed_length</span><span class="p">,</span> <span class="nb">bool</span><span class="p">)</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">nucleotides_uppercase</span><span class="p">,</span> <span class="nb">bool</span><span class="p">)</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">chars_per_line</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">nucleotide_index_range</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">nucleotide_index_range</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">)</span>
            <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">nucleotide_index_range</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">int</span><span class="p">)</span>
            <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">nucleotide_index_range</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="nb">int</span><span class="p">)</span>
            <span class="k">assert</span> <span class="p">(</span>
                <span class="n">nucleotide_index_range</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">1</span>
            <span class="p">),</span> <span class="s2">&quot;genomes are 1-based so the first base must be &gt;=1&quot;</span>
            <span class="k">assert</span> <span class="n">nucleotide_index_range</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">,</span> <span class="s2">&quot;longer than the genome!&quot;</span>
        <span class="k">assert</span> <span class="n">compresslevel</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="s2">&quot;compresslevel must be in range 1-9!&quot;</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="n">chars_per_line</span> <span class="o">&gt;</span> <span class="mi">0</span>
        <span class="p">),</span> <span class="s2">&quot;number of characters per line in the FASTA file must be a positive integer!&quot;</span>
        <span class="k">if</span> <span class="n">description</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">description</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span>

        <span class="c1"># check the specified fileextension to see if the FASTA file needs compressing</span>
        <span class="k">if</span> <span class="n">compression</span><span class="p">:</span>
            <span class="n">OUTPUT</span> <span class="o">=</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">filename</span> <span class="o">+</span> <span class="s2">&quot;.gz&quot;</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">,</span> <span class="n">compresslevel</span><span class="o">=</span><span class="n">compresslevel</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">OUTPUT</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>

        <span class="c1"># create the header line for the FASTA file using &quot;|&quot; as delimiters</span>
        <span class="n">header</span> <span class="o">=</span> <span class="s2">&quot;&gt;&quot;</span>
        <span class="k">if</span> <span class="n">description</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">):</span>
                <span class="n">header</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;|&quot;</span>
            <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;id&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">header</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">+</span> <span class="s2">&quot;|&quot;</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;description&quot;</span><span class="p">)</span>
                <span class="ow">and</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">description</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span>
                <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">description</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span>
            <span class="p">):</span>
                <span class="n">header</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">description</span> <span class="o">+</span> <span class="s2">&quot;|&quot;</span>
            <span class="n">header</span> <span class="o">=</span> <span class="n">header</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">header</span> <span class="o">+=</span> <span class="n">description</span>
        <span class="n">header</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="n">genome_string</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">build_genome_string</span><span class="p">(</span><span class="n">fixed_length</span><span class="p">,</span> <span class="n">nucleotide_index_range</span><span class="p">)</span>

        <span class="c1"># insert carriage returns so it looks pretty in the file...</span>
        <span class="n">output_string</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__insert_newlines</span><span class="p">(</span><span class="n">genome_string</span><span class="p">,</span> <span class="n">every</span><span class="o">=</span><span class="n">chars_per_line</span><span class="p">)</span>
        <span class="n">output_string</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="c1"># set the case accordingly</span>
        <span class="k">if</span> <span class="n">nucleotides_uppercase</span><span class="p">:</span>
            <span class="n">output_string</span> <span class="o">=</span> <span class="n">output_string</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">output_string</span> <span class="o">=</span> <span class="n">output_string</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>

        <span class="c1"># write out the FASTA files</span>
        <span class="k">if</span> <span class="n">compression</span><span class="p">:</span>
            <span class="n">OUTPUT</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="n">header</span><span class="p">))</span>
            <span class="n">OUTPUT</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="n">output_string</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">OUTPUT</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">header</span><span class="p">)</span>
            <span class="n">OUTPUT</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">output_string</span><span class="p">)</span>

        <span class="n">OUTPUT</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">__add_empty_row</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">array</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Private function to add an empty row of the correct type to a numpy array</span>
<span class="sd">        Args:</span>
<span class="sd">            array (numpy.ndarray) : Array to add an empty row to</span>
<span class="sd">        Returns:</span>
<span class="sd">            (numpy.ndarray): The same array with an empty row of the same length and</span>
<span class="sd">                dtype appended</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">empty_row</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="n">array</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">array</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">vstack</span><span class="p">((</span><span class="n">array</span><span class="p">,</span> <span class="n">empty_row</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">__parse_genbank_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genbank_file</span><span class="p">:</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Private function to parse a genbank file</span>
<span class="sd">        Args:</span>
<span class="sd">            genbank_file (Path) : pathlib.Path object of the genbank file</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">genbank_file</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.gz&quot;</span><span class="p">:</span>
            <span class="n">file_handle</span> <span class="o">=</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">genbank_file</span><span class="p">,</span> <span class="s2">&quot;rt&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">genbank_file</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.gbk&quot;</span><span class="p">:</span>
            <span class="n">file_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">genbank_file</span><span class="p">,</span> <span class="s2">&quot;rt&quot;</span><span class="p">)</span>

        <span class="n">reference_genome</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">file_handle</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span>

        <span class="c1"># convert to a numpy array at the first opportunity since slicing BioPython</span>
        <span class="c1">#   is between 10 and 50,000 times slower!</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">str</span><span class="p">(</span><span class="n">reference_genome</span><span class="o">.</span><span class="n">seq</span><span class="p">)]</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">reference_genome</span><span class="o">.</span><span class="n">name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">reference_genome</span><span class="o">.</span><span class="n">id</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="n">reference_genome</span><span class="o">.</span><span class="n">description</span>

        <span class="c1"># store the length of the genome</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">)</span>

        <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;genome length zero!&quot;</span>

        <span class="c1"># create an array of the genome indices</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;int&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span>
            <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;&lt;U&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">max_gene_name_length</span><span class="p">))</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_cds</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">bool</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_promoter</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">bool</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;int&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_reverse_complement</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">bool</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">bool</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">object</span><span class="p">)</span>

        <span class="k">assert</span> <span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">reference_genome</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;accessions&quot;</span><span class="p">])</span> <span class="o">==</span> <span class="mi">1</span>
        <span class="p">),</span> <span class="s2">&quot;only GenBank files with a single accessions currently allowed&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">reference_genome</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">reference_genome</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">:</span> <span class="n">Dict</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="c1"># loop through the features listed in the GenBank File</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Iterating through features in GenBank file...&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span>
            <span class="n">reference_genome</span><span class="o">.</span><span class="n">features</span><span class="p">,</span> <span class="n">disable</span><span class="o">=</span><span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">show_progress_bar</span><span class="p">)</span>
        <span class="p">):</span>
            <span class="c1"># only parse coding sequences and rRNA features</span>
            <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;CDS&quot;</span><span class="p">,</span> <span class="s2">&quot;rRNA&quot;</span><span class="p">]:</span>
                <span class="k">continue</span>

            <span class="n">gene_name</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">type_</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">codes_protein</span> <span class="o">=</span> <span class="kc">True</span>

            <span class="c1"># try and use the gene name if available, otherwise use the locus</span>
            <span class="k">if</span> <span class="s2">&quot;gene&quot;</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">gene_name</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">type_</span> <span class="o">=</span> <span class="s2">&quot;GENE&quot;</span>

            <span class="k">elif</span> <span class="s2">&quot;locus_tag&quot;</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">gene_name</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;locus_tag&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">type_</span> <span class="o">=</span> <span class="s2">&quot;LOCUS&quot;</span>

            <span class="k">if</span> <span class="n">gene_name</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">gene_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span>
            <span class="p">):</span>
                <span class="k">continue</span>

            <span class="c1"># if this is ribosomal RNA, then record as such</span>
            <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;rRNA&quot;</span><span class="p">:</span>
                <span class="n">type_</span> <span class="o">=</span> <span class="s2">&quot;RNA&quot;</span>
                <span class="n">codes_protein</span> <span class="o">=</span> <span class="kc">False</span>

            <span class="c1"># determine if this is a reverse complement gene (only relevant to</span>
            <span class="c1">#   dsDNA genomes)</span>
            <span class="n">rev_comp</span> <span class="o">=</span> <span class="kc">True</span> <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span> <span class="k">else</span> <span class="kc">False</span>

            <span class="c1"># sigh, you can&#39;t assume that a gene_name is unique in a GenBank file</span>
            <span class="c1"># this only allows for duplicates though.</span>
            <span class="c1"># duplicates of duplicates will be foo_2_2</span>
            <span class="n">gene_name</span> <span class="o">+=</span> <span class="s2">&quot;_2&quot;</span> <span class="k">if</span> <span class="n">gene_name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>

            <span class="c1"># check the gene_name will fit in the max gene name length</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_gene_name_length</span><span class="p">,</span> <span class="p">(</span>
                <span class="s2">&quot;Gene &quot;</span>
                <span class="o">+</span> <span class="n">gene_name</span>
                <span class="o">+</span> <span class="s2">&quot; is too long at &quot;</span>
                <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_name</span><span class="p">))</span>
                <span class="o">+</span> <span class="s2">&quot; chars; need to specify max_gene_name_length&quot;</span>
            <span class="p">)</span>

            <span class="c1"># note that BioPython &quot;helpfully&quot; turns these from 1-based into 0-based</span>
            <span class="c1">#   coordinates, hence the +1</span>
            <span class="c1"># gene_end has also been incremented by 1 so that slicing naturally works</span>
            <span class="n">gene_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">gene_end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>

            <span class="c1"># Check for ribosomal shift</span>
            <span class="c1"># This happens when a start position &lt; end position</span>
            <span class="n">positions</span> <span class="o">=</span> <span class="p">[</span>
                <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">loc</span><span class="o">.</span><span class="n">start</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">loc</span><span class="o">.</span><span class="n">end</span><span class="p">))</span> <span class="k">for</span> <span class="n">loc</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">parts</span>
            <span class="p">]</span>
            <span class="n">shifts</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="c1"># Check for -1 PFS</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">positions</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">positions</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]:</span>
                <span class="n">shifts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">positions</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">gene_start</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

            <span class="c1"># record feature metadata in a dict</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;reverse_complement&quot;</span><span class="p">:</span> <span class="n">rev_comp</span><span class="p">,</span>
                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="n">type_</span><span class="p">,</span>
                <span class="s2">&quot;codes_protein&quot;</span><span class="p">:</span> <span class="n">codes_protein</span><span class="p">,</span>
                <span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="n">gene_start</span><span class="p">,</span>
                <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="n">gene_end</span><span class="p">,</span>
                <span class="s2">&quot;ribosomal_shifts&quot;</span><span class="p">:</span> <span class="n">shifts</span><span class="p">,</span>
            <span class="p">}</span>

        <span class="c1"># now we can check that all the genes in the gene_subset exist in the</span>
        <span class="c1">#   GenBank file!</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span><span class="p">:</span>
                <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">contains_gene</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="p">(</span>
                    <span class="s2">&quot;Gene &quot;</span> <span class="o">+</span> <span class="n">i</span> <span class="o">+</span> <span class="s2">&quot; not found in the Genbank file!&quot;</span>
                <span class="p">)</span>

    <span class="k">def</span> <span class="nf">__handle_rev_comp</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">rev_comp</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">end</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">i</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Private function to handle the rev-comp changes required</span>
<span class="sd">        Args:</span>
<span class="sd">            rev_comp (bool) : Boolean to show if rev-comp is required</span>
<span class="sd">            start (int) : Start index of the gene</span>
<span class="sd">            end (int) : End index of the gene</span>
<span class="sd">            i (int) : The index of the stacked row</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Check if the arrays have the correct number of rows and add as required</span>
        <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">i</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_cds</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__add_empty_row</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_cds</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_reverse_complement</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__add_empty_row</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_reverse_complement</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_promoter</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__add_empty_row</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_promoter</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__add_empty_row</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span>
            <span class="p">)</span>
        <span class="c1"># Update the required items for rev_comp</span>
        <span class="c1"># Use of array slicing here introduces speedups as not all of the array should</span>
        <span class="c1">#   be considered</span>
        <span class="c1"># (Previous method applied a mask which requires full array consideration</span>
        <span class="c1">#   rather than direct access)</span>
        <span class="k">if</span> <span class="n">rev_comp</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mod</span><span class="p">(</span>
                <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">end</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_reverse_complement</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mod</span><span class="p">(</span>
                <span class="mi">1</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">start</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">__fit_gene</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">mask</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">genes</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">genes_mask</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">end</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">gene_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">rev_comp</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Private function to fit a gene into the genes based on the dot product of</span>
<span class="sd">            the masks numpy.dot([bool], [bool])-&gt; bool showing if there are collisions</span>
<span class="sd">            of True values within args. This takes 10^-5 seconds which is a significant</span>
<span class="sd">            improvement on use of numpy.all() iteration of 10^-2 seconds for TB</span>
<span class="sd">            length genome</span>

<span class="sd">        Args:</span>
<span class="sd">            mask (numpy.ndarray) : Boolean array showing positions where the gene lies</span>
<span class="sd">            genes (numpy.ndarray) : 2D numpy array of the format used for all stacked</span>
<span class="sd">                values</span>
<span class="sd">            genes_mask (numpy.ndarray) : The corresponding boolean mask arrays for the</span>
<span class="sd">                `genes` arg</span>
<span class="sd">            start (int) : Start index of the gene</span>
<span class="sd">            end (int) : End index of the gene</span>
<span class="sd">            gene_name (str) : Name of the gene</span>
<span class="sd">            rev_comp (bool) : Boolean to show whether the gene required a</span>
<span class="sd">                reverse complement</span>

<span class="sd">        Returns:</span>
<span class="sd">            (numpy_array) : Updated genes_mask array</span>
<span class="sd">            (numpy_array) : Updated genes array</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">genes_mask</span><span class="p">):</span>
            <span class="c1"># use of the dot product of masks allows determining if a gene will fit</span>
            <span class="c1">#   in an array</span>
            <span class="c1"># Disable ruff here as `is False` doesn&#39;t work here</span>
            <span class="k">if</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">mask</span><span class="p">)</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>  <span class="c1"># noqa: E712</span>
                <span class="c1"># there is no collision with this row so add the row</span>
                <span class="c1"># start/end have to be adjusted to account for 0 indexing of arrays and</span>
                <span class="c1">#   1 indexing of genetics</span>
                <span class="n">row</span><span class="p">[</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="n">genes</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_name</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">__handle_rev_comp</span><span class="p">(</span><span class="n">rev_comp</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
                <span class="k">return</span> <span class="n">genes_mask</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">i</span>

        <span class="c1"># if this point is reached, there has been no rows without collisions, so</span>
        <span class="c1">#   add one</span>
        <span class="n">genes_mask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">vstack</span><span class="p">((</span><span class="n">genes_mask</span><span class="p">,</span> <span class="n">mask</span><span class="p">))</span>
        <span class="n">new_row</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">gene_name</span> <span class="k">if</span> <span class="n">m</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">mask</span><span class="p">])</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">vstack</span><span class="p">((</span><span class="n">genes</span><span class="p">,</span> <span class="n">new_row</span><span class="p">))</span>
        <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">__handle_rev_comp</span><span class="p">(</span><span class="n">rev_comp</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">genes_mask</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">i</span>

    <span class="k">def</span> <span class="nf">__find_overlaps</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Private function to find the sections of the genome in which there are</span>
<span class="sd">            overlapping genes. This should be more efficient than the older version</span>
<span class="sd">            as it avoids consistent genome iteration. Use of the dot product on boolean</span>
<span class="sd">            arrays returns a single boolean showing collisions in almost constant time</span>
<span class="sd">            (10^-5 secs for TB size). This can be used to determine which row the gene</span>
<span class="sd">            should be in</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Boolean mask to show gene presence at indicies (default to all False values)</span>
        <span class="n">genes_mask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="kc">False</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">)])])</span>

        <span class="c1"># gene names</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;&quot;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">)])],</span>
            <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;U&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">max_gene_name_length</span><span class="p">),</span>
        <span class="p">)</span>

        <span class="c1"># dict to pull out row indicies for each gene in the stacked arrays</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_rows</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Finding overlaps...&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">gene_name</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">,</span> <span class="n">disable</span><span class="o">=</span><span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">show_progress_bar</span><span class="p">)):</span>
            <span class="c1"># get the start/end/rev_comp values</span>
            <span class="n">start</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;start&quot;</span><span class="p">]</span>
            <span class="n">end</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;end&quot;</span><span class="p">]</span>
            <span class="n">rev_comp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;reverse_complement&quot;</span><span class="p">]</span>

            <span class="c1"># determine the boolean mask for this gene</span>
            <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">&gt;=</span> <span class="n">start</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">&lt;</span> <span class="n">end</span><span class="p">)</span>

            <span class="c1"># fit the gene into the stacked arrays</span>
            <span class="n">genes_mask</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__fit_gene</span><span class="p">(</span>
                <span class="n">mask</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">genes_mask</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">rev_comp</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene_rows</span><span class="p">[</span><span class="n">gene_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span>

        <span class="c1"># Singular array to determine if there are genes in places within the genome</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genes_mask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">genes_mask</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># Set instance variable for the gene names</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">=</span> <span class="n">genes</span>

    <span class="k">def</span> <span class="nf">__setup_arrays</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Private function to initalise all of the required arrays, fitting the gene</span>
<span class="sd">            names into the correct places within stacked arrays</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">__find_overlaps</span><span class="p">()</span>

        <span class="c1"># do as many assignments outside the loop, i.e. in one go to improve performance</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_cds</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">n_rows</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stacked_gene_name</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">,</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_rows</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_sequence</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">,</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_rows</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="c1"># Use a list to track minority populations</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">minor_populations</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Track which nucleotides are deleted</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_deleted</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="kc">False</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">])</span>

        <span class="c1"># Use a dict to track VCF evidence as req</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidence</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">__assign_promoter_regions</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Private function to assign promoter regions to genes</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">max_promoter_length</span><span class="p">,</span> <span class="nb">int</span>
        <span class="p">),</span> <span class="s2">&quot;max_promoter_length must be an integer!&quot;</span>

        <span class="k">assert</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">max_promoter_length</span> <span class="o">&gt;</span> <span class="mi">0</span>
        <span class="p">),</span> <span class="s2">&quot;max_promoter_length must be greater than zero&quot;</span>

        <span class="c1"># labelling promoters is a difficult problem since</span>
        <span class="c1">#  (i)  it is arbitrary and</span>
        <span class="c1">#  (ii) we need to ensure that only unassigned bases can be labelled as</span>
        <span class="c1">#   promoters and each should only &#39;belong&#39; to a single feature</span>
        <span class="c1"># the latter is especially difficult when you have two genes next to one</span>
        <span class="c1">#   another, one reverse complement, since their promoters can &#39;fight&#39; for</span>
        <span class="c1">#   space. It is this problem that means we have to grow each promoter out</span>
        <span class="c1">#   one base at a time</span>

        <span class="c1"># populate a dictionary to store the starts/ends of genes as they grow</span>
        <span class="c1">#   with promoters</span>
        <span class="n">start_end</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">gene_name</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;start&quot;</span><span class="p">],</span>
                <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;end&quot;</span><span class="p">],</span>
            <span class="p">}</span>
            <span class="k">for</span> <span class="n">gene_name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span>
        <span class="p">}</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Assigning promoters...&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">promoter</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span>
            <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_promoter_length</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">disable</span><span class="o">=</span><span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">show_progress_bar</span><span class="p">)</span>
        <span class="p">):</span>
            <span class="c1"># Replacement `start_end` because dictionaries can&#39;t be changed during</span>
            <span class="c1">#   iteration</span>
            <span class="n">new_start_end</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">gene_name</span> <span class="ow">in</span> <span class="n">start_end</span><span class="p">:</span>
                <span class="c1"># Get the associated start/end</span>
                <span class="n">start</span> <span class="o">=</span> <span class="n">start_end</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;start&quot;</span><span class="p">]</span>
                <span class="n">end</span> <span class="o">=</span> <span class="n">start_end</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;end&quot;</span><span class="p">]</span>
                <span class="n">rev_comp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;reverse_complement&quot;</span><span class="p">]</span>
                <span class="c1"># Check if the region which the gene would grow into is empty</span>
                <span class="k">if</span> <span class="n">rev_comp</span><span class="p">:</span>
                    <span class="c1"># Indexing is weird so stacked_array[i][end-2] is the end of</span>
                    <span class="c1">#   the gene making stacked_array[i][end-1] the next item on</span>
                    <span class="c1">#   the right</span>
                    <span class="k">if</span> <span class="n">end</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">):</span>
                        <span class="c1"># If the end would be out of range, loop back around to</span>
                        <span class="c1">#   position 0</span>
                        <span class="n">end</span> <span class="o">=</span> <span class="mi">0</span>
                    <span class="n">pos</span> <span class="o">=</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Similar indexing issue except indexing starts on start-1</span>
                    <span class="c1">#   so start-2 is the next item on the left</span>
                    <span class="n">pos</span> <span class="o">=</span> <span class="n">start</span> <span class="o">-</span> <span class="mi">2</span>
                <span class="c1"># Disable ruff for this line as `is True` doesn&#39;t work here</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_mask</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>  <span class="c1"># noqa: E712</span>
                    <span class="c1"># There is a gene here already so skip it</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># This position is free so set the appropriate values</span>
                    <span class="n">new_start_end</span><span class="p">[</span><span class="n">gene_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">start_end</span><span class="p">[</span>
                        <span class="n">gene_name</span>
                    <span class="p">]</span>  <span class="c1"># Retain gene for future expansions</span>
                    <span class="c1"># Get the row index for stacked arrays</span>
                    <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_rows</span><span class="p">[</span><span class="n">gene_name</span><span class="p">]</span>

                    <span class="c1"># Set appropriate values</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">stacked_gene_name</span><span class="p">[</span><span class="n">row</span><span class="p">][</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_name</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span><span class="n">row</span><span class="p">][</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="n">promoter</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_reverse_complement</span><span class="p">[</span><span class="n">row</span><span class="p">][</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">rev_comp</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_promoter</span><span class="p">[</span><span class="n">row</span><span class="p">][</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">genes_mask</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>

                    <span class="c1"># move the start/end values appropriately</span>
                    <span class="k">if</span> <span class="n">rev_comp</span><span class="p">:</span>
                        <span class="n">new_start_end</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;end&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">end</span> <span class="o">+</span> <span class="mi">1</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">new_start_end</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;start&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">start</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="n">start_end</span> <span class="o">=</span> <span class="n">new_start_end</span>

    <span class="k">def</span> <span class="nf">__insert_newlines</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">string</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">every</span><span class="o">=</span><span class="mi">70</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Simple private method for inserting a carriage return every N characters into</span>
<span class="sd">            a long string.</span>

<span class="sd">        Args:</span>
<span class="sd">            string (str): the string to insert carriage returns</span>
<span class="sd">            every (int): how many characters between each carriage return</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: Same string with &quot;\n&quot; characters inserted</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">assert</span> <span class="n">every</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;every must be an integer greater than zero&quot;</span>

        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">string</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;string is too short!&quot;</span>

        <span class="k">return</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">string</span><span class="p">[</span><span class="n">i</span> <span class="p">:</span> <span class="n">i</span> <span class="o">+</span> <span class="n">every</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">string</span><span class="p">),</span> <span class="n">every</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">build_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Gene</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Public function to build the gumpy.Gene object</span>

<span class="sd">        Args:</span>
<span class="sd">            gene (str) : The name of the gene</span>

<span class="sd">        Returns:</span>
<span class="sd">            gumpy.Gene : The instanciated gene object.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># The mask for all stacked arrays (N-dim)</span>
        <span class="n">stacked_mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>

        <span class="c1"># The mask for singular arrays (1-dim) by collapsing stacked mask to 1-dim</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">stacked_mask</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># check that the genome does contain the gene name</span>
        <span class="k">assert</span> <span class="n">numpy</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">mask</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;gene (&quot;</span> <span class="o">+</span> <span class="n">gene</span> <span class="o">+</span> <span class="s2">&quot;) not found in genome!&quot;</span>

        <span class="c1"># Revert is_cds to all False if this gene is non-coding</span>
        <span class="c1"># Easier to do it like this than to mess with masking</span>
        <span class="n">nucleotide_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]:</span>
            <span class="n">is_cd</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="kc">False</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_seq</span><span class="p">))])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">is_cd</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_cds</span><span class="p">[</span><span class="n">stacked_mask</span><span class="p">]</span>

        <span class="n">gene_nucleotides</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
        <span class="n">gene_minor_populations</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">population</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">minor_populations</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">population</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">gene_nucleotides</span><span class="p">:</span>
                <span class="c1"># This minor population is within the gene</span>
                <span class="n">gene_minor_populations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">population</span><span class="p">)</span>

        <span class="n">vcf_evidence</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">idx</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">gene_nucleotides</span>
        <span class="p">}</span>

        <span class="c1"># instantiate a Gene object</span>
        <span class="n">g</span> <span class="o">=</span> <span class="n">Gene</span><span class="p">(</span>
            <span class="n">gene</span><span class="p">,</span>
            <span class="n">nucleotide_seq</span><span class="p">,</span>
            <span class="n">gene_nucleotides</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span><span class="n">stacked_mask</span><span class="p">],</span>
            <span class="n">is_cd</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_promoter</span><span class="p">[</span><span class="n">stacked_mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;reverse_complement&quot;</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;type&quot;</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;ribosomal_shifts&quot;</span><span class="p">],</span>
            <span class="n">gene_minor_populations</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="n">vcf_evidence</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">g</span>

    <span class="k">def</span> <span class="nf">__assign_deleted</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Assign a boolean array of which nucleotides are deleted by indels.</span>
<span class="sd">        This holds the same 1-1 relationship as nucleotide_index &lt;-&gt; nucleotide_sequence</span>

<span class="sd">        Args:</span>
<span class="sd">            genome (gumpy.Genome): Genome to apply this to</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">marking</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">current_evidence</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">length</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">genome</span><span class="o">.</span><span class="n">indel_length</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">length</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># We found a deletion, so mark the next N bases as deleted</span>
                <span class="c1"># Deletions are -N here</span>
                <span class="n">marking</span> <span class="o">-=</span> <span class="n">length</span>
                <span class="n">current_evidence</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">[</span><span class="n">genome</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">idx</span><span class="p">]]</span>

            <span class="k">if</span> <span class="n">marking</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># We mark this as deleted</span>
                <span class="n">genome</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="c1"># Update evidence too so we can pull out VCF rows from downstream genes</span>
                <span class="n">genome</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">[</span><span class="n">genome</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">idx</span><span class="p">]]</span> <span class="o">=</span> <span class="n">current_evidence</span>
                <span class="n">marking</span> <span class="o">-=</span> <span class="mi">1</span>

    <span class="k">def</span> <span class="fm">__add__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">vcf</span><span class="p">:</span> <span class="n">VCFFile</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Function to apply a VCF file to the genome  - producing a replica genome</span>
<span class="sd">             the specified changes</span>

<span class="sd">        Args:</span>
<span class="sd">            vcf (gumpy.VCFFile): The VCFFile object for the VCF</span>

<span class="sd">        Returns:</span>
<span class="sd">            gumpy.Genome: The resulting Genome object</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">vcf</span><span class="p">,</span> <span class="n">VCFFile</span><span class="p">),</span> <span class="s2">&quot;RHS must be a gumpy.VCFFile object!&quot;</span>

        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span>
            <span class="n">vcf</span><span class="o">.</span><span class="n">calls</span><span class="p">,</span> <span class="nb">dict</span>
        <span class="p">),</span> <span class="s2">&quot;something wrong with the gumpy.VCFFile object!&quot;</span>

        <span class="n">indices</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">vcf</span><span class="o">.</span><span class="n">calls</span><span class="o">.</span><span class="n">keys</span><span class="p">()]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># It&#39;s possible that a VCF only details minority populations, and max()</span>
            <span class="c1">#   complains for empty</span>
            <span class="k">assert</span> <span class="p">(</span>
                <span class="nb">max</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span>
            <span class="p">),</span> <span class="s2">&quot;The VCF file details changes outside of this genome!&quot;</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">minor_populations</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">vcf</span><span class="o">.</span><span class="n">minor_population_indices</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># Both this genome and the VCF have minor populations so for simplicity</span>
            <span class="c1">#   atm, complain</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span>
                <span class="s2">&quot;Both the existing Genome and the VCF have minor populations!&quot;</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Copying the genome...&quot;</span><span class="p">)</span>

        <span class="c1"># Replicate this Genome object</span>
        <span class="n">genome</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Using numpy&#39;s fancy array indexing may provide neat code, and provides some </span>
<span class="sd">            speed in some cases, the constant time access of a standard dictionary </span>
<span class="sd">            results in faster code when the mask only contains a few True values.</span>

<span class="sd">        For TB length arrays with a ~0.1% True mask, applying a mask takes ~10^-3s</span>
<span class="sd">            Applying a dictionary to the same array takes ~10^-4s</span>

<span class="sd">        ~0.1% True mask is a reasonable amount for this task as a VCF file </span>
<span class="sd">            is ~4000 entries</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Updating the genome...&quot;</span><span class="p">)</span>

        <span class="c1"># use the calls dict to change the nucleotide indicies in the copy of the genome</span>
        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">type_</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">vcf</span><span class="o">.</span><span class="n">calls</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">disable</span><span class="o">=</span><span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">show_progress_bar</span><span class="p">)):</span>
            <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">vcf</span><span class="o">.</span><span class="n">calls</span><span class="p">[(</span><span class="n">idx</span><span class="p">,</span> <span class="n">type_</span><span class="p">)]:</span>
                <span class="n">genome</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">[</span><span class="n">genome</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]]</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span>
                    <span class="s2">&quot;original_vcf_row&quot;</span>
                <span class="p">]</span>

                <span class="c1"># deal with changes at a single nucleotide site</span>
                <span class="k">if</span> <span class="n">type_</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;snp&quot;</span><span class="p">,</span> <span class="s2">&quot;null&quot;</span><span class="p">,</span> <span class="s2">&quot;het&quot;</span><span class="p">]:</span>
                    <span class="c1"># only set values if the idx is to a single nucleotide</span>
                    <span class="n">genome</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;call&quot;</span><span class="p">]</span>

                <span class="c1"># deal with insertions and deletions</span>
                <span class="k">elif</span> <span class="n">type_</span> <span class="o">==</span> <span class="s2">&quot;indel&quot;</span><span class="p">:</span>
                    <span class="n">genome</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="n">genome</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;call&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>

                    <span class="k">if</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;call&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;ins&quot;</span><span class="p">:</span>
                        <span class="n">genome</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;call&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">genome</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;call&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span>

                <span class="k">elif</span> <span class="n">type_</span> <span class="o">==</span> <span class="s2">&quot;ref&quot;</span><span class="p">:</span>
                    <span class="c1"># These only exist due to reference calls</span>
                    <span class="c1"># They only made it this far as they are required to pull</span>
                    <span class="c1"># out minors at these positions</span>
                    <span class="k">pass</span>

        <span class="n">genome</span><span class="o">.</span><span class="n">minor_populations</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">minor</span> <span class="ow">in</span> <span class="n">vcf</span><span class="o">.</span><span class="n">minor_populations</span><span class="p">:</span>
            <span class="c1"># Ensure that the VCF evidence for these is also recorded correctly</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="n">minor</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">evidence</span> <span class="o">=</span> <span class="n">minor</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>

            <span class="c1"># Store the VCF evidence of this</span>
            <span class="n">genome</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">[</span><span class="n">genome</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">pos</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]]</span> <span class="o">=</span> <span class="n">evidence</span>

            <span class="c1"># Don&#39;t keep the VCF evidence with this, as it is already stored in the</span>
            <span class="c1">#   main vcf_evidence dict</span>
            <span class="n">genome</span><span class="o">.</span><span class="n">minor_populations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">minor</span><span class="p">[:</span><span class="mi">5</span><span class="p">])</span>

        <span class="n">genome</span><span class="o">.</span><span class="n">vcf_file</span> <span class="o">=</span> <span class="n">vcf</span>

        <span class="c1"># Let&#39;s assign some deleted regions (if exist)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">__assign_deleted</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>

        <span class="c1"># the genome has been altered so not a reference genome</span>
        <span class="n">genome</span><span class="o">.</span><span class="n">is_reference</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="k">return</span> <span class="n">genome</span>

    <span class="k">def</span> <span class="nf">minority_populations_GARC</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">interpretation</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;reads&quot;</span><span class="p">,</span> <span class="n">reference</span><span class="o">=</span><span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get the variants in GARC of the minority populations for this genome.</span>
<span class="sd">        Whether the variants are given in terms of reads or read percentage is</span>
<span class="sd">            controlled by `interpretation`</span>

<span class="sd">        Args:</span>
<span class="sd">            interpretation (str, optional): Which interpretation to use. `reads` for</span>
<span class="sd">                number of reads for this population. `percentage` for the decimal</span>
<span class="sd">                percentage of total reads for this population. Defaults to &#39;reads&#39;.</span>
<span class="sd">            reference (gumpy.Genome, optional): The reference to denote mutations from.</span>
<span class="sd">                Defaults to self</span>
<span class="sd">        Returns:</span>
<span class="sd">            List[str]: List of the variants in GARC</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Use the interpretation type to pull out which index of the minor_populations</span>
        <span class="c1"># Each item of minor_populations is (pos, type, bases, abs_coverage,</span>
        <span class="c1">#   percent_coverage)</span>
        <span class="k">if</span> <span class="n">interpretation</span> <span class="o">==</span> <span class="s2">&quot;percentage&quot;</span><span class="p">:</span>
            <span class="n">coverage</span> <span class="o">=</span> <span class="mi">4</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">coverage</span> <span class="o">=</span> <span class="mi">3</span>

        <span class="k">if</span> <span class="n">reference</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">reference</span> <span class="o">=</span> <span class="bp">self</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Ensure that only one of the two Genomes has minor populations</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">reference</span><span class="o">.</span><span class="n">minor_populations</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">,</span> <span class="p">(</span>
                <span class="s2">&quot;Minority populations can only be compared when 1 Genome &quot;</span>
                <span class="s2">&quot;does not have them!&quot;</span>
            <span class="p">)</span>

        <span class="n">variants</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">minor</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">minor_populations</span><span class="p">:</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="n">minor</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">type_</span> <span class="o">=</span> <span class="n">minor</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">bases</span> <span class="o">=</span> <span class="n">minor</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
            <span class="n">depth</span> <span class="o">=</span> <span class="n">minor</span><span class="p">[</span><span class="n">coverage</span><span class="p">]</span>

            <span class="k">if</span> <span class="n">type_</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;ref&quot;</span><span class="p">,</span> <span class="s2">&quot;snp&quot;</span><span class="p">]:</span>
                <span class="c1"># These are functionally the same</span>
                <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">alt</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">bases</span><span class="p">)):</span>
                    <span class="n">ref</span> <span class="o">=</span> <span class="n">reference</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span>
                        <span class="n">reference</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">==</span> <span class="n">pos</span> <span class="o">+</span> <span class="n">i</span>
                    <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pos</span><span class="o">+</span><span class="n">i</span><span class="si">}{</span><span class="n">ref</span><span class="si">}</span><span class="s2">&gt;</span><span class="si">{</span><span class="n">alt</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">depth</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Indels are the same too</span>
                <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pos</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">type_</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">bases</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">depth</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="genome.Genome.__add__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__add__</span><span class="p">(</span><span class="n">vcf</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Function to apply a VCF file to the genome  - producing a replica genome
     the specified changes</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>vcf</code></td>
            <td>
                  <code><span title="gumpy.VCFFile">VCFFile</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The VCFFile object for the VCF</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>gumpy.Genome: The resulting Genome object</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 947</span>
<span class="normal"> 948</span>
<span class="normal"> 949</span>
<span class="normal"> 950</span>
<span class="normal"> 951</span>
<span class="normal"> 952</span>
<span class="normal"> 953</span>
<span class="normal"> 954</span>
<span class="normal"> 955</span>
<span class="normal"> 956</span>
<span class="normal"> 957</span>
<span class="normal"> 958</span>
<span class="normal"> 959</span>
<span class="normal"> 960</span>
<span class="normal"> 961</span>
<span class="normal"> 962</span>
<span class="normal"> 963</span>
<span class="normal"> 964</span>
<span class="normal"> 965</span>
<span class="normal"> 966</span>
<span class="normal"> 967</span>
<span class="normal"> 968</span>
<span class="normal"> 969</span>
<span class="normal"> 970</span>
<span class="normal"> 971</span>
<span class="normal"> 972</span>
<span class="normal"> 973</span>
<span class="normal"> 974</span>
<span class="normal"> 975</span>
<span class="normal"> 976</span>
<span class="normal"> 977</span>
<span class="normal"> 978</span>
<span class="normal"> 979</span>
<span class="normal"> 980</span>
<span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span>
<span class="normal">1022</span>
<span class="normal">1023</span>
<span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__add__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">vcf</span><span class="p">:</span> <span class="n">VCFFile</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Function to apply a VCF file to the genome  - producing a replica genome</span>
<span class="sd">         the specified changes</span>

<span class="sd">    Args:</span>
<span class="sd">        vcf (gumpy.VCFFile): The VCFFile object for the VCF</span>

<span class="sd">    Returns:</span>
<span class="sd">        gumpy.Genome: The resulting Genome object</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">vcf</span><span class="p">,</span> <span class="n">VCFFile</span><span class="p">),</span> <span class="s2">&quot;RHS must be a gumpy.VCFFile object!&quot;</span>

    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span>
        <span class="n">vcf</span><span class="o">.</span><span class="n">calls</span><span class="p">,</span> <span class="nb">dict</span>
    <span class="p">),</span> <span class="s2">&quot;something wrong with the gumpy.VCFFile object!&quot;</span>

    <span class="n">indices</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">vcf</span><span class="o">.</span><span class="n">calls</span><span class="o">.</span><span class="n">keys</span><span class="p">()]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># It&#39;s possible that a VCF only details minority populations, and max()</span>
        <span class="c1">#   complains for empty</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="nb">max</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span>
        <span class="p">),</span> <span class="s2">&quot;The VCF file details changes outside of this genome!&quot;</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">minor_populations</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">vcf</span><span class="o">.</span><span class="n">minor_population_indices</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Both this genome and the VCF have minor populations so for simplicity</span>
        <span class="c1">#   atm, complain</span>
        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span>
            <span class="s2">&quot;Both the existing Genome and the VCF have minor populations!&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Copying the genome...&quot;</span><span class="p">)</span>

    <span class="c1"># Replicate this Genome object</span>
    <span class="n">genome</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Using numpy&#39;s fancy array indexing may provide neat code, and provides some </span>
<span class="sd">        speed in some cases, the constant time access of a standard dictionary </span>
<span class="sd">        results in faster code when the mask only contains a few True values.</span>

<span class="sd">    For TB length arrays with a ~0.1% True mask, applying a mask takes ~10^-3s</span>
<span class="sd">        Applying a dictionary to the same array takes ~10^-4s</span>

<span class="sd">    ~0.1% True mask is a reasonable amount for this task as a VCF file </span>
<span class="sd">        is ~4000 entries</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Updating the genome...&quot;</span><span class="p">)</span>

    <span class="c1"># use the calls dict to change the nucleotide indicies in the copy of the genome</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">type_</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">vcf</span><span class="o">.</span><span class="n">calls</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">disable</span><span class="o">=</span><span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">show_progress_bar</span><span class="p">)):</span>
        <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">vcf</span><span class="o">.</span><span class="n">calls</span><span class="p">[(</span><span class="n">idx</span><span class="p">,</span> <span class="n">type_</span><span class="p">)]:</span>
            <span class="n">genome</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">[</span><span class="n">genome</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]]</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span>
                <span class="s2">&quot;original_vcf_row&quot;</span>
            <span class="p">]</span>

            <span class="c1"># deal with changes at a single nucleotide site</span>
            <span class="k">if</span> <span class="n">type_</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;snp&quot;</span><span class="p">,</span> <span class="s2">&quot;null&quot;</span><span class="p">,</span> <span class="s2">&quot;het&quot;</span><span class="p">]:</span>
                <span class="c1"># only set values if the idx is to a single nucleotide</span>
                <span class="n">genome</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;call&quot;</span><span class="p">]</span>

            <span class="c1"># deal with insertions and deletions</span>
            <span class="k">elif</span> <span class="n">type_</span> <span class="o">==</span> <span class="s2">&quot;indel&quot;</span><span class="p">:</span>
                <span class="n">genome</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="n">genome</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;call&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>

                <span class="k">if</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;call&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;ins&quot;</span><span class="p">:</span>
                    <span class="n">genome</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;call&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">genome</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;call&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span>

            <span class="k">elif</span> <span class="n">type_</span> <span class="o">==</span> <span class="s2">&quot;ref&quot;</span><span class="p">:</span>
                <span class="c1"># These only exist due to reference calls</span>
                <span class="c1"># They only made it this far as they are required to pull</span>
                <span class="c1"># out minors at these positions</span>
                <span class="k">pass</span>

    <span class="n">genome</span><span class="o">.</span><span class="n">minor_populations</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">minor</span> <span class="ow">in</span> <span class="n">vcf</span><span class="o">.</span><span class="n">minor_populations</span><span class="p">:</span>
        <span class="c1"># Ensure that the VCF evidence for these is also recorded correctly</span>
        <span class="n">pos</span> <span class="o">=</span> <span class="n">minor</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">evidence</span> <span class="o">=</span> <span class="n">minor</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>

        <span class="c1"># Store the VCF evidence of this</span>
        <span class="n">genome</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">[</span><span class="n">genome</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">pos</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]]</span> <span class="o">=</span> <span class="n">evidence</span>

        <span class="c1"># Don&#39;t keep the VCF evidence with this, as it is already stored in the</span>
        <span class="c1">#   main vcf_evidence dict</span>
        <span class="n">genome</span><span class="o">.</span><span class="n">minor_populations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">minor</span><span class="p">[:</span><span class="mi">5</span><span class="p">])</span>

    <span class="n">genome</span><span class="o">.</span><span class="n">vcf_file</span> <span class="o">=</span> <span class="n">vcf</span>

    <span class="c1"># Let&#39;s assign some deleted regions (if exist)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">__assign_deleted</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>

    <span class="c1"># the genome has been altered so not a reference genome</span>
    <span class="n">genome</span><span class="o">.</span><span class="n">is_reference</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">return</span> <span class="n">genome</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.__add_empty_row" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__add_empty_row</span><span class="p">(</span><span class="n">array</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Private function to add an empty row of the correct type to a numpy array
Args:
    array (numpy.ndarray) : Array to add an empty row to
Returns:
    (numpy.ndarray): The same array with an empty row of the same length and
        dtype appended</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__add_empty_row</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">array</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Private function to add an empty row of the correct type to a numpy array</span>
<span class="sd">    Args:</span>
<span class="sd">        array (numpy.ndarray) : Array to add an empty row to</span>
<span class="sd">    Returns:</span>
<span class="sd">        (numpy.ndarray): The same array with an empty row of the same length and</span>
<span class="sd">            dtype appended</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">empty_row</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="n">array</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">array</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">vstack</span><span class="p">((</span><span class="n">array</span><span class="p">,</span> <span class="n">empty_row</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.__assign_deleted" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__assign_deleted</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Assign a boolean array of which nucleotides are deleted by indels.
This holds the same 1-1 relationship as nucleotide_index &lt;-&gt; nucleotide_sequence</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>genome</code></td>
            <td>
                  <code><span title="gumpy.Genome">Genome</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Genome to apply this to</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">924</span>
<span class="normal">925</span>
<span class="normal">926</span>
<span class="normal">927</span>
<span class="normal">928</span>
<span class="normal">929</span>
<span class="normal">930</span>
<span class="normal">931</span>
<span class="normal">932</span>
<span class="normal">933</span>
<span class="normal">934</span>
<span class="normal">935</span>
<span class="normal">936</span>
<span class="normal">937</span>
<span class="normal">938</span>
<span class="normal">939</span>
<span class="normal">940</span>
<span class="normal">941</span>
<span class="normal">942</span>
<span class="normal">943</span>
<span class="normal">944</span>
<span class="normal">945</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__assign_deleted</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Assign a boolean array of which nucleotides are deleted by indels.</span>
<span class="sd">    This holds the same 1-1 relationship as nucleotide_index &lt;-&gt; nucleotide_sequence</span>

<span class="sd">    Args:</span>
<span class="sd">        genome (gumpy.Genome): Genome to apply this to</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">marking</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">current_evidence</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">length</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">genome</span><span class="o">.</span><span class="n">indel_length</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">length</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># We found a deletion, so mark the next N bases as deleted</span>
            <span class="c1"># Deletions are -N here</span>
            <span class="n">marking</span> <span class="o">-=</span> <span class="n">length</span>
            <span class="n">current_evidence</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">[</span><span class="n">genome</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">idx</span><span class="p">]]</span>

        <span class="k">if</span> <span class="n">marking</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># We mark this as deleted</span>
            <span class="n">genome</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="c1"># Update evidence too so we can pull out VCF rows from downstream genes</span>
            <span class="n">genome</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">[</span><span class="n">genome</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">idx</span><span class="p">]]</span> <span class="o">=</span> <span class="n">current_evidence</span>
            <span class="n">marking</span> <span class="o">-=</span> <span class="mi">1</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.__assign_promoter_regions" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__assign_promoter_regions</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Private function to assign promoter regions to genes</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span>
<span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__assign_promoter_regions</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Private function to assign promoter regions to genes</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_promoter_length</span><span class="p">,</span> <span class="nb">int</span>
    <span class="p">),</span> <span class="s2">&quot;max_promoter_length must be an integer!&quot;</span>

    <span class="k">assert</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_promoter_length</span> <span class="o">&gt;</span> <span class="mi">0</span>
    <span class="p">),</span> <span class="s2">&quot;max_promoter_length must be greater than zero&quot;</span>

    <span class="c1"># labelling promoters is a difficult problem since</span>
    <span class="c1">#  (i)  it is arbitrary and</span>
    <span class="c1">#  (ii) we need to ensure that only unassigned bases can be labelled as</span>
    <span class="c1">#   promoters and each should only &#39;belong&#39; to a single feature</span>
    <span class="c1"># the latter is especially difficult when you have two genes next to one</span>
    <span class="c1">#   another, one reverse complement, since their promoters can &#39;fight&#39; for</span>
    <span class="c1">#   space. It is this problem that means we have to grow each promoter out</span>
    <span class="c1">#   one base at a time</span>

    <span class="c1"># populate a dictionary to store the starts/ends of genes as they grow</span>
    <span class="c1">#   with promoters</span>
    <span class="n">start_end</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">gene_name</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;start&quot;</span><span class="p">],</span>
            <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;end&quot;</span><span class="p">],</span>
        <span class="p">}</span>
        <span class="k">for</span> <span class="n">gene_name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span>
    <span class="p">}</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Assigning promoters...&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">promoter</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span>
        <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_promoter_length</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">disable</span><span class="o">=</span><span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">show_progress_bar</span><span class="p">)</span>
    <span class="p">):</span>
        <span class="c1"># Replacement `start_end` because dictionaries can&#39;t be changed during</span>
        <span class="c1">#   iteration</span>
        <span class="n">new_start_end</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">gene_name</span> <span class="ow">in</span> <span class="n">start_end</span><span class="p">:</span>
            <span class="c1"># Get the associated start/end</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">start_end</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;start&quot;</span><span class="p">]</span>
            <span class="n">end</span> <span class="o">=</span> <span class="n">start_end</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;end&quot;</span><span class="p">]</span>
            <span class="n">rev_comp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;reverse_complement&quot;</span><span class="p">]</span>
            <span class="c1"># Check if the region which the gene would grow into is empty</span>
            <span class="k">if</span> <span class="n">rev_comp</span><span class="p">:</span>
                <span class="c1"># Indexing is weird so stacked_array[i][end-2] is the end of</span>
                <span class="c1">#   the gene making stacked_array[i][end-1] the next item on</span>
                <span class="c1">#   the right</span>
                <span class="k">if</span> <span class="n">end</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">):</span>
                    <span class="c1"># If the end would be out of range, loop back around to</span>
                    <span class="c1">#   position 0</span>
                    <span class="n">end</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">pos</span> <span class="o">=</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Similar indexing issue except indexing starts on start-1</span>
                <span class="c1">#   so start-2 is the next item on the left</span>
                <span class="n">pos</span> <span class="o">=</span> <span class="n">start</span> <span class="o">-</span> <span class="mi">2</span>
            <span class="c1"># Disable ruff for this line as `is True` doesn&#39;t work here</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_mask</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>  <span class="c1"># noqa: E712</span>
                <span class="c1"># There is a gene here already so skip it</span>
                <span class="k">continue</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># This position is free so set the appropriate values</span>
                <span class="n">new_start_end</span><span class="p">[</span><span class="n">gene_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">start_end</span><span class="p">[</span>
                    <span class="n">gene_name</span>
                <span class="p">]</span>  <span class="c1"># Retain gene for future expansions</span>
                <span class="c1"># Get the row index for stacked arrays</span>
                <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_rows</span><span class="p">[</span><span class="n">gene_name</span><span class="p">]</span>

                <span class="c1"># Set appropriate values</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">stacked_gene_name</span><span class="p">[</span><span class="n">row</span><span class="p">][</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_name</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span><span class="n">row</span><span class="p">][</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="n">promoter</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_reverse_complement</span><span class="p">[</span><span class="n">row</span><span class="p">][</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">rev_comp</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_promoter</span><span class="p">[</span><span class="n">row</span><span class="p">][</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genes_mask</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>

                <span class="c1"># move the start/end values appropriately</span>
                <span class="k">if</span> <span class="n">rev_comp</span><span class="p">:</span>
                    <span class="n">new_start_end</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;end&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">end</span> <span class="o">+</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">new_start_end</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;start&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">start</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="n">start_end</span> <span class="o">=</span> <span class="n">new_start_end</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.__build_genome_variable_length_string" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__build_genome_variable_length_string</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Build a string of the genome sequence, including indels - resulting in a
    variable length genome</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>indices</code></td>
            <td>
                  <code><span title="typing.List">List</span>[int]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of the indices of indels</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Genome sequence as a string</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__build_genome_variable_length_string</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">indices</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Build a string of the genome sequence, including indels - resulting in a</span>
<span class="sd">        variable length genome</span>

<span class="sd">    Args:</span>
<span class="sd">        indices (List[int]): List of the indices of indels</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Genome sequence as a string</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genome_string</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="c1"># work backwards as easier to deal with insertions/deletions when you&#39;ve</span>
    <span class="c1">#   already gone past them</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">indices</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">==</span> <span class="n">i</span>
        <span class="n">base</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">genome_string</span> <span class="o">=</span> <span class="n">base</span> <span class="o">+</span> <span class="n">genome_string</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[</span><span class="n">mask</span><span class="p">]:</span>
            <span class="n">indel_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">mask</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">indel_length</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">genome_string</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">mask</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">genome_string</span>
            <span class="k">elif</span> <span class="n">indel_length</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">genome_string</span> <span class="o">=</span> <span class="n">genome_string</span><span class="p">[</span><span class="nb">abs</span><span class="p">(</span><span class="n">indel_length</span><span class="p">)</span> <span class="p">:]</span>
    <span class="k">return</span> <span class="n">genome_string</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.__convert_references" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__convert_references</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Convert BIOPython Reference objects to normal dictionaries. They do not
appear to have any greater application than storing structured data, so
removing the object wrappers appears to be a clean way to combat the object's
issues with serialization.</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__convert_references</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert BIOPython Reference objects to normal dictionaries. They do not</span>
<span class="sd">    appear to have any greater application than storing structured data, so</span>
<span class="sd">    removing the object wrappers appears to be a clean way to combat the object&#39;s</span>
<span class="sd">    issues with serialization.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">reference</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;references&quot;</span><span class="p">]):</span>
        <span class="n">new_ref</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">vars</span><span class="p">(</span><span class="n">reference</span><span class="p">):</span>
            <span class="c1"># This key contains unhelpfully structured data</span>
            <span class="k">if</span> <span class="n">key</span> <span class="o">==</span> <span class="s2">&quot;location&quot;</span><span class="p">:</span>
                <span class="n">new_loc</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">vars</span><span class="p">(</span><span class="n">reference</span><span class="p">)[</span><span class="n">key</span><span class="p">]:</span>
                    <span class="n">loc</span> <span class="o">=</span> <span class="p">{}</span>
                    <span class="k">for</span> <span class="n">item_key</span> <span class="ow">in</span> <span class="nb">vars</span><span class="p">(</span><span class="n">item</span><span class="p">):</span>
                        <span class="k">if</span> <span class="n">item_key</span> <span class="o">==</span> <span class="s2">&quot;_start&quot;</span> <span class="ow">or</span> <span class="n">item_key</span> <span class="o">==</span> <span class="s2">&quot;_end&quot;</span><span class="p">:</span>
                            <span class="c1"># These are the only ones we care about</span>
                            <span class="n">loc</span><span class="p">[</span><span class="n">item_key</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="n">item</span><span class="p">,</span> <span class="n">item_key</span><span class="p">))</span>
                    <span class="n">new_loc</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span>
                <span class="n">new_ref</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_loc</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">new_ref</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="nb">vars</span><span class="p">(</span><span class="n">reference</span><span class="p">)[</span><span class="n">key</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;references&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_ref</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.__eq__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__eq__</span><span class="p">(</span><span class="n">other</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Overloading the equality operator so two Genome objects can be compared
directly. Checks for the equality based on fields, but does not check
for filename equality</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>other</code></td>
            <td>
                  <code>gumpy.Genome) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The other Genome object to compare to</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>bool</code></td>            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Boolean showing equality of the objects</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Overloading the equality operator so two Genome objects can be compared</span>
<span class="sd">    directly. Checks for the equality based on fields, but does not check</span>
<span class="sd">    for filename equality</span>

<span class="sd">    Args:</span>
<span class="sd">        other (gumpy.Genome) : The other Genome object to compare to</span>

<span class="sd">    Returns:</span>
<span class="sd">        bool : Boolean showing equality of the objects</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">Genome</span><span class="p">)</span>

    <span class="n">check</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bool_</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bool_</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">name</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">id</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">description</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">description</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">nucleotide_sequence</span>
    <span class="p">)</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">)</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bool_</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">length</span><span class="p">)</span>
    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_gene_name</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">stacked_gene_name</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="nb">bool</span><span class="p">(</span><span class="n">check</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.__find_overlaps" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__find_overlaps</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Private function to find the sections of the genome in which there are
    overlapping genes. This should be more efficient than the older version
    as it avoids consistent genome iteration. Use of the dot product on boolean
    arrays returns a single boolean showing collisions in almost constant time
    (10^-5 secs for TB size). This can be used to determine which row the gene
    should be in</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__find_overlaps</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Private function to find the sections of the genome in which there are</span>
<span class="sd">        overlapping genes. This should be more efficient than the older version</span>
<span class="sd">        as it avoids consistent genome iteration. Use of the dot product on boolean</span>
<span class="sd">        arrays returns a single boolean showing collisions in almost constant time</span>
<span class="sd">        (10^-5 secs for TB size). This can be used to determine which row the gene</span>
<span class="sd">        should be in</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Boolean mask to show gene presence at indicies (default to all False values)</span>
    <span class="n">genes_mask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="kc">False</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">)])])</span>

    <span class="c1"># gene names</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
        <span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;&quot;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">)])],</span>
        <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;U&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">max_gene_name_length</span><span class="p">),</span>
    <span class="p">)</span>

    <span class="c1"># dict to pull out row indicies for each gene in the stacked arrays</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">gene_rows</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Finding overlaps...&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">gene_name</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">,</span> <span class="n">disable</span><span class="o">=</span><span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">show_progress_bar</span><span class="p">)):</span>
        <span class="c1"># get the start/end/rev_comp values</span>
        <span class="n">start</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;start&quot;</span><span class="p">]</span>
        <span class="n">end</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;end&quot;</span><span class="p">]</span>
        <span class="n">rev_comp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene_name</span><span class="p">][</span><span class="s2">&quot;reverse_complement&quot;</span><span class="p">]</span>

        <span class="c1"># determine the boolean mask for this gene</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">&gt;=</span> <span class="n">start</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">&lt;</span> <span class="n">end</span><span class="p">)</span>

        <span class="c1"># fit the gene into the stacked arrays</span>
        <span class="n">genes_mask</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__fit_gene</span><span class="p">(</span>
            <span class="n">mask</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">genes_mask</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">rev_comp</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_rows</span><span class="p">[</span><span class="n">gene_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span>

    <span class="c1"># Singular array to determine if there are genes in places within the genome</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">genes_mask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">genes_mask</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Set instance variable for the gene names</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">=</span> <span class="n">genes</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.__fit_gene" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__fit_gene</span><span class="p">(</span><span class="n">mask</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">genes_mask</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">rev_comp</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Private function to fit a gene into the genes based on the dot product of
    the masks numpy.dot([bool], [bool])-&gt; bool showing if there are collisions
    of True values within args. This takes 10^-5 seconds which is a significant
    improvement on use of numpy.all() iteration of 10^-2 seconds for TB
    length genome</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>mask</code></td>
            <td>
                  <code>numpy.ndarray) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Boolean array showing positions where the gene lies</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>genes</code></td>
            <td>
                  <code>numpy.ndarray) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>2D numpy array of the format used for all stacked
values</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>genes_mask</code></td>
            <td>
                  <code>numpy.ndarray) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The corresponding boolean mask arrays for the
<code>genes</code> arg</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>start</code></td>
            <td>
                  <code>int) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Start index of the gene</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>end</code></td>
            <td>
                  <code>int) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>End index of the gene</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>gene_name</code></td>
            <td>
                  <code>str) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the gene</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>rev_comp</code></td>
            <td>
                  <code>bool) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Boolean to show whether the gene required a
reverse complement</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(numpy_array) : Updated genes_mask array</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(numpy_array) : Updated genes array</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__fit_gene</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">mask</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">genes</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">genes_mask</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">end</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">gene_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">rev_comp</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Private function to fit a gene into the genes based on the dot product of</span>
<span class="sd">        the masks numpy.dot([bool], [bool])-&gt; bool showing if there are collisions</span>
<span class="sd">        of True values within args. This takes 10^-5 seconds which is a significant</span>
<span class="sd">        improvement on use of numpy.all() iteration of 10^-2 seconds for TB</span>
<span class="sd">        length genome</span>

<span class="sd">    Args:</span>
<span class="sd">        mask (numpy.ndarray) : Boolean array showing positions where the gene lies</span>
<span class="sd">        genes (numpy.ndarray) : 2D numpy array of the format used for all stacked</span>
<span class="sd">            values</span>
<span class="sd">        genes_mask (numpy.ndarray) : The corresponding boolean mask arrays for the</span>
<span class="sd">            `genes` arg</span>
<span class="sd">        start (int) : Start index of the gene</span>
<span class="sd">        end (int) : End index of the gene</span>
<span class="sd">        gene_name (str) : Name of the gene</span>
<span class="sd">        rev_comp (bool) : Boolean to show whether the gene required a</span>
<span class="sd">            reverse complement</span>

<span class="sd">    Returns:</span>
<span class="sd">        (numpy_array) : Updated genes_mask array</span>
<span class="sd">        (numpy_array) : Updated genes array</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">genes_mask</span><span class="p">):</span>
        <span class="c1"># use of the dot product of masks allows determining if a gene will fit</span>
        <span class="c1">#   in an array</span>
        <span class="c1"># Disable ruff here as `is False` doesn&#39;t work here</span>
        <span class="k">if</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">mask</span><span class="p">)</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>  <span class="c1"># noqa: E712</span>
            <span class="c1"># there is no collision with this row so add the row</span>
            <span class="c1"># start/end have to be adjusted to account for 0 indexing of arrays and</span>
            <span class="c1">#   1 indexing of genetics</span>
            <span class="n">row</span><span class="p">[</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">genes</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_name</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">__handle_rev_comp</span><span class="p">(</span><span class="n">rev_comp</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">genes_mask</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">i</span>

    <span class="c1"># if this point is reached, there has been no rows without collisions, so</span>
    <span class="c1">#   add one</span>
    <span class="n">genes_mask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">vstack</span><span class="p">((</span><span class="n">genes_mask</span><span class="p">,</span> <span class="n">mask</span><span class="p">))</span>
    <span class="n">new_row</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">gene_name</span> <span class="k">if</span> <span class="n">m</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">mask</span><span class="p">])</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">vstack</span><span class="p">((</span><span class="n">genes</span><span class="p">,</span> <span class="n">new_row</span><span class="p">))</span>
    <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">__handle_rev_comp</span><span class="p">(</span><span class="n">rev_comp</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">genes_mask</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">i</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.__handle_rev_comp" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__handle_rev_comp</span><span class="p">(</span><span class="n">rev_comp</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Private function to handle the rev-comp changes required
Args:
    rev_comp (bool) : Boolean to show if rev-comp is required
    start (int) : Start index of the gene
    end (int) : End index of the gene
    i (int) : The index of the stacked row</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__handle_rev_comp</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">rev_comp</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">end</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">i</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Private function to handle the rev-comp changes required</span>
<span class="sd">    Args:</span>
<span class="sd">        rev_comp (bool) : Boolean to show if rev-comp is required</span>
<span class="sd">        start (int) : Start index of the gene</span>
<span class="sd">        end (int) : End index of the gene</span>
<span class="sd">        i (int) : The index of the stacked row</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Check if the arrays have the correct number of rows and add as required</span>
    <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">i</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_cds</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__add_empty_row</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_cds</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_reverse_complement</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__add_empty_row</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_reverse_complement</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_promoter</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__add_empty_row</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_promoter</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__add_empty_row</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span>
        <span class="p">)</span>
    <span class="c1"># Update the required items for rev_comp</span>
    <span class="c1"># Use of array slicing here introduces speedups as not all of the array should</span>
    <span class="c1">#   be considered</span>
    <span class="c1"># (Previous method applied a mask which requires full array consideration</span>
    <span class="c1">#   rather than direct access)</span>
    <span class="k">if</span> <span class="n">rev_comp</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mod</span><span class="p">(</span>
            <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">end</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_reverse_complement</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mod</span><span class="p">(</span>
            <span class="mi">1</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">start</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">genbank_file_</span><span class="p">,</span> <span class="n">show_progress_bar</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">gene_subset</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">max_promoter_length</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">max_gene_name_length</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">is_reference</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Constructor for the Genome object.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>genbank_file_</code></td>
            <td>
                  <code>str) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the genbank file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>show_progress_bar</code></td>
            <td>
                  <code>bool, optional) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Boolean as whether to show a progress
bar when building Gene objects. Defaults to False.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>gene_subset</code></td>
            <td>
                  <code>list, optional) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of gene names used to extract just a
subset of genes. Defaults to None</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_promoter_length</code></td>
            <td>
                  <code>int, optional) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Size of the default maximum number of
upstream bases to consider the promoter of a gene. Defaults to 100</p>
              </div>
            </td>
            <td>
                  <code>100</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_gene_name_length</code></td>
            <td>
                  <code>int, optional) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Length of the longest gene name.
Defaults to 20</p>
              </div>
            </td>
            <td>
                  <code>20</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>verbose</code></td>
            <td>
                  <code>bool, optional) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Give verbose statements? Defaults to False</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>is_reference</code></td>
            <td>
                  <code>bool, optional) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Is this a reference genome? i.e. mutations
can be derived with respect to it? Defaults to False</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span>
<span class="normal">97</span>
<span class="normal">98</span>
<span class="normal">99</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">genbank_file_</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">show_progress_bar</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">gene_subset</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">max_promoter_length</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
    <span class="n">max_gene_name_length</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>
    <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">is_reference</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Constructor for the Genome object.</span>

<span class="sd">    Args:</span>
<span class="sd">        genbank_file_ (str) : The path to the genbank file.</span>
<span class="sd">        show_progress_bar (bool, optional) : Boolean as whether to show a progress</span>
<span class="sd">            bar when building Gene objects. Defaults to False.</span>
<span class="sd">        gene_subset (list, optional) : List of gene names used to extract just a</span>
<span class="sd">            subset of genes. Defaults to None</span>
<span class="sd">        max_promoter_length (int, optional) : Size of the default maximum number of</span>
<span class="sd">            upstream bases to consider the promoter of a gene. Defaults to 100</span>
<span class="sd">        max_gene_name_length (int, optional) : Length of the longest gene name.</span>
<span class="sd">            Defaults to 20</span>
<span class="sd">        verbose (bool, optional) : Give verbose statements? Defaults to False</span>
<span class="sd">        is_reference (bool, optional) : Is this a reference genome? i.e. mutations</span>
<span class="sd">            can be derived with respect to it? Defaults to False</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">show_progress_bar</span> <span class="o">=</span> <span class="n">show_progress_bar</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span> <span class="o">=</span> <span class="n">gene_subset</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">max_promoter_length</span> <span class="o">=</span> <span class="n">max_promoter_length</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">max_gene_name_length</span> <span class="o">=</span> <span class="n">max_gene_name_length</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span> <span class="o">=</span> <span class="n">verbose</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_reference</span> <span class="o">=</span> <span class="n">is_reference</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">vcf_file</span><span class="p">:</span> <span class="n">VCFFile</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">gumpy_version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="n">genbank_file</span> <span class="o">=</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">genbank_file_</span><span class="p">)</span>

    <span class="k">assert</span> <span class="n">genbank_file</span><span class="o">.</span><span class="n">is_file</span><span class="p">(),</span> <span class="s2">&quot;GenBank file does not exist!&quot;</span>
    <span class="k">assert</span> <span class="p">(</span>
        <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">max_promoter_length</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_promoter_length</span> <span class="o">&gt;=</span> <span class="mi">0</span>
    <span class="p">),</span> <span class="s2">&quot;the promoter length must be zero or a positive integer!&quot;</span>

    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">,</span> <span class="nb">bool</span><span class="p">)</span>
    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_reference</span><span class="p">,</span> <span class="nb">bool</span><span class="p">)</span>
    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">show_progress_bar</span><span class="p">,</span> <span class="nb">bool</span><span class="p">)</span>
    <span class="k">assert</span> <span class="p">(</span>
        <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">max_gene_name_length</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_gene_name_length</span> <span class="o">&gt;</span> <span class="mi">0</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># first check it is a list</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span>
        <span class="c1"># then check all elements in the list are strings</span>
        <span class="k">assert</span> <span class="nb">all</span><span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span><span class="p">)</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="n">timings</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
        <span class="n">start_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">__parse_genbank_file</span><span class="p">(</span><span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">genbank_file</span><span class="p">))</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="n">timings</span><span class="p">[</span><span class="s2">&quot;parse genbank&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">start_time</span><span class="p">)</span>
        <span class="n">start_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">__setup_arrays</span><span class="p">()</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="n">timings</span><span class="p">[</span><span class="s2">&quot;define arrays&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">start_time</span><span class="p">)</span>
        <span class="n">start_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_promoter_length</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">__assign_promoter_regions</span><span class="p">()</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="n">timings</span><span class="p">[</span><span class="s2">&quot;promoter&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">start_time</span><span class="p">)</span>
        <span class="n">start_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">__convert_references</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.__insert_newlines" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__insert_newlines</span><span class="p">(</span><span class="n">string</span><span class="p">,</span> <span class="n">every</span><span class="o">=</span><span class="mi">70</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <pre><code>    Simple private method for inserting a carriage return every N characters into
        a long string.

    Args:
        string (str): the string to insert carriage returns
        every (int): how many characters between each carriage return

    Returns:
        str: Same string with "
</code></pre>
<p>" characters inserted</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span>
<span class="normal">847</span>
<span class="normal">848</span>
<span class="normal">849</span>
<span class="normal">850</span>
<span class="normal">851</span>
<span class="normal">852</span>
<span class="normal">853</span>
<span class="normal">854</span>
<span class="normal">855</span>
<span class="normal">856</span>
<span class="normal">857</span>
<span class="normal">858</span>
<span class="normal">859</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__insert_newlines</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">string</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">every</span><span class="o">=</span><span class="mi">70</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Simple private method for inserting a carriage return every N characters into</span>
<span class="sd">        a long string.</span>

<span class="sd">    Args:</span>
<span class="sd">        string (str): the string to insert carriage returns</span>
<span class="sd">        every (int): how many characters between each carriage return</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Same string with &quot;\n&quot; characters inserted</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">assert</span> <span class="n">every</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;every must be an integer greater than zero&quot;</span>

    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">string</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;string is too short!&quot;</span>

    <span class="k">return</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">string</span><span class="p">[</span><span class="n">i</span> <span class="p">:</span> <span class="n">i</span> <span class="o">+</span> <span class="n">every</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">string</span><span class="p">),</span> <span class="n">every</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.__len__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__len__</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Adding len functionality - len(genome) returns the length of the genome</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>int</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Length of the genome</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adding len functionality - len(genome) returns the length of the genome</span>

<span class="sd">    Returns:</span>
<span class="sd">        int : Length of the genome</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.__parse_genbank_file" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__parse_genbank_file</span><span class="p">(</span><span class="n">genbank_file</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Private function to parse a genbank file
Args:
    genbank_file (Path) : pathlib.Path object of the genbank file</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__parse_genbank_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genbank_file</span><span class="p">:</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Private function to parse a genbank file</span>
<span class="sd">    Args:</span>
<span class="sd">        genbank_file (Path) : pathlib.Path object of the genbank file</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">genbank_file</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.gz&quot;</span><span class="p">:</span>
        <span class="n">file_handle</span> <span class="o">=</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">genbank_file</span><span class="p">,</span> <span class="s2">&quot;rt&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">genbank_file</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.gbk&quot;</span><span class="p">:</span>
        <span class="n">file_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">genbank_file</span><span class="p">,</span> <span class="s2">&quot;rt&quot;</span><span class="p">)</span>

    <span class="n">reference_genome</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">file_handle</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span>

    <span class="c1"># convert to a numpy array at the first opportunity since slicing BioPython</span>
    <span class="c1">#   is between 10 and 50,000 times slower!</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
        <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">str</span><span class="p">(</span><span class="n">reference_genome</span><span class="o">.</span><span class="n">seq</span><span class="p">)]</span>
    <span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">reference_genome</span><span class="o">.</span><span class="n">name</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">reference_genome</span><span class="o">.</span><span class="n">id</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="n">reference_genome</span><span class="o">.</span><span class="n">description</span>

    <span class="c1"># store the length of the genome</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">)</span>

    <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;genome length zero!&quot;</span>

    <span class="c1"># create an array of the genome indices</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;int&quot;</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span>
        <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;&lt;U&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">max_gene_name_length</span><span class="p">))</span>
    <span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_cds</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">bool</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_promoter</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">bool</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;int&quot;</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_reverse_complement</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">bool</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">bool</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">object</span><span class="p">)</span>

    <span class="k">assert</span> <span class="p">(</span>
        <span class="nb">len</span><span class="p">(</span><span class="n">reference_genome</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;accessions&quot;</span><span class="p">])</span> <span class="o">==</span> <span class="mi">1</span>
    <span class="p">),</span> <span class="s2">&quot;only GenBank files with a single accessions currently allowed&quot;</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">reference_genome</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">reference_genome</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">:</span> <span class="n">Dict</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># loop through the features listed in the GenBank File</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Iterating through features in GenBank file...&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span>
        <span class="n">reference_genome</span><span class="o">.</span><span class="n">features</span><span class="p">,</span> <span class="n">disable</span><span class="o">=</span><span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">show_progress_bar</span><span class="p">)</span>
    <span class="p">):</span>
        <span class="c1"># only parse coding sequences and rRNA features</span>
        <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;CDS&quot;</span><span class="p">,</span> <span class="s2">&quot;rRNA&quot;</span><span class="p">]:</span>
            <span class="k">continue</span>

        <span class="n">gene_name</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">type_</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">codes_protein</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="c1"># try and use the gene name if available, otherwise use the locus</span>
        <span class="k">if</span> <span class="s2">&quot;gene&quot;</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">gene_name</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">type_</span> <span class="o">=</span> <span class="s2">&quot;GENE&quot;</span>

        <span class="k">elif</span> <span class="s2">&quot;locus_tag&quot;</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">gene_name</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;locus_tag&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">type_</span> <span class="o">=</span> <span class="s2">&quot;LOCUS&quot;</span>

        <span class="k">if</span> <span class="n">gene_name</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">gene_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span>
        <span class="p">):</span>
            <span class="k">continue</span>

        <span class="c1"># if this is ribosomal RNA, then record as such</span>
        <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;rRNA&quot;</span><span class="p">:</span>
            <span class="n">type_</span> <span class="o">=</span> <span class="s2">&quot;RNA&quot;</span>
            <span class="n">codes_protein</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="c1"># determine if this is a reverse complement gene (only relevant to</span>
        <span class="c1">#   dsDNA genomes)</span>
        <span class="n">rev_comp</span> <span class="o">=</span> <span class="kc">True</span> <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span> <span class="k">else</span> <span class="kc">False</span>

        <span class="c1"># sigh, you can&#39;t assume that a gene_name is unique in a GenBank file</span>
        <span class="c1"># this only allows for duplicates though.</span>
        <span class="c1"># duplicates of duplicates will be foo_2_2</span>
        <span class="n">gene_name</span> <span class="o">+=</span> <span class="s2">&quot;_2&quot;</span> <span class="k">if</span> <span class="n">gene_name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>

        <span class="c1"># check the gene_name will fit in the max gene name length</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_gene_name_length</span><span class="p">,</span> <span class="p">(</span>
            <span class="s2">&quot;Gene &quot;</span>
            <span class="o">+</span> <span class="n">gene_name</span>
            <span class="o">+</span> <span class="s2">&quot; is too long at &quot;</span>
            <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_name</span><span class="p">))</span>
            <span class="o">+</span> <span class="s2">&quot; chars; need to specify max_gene_name_length&quot;</span>
        <span class="p">)</span>

        <span class="c1"># note that BioPython &quot;helpfully&quot; turns these from 1-based into 0-based</span>
        <span class="c1">#   coordinates, hence the +1</span>
        <span class="c1"># gene_end has also been incremented by 1 so that slicing naturally works</span>
        <span class="n">gene_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="n">gene_end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>

        <span class="c1"># Check for ribosomal shift</span>
        <span class="c1"># This happens when a start position &lt; end position</span>
        <span class="n">positions</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">loc</span><span class="o">.</span><span class="n">start</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">loc</span><span class="o">.</span><span class="n">end</span><span class="p">))</span> <span class="k">for</span> <span class="n">loc</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">parts</span>
        <span class="p">]</span>
        <span class="n">shifts</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># Check for -1 PFS</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">positions</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">positions</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]:</span>
            <span class="n">shifts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">positions</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">gene_start</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

        <span class="c1"># record feature metadata in a dict</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;reverse_complement&quot;</span><span class="p">:</span> <span class="n">rev_comp</span><span class="p">,</span>
            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="n">type_</span><span class="p">,</span>
            <span class="s2">&quot;codes_protein&quot;</span><span class="p">:</span> <span class="n">codes_protein</span><span class="p">,</span>
            <span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="n">gene_start</span><span class="p">,</span>
            <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="n">gene_end</span><span class="p">,</span>
            <span class="s2">&quot;ribosomal_shifts&quot;</span><span class="p">:</span> <span class="n">shifts</span><span class="p">,</span>
        <span class="p">}</span>

    <span class="c1"># now we can check that all the genes in the gene_subset exist in the</span>
    <span class="c1">#   GenBank file!</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span><span class="p">:</span>
            <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">contains_gene</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="p">(</span>
                <span class="s2">&quot;Gene &quot;</span> <span class="o">+</span> <span class="n">i</span> <span class="o">+</span> <span class="s2">&quot; not found in the Genbank file!&quot;</span>
            <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.__repr__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__repr__</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Overload the print function to write a summary of the genome.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>String including attributes for the genome</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Overload the print function to write a summary of the genome.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str : String including attributes for the genome</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">):</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;id&quot;</span><span class="p">):</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;description&quot;</span><span class="p">):</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">description</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">output</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot; bases</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">6</span><span class="p">])</span>
    <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;...&quot;</span>
    <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="o">-</span><span class="mi">6</span><span class="p">:])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;metadata for all genes/loci have been included</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">10</span><span class="p">:</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="p">(</span>
            <span class="s2">&quot;the following &quot;</span>
            <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span><span class="p">))</span>
            <span class="o">+</span> <span class="s2">&quot; genes have been included: &quot;</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span><span class="p">:</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_subset</span><span class="p">))</span> <span class="o">+</span> <span class="s2">&quot; gene/loci have been included.&quot;</span>
    <span class="k">return</span> <span class="n">output</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.__setup_arrays" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__setup_arrays</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Private function to initalise all of the required arrays, fitting the gene
    names into the correct places within stacked arrays</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__setup_arrays</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Private function to initalise all of the required arrays, fitting the gene</span>
<span class="sd">        names into the correct places within stacked arrays</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">__find_overlaps</span><span class="p">()</span>

    <span class="c1"># do as many assignments outside the loop, i.e. in one go to improve performance</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_cds</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">n_rows</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stacked_gene_name</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">,</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_rows</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_sequence</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">,</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_rows</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Use a list to track minority populations</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">minor_populations</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Track which nucleotides are deleted</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_deleted</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="kc">False</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">])</span>

    <span class="c1"># Use a dict to track VCF evidence as req</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidence</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.__sub__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__sub__</span><span class="p">(</span><span class="n">other</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Generate a GenomeDifference object for a in-depth difference of the
    two Genomes</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>other</code></td>
            <td>
                  <code>gumpy.Genome) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The other genome used in the subtraction</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>GenomeDifference</code></td>            <td>
                  <code><span title="gumpy.GenomeDifference">GenomeDifference</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>object containing numpy arrays of the
differences (variants)</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__sub__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenomeDifference</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate a GenomeDifference object for a in-depth difference of the</span>
<span class="sd">        two Genomes</span>

<span class="sd">    Args:</span>
<span class="sd">        other (gumpy.Genome) : The other genome used in the subtraction</span>

<span class="sd">    Returns:</span>
<span class="sd">        GenomeDifference: object containing numpy arrays of the</span>
<span class="sd">            differences (variants)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">Genome</span><span class="p">),</span> <span class="s2">&quot;RHS must be a gumpy.Genome object&quot;</span>

    <span class="k">return</span> <span class="n">GenomeDifference</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.at_index" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">at_index</span><span class="p">(</span><span class="n">index</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Returns the name of any genome features (genes, loci) at a specified genome
    index (1-based).</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>index</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Genome index to check for genes at.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.List">List</span>[str] | None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List[str] | None: list of gene_names or locus_tags at that index
in the genome</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">at_index</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns the name of any genome features (genes, loci) at a specified genome</span>
<span class="sd">        index (1-based).</span>

<span class="sd">    Args:</span>
<span class="sd">        index (int): Genome index to check for genes at.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[str] | None: list of gene_names or locus_tags at that index</span>
<span class="sd">            in the genome</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span> <span class="s2">&quot;index must be an integer!&quot;</span>
    <span class="k">assert</span> <span class="n">index</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;index must be a positive integer!&quot;</span>
    <span class="k">assert</span> <span class="n">index</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">,</span> <span class="s2">&quot;index must be less than the length of the genome!&quot;</span>

    <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span> <span class="o">==</span> <span class="n">index</span>

    <span class="n">foo</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stacked_gene_name</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>

    <span class="n">putative_genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">foo</span><span class="p">[</span><span class="n">foo</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">])</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">putative_genes</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">putative_genes</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.build_gene" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">build_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Public function to build the gumpy.Gene object</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>gene</code></td>
            <td>
                  <code>str) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the gene</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="gumpy.Gene">Gene</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>gumpy.Gene : The instanciated gene object.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">861</span>
<span class="normal">862</span>
<span class="normal">863</span>
<span class="normal">864</span>
<span class="normal">865</span>
<span class="normal">866</span>
<span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span>
<span class="normal">873</span>
<span class="normal">874</span>
<span class="normal">875</span>
<span class="normal">876</span>
<span class="normal">877</span>
<span class="normal">878</span>
<span class="normal">879</span>
<span class="normal">880</span>
<span class="normal">881</span>
<span class="normal">882</span>
<span class="normal">883</span>
<span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span>
<span class="normal">899</span>
<span class="normal">900</span>
<span class="normal">901</span>
<span class="normal">902</span>
<span class="normal">903</span>
<span class="normal">904</span>
<span class="normal">905</span>
<span class="normal">906</span>
<span class="normal">907</span>
<span class="normal">908</span>
<span class="normal">909</span>
<span class="normal">910</span>
<span class="normal">911</span>
<span class="normal">912</span>
<span class="normal">913</span>
<span class="normal">914</span>
<span class="normal">915</span>
<span class="normal">916</span>
<span class="normal">917</span>
<span class="normal">918</span>
<span class="normal">919</span>
<span class="normal">920</span>
<span class="normal">921</span>
<span class="normal">922</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">build_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Gene</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Public function to build the gumpy.Gene object</span>

<span class="sd">    Args:</span>
<span class="sd">        gene (str) : The name of the gene</span>

<span class="sd">    Returns:</span>
<span class="sd">        gumpy.Gene : The instanciated gene object.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># The mask for all stacked arrays (N-dim)</span>
    <span class="n">stacked_mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>

    <span class="c1"># The mask for singular arrays (1-dim) by collapsing stacked mask to 1-dim</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">stacked_mask</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># check that the genome does contain the gene name</span>
    <span class="k">assert</span> <span class="n">numpy</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">mask</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;gene (&quot;</span> <span class="o">+</span> <span class="n">gene</span> <span class="o">+</span> <span class="s2">&quot;) not found in genome!&quot;</span>

    <span class="c1"># Revert is_cds to all False if this gene is non-coding</span>
    <span class="c1"># Easier to do it like this than to mess with masking</span>
    <span class="n">nucleotide_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]:</span>
        <span class="n">is_cd</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="kc">False</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">nucleotide_seq</span><span class="p">))])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">is_cd</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_cds</span><span class="p">[</span><span class="n">stacked_mask</span><span class="p">]</span>

    <span class="n">gene_nucleotides</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
    <span class="n">gene_minor_populations</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">population</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">minor_populations</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">population</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">gene_nucleotides</span><span class="p">:</span>
            <span class="c1"># This minor population is within the gene</span>
            <span class="n">gene_minor_populations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">population</span><span class="p">)</span>

    <span class="n">vcf_evidence</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">idx</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">gene_nucleotides</span>
    <span class="p">}</span>

    <span class="c1"># instantiate a Gene object</span>
    <span class="n">g</span> <span class="o">=</span> <span class="n">Gene</span><span class="p">(</span>
        <span class="n">gene</span><span class="p">,</span>
        <span class="n">nucleotide_seq</span><span class="p">,</span>
        <span class="n">gene_nucleotides</span><span class="p">,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span><span class="n">stacked_mask</span><span class="p">],</span>
        <span class="n">is_cd</span><span class="p">,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stacked_is_promoter</span><span class="p">[</span><span class="n">stacked_mask</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;reverse_complement&quot;</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;type&quot;</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;ribosomal_shifts&quot;</span><span class="p">],</span>
        <span class="n">gene_minor_populations</span><span class="p">,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="n">vcf_evidence</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">g</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.build_genome_string" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">build_genome_string</span><span class="p">(</span><span class="n">fixed_length</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">nucleotide_index_range</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Generate a string of the nucleotides in the genome (positive strand if DNA).</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>fixed_length</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>if True, then do not add insertions and deletions.
Default False.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>nucleotide_index_range</code></td>
            <td>
                  <code>(tuple, ints)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the 1-based positions of the sequence
to return with start&lt;=index&lt;end.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the genome as a string.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">build_genome_string</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">fixed_length</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">nucleotide_index_range</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate a string of the nucleotides in the genome (positive strand if DNA).</span>

<span class="sd">    Args:</span>
<span class="sd">        fixed_length (bool): if True, then do not add insertions and deletions.</span>
<span class="sd">            Default False.</span>
<span class="sd">        nucleotide_index_range (tuple, ints): the 1-based positions of the sequence</span>
<span class="sd">            to return with start&lt;=index&lt;end.</span>

<span class="sd">    Returns:</span>
<span class="sd">        (str): the genome as a string.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># create a string of the genome</span>
    <span class="k">if</span> <span class="n">fixed_length</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">nucleotide_index_range</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">nucleotide_index_range</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">)</span>
            <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">nucleotide_index_range</span>
            <span class="n">genome_string</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">genome_string</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">nucleotide_index_range</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">nucleotide_index_range</span>
            <span class="n">genome_string</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__build_genome_variable_length_string</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">genome_string</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__build_genome_variable_length_string</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span>
            <span class="p">)</span>

    <span class="k">return</span> <span class="n">genome_string</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.contains_gene" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">contains_gene</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Simply checks to see if the specified gene exists in the Genome object.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>gene_name</code></td>
            <td>
                  <code>str) </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the gene e.g. katG</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>bool</code></td>            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Boolean showing if the genome contains a gene with that name</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">contains_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Simply checks to see if the specified gene exists in the Genome object.</span>

<span class="sd">    Args:</span>
<span class="sd">        gene_name (str) : Name of the gene e.g. katG</span>

<span class="sd">    Returns:</span>
<span class="sd">        bool : Boolean showing if the genome contains a gene with that name</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span>
        <span class="n">gene_name</span><span class="p">,</span> <span class="nb">str</span>
    <span class="p">),</span> <span class="s2">&quot;Gene name must be string. Gene name provided was of type: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span>
        <span class="nb">type</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="c1"># Use of dict.get(obj) returns an object or None if obj does not exist in dict</span>
    <span class="c1"># bool(None) = False, bool(obj) = True</span>
    <span class="k">return</span> <span class="nb">bool</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene_name</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.minority_populations_GARC" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">minority_populations_GARC</span><span class="p">(</span><span class="n">interpretation</span><span class="o">=</span><span class="s1">&#39;reads&#39;</span><span class="p">,</span> <span class="n">reference</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Get the variants in GARC of the minority populations for this genome.
Whether the variants are given in terms of reads or read percentage is
    controlled by <code>interpretation</code></p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>interpretation</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Which interpretation to use. <code>reads</code> for
number of reads for this population. <code>percentage</code> for the decimal
percentage of total reads for this population. Defaults to 'reads'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;reads&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>reference</code></td>
            <td>
                  <code><span title="gumpy.Genome">Genome</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The reference to denote mutations from.
Defaults to self</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>
      <p>Returns:
    List[str]: List of the variants in GARC</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span>
<span class="normal">1103</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">minority_populations_GARC</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span> <span class="n">interpretation</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;reads&quot;</span><span class="p">,</span> <span class="n">reference</span><span class="o">=</span><span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the variants in GARC of the minority populations for this genome.</span>
<span class="sd">    Whether the variants are given in terms of reads or read percentage is</span>
<span class="sd">        controlled by `interpretation`</span>

<span class="sd">    Args:</span>
<span class="sd">        interpretation (str, optional): Which interpretation to use. `reads` for</span>
<span class="sd">            number of reads for this population. `percentage` for the decimal</span>
<span class="sd">            percentage of total reads for this population. Defaults to &#39;reads&#39;.</span>
<span class="sd">        reference (gumpy.Genome, optional): The reference to denote mutations from.</span>
<span class="sd">            Defaults to self</span>
<span class="sd">    Returns:</span>
<span class="sd">        List[str]: List of the variants in GARC</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Use the interpretation type to pull out which index of the minor_populations</span>
    <span class="c1"># Each item of minor_populations is (pos, type, bases, abs_coverage,</span>
    <span class="c1">#   percent_coverage)</span>
    <span class="k">if</span> <span class="n">interpretation</span> <span class="o">==</span> <span class="s2">&quot;percentage&quot;</span><span class="p">:</span>
        <span class="n">coverage</span> <span class="o">=</span> <span class="mi">4</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">coverage</span> <span class="o">=</span> <span class="mi">3</span>

    <span class="k">if</span> <span class="n">reference</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">reference</span> <span class="o">=</span> <span class="bp">self</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Ensure that only one of the two Genomes has minor populations</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">reference</span><span class="o">.</span><span class="n">minor_populations</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">,</span> <span class="p">(</span>
            <span class="s2">&quot;Minority populations can only be compared when 1 Genome &quot;</span>
            <span class="s2">&quot;does not have them!&quot;</span>
        <span class="p">)</span>

    <span class="n">variants</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">minor</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">minor_populations</span><span class="p">:</span>
        <span class="n">pos</span> <span class="o">=</span> <span class="n">minor</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">type_</span> <span class="o">=</span> <span class="n">minor</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">bases</span> <span class="o">=</span> <span class="n">minor</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
        <span class="n">depth</span> <span class="o">=</span> <span class="n">minor</span><span class="p">[</span><span class="n">coverage</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">type_</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;ref&quot;</span><span class="p">,</span> <span class="s2">&quot;snp&quot;</span><span class="p">]:</span>
            <span class="c1"># These are functionally the same</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">alt</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">bases</span><span class="p">)):</span>
                <span class="n">ref</span> <span class="o">=</span> <span class="n">reference</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span>
                    <span class="n">reference</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">==</span> <span class="n">pos</span> <span class="o">+</span> <span class="n">i</span>
                <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pos</span><span class="o">+</span><span class="n">i</span><span class="si">}{</span><span class="n">ref</span><span class="si">}</span><span class="s2">&gt;</span><span class="si">{</span><span class="n">alt</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">depth</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Indels are the same too</span>
            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pos</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">type_</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">bases</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">depth</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.save_fasta" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">save_fasta</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">fixed_length</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">nucleotide_index_range</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">compresslevel</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">chars_per_line</span><span class="o">=</span><span class="mi">70</span><span class="p">,</span> <span class="n">nucleotides_uppercase</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overwrite_existing</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Save the genome as a FASTA file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>filename</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>path of the output file</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>fixed_length</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, ignore indels and only output a genome the
same length as the reference but with SNPs. This is useful for
phylogeny analyses and relatedness. If false, a genome including indels
is produced. Default is false.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>nucleotide_index_range</code></td>
            <td>
                  <code>tuple</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A tuple of (start,end)
genome indices</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>compression</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, save compressed using gzip. (bzip2 is too slow)</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>compresslevel</code></td>
            <td>
                  <code>0 - 9</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the higher the number, the harder the algorithm tries
to compress but it takes longer. Default is 2.</p>
              </div>
            </td>
            <td>
                  <code>2</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>chars_per_line</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the number of characters per line. Default=70. Must
be either a positive integer or None (i.e. no CRs)</p>
              </div>
            </td>
            <td>
                  <code>70</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>nucleotide_uppercase</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, provide the nucleotides in
UPPER CASE. Default is True.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>description</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>what to write on the header line of the FASTA
file. If not provided, then a description will be automatically
generated from the GenBank file metadata.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>overwrite_existing</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If False, then the code will refuse to overwrite
a FASTA file already on disc. Default is True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">save_fasta</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">filename</span><span class="p">,</span>
    <span class="n">fixed_length</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">nucleotide_index_range</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">compression</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">compresslevel</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
    <span class="n">chars_per_line</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">70</span><span class="p">,</span>
    <span class="n">nucleotides_uppercase</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">description</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">overwrite_existing</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Save the genome as a FASTA file.</span>

<span class="sd">    Args:</span>
<span class="sd">        filename (str): path of the output file</span>
<span class="sd">        fixed_length (bool): If True, ignore indels and only output a genome the</span>
<span class="sd">            same length as the reference but with SNPs. This is useful for</span>
<span class="sd">            phylogeny analyses and relatedness. If false, a genome including indels</span>
<span class="sd">            is produced. Default is false.</span>
<span class="sd">        nucleotide_index_range (tuple, optional): A tuple of (start,end)</span>
<span class="sd">            genome indices</span>
<span class="sd">        compression (bool): If True, save compressed using gzip. (bzip2 is too slow)</span>
<span class="sd">        compresslevel (0-9): the higher the number, the harder the algorithm tries</span>
<span class="sd">            to compress but it takes longer. Default is 2.</span>
<span class="sd">        chars_per_line (int): the number of characters per line. Default=70. Must</span>
<span class="sd">            be either a positive integer or None (i.e. no CRs)</span>
<span class="sd">        nucleotide_uppercase (bool): If True, provide the nucleotides in</span>
<span class="sd">            UPPER CASE. Default is True.</span>
<span class="sd">        description (str, optional): what to write on the header line of the FASTA</span>
<span class="sd">            file. If not provided, then a description will be automatically</span>
<span class="sd">            generated from the GenBank file metadata.</span>
<span class="sd">        overwrite_existing (bool): If False, then the code will refuse to overwrite</span>
<span class="sd">            a FASTA file already on disc. Default is True.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># check the arguments are well formed</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">overwrite_existing</span><span class="p">:</span>
        <span class="k">assert</span> <span class="ow">not</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span><span class="o">.</span><span class="n">is_file</span><span class="p">(),</span> <span class="p">(</span>
            <span class="s2">&quot;filename already exists! &quot;</span> <span class="o">+</span> <span class="n">filename</span>
        <span class="p">)</span>
    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">compression</span><span class="p">,</span> <span class="nb">bool</span><span class="p">)</span>
    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">fixed_length</span><span class="p">,</span> <span class="nb">bool</span><span class="p">)</span>
    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">nucleotides_uppercase</span><span class="p">,</span> <span class="nb">bool</span><span class="p">)</span>
    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">chars_per_line</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">nucleotide_index_range</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">nucleotide_index_range</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">)</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">nucleotide_index_range</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">int</span><span class="p">)</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">nucleotide_index_range</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="nb">int</span><span class="p">)</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="n">nucleotide_index_range</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">1</span>
        <span class="p">),</span> <span class="s2">&quot;genomes are 1-based so the first base must be &gt;=1&quot;</span>
        <span class="k">assert</span> <span class="n">nucleotide_index_range</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">,</span> <span class="s2">&quot;longer than the genome!&quot;</span>
    <span class="k">assert</span> <span class="n">compresslevel</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="s2">&quot;compresslevel must be in range 1-9!&quot;</span>
    <span class="k">assert</span> <span class="p">(</span>
        <span class="n">chars_per_line</span> <span class="o">&gt;</span> <span class="mi">0</span>
    <span class="p">),</span> <span class="s2">&quot;number of characters per line in the FASTA file must be a positive integer!&quot;</span>
    <span class="k">if</span> <span class="n">description</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">description</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span>

    <span class="c1"># check the specified fileextension to see if the FASTA file needs compressing</span>
    <span class="k">if</span> <span class="n">compression</span><span class="p">:</span>
        <span class="n">OUTPUT</span> <span class="o">=</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">filename</span> <span class="o">+</span> <span class="s2">&quot;.gz&quot;</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">,</span> <span class="n">compresslevel</span><span class="o">=</span><span class="n">compresslevel</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">OUTPUT</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>

    <span class="c1"># create the header line for the FASTA file using &quot;|&quot; as delimiters</span>
    <span class="n">header</span> <span class="o">=</span> <span class="s2">&quot;&gt;&quot;</span>
    <span class="k">if</span> <span class="n">description</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">):</span>
            <span class="n">header</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;|&quot;</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;id&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">header</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">+</span> <span class="s2">&quot;|&quot;</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;description&quot;</span><span class="p">)</span>
            <span class="ow">and</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">description</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span>
            <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">description</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span>
        <span class="p">):</span>
            <span class="n">header</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">description</span> <span class="o">+</span> <span class="s2">&quot;|&quot;</span>
        <span class="n">header</span> <span class="o">=</span> <span class="n">header</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">header</span> <span class="o">+=</span> <span class="n">description</span>
    <span class="n">header</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

    <span class="n">genome_string</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">build_genome_string</span><span class="p">(</span><span class="n">fixed_length</span><span class="p">,</span> <span class="n">nucleotide_index_range</span><span class="p">)</span>

    <span class="c1"># insert carriage returns so it looks pretty in the file...</span>
    <span class="n">output_string</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__insert_newlines</span><span class="p">(</span><span class="n">genome_string</span><span class="p">,</span> <span class="n">every</span><span class="o">=</span><span class="n">chars_per_line</span><span class="p">)</span>
    <span class="n">output_string</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

    <span class="c1"># set the case accordingly</span>
    <span class="k">if</span> <span class="n">nucleotides_uppercase</span><span class="p">:</span>
        <span class="n">output_string</span> <span class="o">=</span> <span class="n">output_string</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">output_string</span> <span class="o">=</span> <span class="n">output_string</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>

    <span class="c1"># write out the FASTA files</span>
    <span class="k">if</span> <span class="n">compression</span><span class="p">:</span>
        <span class="n">OUTPUT</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="n">header</span><span class="p">))</span>
        <span class="n">OUTPUT</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="n">output_string</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">OUTPUT</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">header</span><span class="p">)</span>
        <span class="n">OUTPUT</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">output_string</span><span class="p">)</span>

    <span class="n">OUTPUT</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="genome.Genome.save_sequence" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">save_sequence</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Save the genome as a compressed NPZ file (compressed internally using gzip).</p>
<p>This is purely done because loading an NPZ file back into memory is FAST
    (~200s) so this could allow future analyses</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>filename</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>path of the output file without the file extension</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/genome.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">save_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Save the genome as a compressed NPZ file (compressed internally using gzip).</span>

<span class="sd">    This is purely done because loading an NPZ file back into memory is FAST</span>
<span class="sd">        (~200s) so this could allow future analyses</span>

<span class="sd">    Args:</span>
<span class="sd">        filename (str): path of the output file without the file extension</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">numpy</span><span class="o">.</span><span class="n">savez_compressed</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../gene/" class="md-footer__link md-footer__link--prev" aria-label="Previous: gene">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                gene
              </div>
            </div>
          </a>
        
        
          
          <a href="../variantfile/" class="md-footer__link md-footer__link--next" aria-label="Next: variantfile">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                variantfile
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.footer", "navigation.instant", "navigation.sections", "navigation.tabs", "navigation.tracking"], "search": "../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "jeremy"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.fe8b6f2b.min.js"></script>
      
    
  </body>
</html>