
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../">
      
      
        <link rel="next" href="../gene/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.30">
    
    
      
        <title>difference - Gumpy Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.3cba04c6.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#difference" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
      </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Gumpy Documentation" class="md-header__button md-logo" aria-label="Gumpy Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Gumpy Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              difference
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/oxfordmmm/gumpy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    gumpy
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../readme/" class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../" class="md-tabs__link">
          
  
  Code Reference

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Gumpy Documentation" class="md-nav__button md-logo" aria-label="Gumpy Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Gumpy Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/oxfordmmm/gumpy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    gumpy
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../readme/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Code Reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gumpy
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    difference
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    difference
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#difference" class="md-nav__link">
    <span class="md-ellipsis">
      difference
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#difference.Difference" class="md-nav__link">
    <span class="md-ellipsis">
      Difference
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Difference">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#difference.Difference.__full_to_diff" class="md-nav__link">
    <span class="md-ellipsis">
      __full_to_diff
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.Difference.update_view" class="md-nav__link">
    <span class="md-ellipsis">
      update_view
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#difference.FailedComparison" class="md-nav__link">
    <span class="md-ellipsis">
      FailedComparison
    </span>
  </a>
  
    <nav class="md-nav" aria-label="FailedComparison">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#difference.FailedComparison.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#difference.GeneDifference" class="md-nav__link">
    <span class="md-ellipsis">
      GeneDifference
    </span>
  </a>
  
    <nav class="md-nav" aria-label="GeneDifference">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#difference.GeneDifference.__amino_acids" class="md-nav__link">
    <span class="md-ellipsis">
      __amino_acids
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GeneDifference.__assign_vcf_evidence" class="md-nav__link">
    <span class="md-ellipsis">
      __assign_vcf_evidence
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GeneDifference.__codons" class="md-nav__link">
    <span class="md-ellipsis">
      __codons
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GeneDifference.__get_mutations" class="md-nav__link">
    <span class="md-ellipsis">
      __get_mutations
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GeneDifference.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GeneDifference.__large_deletions" class="md-nav__link">
    <span class="md-ellipsis">
      __large_deletions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GeneDifference.__nucleotides" class="md-nav__link">
    <span class="md-ellipsis">
      __nucleotides
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GeneDifference.minor_populations" class="md-nav__link">
    <span class="md-ellipsis">
      minor_populations
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#difference.GenomeDifference" class="md-nav__link">
    <span class="md-ellipsis">
      GenomeDifference
    </span>
  </a>
  
    <nav class="md-nav" aria-label="GenomeDifference">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#difference.GenomeDifference.snp_distance" class="md-nav__link">
    <span class="md-ellipsis">
      snp_distance
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GenomeDifference.__assign_vcf_evidence" class="md-nav__link">
    <span class="md-ellipsis">
      __assign_vcf_evidence
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GenomeDifference.__get_variants" class="md-nav__link">
    <span class="md-ellipsis">
      __get_variants
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GenomeDifference.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GenomeDifference.__nucleotides" class="md-nav__link">
    <span class="md-ellipsis">
      __nucleotides
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GenomeDifference.__snp_distance" class="md-nav__link">
    <span class="md-ellipsis">
      __snp_distance
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GenomeDifference.get_gene_pos" class="md-nav__link">
    <span class="md-ellipsis">
      get_gene_pos
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GenomeDifference.minor_populations" class="md-nav__link">
    <span class="md-ellipsis">
      minor_populations
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#difference.convert_nucleotides_codons" class="md-nav__link">
    <span class="md-ellipsis">
      convert_nucleotides_codons
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#difference.setup_codon_aa_dict" class="md-nav__link">
    <span class="md-ellipsis">
      setup_codon_aa_dict
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gene/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gene
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genome/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    genome
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../variantfile/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    variantfile
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#difference" class="md-nav__link">
    <span class="md-ellipsis">
      difference
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#difference.Difference" class="md-nav__link">
    <span class="md-ellipsis">
      Difference
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Difference">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#difference.Difference.__full_to_diff" class="md-nav__link">
    <span class="md-ellipsis">
      __full_to_diff
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.Difference.update_view" class="md-nav__link">
    <span class="md-ellipsis">
      update_view
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#difference.FailedComparison" class="md-nav__link">
    <span class="md-ellipsis">
      FailedComparison
    </span>
  </a>
  
    <nav class="md-nav" aria-label="FailedComparison">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#difference.FailedComparison.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#difference.GeneDifference" class="md-nav__link">
    <span class="md-ellipsis">
      GeneDifference
    </span>
  </a>
  
    <nav class="md-nav" aria-label="GeneDifference">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#difference.GeneDifference.__amino_acids" class="md-nav__link">
    <span class="md-ellipsis">
      __amino_acids
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GeneDifference.__assign_vcf_evidence" class="md-nav__link">
    <span class="md-ellipsis">
      __assign_vcf_evidence
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GeneDifference.__codons" class="md-nav__link">
    <span class="md-ellipsis">
      __codons
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GeneDifference.__get_mutations" class="md-nav__link">
    <span class="md-ellipsis">
      __get_mutations
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GeneDifference.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GeneDifference.__large_deletions" class="md-nav__link">
    <span class="md-ellipsis">
      __large_deletions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GeneDifference.__nucleotides" class="md-nav__link">
    <span class="md-ellipsis">
      __nucleotides
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GeneDifference.minor_populations" class="md-nav__link">
    <span class="md-ellipsis">
      minor_populations
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#difference.GenomeDifference" class="md-nav__link">
    <span class="md-ellipsis">
      GenomeDifference
    </span>
  </a>
  
    <nav class="md-nav" aria-label="GenomeDifference">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#difference.GenomeDifference.snp_distance" class="md-nav__link">
    <span class="md-ellipsis">
      snp_distance
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GenomeDifference.__assign_vcf_evidence" class="md-nav__link">
    <span class="md-ellipsis">
      __assign_vcf_evidence
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GenomeDifference.__get_variants" class="md-nav__link">
    <span class="md-ellipsis">
      __get_variants
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GenomeDifference.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GenomeDifference.__nucleotides" class="md-nav__link">
    <span class="md-ellipsis">
      __nucleotides
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GenomeDifference.__snp_distance" class="md-nav__link">
    <span class="md-ellipsis">
      __snp_distance
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GenomeDifference.get_gene_pos" class="md-nav__link">
    <span class="md-ellipsis">
      get_gene_pos
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#difference.GenomeDifference.minor_populations" class="md-nav__link">
    <span class="md-ellipsis">
      minor_populations
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#difference.convert_nucleotides_codons" class="md-nav__link">
    <span class="md-ellipsis">
      convert_nucleotides_codons
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#difference.setup_codon_aa_dict" class="md-nav__link">
    <span class="md-ellipsis">
      setup_codon_aa_dict
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>difference</h1>

<div class="doc doc-object doc-module">



<a id="difference"></a>
    <div class="doc doc-contents first">

      <p>Used to find differences between Genomes and Genes, as well as the impact a VCF has on 
a genome.</p>


<details class="abstract-classes" open>
  <summary>Abstract classes</summary>
  <ul>
<li>Difference - Abstract class providing the ability to change data views</li>
</ul>
</details>

<p><span class="doc-section-title">Functions:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="difference.convert_nucleotides_codons" href="#difference.convert_nucleotides_codons">convert_nucleotides_codons</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Converts an array of 
nucleotides to an array of codons.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="difference.setup_codon_aa_dict" href="#difference.setup_codon_aa_dict">setup_codon_aa_dict</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Returns a dictionary mapping codon-&gt;amino_acid</p>
                </div>
              </td>
            </tr>
      </tbody>
    </table>



  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="difference.Difference" class="doc doc-heading">
            <code>Difference</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="abc.ABC">ABC</span></code></p>


      <p>Abstract class used to provide the ability to switch between views.
Inherited by GenomeDifference and GeneDifference.</p>
<p>This should not be instantiated.</p>

              <details class="quote">
                <summary>Source code in <code>gumpy/difference.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Difference</span><span class="p">(</span><span class="n">ABC</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Abstract class used to provide the ability to switch between views.</span>
<span class="sd">    Inherited by GenomeDifference and GeneDifference.</span>

<span class="sd">    This should not be instantiated.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Give some default values to appease the linter</span>
    <span class="n">_nucleotides_full</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([])</span>
    <span class="n">codes_protein</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">_indels_full</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([])</span>

    <span class="k">def</span> <span class="nf">update_view</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">method</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Update the viewing method. Can either be `diff` or `full`:</span>

<span class="sd">        * `diff`: Where applicable, variables return arrays of values from object1</span>
<span class="sd">                where they are not equal to object 2</span>
<span class="sd">        * `full`: Where applicable, variables return arrays of tuples showing</span>
<span class="sd">                (object1_val, object2_val) where values are not equal between objects.</span>

<span class="sd">        Args:</span>
<span class="sd">            method (str): Name of the viewing method. Must be within `[&#39;diff&#39;, &#39;full&#39;]`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">method</span><span class="p">,</span> <span class="nb">str</span><span class="p">),</span> <span class="p">(</span>
            <span class="s2">&quot;Invalid method &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">method</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot; of type &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">method</span><span class="p">))</span>
        <span class="p">)</span>
        <span class="k">assert</span> <span class="n">method</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;diff&quot;</span><span class="p">,</span> <span class="s2">&quot;full&quot;</span><span class="p">],</span> <span class="s2">&quot;Invalid method: &quot;</span> <span class="o">+</span> <span class="n">method</span>

        <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;full&quot;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nucleotides</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_nucleotides_full</span>

        <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;diff&quot;</span><span class="p">:</span>
            <span class="c1"># Convert the full arrays into diff arrays</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nucleotides</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__full_to_diff</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_nucleotides_full</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_view_method</span> <span class="o">=</span> <span class="n">method</span>

    <span class="k">def</span> <span class="nf">__full_to_diff</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">array</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Convert an array from a full view to a diff view</span>

<span class="sd">        Args:</span>
<span class="sd">            array (numpy.ndarray): Array of tuples of values</span>
<span class="sd">        Returns:</span>
<span class="sd">            numpy.ndarray: Array of values from object1</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">array</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">item2</span> <span class="k">for</span> <span class="p">(</span><span class="n">item1</span><span class="p">,</span> <span class="n">item2</span><span class="p">)</span> <span class="ow">in</span> <span class="n">array</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">object</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([])</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="difference.Difference.__full_to_diff" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__full_to_diff</span><span class="p">(</span><span class="n">array</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Convert an array from a full view to a diff view</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>array</code></td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array of tuples of values</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
      <p>Returns:
    numpy.ndarray: Array of values from object1</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__full_to_diff</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">array</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert an array from a full view to a diff view</span>

<span class="sd">    Args:</span>
<span class="sd">        array (numpy.ndarray): Array of tuples of values</span>
<span class="sd">    Returns:</span>
<span class="sd">        numpy.ndarray: Array of values from object1</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">array</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">item2</span> <span class="k">for</span> <span class="p">(</span><span class="n">item1</span><span class="p">,</span> <span class="n">item2</span><span class="p">)</span> <span class="ow">in</span> <span class="n">array</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">object</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([])</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="difference.Difference.update_view" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">update_view</span><span class="p">(</span><span class="n">method</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Update the viewing method. Can either be <code>diff</code> or <code>full</code>:</p>
<ul>
<li><code>diff</code>: Where applicable, variables return arrays of values from object1
        where they are not equal to object 2</li>
<li><code>full</code>: Where applicable, variables return arrays of tuples showing
        (object1_val, object2_val) where values are not equal between objects.</li>
</ul>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>method</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the viewing method. Must be within <code>['diff', 'full']</code></p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">update_view</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">method</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Update the viewing method. Can either be `diff` or `full`:</span>

<span class="sd">    * `diff`: Where applicable, variables return arrays of values from object1</span>
<span class="sd">            where they are not equal to object 2</span>
<span class="sd">    * `full`: Where applicable, variables return arrays of tuples showing</span>
<span class="sd">            (object1_val, object2_val) where values are not equal between objects.</span>

<span class="sd">    Args:</span>
<span class="sd">        method (str): Name of the viewing method. Must be within `[&#39;diff&#39;, &#39;full&#39;]`</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">method</span><span class="p">,</span> <span class="nb">str</span><span class="p">),</span> <span class="p">(</span>
        <span class="s2">&quot;Invalid method &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">method</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot; of type &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">method</span><span class="p">))</span>
    <span class="p">)</span>
    <span class="k">assert</span> <span class="n">method</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;diff&quot;</span><span class="p">,</span> <span class="s2">&quot;full&quot;</span><span class="p">],</span> <span class="s2">&quot;Invalid method: &quot;</span> <span class="o">+</span> <span class="n">method</span>

    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;full&quot;</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotides</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_nucleotides_full</span>

    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;diff&quot;</span><span class="p">:</span>
        <span class="c1"># Convert the full arrays into diff arrays</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotides</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__full_to_diff</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_nucleotides_full</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_view_method</span> <span class="o">=</span> <span class="n">method</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="difference.FailedComparison" class="doc doc-heading">
            <code>FailedComparison</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code>Exception</code></p>


      <p>Exception to be raised in cases of a failed comparion</p>

              <details class="quote">
                <summary>Source code in <code>gumpy/difference.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">FailedComparison</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Exception to be raised in cases of a failed comparion&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">message</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Constructor</span>

<span class="sd">        Args:</span>
<span class="sd">            message (str): Error message</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">message</span> <span class="o">=</span> <span class="n">message</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="difference.FailedComparison.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">message</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Constructor</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>message</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Error message</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">message</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Constructor</span>

<span class="sd">    Args:</span>
<span class="sd">        message (str): Error message</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">message</span> <span class="o">=</span> <span class="n">message</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="difference.GeneDifference" class="doc doc-heading">
            <code>GeneDifference</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="difference.Difference" href="#difference.Difference">Difference</a></code></p>


      <p>Object to store the differences within genes. The view system is inherited from
    the Difference class.</p>
<ul>
<li>Set to <code>full</code> for arrays of tuple values where applicable.</li>
<li>Set to <code>diff</code> for arrays of values from gene1 where the values vary. Default.</li>
<li>
<p>This can be updated using the update_view function.</p>
</li>
<li>
<p>Instance variables:</p>
<ul>
<li>gene1 (gumpy.Gene): Gene object 1</li>
<li>gene2 (gumpy.Gene): Gene object 2</li>
<li>nucleotides (numpy.array): Array of the nucleotides at which the genes differ.
    Format depends on the current view.</li>
<li>mutations (numpy.array): Array of mutations in GARC between the two Gene
    objects.</li>
<li>indels (numpy.array): Array of indel lengths where the indel lengths differ.
    Format depends on the current view.</li>
<li>codons (numpy.array): Array of codons where the two Gene objects have
    different codons. Format depends on the current view.</li>
<li>amino_acid_sequence (numpy.array): Array of amino acids where the two Gene
    objects have different amino acids. Format depends on the current view.</li>
</ul>
</li>
<li>Functions:<ul>
<li>amino_acid_variants(int) -&gt; dict: Takes an amino acid index and returns a
    dictionary containing data from a vcf file (if applicable) for attributes
    such as calls, ref, and alt for all nucleotides within the codons for this
    amino acid index. If these genes do not code protien, returns {}</li>
<li>nucleotide_variants(int) -&gt; dict: Takes a nucleotide index and returns a
    dictionary containing data from a vcf file (if applicable) for attributes
    such as calls, ref, and alt at the given index</li>
<li>minor_populations() -&gt; [str]: Returns a list of minor population mutations
    in GARC</li>
</ul>
</li>
<li>Inherited functions:<ul>
<li>update_view(str) -&gt; None: Used to change the viewing method for instance
    variables. Input values are either <code>diff</code> or <code>full</code></li>
</ul>
</li>
</ul>

              <details class="quote">
                <summary>Source code in <code>gumpy/difference.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 849</span>
<span class="normal"> 850</span>
<span class="normal"> 851</span>
<span class="normal"> 852</span>
<span class="normal"> 853</span>
<span class="normal"> 854</span>
<span class="normal"> 855</span>
<span class="normal"> 856</span>
<span class="normal"> 857</span>
<span class="normal"> 858</span>
<span class="normal"> 859</span>
<span class="normal"> 860</span>
<span class="normal"> 861</span>
<span class="normal"> 862</span>
<span class="normal"> 863</span>
<span class="normal"> 864</span>
<span class="normal"> 865</span>
<span class="normal"> 866</span>
<span class="normal"> 867</span>
<span class="normal"> 868</span>
<span class="normal"> 869</span>
<span class="normal"> 870</span>
<span class="normal"> 871</span>
<span class="normal"> 872</span>
<span class="normal"> 873</span>
<span class="normal"> 874</span>
<span class="normal"> 875</span>
<span class="normal"> 876</span>
<span class="normal"> 877</span>
<span class="normal"> 878</span>
<span class="normal"> 879</span>
<span class="normal"> 880</span>
<span class="normal"> 881</span>
<span class="normal"> 882</span>
<span class="normal"> 883</span>
<span class="normal"> 884</span>
<span class="normal"> 885</span>
<span class="normal"> 886</span>
<span class="normal"> 887</span>
<span class="normal"> 888</span>
<span class="normal"> 889</span>
<span class="normal"> 890</span>
<span class="normal"> 891</span>
<span class="normal"> 892</span>
<span class="normal"> 893</span>
<span class="normal"> 894</span>
<span class="normal"> 895</span>
<span class="normal"> 896</span>
<span class="normal"> 897</span>
<span class="normal"> 898</span>
<span class="normal"> 899</span>
<span class="normal"> 900</span>
<span class="normal"> 901</span>
<span class="normal"> 902</span>
<span class="normal"> 903</span>
<span class="normal"> 904</span>
<span class="normal"> 905</span>
<span class="normal"> 906</span>
<span class="normal"> 907</span>
<span class="normal"> 908</span>
<span class="normal"> 909</span>
<span class="normal"> 910</span>
<span class="normal"> 911</span>
<span class="normal"> 912</span>
<span class="normal"> 913</span>
<span class="normal"> 914</span>
<span class="normal"> 915</span>
<span class="normal"> 916</span>
<span class="normal"> 917</span>
<span class="normal"> 918</span>
<span class="normal"> 919</span>
<span class="normal"> 920</span>
<span class="normal"> 921</span>
<span class="normal"> 922</span>
<span class="normal"> 923</span>
<span class="normal"> 924</span>
<span class="normal"> 925</span>
<span class="normal"> 926</span>
<span class="normal"> 927</span>
<span class="normal"> 928</span>
<span class="normal"> 929</span>
<span class="normal"> 930</span>
<span class="normal"> 931</span>
<span class="normal"> 932</span>
<span class="normal"> 933</span>
<span class="normal"> 934</span>
<span class="normal"> 935</span>
<span class="normal"> 936</span>
<span class="normal"> 937</span>
<span class="normal"> 938</span>
<span class="normal"> 939</span>
<span class="normal"> 940</span>
<span class="normal"> 941</span>
<span class="normal"> 942</span>
<span class="normal"> 943</span>
<span class="normal"> 944</span>
<span class="normal"> 945</span>
<span class="normal"> 946</span>
<span class="normal"> 947</span>
<span class="normal"> 948</span>
<span class="normal"> 949</span>
<span class="normal"> 950</span>
<span class="normal"> 951</span>
<span class="normal"> 952</span>
<span class="normal"> 953</span>
<span class="normal"> 954</span>
<span class="normal"> 955</span>
<span class="normal"> 956</span>
<span class="normal"> 957</span>
<span class="normal"> 958</span>
<span class="normal"> 959</span>
<span class="normal"> 960</span>
<span class="normal"> 961</span>
<span class="normal"> 962</span>
<span class="normal"> 963</span>
<span class="normal"> 964</span>
<span class="normal"> 965</span>
<span class="normal"> 966</span>
<span class="normal"> 967</span>
<span class="normal"> 968</span>
<span class="normal"> 969</span>
<span class="normal"> 970</span>
<span class="normal"> 971</span>
<span class="normal"> 972</span>
<span class="normal"> 973</span>
<span class="normal"> 974</span>
<span class="normal"> 975</span>
<span class="normal"> 976</span>
<span class="normal"> 977</span>
<span class="normal"> 978</span>
<span class="normal"> 979</span>
<span class="normal"> 980</span>
<span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span>
<span class="normal">1022</span>
<span class="normal">1023</span>
<span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span>
<span class="normal">1103</span>
<span class="normal">1104</span>
<span class="normal">1105</span>
<span class="normal">1106</span>
<span class="normal">1107</span>
<span class="normal">1108</span>
<span class="normal">1109</span>
<span class="normal">1110</span>
<span class="normal">1111</span>
<span class="normal">1112</span>
<span class="normal">1113</span>
<span class="normal">1114</span>
<span class="normal">1115</span>
<span class="normal">1116</span>
<span class="normal">1117</span>
<span class="normal">1118</span>
<span class="normal">1119</span>
<span class="normal">1120</span>
<span class="normal">1121</span>
<span class="normal">1122</span>
<span class="normal">1123</span>
<span class="normal">1124</span>
<span class="normal">1125</span>
<span class="normal">1126</span>
<span class="normal">1127</span>
<span class="normal">1128</span>
<span class="normal">1129</span>
<span class="normal">1130</span>
<span class="normal">1131</span>
<span class="normal">1132</span>
<span class="normal">1133</span>
<span class="normal">1134</span>
<span class="normal">1135</span>
<span class="normal">1136</span>
<span class="normal">1137</span>
<span class="normal">1138</span>
<span class="normal">1139</span>
<span class="normal">1140</span>
<span class="normal">1141</span>
<span class="normal">1142</span>
<span class="normal">1143</span>
<span class="normal">1144</span>
<span class="normal">1145</span>
<span class="normal">1146</span>
<span class="normal">1147</span>
<span class="normal">1148</span>
<span class="normal">1149</span>
<span class="normal">1150</span>
<span class="normal">1151</span>
<span class="normal">1152</span>
<span class="normal">1153</span>
<span class="normal">1154</span>
<span class="normal">1155</span>
<span class="normal">1156</span>
<span class="normal">1157</span>
<span class="normal">1158</span>
<span class="normal">1159</span>
<span class="normal">1160</span>
<span class="normal">1161</span>
<span class="normal">1162</span>
<span class="normal">1163</span>
<span class="normal">1164</span>
<span class="normal">1165</span>
<span class="normal">1166</span>
<span class="normal">1167</span>
<span class="normal">1168</span>
<span class="normal">1169</span>
<span class="normal">1170</span>
<span class="normal">1171</span>
<span class="normal">1172</span>
<span class="normal">1173</span>
<span class="normal">1174</span>
<span class="normal">1175</span>
<span class="normal">1176</span>
<span class="normal">1177</span>
<span class="normal">1178</span>
<span class="normal">1179</span>
<span class="normal">1180</span>
<span class="normal">1181</span>
<span class="normal">1182</span>
<span class="normal">1183</span>
<span class="normal">1184</span>
<span class="normal">1185</span>
<span class="normal">1186</span>
<span class="normal">1187</span>
<span class="normal">1188</span>
<span class="normal">1189</span>
<span class="normal">1190</span>
<span class="normal">1191</span>
<span class="normal">1192</span>
<span class="normal">1193</span>
<span class="normal">1194</span>
<span class="normal">1195</span>
<span class="normal">1196</span>
<span class="normal">1197</span>
<span class="normal">1198</span>
<span class="normal">1199</span>
<span class="normal">1200</span>
<span class="normal">1201</span>
<span class="normal">1202</span>
<span class="normal">1203</span>
<span class="normal">1204</span>
<span class="normal">1205</span>
<span class="normal">1206</span>
<span class="normal">1207</span>
<span class="normal">1208</span>
<span class="normal">1209</span>
<span class="normal">1210</span>
<span class="normal">1211</span>
<span class="normal">1212</span>
<span class="normal">1213</span>
<span class="normal">1214</span>
<span class="normal">1215</span>
<span class="normal">1216</span>
<span class="normal">1217</span>
<span class="normal">1218</span>
<span class="normal">1219</span>
<span class="normal">1220</span>
<span class="normal">1221</span>
<span class="normal">1222</span>
<span class="normal">1223</span>
<span class="normal">1224</span>
<span class="normal">1225</span>
<span class="normal">1226</span>
<span class="normal">1227</span>
<span class="normal">1228</span>
<span class="normal">1229</span>
<span class="normal">1230</span>
<span class="normal">1231</span>
<span class="normal">1232</span>
<span class="normal">1233</span>
<span class="normal">1234</span>
<span class="normal">1235</span>
<span class="normal">1236</span>
<span class="normal">1237</span>
<span class="normal">1238</span>
<span class="normal">1239</span>
<span class="normal">1240</span>
<span class="normal">1241</span>
<span class="normal">1242</span>
<span class="normal">1243</span>
<span class="normal">1244</span>
<span class="normal">1245</span>
<span class="normal">1246</span>
<span class="normal">1247</span>
<span class="normal">1248</span>
<span class="normal">1249</span>
<span class="normal">1250</span>
<span class="normal">1251</span>
<span class="normal">1252</span>
<span class="normal">1253</span>
<span class="normal">1254</span>
<span class="normal">1255</span>
<span class="normal">1256</span>
<span class="normal">1257</span>
<span class="normal">1258</span>
<span class="normal">1259</span>
<span class="normal">1260</span>
<span class="normal">1261</span>
<span class="normal">1262</span>
<span class="normal">1263</span>
<span class="normal">1264</span>
<span class="normal">1265</span>
<span class="normal">1266</span>
<span class="normal">1267</span>
<span class="normal">1268</span>
<span class="normal">1269</span>
<span class="normal">1270</span>
<span class="normal">1271</span>
<span class="normal">1272</span>
<span class="normal">1273</span>
<span class="normal">1274</span>
<span class="normal">1275</span>
<span class="normal">1276</span>
<span class="normal">1277</span>
<span class="normal">1278</span>
<span class="normal">1279</span>
<span class="normal">1280</span>
<span class="normal">1281</span>
<span class="normal">1282</span>
<span class="normal">1283</span>
<span class="normal">1284</span>
<span class="normal">1285</span>
<span class="normal">1286</span>
<span class="normal">1287</span>
<span class="normal">1288</span>
<span class="normal">1289</span>
<span class="normal">1290</span>
<span class="normal">1291</span>
<span class="normal">1292</span>
<span class="normal">1293</span>
<span class="normal">1294</span>
<span class="normal">1295</span>
<span class="normal">1296</span>
<span class="normal">1297</span>
<span class="normal">1298</span>
<span class="normal">1299</span>
<span class="normal">1300</span>
<span class="normal">1301</span>
<span class="normal">1302</span>
<span class="normal">1303</span>
<span class="normal">1304</span>
<span class="normal">1305</span>
<span class="normal">1306</span>
<span class="normal">1307</span>
<span class="normal">1308</span>
<span class="normal">1309</span>
<span class="normal">1310</span>
<span class="normal">1311</span>
<span class="normal">1312</span>
<span class="normal">1313</span>
<span class="normal">1314</span>
<span class="normal">1315</span>
<span class="normal">1316</span>
<span class="normal">1317</span>
<span class="normal">1318</span>
<span class="normal">1319</span>
<span class="normal">1320</span>
<span class="normal">1321</span>
<span class="normal">1322</span>
<span class="normal">1323</span>
<span class="normal">1324</span>
<span class="normal">1325</span>
<span class="normal">1326</span>
<span class="normal">1327</span>
<span class="normal">1328</span>
<span class="normal">1329</span>
<span class="normal">1330</span>
<span class="normal">1331</span>
<span class="normal">1332</span>
<span class="normal">1333</span>
<span class="normal">1334</span>
<span class="normal">1335</span>
<span class="normal">1336</span>
<span class="normal">1337</span>
<span class="normal">1338</span>
<span class="normal">1339</span>
<span class="normal">1340</span>
<span class="normal">1341</span>
<span class="normal">1342</span>
<span class="normal">1343</span>
<span class="normal">1344</span>
<span class="normal">1345</span>
<span class="normal">1346</span>
<span class="normal">1347</span>
<span class="normal">1348</span>
<span class="normal">1349</span>
<span class="normal">1350</span>
<span class="normal">1351</span>
<span class="normal">1352</span>
<span class="normal">1353</span>
<span class="normal">1354</span>
<span class="normal">1355</span>
<span class="normal">1356</span>
<span class="normal">1357</span>
<span class="normal">1358</span>
<span class="normal">1359</span>
<span class="normal">1360</span>
<span class="normal">1361</span>
<span class="normal">1362</span>
<span class="normal">1363</span>
<span class="normal">1364</span>
<span class="normal">1365</span>
<span class="normal">1366</span>
<span class="normal">1367</span>
<span class="normal">1368</span>
<span class="normal">1369</span>
<span class="normal">1370</span>
<span class="normal">1371</span>
<span class="normal">1372</span>
<span class="normal">1373</span>
<span class="normal">1374</span>
<span class="normal">1375</span>
<span class="normal">1376</span>
<span class="normal">1377</span>
<span class="normal">1378</span>
<span class="normal">1379</span>
<span class="normal">1380</span>
<span class="normal">1381</span>
<span class="normal">1382</span>
<span class="normal">1383</span>
<span class="normal">1384</span>
<span class="normal">1385</span>
<span class="normal">1386</span>
<span class="normal">1387</span>
<span class="normal">1388</span>
<span class="normal">1389</span>
<span class="normal">1390</span>
<span class="normal">1391</span>
<span class="normal">1392</span>
<span class="normal">1393</span>
<span class="normal">1394</span>
<span class="normal">1395</span>
<span class="normal">1396</span>
<span class="normal">1397</span>
<span class="normal">1398</span>
<span class="normal">1399</span>
<span class="normal">1400</span>
<span class="normal">1401</span>
<span class="normal">1402</span>
<span class="normal">1403</span>
<span class="normal">1404</span>
<span class="normal">1405</span>
<span class="normal">1406</span>
<span class="normal">1407</span>
<span class="normal">1408</span>
<span class="normal">1409</span>
<span class="normal">1410</span>
<span class="normal">1411</span>
<span class="normal">1412</span>
<span class="normal">1413</span>
<span class="normal">1414</span>
<span class="normal">1415</span>
<span class="normal">1416</span>
<span class="normal">1417</span>
<span class="normal">1418</span>
<span class="normal">1419</span>
<span class="normal">1420</span>
<span class="normal">1421</span>
<span class="normal">1422</span>
<span class="normal">1423</span>
<span class="normal">1424</span>
<span class="normal">1425</span>
<span class="normal">1426</span>
<span class="normal">1427</span>
<span class="normal">1428</span>
<span class="normal">1429</span>
<span class="normal">1430</span>
<span class="normal">1431</span>
<span class="normal">1432</span>
<span class="normal">1433</span>
<span class="normal">1434</span>
<span class="normal">1435</span>
<span class="normal">1436</span>
<span class="normal">1437</span>
<span class="normal">1438</span>
<span class="normal">1439</span>
<span class="normal">1440</span>
<span class="normal">1441</span>
<span class="normal">1442</span>
<span class="normal">1443</span>
<span class="normal">1444</span>
<span class="normal">1445</span>
<span class="normal">1446</span>
<span class="normal">1447</span>
<span class="normal">1448</span>
<span class="normal">1449</span>
<span class="normal">1450</span>
<span class="normal">1451</span>
<span class="normal">1452</span>
<span class="normal">1453</span>
<span class="normal">1454</span>
<span class="normal">1455</span>
<span class="normal">1456</span>
<span class="normal">1457</span>
<span class="normal">1458</span>
<span class="normal">1459</span>
<span class="normal">1460</span>
<span class="normal">1461</span>
<span class="normal">1462</span>
<span class="normal">1463</span>
<span class="normal">1464</span>
<span class="normal">1465</span>
<span class="normal">1466</span>
<span class="normal">1467</span>
<span class="normal">1468</span>
<span class="normal">1469</span>
<span class="normal">1470</span>
<span class="normal">1471</span>
<span class="normal">1472</span>
<span class="normal">1473</span>
<span class="normal">1474</span>
<span class="normal">1475</span>
<span class="normal">1476</span>
<span class="normal">1477</span>
<span class="normal">1478</span>
<span class="normal">1479</span>
<span class="normal">1480</span>
<span class="normal">1481</span>
<span class="normal">1482</span>
<span class="normal">1483</span>
<span class="normal">1484</span>
<span class="normal">1485</span>
<span class="normal">1486</span>
<span class="normal">1487</span>
<span class="normal">1488</span>
<span class="normal">1489</span>
<span class="normal">1490</span>
<span class="normal">1491</span>
<span class="normal">1492</span>
<span class="normal">1493</span>
<span class="normal">1494</span>
<span class="normal">1495</span>
<span class="normal">1496</span>
<span class="normal">1497</span>
<span class="normal">1498</span>
<span class="normal">1499</span>
<span class="normal">1500</span>
<span class="normal">1501</span>
<span class="normal">1502</span>
<span class="normal">1503</span>
<span class="normal">1504</span>
<span class="normal">1505</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">GeneDifference</span><span class="p">(</span><span class="n">Difference</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Object to store the differences within genes. The view system is inherited from</span>
<span class="sd">        the Difference class.</span>

<span class="sd">    * Set to `full` for arrays of tuple values where applicable.</span>
<span class="sd">    * Set to `diff` for arrays of values from gene1 where the values vary. Default.</span>
<span class="sd">    * This can be updated using the update_view function.</span>

<span class="sd">    * Instance variables:</span>
<span class="sd">        * gene1 (gumpy.Gene): Gene object 1</span>
<span class="sd">        * gene2 (gumpy.Gene): Gene object 2</span>
<span class="sd">        * nucleotides (numpy.array): Array of the nucleotides at which the genes differ.</span>
<span class="sd">            Format depends on the current view.</span>
<span class="sd">        * mutations (numpy.array): Array of mutations in GARC between the two Gene</span>
<span class="sd">            objects.</span>
<span class="sd">        * indels (numpy.array): Array of indel lengths where the indel lengths differ.</span>
<span class="sd">            Format depends on the current view.</span>
<span class="sd">        * codons (numpy.array): Array of codons where the two Gene objects have</span>
<span class="sd">            different codons. Format depends on the current view.</span>
<span class="sd">        * amino_acid_sequence (numpy.array): Array of amino acids where the two Gene</span>
<span class="sd">            objects have different amino acids. Format depends on the current view.</span>
<span class="sd">    * Functions:</span>
<span class="sd">        * amino_acid_variants(int) -&gt; dict: Takes an amino acid index and returns a</span>
<span class="sd">            dictionary containing data from a vcf file (if applicable) for attributes</span>
<span class="sd">            such as calls, ref, and alt for all nucleotides within the codons for this</span>
<span class="sd">            amino acid index. If these genes do not code protien, returns {}</span>
<span class="sd">        * nucleotide_variants(int) -&gt; dict: Takes a nucleotide index and returns a</span>
<span class="sd">            dictionary containing data from a vcf file (if applicable) for attributes</span>
<span class="sd">            such as calls, ref, and alt at the given index</span>
<span class="sd">        * minor_populations() -&gt; [str]: Returns a list of minor population mutations</span>
<span class="sd">            in GARC</span>
<span class="sd">    * Inherited functions:</span>
<span class="sd">        * update_view(str) -&gt; None: Used to change the viewing method for instance</span>
<span class="sd">            variables. Input values are either `diff` or `full`</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene1</span><span class="p">,</span> <span class="n">gene2</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Constructor. Takes in two gene objects and calculates the difference in a few</span>
<span class="sd">            areas such as nucleotides, codons, and amino acids.</span>

<span class="sd">        Args:</span>
<span class="sd">            gene1 (gumpy.Gene): Gene object 1</span>
<span class="sd">            gene2 (gumpy.Gene): Gene object 2</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gene1</span><span class="p">,</span> <span class="n">gumpy</span><span class="o">.</span><span class="n">Gene</span><span class="p">)</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gene2</span><span class="p">,</span> <span class="n">gumpy</span><span class="o">.</span><span class="n">Gene</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">gene1</span><span class="o">.</span><span class="n">total_number_nucleotides</span> <span class="o">!=</span> <span class="n">gene2</span><span class="o">.</span><span class="n">total_number_nucleotides</span><span class="p">:</span>
            <span class="c1"># The lengths of the genes are different so comparing them is meaningless</span>
            <span class="k">raise</span> <span class="n">FailedComparison</span><span class="p">(</span>
                <span class="s2">&quot;The two genes (&quot;</span>
                <span class="o">+</span> <span class="n">gene1</span><span class="o">.</span><span class="n">name</span>
                <span class="o">+</span> <span class="s2">&quot; and &quot;</span>
                <span class="o">+</span> <span class="n">gene2</span><span class="o">.</span><span class="n">name</span>
                <span class="o">+</span> <span class="s2">&quot;) are different lengths, so comparision failed...&quot;</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">gene1</span><span class="o">.</span><span class="n">name</span> <span class="o">!=</span> <span class="n">gene2</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
                <span class="s2">&quot;The two genes given have different names (&quot;</span>
                <span class="o">+</span> <span class="n">gene1</span><span class="o">.</span><span class="n">name</span>
                <span class="o">+</span> <span class="s2">&quot;, &quot;</span>
                <span class="o">+</span> <span class="n">gene2</span><span class="o">.</span><span class="n">name</span>
                <span class="o">+</span> <span class="s2">&quot;) but the same length, continuing...&quot;</span><span class="p">,</span>
                <span class="ne">UserWarning</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">gene1</span><span class="o">.</span><span class="n">codes_protein</span> <span class="o">!=</span> <span class="n">gene2</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">FailedComparison</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;The two genes given do not have the same protein coding &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;for </span><span class="si">{</span><span class="n">gene1</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">: Gene1 = </span><span class="si">{</span><span class="n">gene1</span><span class="o">.</span><span class="n">codes_protein</span><span class="si">}</span><span class="s2">, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;Gene2 = </span><span class="si">{</span><span class="n">gene2</span><span class="o">.</span><span class="n">codes_protein</span><span class="si">}</span><span class="s2">, so comparison failed...&quot;</span>
            <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span> <span class="o">=</span> <span class="n">gene1</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span> <span class="o">=</span> <span class="n">gene2</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span> <span class="o">=</span> <span class="n">gene1</span><span class="o">.</span><span class="n">codes_protein</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_view_method</span> <span class="o">=</span> <span class="s2">&quot;diff&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_nucleotides_full</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__nucleotides</span><span class="p">()</span>
        <span class="c1"># self.mutations = self.__mutations()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">__get_mutations</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_codons_full</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__codons</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_amino_acids_full</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__amino_acids</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">__large_deletions</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">__assign_vcf_evidence</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">update_view</span><span class="p">(</span><span class="s2">&quot;diff&quot;</span><span class="p">)</span>  <span class="c1"># Use inherited method to set the view</span>

    <span class="k">def</span> <span class="nf">__assign_vcf_evidence</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Once we have pulled out all of the mutations, find the VCF evidence</span>
<span class="sd">        (if existing) for each</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">evidences</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">:</span>
            <span class="n">evidence1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
            <span class="n">evidence2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">evidence1</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">evidence2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="c1"># We have a collision. For now, just concat FIXME</span>
                <span class="n">evidences</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">evidence1</span><span class="p">,</span> <span class="n">evidence2</span><span class="p">])</span>
            <span class="k">elif</span> <span class="n">evidence1</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">evidences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">evidence1</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">evidence2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">evidences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">evidence2</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">evidences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidences</span> <span class="o">=</span> <span class="n">evidences</span>

    <span class="k">def</span> <span class="nf">__large_deletions</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check to see what proportion of the gene has been deleted. Report separately</span>
<span class="sd">            than other mutations if &gt;=50% deleted</span>

<span class="sd">        Updates internal arrays to denote a new mutation if deletion above threshold.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Find everywhere that we have different deletion within coding regions</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_deleted</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_deleted</span>

        <span class="c1"># Find out how much of the gene we have deleted</span>
        <span class="n">total</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">[</span><span class="n">mask</span><span class="p">])</span> <span class="o">+</span> <span class="nb">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">[</span><span class="n">mask</span><span class="p">])</span>

        <span class="k">if</span> <span class="n">total</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># We have some deletions</span>
            <span class="n">percentage</span> <span class="o">=</span> <span class="n">total</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">percentage</span> <span class="o">&gt;=</span> <span class="mf">0.5</span><span class="p">:</span>
                <span class="c1"># More than 50% deleted, so give a percentage</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">,</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;del_</span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="n">percentage</span><span class="p">,</span><span class="w"> </span><span class="mi">2</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>
                <span class="p">)</span>
                <span class="c1"># Pull out the start of the deletion for vcf evidence</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">,</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">mask</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span>
                <span class="p">)</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span><span class="p">,</span> <span class="p">[</span><span class="kc">None</span><span class="p">])</span>

                <span class="c1"># Get the nucleotide number of the first base of the gene</span>
                <span class="c1"># Not predictable as promoters are variable length</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">,</span>
                    <span class="p">[</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_deleted</span> <span class="o">|</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_deleted</span>
                        <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                    <span class="p">],</span>
                <span class="p">)</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span><span class="p">,</span> <span class="p">[</span><span class="kc">None</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">,</span> <span class="p">[</span><span class="kc">True</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">,</span> <span class="p">[</span><span class="kc">False</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span><span class="p">,</span> <span class="p">[</span><span class="kc">True</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">,</span> <span class="p">[</span><span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="n">total</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">,</span>
                    <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">])],</span>
                <span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">ref_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ref_nucleotides</span><span class="p">,</span> <span class="p">[</span><span class="kc">None</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">alt_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alt_nucleotides</span><span class="p">,</span> <span class="p">[</span><span class="kc">None</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">is_snp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_snp</span><span class="p">,</span> <span class="p">[</span><span class="kc">False</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">is_het</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_het</span><span class="p">,</span> <span class="p">[</span><span class="kc">False</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">is_null</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_null</span><span class="p">,</span> <span class="p">[</span><span class="kc">False</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Check for a deletion at the start of the gene which is the result of</span>
                <span class="c1">#   an upstream del</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">:</span>
                    <span class="c1"># This is difficult as revcomp shifts dels upstream by the length</span>
                    <span class="c1">#   of the del</span>
                    <span class="c1"># So we have to find a del in the RHS of the gene which would be</span>
                    <span class="c1">#   the last item if not shifted</span>
                    <span class="n">pos</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
                    <span class="k">while</span> <span class="n">pos</span> <span class="o">*</span> <span class="o">-</span><span class="mi">1</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">:</span>
                        <span class="k">if</span> <span class="p">(</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span>
                            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                        <span class="p">):</span>
                            <span class="c1"># Del here in gene1 so check for validity</span>
                            <span class="k">if</span> <span class="p">(</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                                <span class="o">+</span> <span class="mi">1</span>
                                <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                            <span class="p">):</span>
                                <span class="c1"># Re-adusting for deletion, this matches</span>
                                <span class="k">break</span>
                        <span class="k">if</span> <span class="p">(</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span>
                            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                        <span class="p">):</span>
                            <span class="c1"># Del here in gene2 so check for validity</span>
                            <span class="k">if</span> <span class="p">(</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                                <span class="o">+</span> <span class="mi">1</span>
                                <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                            <span class="p">):</span>
                                <span class="c1"># Re-adusting for deletion, this matches</span>
                                <span class="k">break</span>
                        <span class="n">pos</span> <span class="o">-=</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># If not revcomp, the del is the first item</span>
                    <span class="n">pos</span> <span class="o">=</span> <span class="mi">0</span>

                <span class="k">if</span> <span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                <span class="p">)</span> <span class="ow">and</span> <span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                <span class="p">):</span>
                    <span class="c1"># Both have a del, so no difference</span>
                    <span class="k">return</span>

                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[</span><span class="n">pos</span><span class="p">]:</span>
                    <span class="n">deleted</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[</span><span class="n">pos</span><span class="p">]:</span>
                    <span class="n">deleted</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Neither, so no difference either</span>
                    <span class="k">return</span>

                <span class="c1"># Adjust idx to point to the  right VCF evidence</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">:</span>
                    <span class="n">idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">pos</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">deleted</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>

                <span class="n">pos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">,</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pos</span><span class="si">}</span><span class="s2">_del_</span><span class="si">{</span><span class="n">deleted</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">])</span>
                <span class="c1"># Pull out the start of the deletion for vcf evidence</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">,</span> <span class="p">[</span><span class="n">idx</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span><span class="p">,</span> <span class="p">[</span><span class="kc">None</span><span class="p">])</span>

                <span class="c1"># Get the nucleotide number of the first base of the gene</span>
                <span class="c1"># Not predictable as promoters are variable length</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">,</span>
                    <span class="p">[</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_deleted</span> <span class="o">|</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_deleted</span>
                        <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                    <span class="p">],</span>
                <span class="p">)</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span><span class="p">,</span> <span class="p">[</span><span class="n">pos</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">,</span> <span class="p">[</span><span class="kc">True</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">,</span> <span class="p">[</span><span class="kc">False</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span><span class="p">,</span> <span class="p">[</span><span class="kc">True</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">,</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">deleted</span><span class="p">)])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">,</span> <span class="p">[</span><span class="n">deleted</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">ref_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ref_nucleotides</span><span class="p">,</span> <span class="p">[</span><span class="kc">None</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">alt_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alt_nucleotides</span><span class="p">,</span> <span class="p">[</span><span class="kc">None</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">is_snp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_snp</span><span class="p">,</span> <span class="p">[</span><span class="kc">False</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">is_het</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_het</span><span class="p">,</span> <span class="p">[</span><span class="kc">False</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">is_null</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_null</span><span class="p">,</span> <span class="p">[</span><span class="kc">False</span><span class="p">])</span>

    <span class="k">def</span> <span class="nf">minor_populations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">interpretation</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;reads&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get the minor population mutations in GARC</span>

<span class="sd">        Args:</span>
<span class="sd">            interpretation (str, optional): How to report minor population. &#39;reads&#39;</span>
<span class="sd">                reports number of reads. &#39;percentage&#39; reports fractional read support.</span>
<span class="sd">                Defaults to &#39;reads&#39;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            [str]: List of mutations in GARC</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">minority_populations_GARC</span><span class="p">(</span>
            <span class="n">interpretation</span><span class="o">=</span><span class="n">interpretation</span><span class="p">,</span> <span class="n">reference</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">__nucleotides</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Find the differences in nucleotides</span>

<span class="sd">        Returns:</span>
<span class="sd">            numpy.ndarray: Array of tuples (gene1_nucleotide, gene2_nucleotide)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="p">(</span><span class="n">n1</span><span class="p">,</span> <span class="n">n2</span><span class="p">)</span>
                <span class="k">for</span> <span class="p">(</span><span class="n">n1</span><span class="p">,</span> <span class="n">n2</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_sequence</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="n">n1</span> <span class="o">!=</span> <span class="n">n2</span>
            <span class="p">]</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">__get_mutations</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get the mutations between the two genes, populating internal arrays&quot;&quot;&quot;</span>

        <span class="n">mutations</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">amino_acid_number</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">nucleotide_number</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">nucleotide_index</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">gene_position</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">is_cds</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">is_indel</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">is_promoter</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">indel_length</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">ref_nucleotides</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">alt_nucleotides</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">is_snp</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">is_het</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">is_null</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">variants</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">:</span>
            <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">codons</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">codons</span>
            <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">num</span><span class="p">,</span> <span class="n">a</span><span class="p">,</span> <span class="n">codon1</span><span class="p">,</span> <span class="n">codon2</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">amino_acid_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">amino_acid_number</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">amino_acid_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="p">):</span>
                <span class="c1"># ref.append(r)</span>
                <span class="c1"># alt.append(a)</span>
                <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)</span> <span class="o">+</span> <span class="n">a</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">a</span> <span class="o">==</span> <span class="s2">&quot;X&quot;</span><span class="p">:</span>
                    <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                    <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                    <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">a</span> <span class="o">==</span> <span class="s2">&quot;Z&quot;</span><span class="p">:</span>
                    <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                    <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                    <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                    <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                    <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">amino_acid_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
                <span class="n">nucleotide_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="n">nucleotide_index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="n">gene_position</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
                <span class="n">is_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_promoter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="n">ref_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">codon1</span><span class="p">)</span>
                <span class="n">alt_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">codon2</span><span class="p">)</span>

                <span class="c1"># Reconstruct the variant which caused this mutation to allow joining</span>
                <span class="c1">#   mutation and variant tables</span>
                <span class="n">codon_nucleotide_indices</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_cds</span>
                <span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">codon_number</span> <span class="o">==</span> <span class="n">num</span><span class="p">]</span>
                <span class="n">v</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">r_</span><span class="p">,</span> <span class="n">a_</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">codon1</span><span class="p">,</span> <span class="n">codon2</span><span class="p">)):</span>
                    <span class="k">if</span> <span class="n">r_</span> <span class="o">!=</span> <span class="n">a_</span><span class="p">:</span>
                        <span class="n">v</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="n">codon_nucleotide_indices</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">+</span> <span class="n">r_</span> <span class="o">+</span> <span class="s2">&quot;&gt;&quot;</span> <span class="o">+</span> <span class="n">a_</span> <span class="o">+</span> <span class="s2">&quot;&amp;&quot;</span>
                <span class="k">if</span> <span class="n">v</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;&amp;&quot;</span><span class="p">:</span>
                    <span class="n">v</span> <span class="o">=</span> <span class="n">v</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>

                <span class="c1"># If synonymous mutation, pull out nucleotide variants too</span>
                <span class="c1"># This lets us determine effects of mutations such as fabG1@L203L more</span>
                <span class="c1">#   precisely</span>
                <span class="k">if</span> <span class="n">r</span> <span class="o">==</span> <span class="n">a</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">rn</span><span class="p">,</span> <span class="n">an</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">codon1</span><span class="p">,</span> <span class="n">codon2</span><span class="p">)):</span>
                        <span class="k">if</span> <span class="n">rn</span> <span class="o">!=</span> <span class="n">an</span><span class="p">:</span>
                            <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rn</span> <span class="o">+</span> <span class="nb">str</span><span class="p">((</span><span class="n">num</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">3</span> <span class="o">+</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="n">an</span><span class="p">)</span>
                            <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">an</span> <span class="o">==</span> <span class="s2">&quot;x&quot;</span><span class="p">)</span>
                            <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">an</span> <span class="o">==</span> <span class="s2">&quot;z&quot;</span><span class="p">)</span>
                            <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">an</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">])</span>
                            <span class="n">amino_acid_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                            <span class="n">nucleotide_number</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">num</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">3</span> <span class="o">+</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
                            <span class="n">nucleotide_index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span>
                                    <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_number</span>
                                    <span class="o">==</span> <span class="p">(</span><span class="n">num</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">3</span> <span class="o">+</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span>
                                <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                            <span class="p">)</span>
                            <span class="n">gene_position</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">num</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">3</span> <span class="o">+</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
                            <span class="n">is_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                            <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                            <span class="n">is_promoter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                            <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                            <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                            <span class="n">ref_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rn</span><span class="p">)</span>
                            <span class="n">alt_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">an</span><span class="p">)</span>

                            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                                <span class="nb">str</span><span class="p">(</span><span class="n">codon_nucleotide_indices</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">+</span> <span class="n">rn</span> <span class="o">+</span> <span class="s2">&quot;&gt;&quot;</span> <span class="o">+</span> <span class="n">an</span>
                            <span class="p">)</span>

            <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_sequence</span>
            <span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">num</span><span class="p">,</span> <span class="n">a</span><span class="p">,</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="p">):</span>
                <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)</span> <span class="o">+</span> <span class="n">a</span><span class="p">)</span>
                <span class="c1"># ref.append(r)</span>
                <span class="c1"># alt.append(a)</span>
                <span class="n">amino_acid_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="n">nucleotide_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
                <span class="n">nucleotide_index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
                <span class="n">gene_position</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
                <span class="n">is_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_promoter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="n">ref_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
                <span class="n">alt_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">a</span> <span class="o">==</span> <span class="s2">&quot;x&quot;</span><span class="p">:</span>
                    <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                    <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                    <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">a</span> <span class="o">==</span> <span class="s2">&quot;z&quot;</span><span class="p">:</span>
                    <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                    <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                    <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                    <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                    <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>

                <span class="c1"># Reconstruct the variant which caused this mutation to allow joining</span>
                <span class="c1">#   mutation and variant tables</span>
                <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span> <span class="o">+</span> <span class="n">r</span> <span class="o">+</span> <span class="s2">&quot;&gt;&quot;</span> <span class="o">+</span> <span class="n">a</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_sequence</span>

            <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">num</span><span class="p">,</span> <span class="n">a</span><span class="p">,</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="p">):</span>
                <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)</span> <span class="o">+</span> <span class="n">a</span><span class="p">)</span>
                <span class="c1"># ref.append(r)</span>
                <span class="c1"># alt.append(a)</span>
                <span class="n">amino_acid_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="n">nucleotide_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
                <span class="n">nucleotide_index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
                <span class="n">gene_position</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">num</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">is_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                    <span class="n">is_promoter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">is_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                    <span class="n">is_promoter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="n">ref_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
                <span class="n">alt_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">a</span> <span class="o">==</span> <span class="s2">&quot;x&quot;</span><span class="p">:</span>
                    <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                    <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                    <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">a</span> <span class="o">==</span> <span class="s2">&quot;z&quot;</span><span class="p">:</span>
                    <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                    <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                    <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                    <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                    <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>

                <span class="c1"># Reconstruct the variant which caused this mutation to allow joining</span>
                <span class="c1">#   mutation and variant tables</span>
                <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span> <span class="o">+</span> <span class="n">r</span> <span class="o">+</span> <span class="s2">&quot;&gt;&quot;</span> <span class="o">+</span> <span class="n">a</span><span class="p">)</span>

        <span class="c1"># now let&#39;s do indels</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
                <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">&amp;</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_indel</span><span class="p">)</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="o">==</span> <span class="mi">0</span>
        <span class="p">),</span> <span class="p">(</span>
            <span class="s2">&quot;both genes have different indels at one or more of the same positions -- &quot;</span>
            <span class="s2">&quot;this cannot be easily be resolved!&quot;</span>
        <span class="p">)</span>

        <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">&amp;</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_nucleotides</span>
        <span class="p">)</span>

        <span class="k">for</span> <span class="n">num</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">alt</span><span class="p">,</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_index</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="p">):</span>
            <span class="c1"># ref.append(None)</span>
            <span class="c1"># alt.append(None)</span>
            <span class="n">amino_acid_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">nucleotide_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
            <span class="n">nucleotide_index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
            <span class="n">gene_position</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">length</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_ins_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">alt</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_del_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">alt</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">num</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">is_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">is_promoter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">is_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_promoter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">length</span><span class="p">)</span>
            <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt</span><span class="p">)</span>
            <span class="n">ref_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">alt_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

            <span class="c1"># Reconstruct the variant which caused this mutation to allow joining</span>
            <span class="c1">#   mutation and variant tables</span>
            <span class="n">m</span> <span class="o">=</span> <span class="n">mutations</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:]</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">:</span>
                <span class="c1"># These are slightly different as revcomp need changes</span>
                <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">revcomp_indel_nc_index</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)]]</span> <span class="o">+</span> <span class="n">m</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                        <span class="p">[</span>
                            <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
                            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">==</span> <span class="n">num</span>
                            <span class="p">]</span>
                        <span class="p">]</span>
                        <span class="o">+</span> <span class="n">m</span>
                    <span class="p">)</span>
                <span class="p">)</span>

        <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">&amp;</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_nucleotides</span>
        <span class="p">)</span>

        <span class="k">for</span> <span class="n">num</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">alt</span><span class="p">,</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_index</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="p">):</span>
            <span class="c1"># ref.append(None)</span>
            <span class="c1"># alt.append(None)</span>
            <span class="n">amino_acid_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">nucleotide_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
            <span class="n">nucleotide_index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
            <span class="n">gene_position</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
            <span class="n">length</span> <span class="o">*=</span> <span class="o">-</span><span class="mi">1</span>
            <span class="k">if</span> <span class="n">length</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_ins_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">alt</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_del_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">alt</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">num</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">is_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">is_promoter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">is_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_promoter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">length</span><span class="p">)</span>
            <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt</span><span class="p">)</span>
            <span class="n">ref_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">alt_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

            <span class="c1"># Reconstruct the variant which caused this mutation to allow joining</span>
            <span class="c1">#   mutation and variant tables</span>
            <span class="n">m</span> <span class="o">=</span> <span class="n">mutations</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:]</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">:</span>
                <span class="c1"># These are slightly different as revcomp need changes</span>
                <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">revcomp_indel_nc_index</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)]]</span> <span class="o">+</span> <span class="n">m</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                        <span class="p">[</span>
                            <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
                            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">==</span> <span class="n">num</span>
                            <span class="p">]</span>
                        <span class="p">]</span>
                        <span class="o">+</span> <span class="n">m</span>
                    <span class="p">)</span>
                <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mutations</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">amino_acid_number</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">nucleotide_number</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">gene_position</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_cds</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_promoter</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_indel</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">indel_length</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ref_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">ref_nucleotides</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">alt_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">alt_nucleotides</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_snp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_snp</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_het</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_het</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_null</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_null</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">variants</span> <span class="o">=</span> <span class="n">variants</span>

    <span class="k">def</span> <span class="nf">__codons</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Find the codon positions which are different within the genes</span>
<span class="sd">            (within codon regions)</span>

<span class="sd">        Returns:</span>
<span class="sd">            numpy.ndarray: Array of codons which differ of the form</span>
<span class="sd">                [(gene1_codon, gene2_codon)]</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">codons1</span> <span class="o">=</span> <span class="n">convert_nucleotides_codons</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_cds</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">codons2</span> <span class="o">=</span> <span class="n">convert_nucleotides_codons</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_cds</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="n">codons1</span> <span class="o">!=</span> <span class="n">codons2</span>
        <span class="n">codons1</span> <span class="o">=</span> <span class="n">codons1</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
        <span class="n">codons2</span> <span class="o">=</span> <span class="n">codons2</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">codons1</span><span class="p">,</span> <span class="n">codons2</span><span class="p">)))</span>

    <span class="k">def</span> <span class="nf">__amino_acids</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate the difference in amino acid sequences (within codon regions)</span>

<span class="sd">        Returns:</span>
<span class="sd">            numpy.ndarray: Array of tuples showing (amino_acid_1, amino_acid_2)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">codon_to_amino_acid</span> <span class="o">=</span> <span class="n">setup_codon_aa_dict</span><span class="p">()</span>

        <span class="n">aa_diff</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">codon1</span><span class="p">,</span> <span class="n">codon2</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_codons_full</span><span class="p">:</span>
            <span class="n">aa1</span> <span class="o">=</span> <span class="n">codon_to_amino_acid</span><span class="p">[</span><span class="n">codon1</span><span class="p">]</span>
            <span class="n">aa2</span> <span class="o">=</span> <span class="n">codon_to_amino_acid</span><span class="p">[</span><span class="n">codon2</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">codon1</span> <span class="o">!=</span> <span class="n">codon2</span><span class="p">:</span>
                <span class="n">aa_diff</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">aa1</span><span class="p">,</span> <span class="n">aa2</span><span class="p">))</span>
                <span class="c1"># Adding nucleotide variants for synon</span>
                <span class="k">if</span> <span class="n">aa1</span> <span class="o">==</span> <span class="n">aa2</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">c1</span><span class="p">,</span> <span class="n">c2</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">codon1</span><span class="p">,</span> <span class="n">codon2</span><span class="p">):</span>
                        <span class="k">if</span> <span class="n">c1</span> <span class="o">!=</span> <span class="n">c2</span><span class="p">:</span>
                            <span class="n">aa_diff</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">))</span>
        <span class="c1"># Mutations should be in order of aa-&gt;indel/promoter, so pad to match</span>
        <span class="n">fixed</span> <span class="o">=</span> <span class="n">aa_diff</span>
        <span class="c1"># Promoters</span>
        <span class="n">to_add</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">numpy</span><span class="o">.</span><span class="n">logical_or</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span><span class="p">)</span> <span class="k">if</span> <span class="n">p</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">to_add</span><span class="p">):</span>
            <span class="n">fixed</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">fixed</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="difference.GeneDifference.__amino_acids" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__amino_acids</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Calculate the difference in amino acid sequences (within codon regions)</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>numpy.ndarray: Array of tuples showing (amino_acid_1, amino_acid_2)</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1480</span>
<span class="normal">1481</span>
<span class="normal">1482</span>
<span class="normal">1483</span>
<span class="normal">1484</span>
<span class="normal">1485</span>
<span class="normal">1486</span>
<span class="normal">1487</span>
<span class="normal">1488</span>
<span class="normal">1489</span>
<span class="normal">1490</span>
<span class="normal">1491</span>
<span class="normal">1492</span>
<span class="normal">1493</span>
<span class="normal">1494</span>
<span class="normal">1495</span>
<span class="normal">1496</span>
<span class="normal">1497</span>
<span class="normal">1498</span>
<span class="normal">1499</span>
<span class="normal">1500</span>
<span class="normal">1501</span>
<span class="normal">1502</span>
<span class="normal">1503</span>
<span class="normal">1504</span>
<span class="normal">1505</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__amino_acids</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate the difference in amino acid sequences (within codon regions)</span>

<span class="sd">    Returns:</span>
<span class="sd">        numpy.ndarray: Array of tuples showing (amino_acid_1, amino_acid_2)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">codon_to_amino_acid</span> <span class="o">=</span> <span class="n">setup_codon_aa_dict</span><span class="p">()</span>

    <span class="n">aa_diff</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">codon1</span><span class="p">,</span> <span class="n">codon2</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_codons_full</span><span class="p">:</span>
        <span class="n">aa1</span> <span class="o">=</span> <span class="n">codon_to_amino_acid</span><span class="p">[</span><span class="n">codon1</span><span class="p">]</span>
        <span class="n">aa2</span> <span class="o">=</span> <span class="n">codon_to_amino_acid</span><span class="p">[</span><span class="n">codon2</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">codon1</span> <span class="o">!=</span> <span class="n">codon2</span><span class="p">:</span>
            <span class="n">aa_diff</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">aa1</span><span class="p">,</span> <span class="n">aa2</span><span class="p">))</span>
            <span class="c1"># Adding nucleotide variants for synon</span>
            <span class="k">if</span> <span class="n">aa1</span> <span class="o">==</span> <span class="n">aa2</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">c1</span><span class="p">,</span> <span class="n">c2</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">codon1</span><span class="p">,</span> <span class="n">codon2</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">c1</span> <span class="o">!=</span> <span class="n">c2</span><span class="p">:</span>
                        <span class="n">aa_diff</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">))</span>
    <span class="c1"># Mutations should be in order of aa-&gt;indel/promoter, so pad to match</span>
    <span class="n">fixed</span> <span class="o">=</span> <span class="n">aa_diff</span>
    <span class="c1"># Promoters</span>
    <span class="n">to_add</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">numpy</span><span class="o">.</span><span class="n">logical_or</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span><span class="p">)</span> <span class="k">if</span> <span class="n">p</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">to_add</span><span class="p">):</span>
        <span class="n">fixed</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">fixed</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="difference.GeneDifference.__assign_vcf_evidence" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__assign_vcf_evidence</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Once we have pulled out all of the mutations, find the VCF evidence
(if existing) for each</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">939</span>
<span class="normal">940</span>
<span class="normal">941</span>
<span class="normal">942</span>
<span class="normal">943</span>
<span class="normal">944</span>
<span class="normal">945</span>
<span class="normal">946</span>
<span class="normal">947</span>
<span class="normal">948</span>
<span class="normal">949</span>
<span class="normal">950</span>
<span class="normal">951</span>
<span class="normal">952</span>
<span class="normal">953</span>
<span class="normal">954</span>
<span class="normal">955</span>
<span class="normal">956</span>
<span class="normal">957</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__assign_vcf_evidence</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Once we have pulled out all of the mutations, find the VCF evidence</span>
<span class="sd">    (if existing) for each</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">evidences</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">:</span>
        <span class="n">evidence1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
        <span class="n">evidence2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">evidence1</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">evidence2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># We have a collision. For now, just concat FIXME</span>
            <span class="n">evidences</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">evidence1</span><span class="p">,</span> <span class="n">evidence2</span><span class="p">])</span>
        <span class="k">elif</span> <span class="n">evidence1</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">evidences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">evidence1</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">evidence2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">evidences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">evidence2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">evidences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidences</span> <span class="o">=</span> <span class="n">evidences</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="difference.GeneDifference.__codons" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__codons</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Find the codon positions which are different within the genes
    (within codon regions)</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>numpy.ndarray: Array of codons which differ of the form
[(gene1_codon, gene2_codon)]</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1461</span>
<span class="normal">1462</span>
<span class="normal">1463</span>
<span class="normal">1464</span>
<span class="normal">1465</span>
<span class="normal">1466</span>
<span class="normal">1467</span>
<span class="normal">1468</span>
<span class="normal">1469</span>
<span class="normal">1470</span>
<span class="normal">1471</span>
<span class="normal">1472</span>
<span class="normal">1473</span>
<span class="normal">1474</span>
<span class="normal">1475</span>
<span class="normal">1476</span>
<span class="normal">1477</span>
<span class="normal">1478</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__codons</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Find the codon positions which are different within the genes</span>
<span class="sd">        (within codon regions)</span>

<span class="sd">    Returns:</span>
<span class="sd">        numpy.ndarray: Array of codons which differ of the form</span>
<span class="sd">            [(gene1_codon, gene2_codon)]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">codons1</span> <span class="o">=</span> <span class="n">convert_nucleotides_codons</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_cds</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">codons2</span> <span class="o">=</span> <span class="n">convert_nucleotides_codons</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_cds</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="n">codons1</span> <span class="o">!=</span> <span class="n">codons2</span>
    <span class="n">codons1</span> <span class="o">=</span> <span class="n">codons1</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
    <span class="n">codons2</span> <span class="o">=</span> <span class="n">codons2</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">codons1</span><span class="p">,</span> <span class="n">codons2</span><span class="p">)))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="difference.GeneDifference.__get_mutations" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__get_mutations</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Get the mutations between the two genes, populating internal arrays</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1135</span>
<span class="normal">1136</span>
<span class="normal">1137</span>
<span class="normal">1138</span>
<span class="normal">1139</span>
<span class="normal">1140</span>
<span class="normal">1141</span>
<span class="normal">1142</span>
<span class="normal">1143</span>
<span class="normal">1144</span>
<span class="normal">1145</span>
<span class="normal">1146</span>
<span class="normal">1147</span>
<span class="normal">1148</span>
<span class="normal">1149</span>
<span class="normal">1150</span>
<span class="normal">1151</span>
<span class="normal">1152</span>
<span class="normal">1153</span>
<span class="normal">1154</span>
<span class="normal">1155</span>
<span class="normal">1156</span>
<span class="normal">1157</span>
<span class="normal">1158</span>
<span class="normal">1159</span>
<span class="normal">1160</span>
<span class="normal">1161</span>
<span class="normal">1162</span>
<span class="normal">1163</span>
<span class="normal">1164</span>
<span class="normal">1165</span>
<span class="normal">1166</span>
<span class="normal">1167</span>
<span class="normal">1168</span>
<span class="normal">1169</span>
<span class="normal">1170</span>
<span class="normal">1171</span>
<span class="normal">1172</span>
<span class="normal">1173</span>
<span class="normal">1174</span>
<span class="normal">1175</span>
<span class="normal">1176</span>
<span class="normal">1177</span>
<span class="normal">1178</span>
<span class="normal">1179</span>
<span class="normal">1180</span>
<span class="normal">1181</span>
<span class="normal">1182</span>
<span class="normal">1183</span>
<span class="normal">1184</span>
<span class="normal">1185</span>
<span class="normal">1186</span>
<span class="normal">1187</span>
<span class="normal">1188</span>
<span class="normal">1189</span>
<span class="normal">1190</span>
<span class="normal">1191</span>
<span class="normal">1192</span>
<span class="normal">1193</span>
<span class="normal">1194</span>
<span class="normal">1195</span>
<span class="normal">1196</span>
<span class="normal">1197</span>
<span class="normal">1198</span>
<span class="normal">1199</span>
<span class="normal">1200</span>
<span class="normal">1201</span>
<span class="normal">1202</span>
<span class="normal">1203</span>
<span class="normal">1204</span>
<span class="normal">1205</span>
<span class="normal">1206</span>
<span class="normal">1207</span>
<span class="normal">1208</span>
<span class="normal">1209</span>
<span class="normal">1210</span>
<span class="normal">1211</span>
<span class="normal">1212</span>
<span class="normal">1213</span>
<span class="normal">1214</span>
<span class="normal">1215</span>
<span class="normal">1216</span>
<span class="normal">1217</span>
<span class="normal">1218</span>
<span class="normal">1219</span>
<span class="normal">1220</span>
<span class="normal">1221</span>
<span class="normal">1222</span>
<span class="normal">1223</span>
<span class="normal">1224</span>
<span class="normal">1225</span>
<span class="normal">1226</span>
<span class="normal">1227</span>
<span class="normal">1228</span>
<span class="normal">1229</span>
<span class="normal">1230</span>
<span class="normal">1231</span>
<span class="normal">1232</span>
<span class="normal">1233</span>
<span class="normal">1234</span>
<span class="normal">1235</span>
<span class="normal">1236</span>
<span class="normal">1237</span>
<span class="normal">1238</span>
<span class="normal">1239</span>
<span class="normal">1240</span>
<span class="normal">1241</span>
<span class="normal">1242</span>
<span class="normal">1243</span>
<span class="normal">1244</span>
<span class="normal">1245</span>
<span class="normal">1246</span>
<span class="normal">1247</span>
<span class="normal">1248</span>
<span class="normal">1249</span>
<span class="normal">1250</span>
<span class="normal">1251</span>
<span class="normal">1252</span>
<span class="normal">1253</span>
<span class="normal">1254</span>
<span class="normal">1255</span>
<span class="normal">1256</span>
<span class="normal">1257</span>
<span class="normal">1258</span>
<span class="normal">1259</span>
<span class="normal">1260</span>
<span class="normal">1261</span>
<span class="normal">1262</span>
<span class="normal">1263</span>
<span class="normal">1264</span>
<span class="normal">1265</span>
<span class="normal">1266</span>
<span class="normal">1267</span>
<span class="normal">1268</span>
<span class="normal">1269</span>
<span class="normal">1270</span>
<span class="normal">1271</span>
<span class="normal">1272</span>
<span class="normal">1273</span>
<span class="normal">1274</span>
<span class="normal">1275</span>
<span class="normal">1276</span>
<span class="normal">1277</span>
<span class="normal">1278</span>
<span class="normal">1279</span>
<span class="normal">1280</span>
<span class="normal">1281</span>
<span class="normal">1282</span>
<span class="normal">1283</span>
<span class="normal">1284</span>
<span class="normal">1285</span>
<span class="normal">1286</span>
<span class="normal">1287</span>
<span class="normal">1288</span>
<span class="normal">1289</span>
<span class="normal">1290</span>
<span class="normal">1291</span>
<span class="normal">1292</span>
<span class="normal">1293</span>
<span class="normal">1294</span>
<span class="normal">1295</span>
<span class="normal">1296</span>
<span class="normal">1297</span>
<span class="normal">1298</span>
<span class="normal">1299</span>
<span class="normal">1300</span>
<span class="normal">1301</span>
<span class="normal">1302</span>
<span class="normal">1303</span>
<span class="normal">1304</span>
<span class="normal">1305</span>
<span class="normal">1306</span>
<span class="normal">1307</span>
<span class="normal">1308</span>
<span class="normal">1309</span>
<span class="normal">1310</span>
<span class="normal">1311</span>
<span class="normal">1312</span>
<span class="normal">1313</span>
<span class="normal">1314</span>
<span class="normal">1315</span>
<span class="normal">1316</span>
<span class="normal">1317</span>
<span class="normal">1318</span>
<span class="normal">1319</span>
<span class="normal">1320</span>
<span class="normal">1321</span>
<span class="normal">1322</span>
<span class="normal">1323</span>
<span class="normal">1324</span>
<span class="normal">1325</span>
<span class="normal">1326</span>
<span class="normal">1327</span>
<span class="normal">1328</span>
<span class="normal">1329</span>
<span class="normal">1330</span>
<span class="normal">1331</span>
<span class="normal">1332</span>
<span class="normal">1333</span>
<span class="normal">1334</span>
<span class="normal">1335</span>
<span class="normal">1336</span>
<span class="normal">1337</span>
<span class="normal">1338</span>
<span class="normal">1339</span>
<span class="normal">1340</span>
<span class="normal">1341</span>
<span class="normal">1342</span>
<span class="normal">1343</span>
<span class="normal">1344</span>
<span class="normal">1345</span>
<span class="normal">1346</span>
<span class="normal">1347</span>
<span class="normal">1348</span>
<span class="normal">1349</span>
<span class="normal">1350</span>
<span class="normal">1351</span>
<span class="normal">1352</span>
<span class="normal">1353</span>
<span class="normal">1354</span>
<span class="normal">1355</span>
<span class="normal">1356</span>
<span class="normal">1357</span>
<span class="normal">1358</span>
<span class="normal">1359</span>
<span class="normal">1360</span>
<span class="normal">1361</span>
<span class="normal">1362</span>
<span class="normal">1363</span>
<span class="normal">1364</span>
<span class="normal">1365</span>
<span class="normal">1366</span>
<span class="normal">1367</span>
<span class="normal">1368</span>
<span class="normal">1369</span>
<span class="normal">1370</span>
<span class="normal">1371</span>
<span class="normal">1372</span>
<span class="normal">1373</span>
<span class="normal">1374</span>
<span class="normal">1375</span>
<span class="normal">1376</span>
<span class="normal">1377</span>
<span class="normal">1378</span>
<span class="normal">1379</span>
<span class="normal">1380</span>
<span class="normal">1381</span>
<span class="normal">1382</span>
<span class="normal">1383</span>
<span class="normal">1384</span>
<span class="normal">1385</span>
<span class="normal">1386</span>
<span class="normal">1387</span>
<span class="normal">1388</span>
<span class="normal">1389</span>
<span class="normal">1390</span>
<span class="normal">1391</span>
<span class="normal">1392</span>
<span class="normal">1393</span>
<span class="normal">1394</span>
<span class="normal">1395</span>
<span class="normal">1396</span>
<span class="normal">1397</span>
<span class="normal">1398</span>
<span class="normal">1399</span>
<span class="normal">1400</span>
<span class="normal">1401</span>
<span class="normal">1402</span>
<span class="normal">1403</span>
<span class="normal">1404</span>
<span class="normal">1405</span>
<span class="normal">1406</span>
<span class="normal">1407</span>
<span class="normal">1408</span>
<span class="normal">1409</span>
<span class="normal">1410</span>
<span class="normal">1411</span>
<span class="normal">1412</span>
<span class="normal">1413</span>
<span class="normal">1414</span>
<span class="normal">1415</span>
<span class="normal">1416</span>
<span class="normal">1417</span>
<span class="normal">1418</span>
<span class="normal">1419</span>
<span class="normal">1420</span>
<span class="normal">1421</span>
<span class="normal">1422</span>
<span class="normal">1423</span>
<span class="normal">1424</span>
<span class="normal">1425</span>
<span class="normal">1426</span>
<span class="normal">1427</span>
<span class="normal">1428</span>
<span class="normal">1429</span>
<span class="normal">1430</span>
<span class="normal">1431</span>
<span class="normal">1432</span>
<span class="normal">1433</span>
<span class="normal">1434</span>
<span class="normal">1435</span>
<span class="normal">1436</span>
<span class="normal">1437</span>
<span class="normal">1438</span>
<span class="normal">1439</span>
<span class="normal">1440</span>
<span class="normal">1441</span>
<span class="normal">1442</span>
<span class="normal">1443</span>
<span class="normal">1444</span>
<span class="normal">1445</span>
<span class="normal">1446</span>
<span class="normal">1447</span>
<span class="normal">1448</span>
<span class="normal">1449</span>
<span class="normal">1450</span>
<span class="normal">1451</span>
<span class="normal">1452</span>
<span class="normal">1453</span>
<span class="normal">1454</span>
<span class="normal">1455</span>
<span class="normal">1456</span>
<span class="normal">1457</span>
<span class="normal">1458</span>
<span class="normal">1459</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__get_mutations</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the mutations between the two genes, populating internal arrays&quot;&quot;&quot;</span>

    <span class="n">mutations</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">amino_acid_number</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">nucleotide_number</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">nucleotide_index</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">gene_position</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">is_cds</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">is_indel</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">is_promoter</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">indel_length</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">ref_nucleotides</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">alt_nucleotides</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">is_snp</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">is_het</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">is_null</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">variants</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">:</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">codons</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">codons</span>
        <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">num</span><span class="p">,</span> <span class="n">a</span><span class="p">,</span> <span class="n">codon1</span><span class="p">,</span> <span class="n">codon2</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">amino_acid_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">amino_acid_number</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">amino_acid_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="p">):</span>
            <span class="c1"># ref.append(r)</span>
            <span class="c1"># alt.append(a)</span>
            <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)</span> <span class="o">+</span> <span class="n">a</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">a</span> <span class="o">==</span> <span class="s2">&quot;X&quot;</span><span class="p">:</span>
                <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">a</span> <span class="o">==</span> <span class="s2">&quot;Z&quot;</span><span class="p">:</span>
                <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">amino_acid_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
            <span class="n">nucleotide_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">nucleotide_index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">gene_position</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
            <span class="n">is_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">is_promoter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">ref_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">codon1</span><span class="p">)</span>
            <span class="n">alt_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">codon2</span><span class="p">)</span>

            <span class="c1"># Reconstruct the variant which caused this mutation to allow joining</span>
            <span class="c1">#   mutation and variant tables</span>
            <span class="n">codon_nucleotide_indices</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_cds</span>
            <span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">codon_number</span> <span class="o">==</span> <span class="n">num</span><span class="p">]</span>
            <span class="n">v</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">r_</span><span class="p">,</span> <span class="n">a_</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">codon1</span><span class="p">,</span> <span class="n">codon2</span><span class="p">)):</span>
                <span class="k">if</span> <span class="n">r_</span> <span class="o">!=</span> <span class="n">a_</span><span class="p">:</span>
                    <span class="n">v</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="n">codon_nucleotide_indices</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">+</span> <span class="n">r_</span> <span class="o">+</span> <span class="s2">&quot;&gt;&quot;</span> <span class="o">+</span> <span class="n">a_</span> <span class="o">+</span> <span class="s2">&quot;&amp;&quot;</span>
            <span class="k">if</span> <span class="n">v</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;&amp;&quot;</span><span class="p">:</span>
                <span class="n">v</span> <span class="o">=</span> <span class="n">v</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>

            <span class="c1"># If synonymous mutation, pull out nucleotide variants too</span>
            <span class="c1"># This lets us determine effects of mutations such as fabG1@L203L more</span>
            <span class="c1">#   precisely</span>
            <span class="k">if</span> <span class="n">r</span> <span class="o">==</span> <span class="n">a</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">rn</span><span class="p">,</span> <span class="n">an</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">codon1</span><span class="p">,</span> <span class="n">codon2</span><span class="p">)):</span>
                    <span class="k">if</span> <span class="n">rn</span> <span class="o">!=</span> <span class="n">an</span><span class="p">:</span>
                        <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rn</span> <span class="o">+</span> <span class="nb">str</span><span class="p">((</span><span class="n">num</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">3</span> <span class="o">+</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="n">an</span><span class="p">)</span>
                        <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">an</span> <span class="o">==</span> <span class="s2">&quot;x&quot;</span><span class="p">)</span>
                        <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">an</span> <span class="o">==</span> <span class="s2">&quot;z&quot;</span><span class="p">)</span>
                        <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">an</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">])</span>
                        <span class="n">amino_acid_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                        <span class="n">nucleotide_number</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">num</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">3</span> <span class="o">+</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
                        <span class="n">nucleotide_index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_number</span>
                                <span class="o">==</span> <span class="p">(</span><span class="n">num</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">3</span> <span class="o">+</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span>
                            <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                        <span class="p">)</span>
                        <span class="n">gene_position</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">num</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">3</span> <span class="o">+</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
                        <span class="n">is_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                        <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                        <span class="n">is_promoter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                        <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                        <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                        <span class="n">ref_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rn</span><span class="p">)</span>
                        <span class="n">alt_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">an</span><span class="p">)</span>

                        <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                            <span class="nb">str</span><span class="p">(</span><span class="n">codon_nucleotide_indices</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">+</span> <span class="n">rn</span> <span class="o">+</span> <span class="s2">&quot;&gt;&quot;</span> <span class="o">+</span> <span class="n">an</span>
                        <span class="p">)</span>

        <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_sequence</span>
        <span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">num</span><span class="p">,</span> <span class="n">a</span><span class="p">,</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="p">):</span>
            <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)</span> <span class="o">+</span> <span class="n">a</span><span class="p">)</span>
            <span class="c1"># ref.append(r)</span>
            <span class="c1"># alt.append(a)</span>
            <span class="n">amino_acid_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">nucleotide_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
            <span class="n">nucleotide_index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
            <span class="n">gene_position</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
            <span class="n">is_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">is_promoter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">ref_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
            <span class="n">alt_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">a</span> <span class="o">==</span> <span class="s2">&quot;x&quot;</span><span class="p">:</span>
                <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">a</span> <span class="o">==</span> <span class="s2">&quot;z&quot;</span><span class="p">:</span>
                <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>

            <span class="c1"># Reconstruct the variant which caused this mutation to allow joining</span>
            <span class="c1">#   mutation and variant tables</span>
            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span> <span class="o">+</span> <span class="n">r</span> <span class="o">+</span> <span class="s2">&quot;&gt;&quot;</span> <span class="o">+</span> <span class="n">a</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_sequence</span>

        <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">num</span><span class="p">,</span> <span class="n">a</span><span class="p">,</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="p">):</span>
            <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)</span> <span class="o">+</span> <span class="n">a</span><span class="p">)</span>
            <span class="c1"># ref.append(r)</span>
            <span class="c1"># alt.append(a)</span>
            <span class="n">amino_acid_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">nucleotide_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
            <span class="n">nucleotide_index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
            <span class="n">gene_position</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">num</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">is_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_promoter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">is_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_promoter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">ref_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
            <span class="n">alt_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">a</span> <span class="o">==</span> <span class="s2">&quot;x&quot;</span><span class="p">:</span>
                <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">a</span> <span class="o">==</span> <span class="s2">&quot;z&quot;</span><span class="p">:</span>
                <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>

            <span class="c1"># Reconstruct the variant which caused this mutation to allow joining</span>
            <span class="c1">#   mutation and variant tables</span>
            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span> <span class="o">+</span> <span class="n">r</span> <span class="o">+</span> <span class="s2">&quot;&gt;&quot;</span> <span class="o">+</span> <span class="n">a</span><span class="p">)</span>

    <span class="c1"># now let&#39;s do indels</span>
    <span class="k">assert</span> <span class="p">(</span>
        <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
            <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">&amp;</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_indel</span><span class="p">)</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="o">==</span> <span class="mi">0</span>
    <span class="p">),</span> <span class="p">(</span>
        <span class="s2">&quot;both genes have different indels at one or more of the same positions -- &quot;</span>
        <span class="s2">&quot;this cannot be easily be resolved!&quot;</span>
    <span class="p">)</span>

    <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">&amp;</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_nucleotides</span>
    <span class="p">)</span>

    <span class="k">for</span> <span class="n">num</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">alt</span><span class="p">,</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_index</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
    <span class="p">):</span>
        <span class="c1"># ref.append(None)</span>
        <span class="c1"># alt.append(None)</span>
        <span class="n">amino_acid_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">nucleotide_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
        <span class="n">nucleotide_index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
        <span class="n">gene_position</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">length</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_ins_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">alt</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_del_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">alt</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">num</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">is_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">is_promoter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">is_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">is_promoter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">length</span><span class="p">)</span>
        <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt</span><span class="p">)</span>
        <span class="n">ref_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">alt_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># Reconstruct the variant which caused this mutation to allow joining</span>
        <span class="c1">#   mutation and variant tables</span>
        <span class="n">m</span> <span class="o">=</span> <span class="n">mutations</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:]</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">:</span>
            <span class="c1"># These are slightly different as revcomp need changes</span>
            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">revcomp_indel_nc_index</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)]]</span> <span class="o">+</span> <span class="n">m</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="p">[</span>
                        <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
                        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">==</span> <span class="n">num</span>
                        <span class="p">]</span>
                    <span class="p">]</span>
                    <span class="o">+</span> <span class="n">m</span>
                <span class="p">)</span>
            <span class="p">)</span>

    <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">&amp;</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_nucleotides</span>
    <span class="p">)</span>

    <span class="k">for</span> <span class="n">num</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">alt</span><span class="p">,</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_index</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
    <span class="p">):</span>
        <span class="c1"># ref.append(None)</span>
        <span class="c1"># alt.append(None)</span>
        <span class="n">amino_acid_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">nucleotide_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
        <span class="n">nucleotide_index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
        <span class="n">gene_position</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num</span><span class="p">)</span>
        <span class="n">length</span> <span class="o">*=</span> <span class="o">-</span><span class="mi">1</span>
        <span class="k">if</span> <span class="n">length</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_ins_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">alt</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_del_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">alt</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">num</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">is_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">is_promoter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">is_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">is_promoter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">length</span><span class="p">)</span>
        <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt</span><span class="p">)</span>
        <span class="n">ref_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">alt_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># Reconstruct the variant which caused this mutation to allow joining</span>
        <span class="c1">#   mutation and variant tables</span>
        <span class="n">m</span> <span class="o">=</span> <span class="n">mutations</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:]</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">:</span>
            <span class="c1"># These are slightly different as revcomp need changes</span>
            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">revcomp_indel_nc_index</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)]]</span> <span class="o">+</span> <span class="n">m</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="p">[</span>
                        <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
                        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">==</span> <span class="n">num</span>
                        <span class="p">]</span>
                    <span class="p">]</span>
                    <span class="o">+</span> <span class="n">m</span>
                <span class="p">)</span>
            <span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mutations</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">amino_acid_number</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">nucleotide_number</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">nucleotide_index</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">gene_position</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_cds</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_promoter</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_indel</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">indel_length</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">ref_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">ref_nucleotides</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">alt_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">alt_nucleotides</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_snp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_snp</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_het</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_het</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_null</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_null</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">variants</span> <span class="o">=</span> <span class="n">variants</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="difference.GeneDifference.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">gene1</span><span class="p">,</span> <span class="n">gene2</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Constructor. Takes in two gene objects and calculates the difference in a few
    areas such as nucleotides, codons, and amino acids.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>gene1</code></td>
            <td>
                  <code><span title="gumpy.Gene">Gene</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Gene object 1</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>gene2</code></td>
            <td>
                  <code><span title="gumpy.Gene">Gene</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Gene object 2</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span>
<span class="normal">899</span>
<span class="normal">900</span>
<span class="normal">901</span>
<span class="normal">902</span>
<span class="normal">903</span>
<span class="normal">904</span>
<span class="normal">905</span>
<span class="normal">906</span>
<span class="normal">907</span>
<span class="normal">908</span>
<span class="normal">909</span>
<span class="normal">910</span>
<span class="normal">911</span>
<span class="normal">912</span>
<span class="normal">913</span>
<span class="normal">914</span>
<span class="normal">915</span>
<span class="normal">916</span>
<span class="normal">917</span>
<span class="normal">918</span>
<span class="normal">919</span>
<span class="normal">920</span>
<span class="normal">921</span>
<span class="normal">922</span>
<span class="normal">923</span>
<span class="normal">924</span>
<span class="normal">925</span>
<span class="normal">926</span>
<span class="normal">927</span>
<span class="normal">928</span>
<span class="normal">929</span>
<span class="normal">930</span>
<span class="normal">931</span>
<span class="normal">932</span>
<span class="normal">933</span>
<span class="normal">934</span>
<span class="normal">935</span>
<span class="normal">936</span>
<span class="normal">937</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene1</span><span class="p">,</span> <span class="n">gene2</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Constructor. Takes in two gene objects and calculates the difference in a few</span>
<span class="sd">        areas such as nucleotides, codons, and amino acids.</span>

<span class="sd">    Args:</span>
<span class="sd">        gene1 (gumpy.Gene): Gene object 1</span>
<span class="sd">        gene2 (gumpy.Gene): Gene object 2</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gene1</span><span class="p">,</span> <span class="n">gumpy</span><span class="o">.</span><span class="n">Gene</span><span class="p">)</span>
    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gene2</span><span class="p">,</span> <span class="n">gumpy</span><span class="o">.</span><span class="n">Gene</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">gene1</span><span class="o">.</span><span class="n">total_number_nucleotides</span> <span class="o">!=</span> <span class="n">gene2</span><span class="o">.</span><span class="n">total_number_nucleotides</span><span class="p">:</span>
        <span class="c1"># The lengths of the genes are different so comparing them is meaningless</span>
        <span class="k">raise</span> <span class="n">FailedComparison</span><span class="p">(</span>
            <span class="s2">&quot;The two genes (&quot;</span>
            <span class="o">+</span> <span class="n">gene1</span><span class="o">.</span><span class="n">name</span>
            <span class="o">+</span> <span class="s2">&quot; and &quot;</span>
            <span class="o">+</span> <span class="n">gene2</span><span class="o">.</span><span class="n">name</span>
            <span class="o">+</span> <span class="s2">&quot;) are different lengths, so comparision failed...&quot;</span>
        <span class="p">)</span>
    <span class="k">if</span> <span class="n">gene1</span><span class="o">.</span><span class="n">name</span> <span class="o">!=</span> <span class="n">gene2</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
            <span class="s2">&quot;The two genes given have different names (&quot;</span>
            <span class="o">+</span> <span class="n">gene1</span><span class="o">.</span><span class="n">name</span>
            <span class="o">+</span> <span class="s2">&quot;, &quot;</span>
            <span class="o">+</span> <span class="n">gene2</span><span class="o">.</span><span class="n">name</span>
            <span class="o">+</span> <span class="s2">&quot;) but the same length, continuing...&quot;</span><span class="p">,</span>
            <span class="ne">UserWarning</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">if</span> <span class="n">gene1</span><span class="o">.</span><span class="n">codes_protein</span> <span class="o">!=</span> <span class="n">gene2</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">FailedComparison</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;The two genes given do not have the same protein coding &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;for </span><span class="si">{</span><span class="n">gene1</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">: Gene1 = </span><span class="si">{</span><span class="n">gene1</span><span class="o">.</span><span class="n">codes_protein</span><span class="si">}</span><span class="s2">, &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;Gene2 = </span><span class="si">{</span><span class="n">gene2</span><span class="o">.</span><span class="n">codes_protein</span><span class="si">}</span><span class="s2">, so comparison failed...&quot;</span>
        <span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span> <span class="o">=</span> <span class="n">gene1</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span> <span class="o">=</span> <span class="n">gene2</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">codes_protein</span> <span class="o">=</span> <span class="n">gene1</span><span class="o">.</span><span class="n">codes_protein</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_view_method</span> <span class="o">=</span> <span class="s2">&quot;diff&quot;</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_nucleotides_full</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__nucleotides</span><span class="p">()</span>
    <span class="c1"># self.mutations = self.__mutations()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">__get_mutations</span><span class="p">()</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_codons_full</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__codons</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_amino_acids_full</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__amino_acids</span><span class="p">()</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">__large_deletions</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">__assign_vcf_evidence</span><span class="p">()</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">update_view</span><span class="p">(</span><span class="s2">&quot;diff&quot;</span><span class="p">)</span>  <span class="c1"># Use inherited method to set the view</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="difference.GeneDifference.__large_deletions" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__large_deletions</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Check to see what proportion of the gene has been deleted. Report separately
    than other mutations if &gt;=50% deleted</p>
<p>Updates internal arrays to denote a new mutation if deletion above threshold.</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 959</span>
<span class="normal"> 960</span>
<span class="normal"> 961</span>
<span class="normal"> 962</span>
<span class="normal"> 963</span>
<span class="normal"> 964</span>
<span class="normal"> 965</span>
<span class="normal"> 966</span>
<span class="normal"> 967</span>
<span class="normal"> 968</span>
<span class="normal"> 969</span>
<span class="normal"> 970</span>
<span class="normal"> 971</span>
<span class="normal"> 972</span>
<span class="normal"> 973</span>
<span class="normal"> 974</span>
<span class="normal"> 975</span>
<span class="normal"> 976</span>
<span class="normal"> 977</span>
<span class="normal"> 978</span>
<span class="normal"> 979</span>
<span class="normal"> 980</span>
<span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span>
<span class="normal">1022</span>
<span class="normal">1023</span>
<span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__large_deletions</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check to see what proportion of the gene has been deleted. Report separately</span>
<span class="sd">        than other mutations if &gt;=50% deleted</span>

<span class="sd">    Updates internal arrays to denote a new mutation if deletion above threshold.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Find everywhere that we have different deletion within coding regions</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_deleted</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_deleted</span>

    <span class="c1"># Find out how much of the gene we have deleted</span>
    <span class="n">total</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">[</span><span class="n">mask</span><span class="p">])</span> <span class="o">+</span> <span class="nb">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">[</span><span class="n">mask</span><span class="p">])</span>

    <span class="k">if</span> <span class="n">total</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># We have some deletions</span>
        <span class="n">percentage</span> <span class="o">=</span> <span class="n">total</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">percentage</span> <span class="o">&gt;=</span> <span class="mf">0.5</span><span class="p">:</span>
            <span class="c1"># More than 50% deleted, so give a percentage</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">,</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;del_</span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="n">percentage</span><span class="p">,</span><span class="w"> </span><span class="mi">2</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="c1"># Pull out the start of the deletion for vcf evidence</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">,</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">mask</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span>
            <span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span><span class="p">,</span> <span class="p">[</span><span class="kc">None</span><span class="p">])</span>

            <span class="c1"># Get the nucleotide number of the first base of the gene</span>
            <span class="c1"># Not predictable as promoters are variable length</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">,</span>
                <span class="p">[</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_deleted</span> <span class="o">|</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_deleted</span>
                    <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="p">],</span>
            <span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span><span class="p">,</span> <span class="p">[</span><span class="kc">None</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">,</span> <span class="p">[</span><span class="kc">True</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">,</span> <span class="p">[</span><span class="kc">False</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span><span class="p">,</span> <span class="p">[</span><span class="kc">True</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">,</span> <span class="p">[</span><span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="n">total</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">,</span>
                <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">])],</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ref_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ref_nucleotides</span><span class="p">,</span> <span class="p">[</span><span class="kc">None</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">alt_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alt_nucleotides</span><span class="p">,</span> <span class="p">[</span><span class="kc">None</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">is_snp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_snp</span><span class="p">,</span> <span class="p">[</span><span class="kc">False</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">is_het</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_het</span><span class="p">,</span> <span class="p">[</span><span class="kc">False</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">is_null</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_null</span><span class="p">,</span> <span class="p">[</span><span class="kc">False</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Check for a deletion at the start of the gene which is the result of</span>
            <span class="c1">#   an upstream del</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">:</span>
                <span class="c1"># This is difficult as revcomp shifts dels upstream by the length</span>
                <span class="c1">#   of the del</span>
                <span class="c1"># So we have to find a del in the RHS of the gene which would be</span>
                <span class="c1">#   the last item if not shifted</span>
                <span class="n">pos</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
                <span class="k">while</span> <span class="n">pos</span> <span class="o">*</span> <span class="o">-</span><span class="mi">1</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">:</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span>
                        <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                    <span class="p">):</span>
                        <span class="c1"># Del here in gene1 so check for validity</span>
                        <span class="k">if</span> <span class="p">(</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                            <span class="o">+</span> <span class="mi">1</span>
                            <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                        <span class="p">):</span>
                            <span class="c1"># Re-adusting for deletion, this matches</span>
                            <span class="k">break</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span>
                        <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                    <span class="p">):</span>
                        <span class="c1"># Del here in gene2 so check for validity</span>
                        <span class="k">if</span> <span class="p">(</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                            <span class="o">+</span> <span class="mi">1</span>
                            <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                        <span class="p">):</span>
                            <span class="c1"># Re-adusting for deletion, this matches</span>
                            <span class="k">break</span>
                    <span class="n">pos</span> <span class="o">-=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># If not revcomp, the del is the first item</span>
                <span class="n">pos</span> <span class="o">=</span> <span class="mi">0</span>

            <span class="k">if</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
            <span class="p">)</span> <span class="ow">and</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
            <span class="p">):</span>
                <span class="c1"># Both have a del, so no difference</span>
                <span class="k">return</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[</span><span class="n">pos</span><span class="p">]:</span>
                <span class="n">deleted</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
            <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_indel</span><span class="p">[</span><span class="n">pos</span><span class="p">]:</span>
                <span class="n">deleted</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Neither, so no difference either</span>
                <span class="k">return</span>

            <span class="c1"># Adjust idx to point to the  right VCF evidence</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">:</span>
                <span class="n">idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">pos</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">deleted</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>

            <span class="n">pos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">,</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pos</span><span class="si">}</span><span class="s2">_del_</span><span class="si">{</span><span class="n">deleted</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">])</span>
            <span class="c1"># Pull out the start of the deletion for vcf evidence</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">,</span> <span class="p">[</span><span class="n">idx</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">amino_acid_number</span><span class="p">,</span> <span class="p">[</span><span class="kc">None</span><span class="p">])</span>

            <span class="c1"># Get the nucleotide number of the first base of the gene</span>
            <span class="c1"># Not predictable as promoters are variable length</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">,</span>
                <span class="p">[</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">[</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">is_deleted</span> <span class="o">|</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">is_deleted</span>
                    <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="p">],</span>
            <span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_position</span><span class="p">,</span> <span class="p">[</span><span class="n">pos</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_cds</span><span class="p">,</span> <span class="p">[</span><span class="kc">True</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_promoter</span><span class="p">,</span> <span class="p">[</span><span class="kc">False</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span><span class="p">,</span> <span class="p">[</span><span class="kc">True</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span><span class="p">,</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">deleted</span><span class="p">)])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">,</span> <span class="p">[</span><span class="n">deleted</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ref_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ref_nucleotides</span><span class="p">,</span> <span class="p">[</span><span class="kc">None</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">alt_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alt_nucleotides</span><span class="p">,</span> <span class="p">[</span><span class="kc">None</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">is_snp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_snp</span><span class="p">,</span> <span class="p">[</span><span class="kc">False</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">is_het</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_het</span><span class="p">,</span> <span class="p">[</span><span class="kc">False</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">is_null</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_null</span><span class="p">,</span> <span class="p">[</span><span class="kc">False</span><span class="p">])</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="difference.GeneDifference.__nucleotides" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__nucleotides</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Find the differences in nucleotides</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>numpy.ndarray: Array of tuples (gene1_nucleotide, gene2_nucleotide)</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1119</span>
<span class="normal">1120</span>
<span class="normal">1121</span>
<span class="normal">1122</span>
<span class="normal">1123</span>
<span class="normal">1124</span>
<span class="normal">1125</span>
<span class="normal">1126</span>
<span class="normal">1127</span>
<span class="normal">1128</span>
<span class="normal">1129</span>
<span class="normal">1130</span>
<span class="normal">1131</span>
<span class="normal">1132</span>
<span class="normal">1133</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__nucleotides</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Find the differences in nucleotides</span>

<span class="sd">    Returns:</span>
<span class="sd">        numpy.ndarray: Array of tuples (gene1_nucleotide, gene2_nucleotide)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="p">(</span><span class="n">n1</span><span class="p">,</span> <span class="n">n2</span><span class="p">)</span>
            <span class="k">for</span> <span class="p">(</span><span class="n">n1</span><span class="p">,</span> <span class="n">n2</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">nucleotide_sequence</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">n1</span> <span class="o">!=</span> <span class="n">n2</span>
        <span class="p">]</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="difference.GeneDifference.minor_populations" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">minor_populations</span><span class="p">(</span><span class="n">interpretation</span><span class="o">=</span><span class="s1">&#39;reads&#39;</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Get the minor population mutations in GARC</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>interpretation</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>How to report minor population. 'reads'
reports number of reads. 'percentage' reports fractional read support.
Defaults to 'reads'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;reads&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.List">List</span>[str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>[str]: List of mutations in GARC</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1104</span>
<span class="normal">1105</span>
<span class="normal">1106</span>
<span class="normal">1107</span>
<span class="normal">1108</span>
<span class="normal">1109</span>
<span class="normal">1110</span>
<span class="normal">1111</span>
<span class="normal">1112</span>
<span class="normal">1113</span>
<span class="normal">1114</span>
<span class="normal">1115</span>
<span class="normal">1116</span>
<span class="normal">1117</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">minor_populations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">interpretation</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;reads&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the minor population mutations in GARC</span>

<span class="sd">    Args:</span>
<span class="sd">        interpretation (str, optional): How to report minor population. &#39;reads&#39;</span>
<span class="sd">            reports number of reads. &#39;percentage&#39; reports fractional read support.</span>
<span class="sd">            Defaults to &#39;reads&#39;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        [str]: List of mutations in GARC</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2</span><span class="o">.</span><span class="n">minority_populations_GARC</span><span class="p">(</span>
        <span class="n">interpretation</span><span class="o">=</span><span class="n">interpretation</span><span class="p">,</span> <span class="n">reference</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">gene1</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="difference.GenomeDifference" class="doc doc-heading">
            <code>GenomeDifference</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="difference.Difference" href="#difference.Difference">Difference</a></code></p>


      <p>GenomeDifference object captures the difference between two genomes. Other than
<code>snp_distance</code>, all public instance variables should be arrays
which align so <code>x[N] &lt;--&gt; y[N]</code></p>
<ul>
<li>
<p>The difference can be viewed in one of two ways:</p>
<ul>
<li><code>diff</code>: Arrays of the values from genome1 where there are different values
    in genome2, or values which exist in genome1 but not genome2 depending on
    which is more appropriate (closer to classical subtraction).
    This is the default view.</li>
<li><code>full</code>: Arrays of tuples consisting of (genome1_val, genome2_val) -
    more information but less like classical subtraction
    and more difficult to wrangle meaningful information from.</li>
<li>This option can be set by using the update_view() method</li>
</ul>
</li>
<li>
<p>Instance variables:</p>
<ul>
<li>snp_distance (int): SNP distance between the two genomes</li>
<li>variants (numpy.array): Variants in GARC</li>
<li>nucleotide_index (numpy.array): Genome index (1 based) of this variant</li>
<li>indel_length (numpy.array): Length of the indel of this variant (None is
    not an indel)</li>
<li>indel_nucleotides (numpy.array): Nucleotides of this indel of this variant
    (None if not an indel)</li>
<li>vcf_evidence (numpy.array): VCF evidence to support this call</li>
<li>gene_name (numpy.array): Name of the gene this variant affects
    (None if intergene)</li>
<li>gene_position (numpy.array): Position of this variant within this gene.
    This refers to codon number if applicable (None if intergene)</li>
<li>codon_idx (numpy.array): 0-based index of the nucleotide within the codon
    this affects if applicable. None otherwise</li>
</ul>
</li>
<li>Functions:<ul>
<li>variants(int) -&gt; dict: Takes a genome index and returns a dictionary mapping
    field-&gt;(genome1_val, genome2_val) for all fields of a vcf file
    (if applicable)</li>
<li>gene_differences() -&gt; [GeneDifference]: Returns a list of GeneDifference
    objects</li>
<li>minor_populations() -&gt; [str]: Returns a list of minor population mutations
    in GARC</li>
</ul>
</li>
<li>Inherited functions:<ul>
<li>update_view(str) -&gt; None: Used to change the viewing method for instance
    variables. Input values are either <code>diff</code> or <code>full</code></li>
</ul>
</li>
</ul>

              <details class="quote">
                <summary>Source code in <code>gumpy/difference.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span>
<span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">GenomeDifference</span><span class="p">(</span><span class="n">Difference</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    GenomeDifference object captures the difference between two genomes. Other than</span>
<span class="sd">    `snp_distance`, all public instance variables should be arrays</span>
<span class="sd">    which align so `x[N] &lt;--&gt; y[N]`</span>

<span class="sd">    * The difference can be viewed in one of two ways:</span>
<span class="sd">        * `diff`: Arrays of the values from genome1 where there are different values</span>
<span class="sd">            in genome2, or values which exist in genome1 but not genome2 depending on</span>
<span class="sd">            which is more appropriate (closer to classical subtraction).</span>
<span class="sd">            This is the default view.</span>
<span class="sd">        * `full`: Arrays of tuples consisting of (genome1_val, genome2_val) -</span>
<span class="sd">            more information but less like classical subtraction</span>
<span class="sd">            and more difficult to wrangle meaningful information from.</span>
<span class="sd">        * This option can be set by using the update_view() method</span>

<span class="sd">    * Instance variables:</span>
<span class="sd">        * snp_distance (int): SNP distance between the two genomes</span>
<span class="sd">        * variants (numpy.array): Variants in GARC</span>
<span class="sd">        * nucleotide_index (numpy.array): Genome index (1 based) of this variant</span>
<span class="sd">        * indel_length (numpy.array): Length of the indel of this variant (None is</span>
<span class="sd">            not an indel)</span>
<span class="sd">        * indel_nucleotides (numpy.array): Nucleotides of this indel of this variant</span>
<span class="sd">            (None if not an indel)</span>
<span class="sd">        * vcf_evidence (numpy.array): VCF evidence to support this call</span>
<span class="sd">        * gene_name (numpy.array): Name of the gene this variant affects</span>
<span class="sd">            (None if intergene)</span>
<span class="sd">        * gene_position (numpy.array): Position of this variant within this gene.</span>
<span class="sd">            This refers to codon number if applicable (None if intergene)</span>
<span class="sd">        * codon_idx (numpy.array): 0-based index of the nucleotide within the codon</span>
<span class="sd">            this affects if applicable. None otherwise</span>
<span class="sd">    * Functions:</span>
<span class="sd">        * variants(int) -&gt; dict: Takes a genome index and returns a dictionary mapping</span>
<span class="sd">            field-&gt;(genome1_val, genome2_val) for all fields of a vcf file</span>
<span class="sd">            (if applicable)</span>
<span class="sd">        * gene_differences() -&gt; [GeneDifference]: Returns a list of GeneDifference</span>
<span class="sd">            objects</span>
<span class="sd">        * minor_populations() -&gt; [str]: Returns a list of minor population mutations</span>
<span class="sd">            in GARC</span>
<span class="sd">    * Inherited functions:</span>
<span class="sd">        * update_view(str) -&gt; None: Used to change the viewing method for instance</span>
<span class="sd">            variables. Input values are either `diff` or `full`</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome1</span><span class="p">,</span> <span class="n">genome2</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Constructor for the GenomeDifference object.</span>

<span class="sd">        Called implictly when `genome1.difference(genome2)` is performed.</span>
<span class="sd">        Args:</span>
<span class="sd">            genome1 (gumpy.Genome): A Genome object to compare against</span>
<span class="sd">            genome2 (gumpy.Genome): The other Genome object</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># insist that both must be Genome objects</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">genome1</span><span class="p">,</span> <span class="n">gumpy</span><span class="o">.</span><span class="n">Genome</span><span class="p">)</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">genome2</span><span class="p">,</span> <span class="n">gumpy</span><span class="o">.</span><span class="n">Genome</span><span class="p">)</span>

        <span class="c1"># Checking for the same genes, give an error if the genes are different</span>
        <span class="k">if</span> <span class="n">genome1</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="o">!=</span> <span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">genes_in_1_not_2</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">genome1</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span>
                <span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
            <span class="p">)</span>
            <span class="n">genes_in_2_not_1</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span>
                <span class="n">genome1</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
            <span class="p">)</span>
            <span class="n">message</span> <span class="o">=</span> <span class="s2">&quot;The two genomes have different genes.&quot;</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes_in_1_not_2</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">message</span> <span class="o">+=</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot; Genome 1 has </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">genes_in_1_not_2</span><span class="p">)</span><span class="si">}</span><span class="s2"> more &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;genes: </span><span class="si">{</span><span class="n">genes_in_1_not_2</span><span class="si">}</span><span class="s2">.&quot;</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes_in_2_not_1</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">message</span> <span class="o">+=</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot; Genome2 has </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">genes_in_2_not_1</span><span class="p">)</span><span class="si">}</span><span class="s2"> more &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;genes: </span><span class="si">{</span><span class="n">genes_in_2_not_1</span><span class="si">}</span><span class="s2">.&quot;</span>
                <span class="p">)</span>
            <span class="k">raise</span> <span class="n">FailedComparison</span><span class="p">(</span><span class="n">message</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span> <span class="o">=</span> <span class="n">genome1</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span> <span class="o">=</span> <span class="n">genome2</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_view_method</span> <span class="o">=</span> <span class="s2">&quot;diff&quot;</span>

        <span class="c1"># Calculate SNPs</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">snp_distance</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__snp_distance</span><span class="p">()</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Where applicable, the `full` difference arrays are stored as these can be </span>
<span class="sd">        easily converted into `diff` arrays but not the other way around.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">__get_variants</span><span class="p">()</span>

        <span class="c1"># Calculate differences</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_nucleotides_full</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__nucleotides</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">__assign_vcf_evidence</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">update_view</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_view_method</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_gene_pos</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">idx</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">variant</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">start</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span> <span class="o">|</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Find the gene position of a given nucleotide index.</span>
<span class="sd">        This is considerably faster than building a whole stacked_gene_pos array</span>
<span class="sd">            (takes ~4.5mins for tb)</span>

<span class="sd">        Args:</span>
<span class="sd">            gene (str): Name of the gene to search</span>
<span class="sd">            idx (int): Nucleotide index we want the gene position of</span>
<span class="sd">            variant (str): Variant we&#39;re looking for in GARC</span>
<span class="sd">            start (int): Start position. Defaults to None</span>

<span class="sd">        Returns:</span>
<span class="sd">            int | None: Gene position of this nucleotide index</span>
<span class="sd">                        (or None if it lies outside of the gene)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">stacked_gene_mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
        <span class="n">nc_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span><span class="p">[</span><span class="n">stacked_gene_mask</span><span class="p">]</span>
        <span class="n">nc_nums</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span><span class="n">stacked_gene_mask</span><span class="p">]</span>

        <span class="c1"># Get the gene&#39;s nucleotide number</span>
        <span class="n">nc_num</span> <span class="o">=</span> <span class="n">nc_nums</span><span class="p">[</span><span class="n">nc_idx</span> <span class="o">==</span> <span class="n">idx</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

        <span class="c1"># If coding SNP, pull out the codon number instead</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]</span>
            <span class="ow">and</span> <span class="n">nc_num</span> <span class="o">&gt;</span> <span class="mi">0</span>
            <span class="ow">and</span> <span class="s2">&quot;ins&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variant</span>
            <span class="ow">and</span> <span class="s2">&quot;del&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variant</span>
        <span class="p">):</span>
            <span class="c1"># Use floor division (//)</span>
            <span class="n">codon_number</span> <span class="o">=</span> <span class="p">(</span><span class="n">nc_num</span> <span class="o">+</span> <span class="mi">2</span><span class="p">)</span> <span class="o">//</span> <span class="mi">3</span>
            <span class="k">return</span> <span class="n">codon_number</span>
        <span class="k">elif</span> <span class="s2">&quot;ins&quot;</span> <span class="ow">in</span> <span class="n">variant</span><span class="p">:</span>
            <span class="c1"># Insertions need a little nudge in revcomp because of `ins at, del after`</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;reverse_complement&quot;</span><span class="p">]:</span>
                <span class="n">nc_num</span> <span class="o">-=</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="s2">&quot;del&quot;</span> <span class="ow">in</span> <span class="n">variant</span><span class="p">:</span>
            <span class="n">pos</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="n">bases</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">start</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">idx</span> <span class="o">!=</span> <span class="n">start</span><span class="p">:</span>
                <span class="c1"># This is not the start of the deletion, so truncate the bases as</span>
                <span class="c1">#   appropriate</span>
                <span class="n">dels</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">bases</span><span class="p">)</span> <span class="o">-</span> <span class="p">(</span><span class="n">idx</span> <span class="o">-</span> <span class="n">start</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">dels</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">bases</span><span class="p">)</span>
            <span class="c1"># Deletions need even more nudging in revcomp because the entire deletion</span>
            <span class="c1">#   is reversed so starts at the end</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;reverse_complement&quot;</span><span class="p">]:</span>
                <span class="n">nc_num</span> <span class="o">=</span> <span class="n">nc_num</span> <span class="o">-</span> <span class="n">dels</span> <span class="o">+</span> <span class="mi">1</span>

                <span class="c1"># Edge case of deletion starting in revcomp gene and extending</span>
                <span class="c1"># past gene start, so return None</span>
                <span class="k">if</span> <span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                    <span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                    <span class="o">&gt;</span> <span class="n">nc_num</span>
                <span class="p">):</span>
                    <span class="k">return</span> <span class="kc">None</span>

        <span class="k">return</span> <span class="n">nc_num</span>

    <span class="k">def</span> <span class="nf">_get_vcf_idx</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">vcf_row</span><span class="p">:</span> <span class="n">Dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span> <span class="o">|</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Given a vcf row, figure out which alt it refers to. Should be available in</span>
<span class="sd">            the GT field</span>

<span class="sd">        Args:</span>
<span class="sd">            vcf_row (dict): Internal representation of the VCF row for this variant</span>

<span class="sd">        Returns:</span>
<span class="sd">            int | None: 0-based index of the alts this refers to</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Use `GT` field to figure out which one this call is</span>
        <span class="c1"># (This will be **much** messier for minor populations)</span>
        <span class="n">gt</span> <span class="o">=</span> <span class="n">vcf_row</span><span class="p">[</span><span class="s2">&quot;GT&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">gt</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">gt</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
            <span class="c1"># Actual call so return the 1-based idx</span>
            <span class="k">return</span> <span class="n">gt</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Het call so return None as it involves &gt;1</span>
            <span class="k">return</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">__assign_vcf_evidence</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Once we have pulled out all of the variants, find the VCF evidence</span>
<span class="sd">        (if existing) for each.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">evidences</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">indices</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">):</span>
            <span class="n">evidence1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
            <span class="n">evidence2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>

            <span class="c1"># As we add the vcf evidence for every deleted base</span>
            <span class="c1"># This should only be reported for the actual start</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
                <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">0</span>
            <span class="p">):</span>
                <span class="n">evidence1</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
                <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">0</span>
            <span class="p">):</span>
                <span class="n">evidence2</span> <span class="o">=</span> <span class="kc">None</span>

            <span class="k">if</span> <span class="n">evidence1</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">evidence2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="c1"># We have a collision. For now, just concat FIXME</span>
                <span class="n">evidences</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">evidence1</span><span class="p">,</span> <span class="n">evidence2</span><span class="p">])</span>
                <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_get_vcf_idx</span><span class="p">(</span><span class="n">evidence1</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_vcf_idx</span><span class="p">(</span><span class="n">evidence2</span><span class="p">)]</span>
                <span class="p">)</span>
            <span class="k">elif</span> <span class="n">evidence1</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">evidences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">evidence1</span><span class="p">)</span>
                <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_get_vcf_idx</span><span class="p">(</span><span class="n">evidence1</span><span class="p">))</span>
            <span class="k">elif</span> <span class="n">evidence2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">evidences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">evidence2</span><span class="p">)</span>
                <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_get_vcf_idx</span><span class="p">(</span><span class="n">evidence2</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">evidences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidences</span> <span class="o">=</span> <span class="n">evidences</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vcf_idx</span> <span class="o">=</span> <span class="n">indices</span>

    <span class="k">def</span> <span class="nf">minor_populations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">interpretation</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;reads&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get the minor population mutations in GARC</span>

<span class="sd">        Args:</span>
<span class="sd">            interpretation (str, optional): How to report minor population.</span>
<span class="sd">                &#39;reads&#39; reports number of reads. &#39;percentage&#39; reports fractional</span>
<span class="sd">                read support. Defaults to &#39;reads&#39;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[str]: List of mutations in GARC</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">minority_populations_GARC</span><span class="p">(</span>
            <span class="n">interpretation</span><span class="o">=</span><span class="n">interpretation</span><span class="p">,</span> <span class="n">reference</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">genome1</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">__snp_distance</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculates the SNP distance between the two genomes</span>

<span class="sd">        Returns:</span>
<span class="sd">            int: The SNP distance between the two genomes</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Ignores `z` and `x`</span>
        <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="mi">1</span>
                <span class="k">for</span> <span class="p">(</span><span class="n">base1</span><span class="p">,</span> <span class="n">base2</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">nucleotide_sequence</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="n">base1</span> <span class="o">!=</span> <span class="n">base2</span>
                <span class="ow">and</span> <span class="n">base1</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;t&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">,</span> <span class="s2">&quot;g&quot;</span><span class="p">]</span>
                <span class="ow">and</span> <span class="n">base2</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;t&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">,</span> <span class="s2">&quot;g&quot;</span><span class="p">]</span>
            <span class="p">]</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">__get_variants</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get the variants between the two genomes, populating the internal arrays&quot;&quot;&quot;</span>

        <span class="n">variants</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">indices</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">is_snp</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">is_het</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">is_null</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">is_indel</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">indel_length</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">refs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">alts</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">gene_name</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">gene_pos</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">codon_idx</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># first do the SNPs, HETs and NULLs</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">nucleotide_sequence</span>

        <span class="c1"># for now we simply allow the ref and alt to be different e.g. can have a</span>
        <span class="c1">#   SNP on a NULL (x&gt;t)</span>
        <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">idx</span><span class="p">,</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span>
            <span class="nb">zip</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="p">),</span>
            <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">]),</span>
        <span class="p">):</span>
            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span> <span class="o">+</span> <span class="n">r</span> <span class="o">+</span> <span class="s2">&quot;&gt;&quot;</span> <span class="o">+</span> <span class="n">a</span><span class="p">)</span>
            <span class="n">refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
            <span class="n">alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
            <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
            <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
            <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">a</span> <span class="o">==</span> <span class="s2">&quot;z&quot;</span><span class="p">:</span>
                <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">a</span> <span class="o">==</span> <span class="s2">&quot;x&quot;</span><span class="p">:</span>
                <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">a</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;t&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">,</span> <span class="s2">&quot;g&quot;</span><span class="p">]:</span>
                <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

            <span class="c1"># Find the genes at this pos</span>
            <span class="n">genes</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                <span class="nb">list</span><span class="p">(</span>
                    <span class="nb">set</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span> <span class="o">==</span> <span class="n">idx</span>
                        <span class="p">]</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="c1"># If we have genes, we need to duplicate some info</span>
                <span class="n">first</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">gene</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                        <span class="c1"># If we have genes, we don&#39;t care about this one</span>
                        <span class="k">continue</span>
                    <span class="n">gene_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                    <span class="n">gene_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_gene_pos</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">idx</span><span class="p">,</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]</span>
                        <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                        <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span>
                    <span class="p">):</span>
                        <span class="c1"># Get codon idx</span>
                        <span class="n">nc_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                        <span class="p">]</span>
                        <span class="n">nc_num</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                        <span class="p">]</span>
                        <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">nc_num</span><span class="p">[</span><span class="n">nc_idx</span> <span class="o">==</span> <span class="n">idx</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>

                    <span class="c1"># If this isn&#39;t the first one, we need to duplicate the row</span>
                    <span class="k">if</span> <span class="n">first</span><span class="p">:</span>
                        <span class="n">first</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">refs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alts</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indices</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_indel</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_length</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_het</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_snp</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_null</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># We have 1 gene or none, so set to None if no gene is present</span>
                <span class="n">gene</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">genes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="k">else</span> <span class="kc">None</span>
                <span class="n">gene_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">gene</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="c1"># Single gene, so pull out data</span>
                    <span class="n">gene_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_gene_pos</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">idx</span><span class="p">,</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>

                    <span class="k">if</span> <span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]</span>
                        <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                        <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span>
                    <span class="p">):</span>
                        <span class="c1"># Get codon idx</span>
                        <span class="n">nc_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                        <span class="p">]</span>
                        <span class="n">nc_num</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                        <span class="p">]</span>
                        <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">nc_num</span><span class="p">[</span><span class="n">nc_idx</span> <span class="o">==</span> <span class="n">idx</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">gene_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                    <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>

        <span class="c1"># INDELs are trickier: we have to deal with the case where both genomes have</span>
        <span class="c1">#   an indel at the same position</span>
        <span class="c1"># if they are different we catch fire since that is too difficult to parse</span>
        <span class="c1">#   right now</span>
        <span class="c1"># if they are the same, then there is no difference</span>
        <span class="c1"># the other cases are the usual one of there being an indel on the RHS and then</span>
        <span class="c1">#   an indel on the LHS (but not the RHS)</span>
        <span class="c1"># for the latter we &#39;reverse&#39; the indel e.g. 1000_ins_at - None becomes 1000_del</span>
        <span class="c1">#   at</span>
        <span class="c1"># subtraction becomes &quot;how do we go from the LHS to the RHS in a-b?&quot;</span>

        <span class="c1"># for indels catch fire if both genomes have different indels at the same</span>
        <span class="c1">#   position</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">is_indel</span>
                <span class="o">&amp;</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">is_indel</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="o">==</span> <span class="mi">0</span>
        <span class="p">),</span> <span class="p">(</span>
            <span class="s2">&quot;both genomes have indels of different lengths at one or more of the &quot;</span>
            <span class="s2">&quot;same positions -- this cannot be easily resolved!&quot;</span>
        <span class="p">)</span>

        <span class="c1"># the other case is where there is an identical indel at a position but</span>
        <span class="c1">#   this leads to a difference of zero!</span>

        <span class="c1"># if the indel is on the RHS, then it is unchanged</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">&amp;</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">indel_nucleotides</span>
        <span class="p">)</span>

        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">alt</span><span class="p">,</span> <span class="n">r</span><span class="p">,</span> <span class="n">a</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">,</span>
        <span class="p">):</span>
            <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
            <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">length</span><span class="p">)</span>
            <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">length</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_ins_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">alt</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_del_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">alt</span><span class="p">))</span>

            <span class="c1"># Find the genes at this pos</span>
            <span class="n">positions</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">seen_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="k">if</span> <span class="s2">&quot;ins&quot;</span> <span class="ow">in</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
                <span class="c1"># Insertion, so only need to check a single index</span>
                <span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
                <span class="n">start</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Deletion, so we need to check every deleted position for genes</span>
                <span class="n">pos</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">bases</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
                <span class="n">start</span> <span class="o">=</span> <span class="n">idx</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">bases</span><span class="p">)):</span>
                    <span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span>
            <span class="n">first</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">positions</span><span class="p">:</span>
                <span class="n">genes</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                    <span class="nb">list</span><span class="p">(</span>
                        <span class="nb">set</span><span class="p">(</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span><span class="p">[</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span> <span class="o">==</span> <span class="n">pos</span>
                            <span class="p">]</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="n">adding</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">multi</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="c1"># If we have genes, we need to duplicate some info</span>
                    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">gene</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                            <span class="c1"># If we have genes, we don&#39;t care about this one</span>
                            <span class="k">continue</span>

                        <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">seen_genes</span><span class="p">:</span>
                            <span class="c1"># If we&#39;ve already seen it, skip</span>
                            <span class="k">continue</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="c1"># Mark as seen now we have seen it</span>
                            <span class="n">seen_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                            <span class="n">adding</span> <span class="o">=</span> <span class="kc">True</span>
                        <span class="n">gene_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                        <span class="n">gene_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">get_gene_pos</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">)</span>
                        <span class="p">)</span>
                        <span class="k">if</span> <span class="p">(</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]</span>
                            <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                            <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span>
                        <span class="p">):</span>
                            <span class="c1"># Get codon pos</span>
                            <span class="n">nc_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span><span class="p">[</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                            <span class="p">]</span>
                            <span class="n">nc_num</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                            <span class="p">]</span>
                            <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">nc_num</span><span class="p">[</span><span class="n">nc_idx</span> <span class="o">==</span> <span class="n">pos</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>

                        <span class="c1"># If this isn&#39;t the first one, we need to duplicate the row</span>
                        <span class="k">if</span> <span class="n">first</span><span class="p">:</span>
                            <span class="n">first</span> <span class="o">=</span> <span class="kc">False</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                            <span class="n">refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">refs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                            <span class="n">alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                            <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indices</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                            <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_indel</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                            <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_length</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                            <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                            <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_het</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                            <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_snp</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                            <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_null</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># We have 1 gene or none, so set to None if no gene is present</span>
                    <span class="n">gene</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">genes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="k">else</span> <span class="kc">None</span>
                    <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">seen_genes</span><span class="p">:</span>
                        <span class="c1"># If we&#39;ve already seen it, or it&#39;s intragene so skip</span>
                        <span class="k">continue</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># Mark as seen now we have seen it</span>
                        <span class="n">seen_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                        <span class="n">adding</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="k">if</span> <span class="n">gene</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="c1"># Single gene, so pull out data</span>
                        <span class="n">gene_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                        <span class="n">gene_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">get_gene_pos</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">)</span>
                        <span class="p">)</span>

                        <span class="k">if</span> <span class="p">(</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]</span>
                            <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                            <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span>
                        <span class="p">):</span>
                            <span class="c1"># Get codon pos</span>
                            <span class="n">nc_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span><span class="p">[</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                            <span class="p">]</span>
                            <span class="n">nc_num</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                            <span class="p">]</span>
                            <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">nc_num</span><span class="p">[</span><span class="n">nc_idx</span> <span class="o">==</span> <span class="n">pos</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># We don&#39;t care about intragene values if its not the first item</span>
                        <span class="k">if</span> <span class="n">pos</span> <span class="o">!=</span> <span class="n">idx</span><span class="p">:</span>
                            <span class="k">continue</span>
                        <span class="n">gene_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                        <span class="n">gene_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                        <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="c1"># If this isn&#39;t the first one, we need to duplicate the row</span>
                <span class="k">if</span> <span class="n">first</span><span class="p">:</span>
                    <span class="n">first</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">elif</span> <span class="n">adding</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">multi</span><span class="p">:</span>
                    <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">refs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indices</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_indel</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_length</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_het</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_snp</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_null</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

        <span class="c1"># if the indel is on the LHS, then it is unchanged, then it needs &#39;reversing&#39;</span>
        <span class="c1">#   since we are returning how to get to the RHS from the LHS hence we</span>
        <span class="c1">#   delete an insertion etc</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">&amp;</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">indel_nucleotides</span>
        <span class="p">)</span>

        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">alt</span><span class="p">,</span> <span class="n">r</span><span class="p">,</span> <span class="n">a</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">,</span>
        <span class="p">):</span>
            <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
            <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">length</span> <span class="o">*=</span> <span class="o">-</span><span class="mi">1</span>
            <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">length</span><span class="p">)</span>
            <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">length</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_ins_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">alt</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_del_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">alt</span><span class="p">))</span>

            <span class="c1"># Find the genes at this pos</span>
            <span class="n">positions</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">seen_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="k">if</span> <span class="s2">&quot;ins&quot;</span> <span class="ow">in</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
                <span class="c1"># Insertion, so only need to check a single index</span>
                <span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
                <span class="n">start</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Deletion, so we need to check every deleted position for genes</span>
                <span class="n">pos</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">bases</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
                <span class="n">start</span> <span class="o">=</span> <span class="n">idx</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">bases</span><span class="p">)):</span>
                    <span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span>
            <span class="n">first</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">positions</span><span class="p">:</span>
                <span class="n">genes</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                    <span class="nb">list</span><span class="p">(</span>
                        <span class="nb">set</span><span class="p">(</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span><span class="p">[</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span> <span class="o">==</span> <span class="n">pos</span>
                            <span class="p">]</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="n">adding</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="n">multi</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">multi</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="c1"># If we have genes, we need to duplicate some info</span>
                    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">gene</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                            <span class="c1"># If we have genes, we don&#39;t care about this one</span>
                            <span class="k">continue</span>

                        <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">seen_genes</span><span class="p">:</span>
                            <span class="c1"># If we&#39;ve already seen it, skip</span>
                            <span class="k">continue</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="c1"># Mark as seen now we have seen it</span>
                            <span class="n">seen_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                            <span class="n">adding</span> <span class="o">=</span> <span class="kc">True</span>

                        <span class="n">gene_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                        <span class="n">gene_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">get_gene_pos</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">)</span>
                        <span class="p">)</span>
                        <span class="k">if</span> <span class="p">(</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]</span>
                            <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                            <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span>
                        <span class="p">):</span>
                            <span class="c1"># Get codon pos</span>
                            <span class="n">nc_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span><span class="p">[</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                            <span class="p">]</span>
                            <span class="n">nc_num</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                            <span class="p">]</span>
                            <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">nc_num</span><span class="p">[</span><span class="n">nc_idx</span> <span class="o">==</span> <span class="n">pos</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                        <span class="c1"># If this isn&#39;t the first one, we need to duplicate the row</span>
                        <span class="k">if</span> <span class="n">first</span><span class="p">:</span>
                            <span class="n">first</span> <span class="o">=</span> <span class="kc">False</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                            <span class="n">refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">refs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                            <span class="n">alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                            <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indices</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                            <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_indel</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                            <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_length</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                            <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                            <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_het</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                            <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_snp</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                            <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_null</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># We have 1 gene or none, so set to None if no gene is present</span>
                    <span class="n">gene</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">genes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="k">else</span> <span class="kc">None</span>
                    <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">seen_genes</span><span class="p">:</span>
                        <span class="c1"># If we&#39;ve already seen it, or it&#39;s intragene so skip</span>
                        <span class="k">continue</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># Mark as seen now we have seen it</span>
                        <span class="n">seen_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                        <span class="n">adding</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="k">if</span> <span class="n">gene</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="c1"># Single gene, so pull out data</span>
                        <span class="n">gene_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                        <span class="n">gene_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">get_gene_pos</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">)</span>
                        <span class="p">)</span>

                        <span class="k">if</span> <span class="p">(</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]</span>
                            <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                            <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span>
                        <span class="p">):</span>
                            <span class="c1"># Get codon pos</span>
                            <span class="n">nc_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span><span class="p">[</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                            <span class="p">]</span>
                            <span class="n">nc_num</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                            <span class="p">]</span>
                            <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">nc_num</span><span class="p">[</span><span class="n">nc_idx</span> <span class="o">==</span> <span class="n">pos</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># We don&#39;t care about intragene values if its not the first item</span>
                        <span class="k">if</span> <span class="n">pos</span> <span class="o">!=</span> <span class="n">idx</span><span class="p">:</span>
                            <span class="k">continue</span>
                        <span class="n">gene_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                        <span class="n">gene_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                        <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>

                <span class="c1"># If this isn&#39;t the first one, we need to duplicate the row</span>
                <span class="k">if</span> <span class="n">first</span><span class="p">:</span>
                    <span class="n">first</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">elif</span> <span class="n">adding</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">multi</span><span class="p">:</span>
                    <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">refs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indices</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_indel</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_length</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_het</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_snp</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_null</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">variants</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_indel</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">indel_length</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_snp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_snp</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_het</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_het</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_null</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_null</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_pos</span> <span class="o">=</span> <span class="n">gene_pos</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">codon_idx</span> <span class="o">=</span> <span class="n">codon_idx</span>

    <span class="k">def</span> <span class="nf">__nucleotides</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate the difference in nucleotides</span>
<span class="sd">        Returns:</span>
<span class="sd">            numpy.ndarray: Numpy array of tuples of (genome1_nucleotide,</span>
<span class="sd">                genome2_nucleotide)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">nucleotide_sequence</span>
        <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="nb">list</span><span class="p">(</span>
                <span class="nb">zip</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="difference.GenomeDifference.snp_distance" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">snp_distance</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__snp_distance</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

      <p>Where applicable, the <code>full</code> difference arrays are stored as these can be 
easily converted into <code>diff</code> arrays but not the other way around.</p>
    </div>

</div>



<div class="doc doc-object doc-function">


<h3 id="difference.GenomeDifference.__assign_vcf_evidence" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__assign_vcf_evidence</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Once we have pulled out all of the variants, find the VCF evidence
(if existing) for each.</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__assign_vcf_evidence</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Once we have pulled out all of the variants, find the VCF evidence</span>
<span class="sd">    (if existing) for each.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">evidences</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">indices</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">):</span>
        <span class="n">evidence1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
        <span class="n">evidence2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">vcf_evidence</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>

        <span class="c1"># As we add the vcf evidence for every deleted base</span>
        <span class="c1"># This should only be reported for the actual start</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
            <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">0</span>
        <span class="p">):</span>
            <span class="n">evidence1</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">is_deleted</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
            <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">0</span>
        <span class="p">):</span>
            <span class="n">evidence2</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="n">evidence1</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">evidence2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># We have a collision. For now, just concat FIXME</span>
            <span class="n">evidences</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">evidence1</span><span class="p">,</span> <span class="n">evidence2</span><span class="p">])</span>
            <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_get_vcf_idx</span><span class="p">(</span><span class="n">evidence1</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_vcf_idx</span><span class="p">(</span><span class="n">evidence2</span><span class="p">)]</span>
            <span class="p">)</span>
        <span class="k">elif</span> <span class="n">evidence1</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">evidences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">evidence1</span><span class="p">)</span>
            <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_get_vcf_idx</span><span class="p">(</span><span class="n">evidence1</span><span class="p">))</span>
        <span class="k">elif</span> <span class="n">evidence2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">evidences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">evidence2</span><span class="p">)</span>
            <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_get_vcf_idx</span><span class="p">(</span><span class="n">evidence2</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">evidences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">vcf_evidences</span> <span class="o">=</span> <span class="n">evidences</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">vcf_idx</span> <span class="o">=</span> <span class="n">indices</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="difference.GenomeDifference.__get_variants" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__get_variants</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Get the variants between the two genomes, populating the internal arrays</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__get_variants</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the variants between the two genomes, populating the internal arrays&quot;&quot;&quot;</span>

    <span class="n">variants</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">is_snp</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">is_het</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">is_null</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">is_indel</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">indel_length</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">refs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">alts</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">gene_name</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">gene_pos</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">codon_idx</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># first do the SNPs, HETs and NULLs</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">nucleotide_sequence</span>

    <span class="c1"># for now we simply allow the ref and alt to be different e.g. can have a</span>
    <span class="c1">#   SNP on a NULL (x&gt;t)</span>
    <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">idx</span><span class="p">,</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span>
        <span class="nb">zip</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="p">),</span>
        <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">]),</span>
    <span class="p">):</span>
        <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span> <span class="o">+</span> <span class="n">r</span> <span class="o">+</span> <span class="s2">&quot;&gt;&quot;</span> <span class="o">+</span> <span class="n">a</span><span class="p">)</span>
        <span class="n">refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
        <span class="n">alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
        <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
        <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">a</span> <span class="o">==</span> <span class="s2">&quot;z&quot;</span><span class="p">:</span>
            <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">a</span> <span class="o">==</span> <span class="s2">&quot;x&quot;</span><span class="p">:</span>
            <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">a</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;t&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">,</span> <span class="s2">&quot;g&quot;</span><span class="p">]:</span>
            <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># Find the genes at this pos</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="nb">list</span><span class="p">(</span>
                <span class="nb">set</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span><span class="p">[</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span> <span class="o">==</span> <span class="n">idx</span>
                    <span class="p">]</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="c1"># If we have genes, we need to duplicate some info</span>
            <span class="n">first</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">gene</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                    <span class="c1"># If we have genes, we don&#39;t care about this one</span>
                    <span class="k">continue</span>
                <span class="n">gene_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                <span class="n">gene_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_gene_pos</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">idx</span><span class="p">,</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>
                <span class="k">if</span> <span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]</span>
                    <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                    <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span>
                <span class="p">):</span>
                    <span class="c1"># Get codon idx</span>
                    <span class="n">nc_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span><span class="p">[</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                    <span class="p">]</span>
                    <span class="n">nc_num</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                    <span class="p">]</span>
                    <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">nc_num</span><span class="p">[</span><span class="n">nc_idx</span> <span class="o">==</span> <span class="n">idx</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>

                <span class="c1"># If this isn&#39;t the first one, we need to duplicate the row</span>
                <span class="k">if</span> <span class="n">first</span><span class="p">:</span>
                    <span class="n">first</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">refs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alts</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indices</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_indel</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_length</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_het</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_snp</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_null</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># We have 1 gene or none, so set to None if no gene is present</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">genes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="k">else</span> <span class="kc">None</span>
            <span class="n">gene_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">gene</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="c1"># Single gene, so pull out data</span>
                <span class="n">gene_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_gene_pos</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">idx</span><span class="p">,</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>

                <span class="k">if</span> <span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]</span>
                    <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                    <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span>
                <span class="p">):</span>
                    <span class="c1"># Get codon idx</span>
                    <span class="n">nc_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span><span class="p">[</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                    <span class="p">]</span>
                    <span class="n">nc_num</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                    <span class="p">]</span>
                    <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">nc_num</span><span class="p">[</span><span class="n">nc_idx</span> <span class="o">==</span> <span class="n">idx</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">gene_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>

    <span class="c1"># INDELs are trickier: we have to deal with the case where both genomes have</span>
    <span class="c1">#   an indel at the same position</span>
    <span class="c1"># if they are different we catch fire since that is too difficult to parse</span>
    <span class="c1">#   right now</span>
    <span class="c1"># if they are the same, then there is no difference</span>
    <span class="c1"># the other cases are the usual one of there being an indel on the RHS and then</span>
    <span class="c1">#   an indel on the LHS (but not the RHS)</span>
    <span class="c1"># for the latter we &#39;reverse&#39; the indel e.g. 1000_ins_at - None becomes 1000_del</span>
    <span class="c1">#   at</span>
    <span class="c1"># subtraction becomes &quot;how do we go from the LHS to the RHS in a-b?&quot;</span>

    <span class="c1"># for indels catch fire if both genomes have different indels at the same</span>
    <span class="c1">#   position</span>
    <span class="k">assert</span> <span class="p">(</span>
        <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">is_indel</span>
            <span class="o">&amp;</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">is_indel</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="o">==</span> <span class="mi">0</span>
    <span class="p">),</span> <span class="p">(</span>
        <span class="s2">&quot;both genomes have indels of different lengths at one or more of the &quot;</span>
        <span class="s2">&quot;same positions -- this cannot be easily resolved!&quot;</span>
    <span class="p">)</span>

    <span class="c1"># the other case is where there is an identical indel at a position but</span>
    <span class="c1">#   this leads to a difference of zero!</span>

    <span class="c1"># if the indel is on the RHS, then it is unchanged</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">&amp;</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">indel_nucleotides</span>
    <span class="p">)</span>

    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">alt</span><span class="p">,</span> <span class="n">r</span><span class="p">,</span> <span class="n">a</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
        <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">length</span><span class="p">)</span>
        <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">length</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_ins_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">alt</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_del_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">alt</span><span class="p">))</span>

        <span class="c1"># Find the genes at this pos</span>
        <span class="n">positions</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">seen_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">if</span> <span class="s2">&quot;ins&quot;</span> <span class="ow">in</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
            <span class="c1"># Insertion, so only need to check a single index</span>
            <span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
            <span class="n">start</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Deletion, so we need to check every deleted position for genes</span>
            <span class="n">pos</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">bases</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">idx</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">bases</span><span class="p">)):</span>
                <span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span>
        <span class="n">first</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">positions</span><span class="p">:</span>
            <span class="n">genes</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                <span class="nb">list</span><span class="p">(</span>
                    <span class="nb">set</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span> <span class="o">==</span> <span class="n">pos</span>
                        <span class="p">]</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="n">adding</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">multi</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="c1"># If we have genes, we need to duplicate some info</span>
                <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">gene</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                        <span class="c1"># If we have genes, we don&#39;t care about this one</span>
                        <span class="k">continue</span>

                    <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">seen_genes</span><span class="p">:</span>
                        <span class="c1"># If we&#39;ve already seen it, skip</span>
                        <span class="k">continue</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># Mark as seen now we have seen it</span>
                        <span class="n">seen_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                        <span class="n">adding</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="n">gene_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                    <span class="n">gene_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">get_gene_pos</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">)</span>
                    <span class="p">)</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]</span>
                        <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                        <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span>
                    <span class="p">):</span>
                        <span class="c1"># Get codon pos</span>
                        <span class="n">nc_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                        <span class="p">]</span>
                        <span class="n">nc_num</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                        <span class="p">]</span>
                        <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">nc_num</span><span class="p">[</span><span class="n">nc_idx</span> <span class="o">==</span> <span class="n">pos</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>

                    <span class="c1"># If this isn&#39;t the first one, we need to duplicate the row</span>
                    <span class="k">if</span> <span class="n">first</span><span class="p">:</span>
                        <span class="n">first</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">refs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indices</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_indel</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_length</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_het</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_snp</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_null</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># We have 1 gene or none, so set to None if no gene is present</span>
                <span class="n">gene</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">genes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="k">else</span> <span class="kc">None</span>
                <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">seen_genes</span><span class="p">:</span>
                    <span class="c1"># If we&#39;ve already seen it, or it&#39;s intragene so skip</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Mark as seen now we have seen it</span>
                    <span class="n">seen_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                    <span class="n">adding</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">if</span> <span class="n">gene</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="c1"># Single gene, so pull out data</span>
                    <span class="n">gene_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                    <span class="n">gene_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">get_gene_pos</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">)</span>
                    <span class="p">)</span>

                    <span class="k">if</span> <span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]</span>
                        <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                        <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span>
                    <span class="p">):</span>
                        <span class="c1"># Get codon pos</span>
                        <span class="n">nc_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                        <span class="p">]</span>
                        <span class="n">nc_num</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                        <span class="p">]</span>
                        <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">nc_num</span><span class="p">[</span><span class="n">nc_idx</span> <span class="o">==</span> <span class="n">pos</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># We don&#39;t care about intragene values if its not the first item</span>
                    <span class="k">if</span> <span class="n">pos</span> <span class="o">!=</span> <span class="n">idx</span><span class="p">:</span>
                        <span class="k">continue</span>
                    <span class="n">gene_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                    <span class="n">gene_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                    <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="c1"># If this isn&#39;t the first one, we need to duplicate the row</span>
            <span class="k">if</span> <span class="n">first</span><span class="p">:</span>
                <span class="n">first</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">elif</span> <span class="n">adding</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">multi</span><span class="p">:</span>
                <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">refs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indices</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_indel</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_length</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_het</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_snp</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_null</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

    <span class="c1"># if the indel is on the LHS, then it is unchanged, then it needs &#39;reversing&#39;</span>
    <span class="c1">#   since we are returning how to get to the RHS from the LHS hence we</span>
    <span class="c1">#   delete an insertion etc</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">&amp;</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">indel_nucleotides</span>
    <span class="p">)</span>

    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">alt</span><span class="p">,</span> <span class="n">r</span><span class="p">,</span> <span class="n">a</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">indel_length</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
        <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">length</span> <span class="o">*=</span> <span class="o">-</span><span class="mi">1</span>
        <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">length</span><span class="p">)</span>
        <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">length</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_ins_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">alt</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_del_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">alt</span><span class="p">))</span>

        <span class="c1"># Find the genes at this pos</span>
        <span class="n">positions</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">seen_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">if</span> <span class="s2">&quot;ins&quot;</span> <span class="ow">in</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
            <span class="c1"># Insertion, so only need to check a single index</span>
            <span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
            <span class="n">start</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Deletion, so we need to check every deleted position for genes</span>
            <span class="n">pos</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">bases</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">idx</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">bases</span><span class="p">)):</span>
                <span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span>
        <span class="n">first</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">positions</span><span class="p">:</span>
            <span class="n">genes</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                <span class="nb">list</span><span class="p">(</span>
                    <span class="nb">set</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span> <span class="o">==</span> <span class="n">pos</span>
                        <span class="p">]</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="n">adding</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="n">multi</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">multi</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="c1"># If we have genes, we need to duplicate some info</span>
                <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">gene</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                        <span class="c1"># If we have genes, we don&#39;t care about this one</span>
                        <span class="k">continue</span>

                    <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">seen_genes</span><span class="p">:</span>
                        <span class="c1"># If we&#39;ve already seen it, skip</span>
                        <span class="k">continue</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># Mark as seen now we have seen it</span>
                        <span class="n">seen_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                        <span class="n">adding</span> <span class="o">=</span> <span class="kc">True</span>

                    <span class="n">gene_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                    <span class="n">gene_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">get_gene_pos</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">)</span>
                    <span class="p">)</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]</span>
                        <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                        <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span>
                    <span class="p">):</span>
                        <span class="c1"># Get codon pos</span>
                        <span class="n">nc_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                        <span class="p">]</span>
                        <span class="n">nc_num</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                        <span class="p">]</span>
                        <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">nc_num</span><span class="p">[</span><span class="n">nc_idx</span> <span class="o">==</span> <span class="n">pos</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                    <span class="c1"># If this isn&#39;t the first one, we need to duplicate the row</span>
                    <span class="k">if</span> <span class="n">first</span><span class="p">:</span>
                        <span class="n">first</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">refs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indices</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_indel</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_length</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_het</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_snp</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_null</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># We have 1 gene or none, so set to None if no gene is present</span>
                <span class="n">gene</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">genes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="k">else</span> <span class="kc">None</span>
                <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">seen_genes</span><span class="p">:</span>
                    <span class="c1"># If we&#39;ve already seen it, or it&#39;s intragene so skip</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Mark as seen now we have seen it</span>
                    <span class="n">seen_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                    <span class="n">adding</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">if</span> <span class="n">gene</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="c1"># Single gene, so pull out data</span>
                    <span class="n">gene_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                    <span class="n">gene_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">get_gene_pos</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">)</span>
                    <span class="p">)</span>

                    <span class="k">if</span> <span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]</span>
                        <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                        <span class="ow">and</span> <span class="n">gene_pos</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span>
                    <span class="p">):</span>
                        <span class="c1"># Get codon pos</span>
                        <span class="n">nc_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                        <span class="p">]</span>
                        <span class="n">nc_num</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                        <span class="p">]</span>
                        <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">nc_num</span><span class="p">[</span><span class="n">nc_idx</span> <span class="o">==</span> <span class="n">pos</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># We don&#39;t care about intragene values if its not the first item</span>
                    <span class="k">if</span> <span class="n">pos</span> <span class="o">!=</span> <span class="n">idx</span><span class="p">:</span>
                        <span class="k">continue</span>
                    <span class="n">gene_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                    <span class="n">gene_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                    <span class="n">codon_idx</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>

            <span class="c1"># If this isn&#39;t the first one, we need to duplicate the row</span>
            <span class="k">if</span> <span class="n">first</span><span class="p">:</span>
                <span class="n">first</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">elif</span> <span class="n">adding</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">multi</span><span class="p">:</span>
                <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variants</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">refs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indices</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">is_indel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_indel</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">indel_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_length</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">indel_nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">is_het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_het</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">is_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_snp</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">is_null</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">is_null</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">variants</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_index</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_indel</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_indel</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">indel_length</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">indel_length</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">indel_nucleotides</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">indel_nucleotides</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_snp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_snp</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_het</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_het</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_null</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">is_null</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene_name</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">gene_pos</span> <span class="o">=</span> <span class="n">gene_pos</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">codon_idx</span> <span class="o">=</span> <span class="n">codon_idx</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="difference.GenomeDifference.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">genome1</span><span class="p">,</span> <span class="n">genome2</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Constructor for the GenomeDifference object.</p>
<p>Called implictly when <code>genome1.difference(genome2)</code> is performed.
Args:
    genome1 (gumpy.Genome): A Genome object to compare against
    genome2 (gumpy.Genome): The other Genome object</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome1</span><span class="p">,</span> <span class="n">genome2</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Constructor for the GenomeDifference object.</span>

<span class="sd">    Called implictly when `genome1.difference(genome2)` is performed.</span>
<span class="sd">    Args:</span>
<span class="sd">        genome1 (gumpy.Genome): A Genome object to compare against</span>
<span class="sd">        genome2 (gumpy.Genome): The other Genome object</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># insist that both must be Genome objects</span>
    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">genome1</span><span class="p">,</span> <span class="n">gumpy</span><span class="o">.</span><span class="n">Genome</span><span class="p">)</span>
    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">genome2</span><span class="p">,</span> <span class="n">gumpy</span><span class="o">.</span><span class="n">Genome</span><span class="p">)</span>

    <span class="c1"># Checking for the same genes, give an error if the genes are different</span>
    <span class="k">if</span> <span class="n">genome1</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="o">!=</span> <span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">genes_in_1_not_2</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">genome1</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span>
            <span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="n">genes_in_2_not_1</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span>
            <span class="n">genome1</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="n">message</span> <span class="o">=</span> <span class="s2">&quot;The two genomes have different genes.&quot;</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes_in_1_not_2</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">message</span> <span class="o">+=</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot; Genome 1 has </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">genes_in_1_not_2</span><span class="p">)</span><span class="si">}</span><span class="s2"> more &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;genes: </span><span class="si">{</span><span class="n">genes_in_1_not_2</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes_in_2_not_1</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">message</span> <span class="o">+=</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot; Genome2 has </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">genes_in_2_not_1</span><span class="p">)</span><span class="si">}</span><span class="s2"> more &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;genes: </span><span class="si">{</span><span class="n">genes_in_2_not_1</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>
        <span class="k">raise</span> <span class="n">FailedComparison</span><span class="p">(</span><span class="n">message</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span> <span class="o">=</span> <span class="n">genome1</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span> <span class="o">=</span> <span class="n">genome2</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_view_method</span> <span class="o">=</span> <span class="s2">&quot;diff&quot;</span>

    <span class="c1"># Calculate SNPs</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">snp_distance</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__snp_distance</span><span class="p">()</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Where applicable, the `full` difference arrays are stored as these can be </span>
<span class="sd">    easily converted into `diff` arrays but not the other way around.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">__get_variants</span><span class="p">()</span>

    <span class="c1"># Calculate differences</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_nucleotides_full</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__nucleotides</span><span class="p">()</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">__assign_vcf_evidence</span><span class="p">()</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">update_view</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_view_method</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="difference.GenomeDifference.__nucleotides" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__nucleotides</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Calculate the difference in nucleotides
Returns:
    numpy.ndarray: Numpy array of tuples of (genome1_nucleotide,
        genome2_nucleotide)</p>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__nucleotides</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate the difference in nucleotides</span>
<span class="sd">    Returns:</span>
<span class="sd">        numpy.ndarray: Numpy array of tuples of (genome1_nucleotide,</span>
<span class="sd">            genome2_nucleotide)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_sequence</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">nucleotide_sequence</span>
    <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
        <span class="nb">list</span><span class="p">(</span>
            <span class="nb">zip</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
            <span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="difference.GenomeDifference.__snp_distance" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">__snp_distance</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Calculates the SNP distance between the two genomes</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>int</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The SNP distance between the two genomes</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">__snp_distance</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculates the SNP distance between the two genomes</span>

<span class="sd">    Returns:</span>
<span class="sd">        int: The SNP distance between the two genomes</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Ignores `z` and `x`</span>
    <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="mi">1</span>
            <span class="k">for</span> <span class="p">(</span><span class="n">base1</span><span class="p">,</span> <span class="n">base2</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">nucleotide_sequence</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">nucleotide_sequence</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">base1</span> <span class="o">!=</span> <span class="n">base2</span>
            <span class="ow">and</span> <span class="n">base1</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;t&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">,</span> <span class="s2">&quot;g&quot;</span><span class="p">]</span>
            <span class="ow">and</span> <span class="n">base2</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;t&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">,</span> <span class="s2">&quot;g&quot;</span><span class="p">]</span>
        <span class="p">]</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="difference.GenomeDifference.get_gene_pos" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_gene_pos</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">idx</span><span class="p">,</span> <span class="n">variant</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Find the gene position of a given nucleotide index.
This is considerably faster than building a whole stacked_gene_pos array
    (takes ~4.5mins for tb)</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>gene</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the gene to search</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>idx</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Nucleotide index we want the gene position of</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>variant</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Variant we're looking for in GARC</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>start</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Start position. Defaults to None</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>int | None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>int | None: Gene position of this nucleotide index
        (or None if it lies outside of the gene)</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_gene_pos</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">idx</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">variant</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">start</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span> <span class="o">|</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Find the gene position of a given nucleotide index.</span>
<span class="sd">    This is considerably faster than building a whole stacked_gene_pos array</span>
<span class="sd">        (takes ~4.5mins for tb)</span>

<span class="sd">    Args:</span>
<span class="sd">        gene (str): Name of the gene to search</span>
<span class="sd">        idx (int): Nucleotide index we want the gene position of</span>
<span class="sd">        variant (str): Variant we&#39;re looking for in GARC</span>
<span class="sd">        start (int): Start position. Defaults to None</span>

<span class="sd">    Returns:</span>
<span class="sd">        int | None: Gene position of this nucleotide index</span>
<span class="sd">                    (or None if it lies outside of the gene)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">stacked_gene_mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
    <span class="n">nc_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_index</span><span class="p">[</span><span class="n">stacked_gene_mask</span><span class="p">]</span>
    <span class="n">nc_nums</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome1</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span><span class="n">stacked_gene_mask</span><span class="p">]</span>

    <span class="c1"># Get the gene&#39;s nucleotide number</span>
    <span class="n">nc_num</span> <span class="o">=</span> <span class="n">nc_nums</span><span class="p">[</span><span class="n">nc_idx</span> <span class="o">==</span> <span class="n">idx</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># If coding SNP, pull out the codon number instead</span>
    <span class="k">if</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]</span>
        <span class="ow">and</span> <span class="n">nc_num</span> <span class="o">&gt;</span> <span class="mi">0</span>
        <span class="ow">and</span> <span class="s2">&quot;ins&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variant</span>
        <span class="ow">and</span> <span class="s2">&quot;del&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variant</span>
    <span class="p">):</span>
        <span class="c1"># Use floor division (//)</span>
        <span class="n">codon_number</span> <span class="o">=</span> <span class="p">(</span><span class="n">nc_num</span> <span class="o">+</span> <span class="mi">2</span><span class="p">)</span> <span class="o">//</span> <span class="mi">3</span>
        <span class="k">return</span> <span class="n">codon_number</span>
    <span class="k">elif</span> <span class="s2">&quot;ins&quot;</span> <span class="ow">in</span> <span class="n">variant</span><span class="p">:</span>
        <span class="c1"># Insertions need a little nudge in revcomp because of `ins at, del after`</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;reverse_complement&quot;</span><span class="p">]:</span>
            <span class="n">nc_num</span> <span class="o">-=</span> <span class="mi">1</span>
    <span class="k">elif</span> <span class="s2">&quot;del&quot;</span> <span class="ow">in</span> <span class="n">variant</span><span class="p">:</span>
        <span class="n">pos</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="n">bases</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">start</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">idx</span> <span class="o">!=</span> <span class="n">start</span><span class="p">:</span>
            <span class="c1"># This is not the start of the deletion, so truncate the bases as</span>
            <span class="c1">#   appropriate</span>
            <span class="n">dels</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">bases</span><span class="p">)</span> <span class="o">-</span> <span class="p">(</span><span class="n">idx</span> <span class="o">-</span> <span class="n">start</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">dels</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">bases</span><span class="p">)</span>
        <span class="c1"># Deletions need even more nudging in revcomp because the entire deletion</span>
        <span class="c1">#   is reversed so starts at the end</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s2">&quot;reverse_complement&quot;</span><span class="p">]:</span>
            <span class="n">nc_num</span> <span class="o">=</span> <span class="n">nc_num</span> <span class="o">-</span> <span class="n">dels</span> <span class="o">+</span> <span class="mi">1</span>

            <span class="c1"># Edge case of deletion starting in revcomp gene and extending</span>
            <span class="c1"># past gene start, so return None</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">stacked_nucleotide_number</span><span class="p">[</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">stacked_gene_name</span> <span class="o">==</span> <span class="n">gene</span>
                <span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                <span class="o">&gt;</span> <span class="n">nc_num</span>
            <span class="p">):</span>
                <span class="k">return</span> <span class="kc">None</span>

    <span class="k">return</span> <span class="n">nc_num</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="difference.GenomeDifference.minor_populations" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">minor_populations</span><span class="p">(</span><span class="n">interpretation</span><span class="o">=</span><span class="s1">&#39;reads&#39;</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Get the minor population mutations in GARC</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>interpretation</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>How to report minor population.
'reads' reports number of reads. 'percentage' reports fractional
read support. Defaults to 'reads'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;reads&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.List">List</span>[str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List[str]: List of mutations in GARC</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">minor_populations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">interpretation</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;reads&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the minor population mutations in GARC</span>

<span class="sd">    Args:</span>
<span class="sd">        interpretation (str, optional): How to report minor population.</span>
<span class="sd">            &#39;reads&#39; reports number of reads. &#39;percentage&#39; reports fractional</span>
<span class="sd">            read support. Defaults to &#39;reads&#39;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[str]: List of mutations in GARC</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome2</span><span class="o">.</span><span class="n">minority_populations_GARC</span><span class="p">(</span>
        <span class="n">interpretation</span><span class="o">=</span><span class="n">interpretation</span><span class="p">,</span> <span class="n">reference</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">genome1</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h2 id="difference.convert_nucleotides_codons" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">convert_nucleotides_codons</span><span class="p">(</span><span class="n">nucleotides</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

      <p>Helper function to convert an array of nucleotides into an array of codons</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>nucleotides</code></td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array of nucleotides</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>numpy.ndarray: Array of codons</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1513</span>
<span class="normal">1514</span>
<span class="normal">1515</span>
<span class="normal">1516</span>
<span class="normal">1517</span>
<span class="normal">1518</span>
<span class="normal">1519</span>
<span class="normal">1520</span>
<span class="normal">1521</span>
<span class="normal">1522</span>
<span class="normal">1523</span>
<span class="normal">1524</span>
<span class="normal">1525</span>
<span class="normal">1526</span>
<span class="normal">1527</span>
<span class="normal">1528</span>
<span class="normal">1529</span>
<span class="normal">1530</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">convert_nucleotides_codons</span><span class="p">(</span><span class="n">nucleotides</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Helper function to convert an array of nucleotides into an array of codons</span>

<span class="sd">    Args:</span>
<span class="sd">        nucleotides (numpy.ndarray): Array of nucleotides</span>

<span class="sd">    Returns:</span>
<span class="sd">        numpy.ndarray: Array of codons</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">codons</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">c</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="k">for</span> <span class="n">index</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotides</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
        <span class="n">c</span> <span class="o">+=</span> <span class="n">nucleotides</span><span class="p">[</span><span class="n">index</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">index</span> <span class="o">%</span> <span class="mi">3</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># There have been 3 bases seen so add the codon</span>
            <span class="n">codons</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">c</span><span class="p">)</span>
            <span class="n">c</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">codons</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="difference.setup_codon_aa_dict" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">setup_codon_aa_dict</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

      <p>Setup a conversion dictionary to convert codons to amino acids</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
                  <code><span title="typing.Dict">Dict</span>[str, str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary mapping codons of form 'xyz' to amino acids of form 'X'</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gumpy/difference.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1533</span>
<span class="normal">1534</span>
<span class="normal">1535</span>
<span class="normal">1536</span>
<span class="normal">1537</span>
<span class="normal">1538</span>
<span class="normal">1539</span>
<span class="normal">1540</span>
<span class="normal">1541</span>
<span class="normal">1542</span>
<span class="normal">1543</span>
<span class="normal">1544</span>
<span class="normal">1545</span>
<span class="normal">1546</span>
<span class="normal">1547</span>
<span class="normal">1548</span>
<span class="normal">1549</span>
<span class="normal">1550</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">setup_codon_aa_dict</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Setup a conversion dictionary to convert codons to amino acids</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Dictionary mapping codons of form &#39;xyz&#39; to amino acids of form &#39;X&#39;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Defined bases</span>
    <span class="n">bases</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;t&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;g&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">,</span> <span class="s2">&quot;o&quot;</span><span class="p">]</span>
    <span class="c1"># Defined amino acids in correct order</span>
    <span class="n">aminoacids</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;FFLLXZOSSSSXZOYY!!XZOCC!WXZOXXXXXXXZZZZXZOOOOOXOOLLLLXZOPPPPXZOHHQQXZORRRRXZOXXX&quot;</span>
        <span class="s2">&quot;XXXXZZZZXZOOOOOXOOIIIMXZOTTTTXZONNKKXZOSSRRXZOXXXXXXXZZZZXZOOOOOXOOVVVVXZOAAAAXZ&quot;</span>
        <span class="s2">&quot;ODDEEXZOGGGGXZOXXXXXXXZZZZXZOOOOOXOOXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX&quot;</span>
        <span class="s2">&quot;XXXXXZZZZXZOZZZZXZOZZZZXZOZZZZXZOXXXXXXXZZZZXZOOOOOXOOOOOOXOOOOOOXOOOOOOXOOOOOOX&quot;</span>
        <span class="s2">&quot;OOXXXXXXXOOOOXOOOOOOXOO&quot;</span>
    <span class="p">)</span>
    <span class="n">all_codons</span> <span class="o">=</span> <span class="p">[</span><span class="n">a</span> <span class="o">+</span> <span class="n">b</span> <span class="o">+</span> <span class="n">c</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">bases</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">bases</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">bases</span><span class="p">]</span>
    <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">all_codons</span><span class="p">,</span> <span class="n">aminoacids</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../" class="md-footer__link md-footer__link--prev" aria-label="Previous: gumpy">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                gumpy
              </div>
            </div>
          </a>
        
        
          
          <a href="../gene/" class="md-footer__link md-footer__link--next" aria-label="Next: gene">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                gene
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.footer", "navigation.instant", "navigation.sections", "navigation.tabs", "navigation.tracking"], "search": "../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "jeremy"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.fe8b6f2b.min.js"></script>
      
    
  </body>
</html>